{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from copy import deepcopy\n",
    "from sklearn.linear_model import Lasso, LassoCV\n",
    "from sklearn.linear_model import ElasticNet, ElasticNetCV\n",
    "from sklearn.linear_model import Ridge, RidgeCV\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.model_selection import RandomizedSearchCV, GridSearchCV\n",
    "import pprint\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "from scipy import stats\n",
    "from scipy.interpolate import CubicSpline\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "import statsmodels.stats.api as sms\n",
    "from sklearn.metrics import r2_score\n",
    "import scipy.stats.kde as kde"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read meta data (remove mice with suffix 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MiceID</th>\n",
       "      <th>Vendor</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample141</th>\n",
       "      <td>BI1</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>Inulin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample142</th>\n",
       "      <td>BI2</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>Inulin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample143</th>\n",
       "      <td>BI3</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>Inulin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample145</th>\n",
       "      <td>BI5</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>Inulin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample156</th>\n",
       "      <td>BI1</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>1</td>\n",
       "      <td>Inulin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          MiceID   Vendor  Day    Diet\n",
       "SampleID                              \n",
       "sample141    BI1  Beijing    0  Inulin\n",
       "sample142    BI2  Beijing    0  Inulin\n",
       "sample143    BI3  Beijing    0  Inulin\n",
       "sample145    BI5  Beijing    0  Inulin\n",
       "sample156    BI1  Beijing    1  Inulin"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_excel('../../../../../../../data/meta_data.xlsx', index_col=0)\n",
    "\n",
    "# change HI5 to HI3\n",
    "df_meta.loc[df_meta.MiceID=='HI5','MiceID'] = 'HI3'\n",
    "\n",
    "df_meta = df_meta[~df_meta.MiceID.str.contains('4')]\n",
    "df_meta = df_meta[df_meta.Diet == 'Inulin']\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read SCFA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Acetate</th>\n",
       "      <th>Butyrate</th>\n",
       "      <th>Propionate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample141</th>\n",
       "      <td>5.233899</td>\n",
       "      <td>1.109582</td>\n",
       "      <td>1.930933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample142</th>\n",
       "      <td>13.459823</td>\n",
       "      <td>1.546647</td>\n",
       "      <td>3.621771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample143</th>\n",
       "      <td>18.075378</td>\n",
       "      <td>1.682656</td>\n",
       "      <td>2.859916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample145</th>\n",
       "      <td>8.065127</td>\n",
       "      <td>1.835172</td>\n",
       "      <td>3.134209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample156</th>\n",
       "      <td>14.944810</td>\n",
       "      <td>5.382770</td>\n",
       "      <td>11.384803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Acetate  Butyrate  Propionate\n",
       "SampleID                                  \n",
       "sample141   5.233899  1.109582    1.930933\n",
       "sample142  13.459823  1.546647    3.621771\n",
       "sample143  18.075378  1.682656    2.859916\n",
       "sample145   8.065127  1.835172    3.134209\n",
       "sample156  14.944810  5.382770   11.384803"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_scfa = pd.read_excel('../../../../../../../data/SCFA.xlsx', index_col=0)\n",
    "df_scfa = df_scfa[['Acetate','Butyrate','Propionate']]\n",
    "common_samples = [x for x in df_meta.index if x in list(df_scfa.index)]\n",
    "df_scfa = df_scfa.loc[common_samples]\n",
    "df_meta = df_meta.loc[common_samples]\n",
    "df_scfa.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Acetate</th>\n",
       "      <th>Butyrate</th>\n",
       "      <th>Propionate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Acetate</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.549807</td>\n",
       "      <td>0.518400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Butyrate</th>\n",
       "      <td>0.549807</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.243709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Propionate</th>\n",
       "      <td>0.518400</td>\n",
       "      <td>0.243709</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Acetate  Butyrate  Propionate\n",
       "Acetate     1.000000  0.549807    0.518400\n",
       "Butyrate    0.549807  1.000000    0.243709\n",
       "Propionate  0.518400  0.243709    1.000000"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_scfa.corr()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read 16S data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2</th>\n",
       "      <th>ASF356</th>\n",
       "      <th>Acetatifactor</th>\n",
       "      <th>Acetatifactor_dash_muris</th>\n",
       "      <th>Acidovorax</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Acinetobacter_dash_sp_dot__dash_CIP_dash_56_dot_2</th>\n",
       "      <th>Acutalibacter_dash_muris</th>\n",
       "      <th>Adlercreutzia_dash_mucosicola</th>\n",
       "      <th>Adlercreutzia_dash_muris</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_ventriosum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_gnavus_dash_group</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_M153</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>marine_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample141</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>58318.686780</td>\n",
       "      <td>98146.082630</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49784.244812</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>200559.386244</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.076714e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample142</th>\n",
       "      <td>79938.738108</td>\n",
       "      <td>25274.748078</td>\n",
       "      <td>33503.735825</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18809.114849</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48198.356800</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1763.354517</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.081524e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample143</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>137987.686312</td>\n",
       "      <td>64115.490610</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>58540.230557</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>167257.801590</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.717939e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample145</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>122899.454628</td>\n",
       "      <td>119532.346282</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>69025.721092</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>368698.363883</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.407638e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample156</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>182420.266467</td>\n",
       "      <td>194421.599787</td>\n",
       "      <td>4800.533328</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24002.666640</td>\n",
       "      <td>2400.266664</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.101722e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 304 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     A2         ASF356  Acetatifactor  \\\n",
       "SampleID                                                \n",
       "sample141      0.000000   58318.686780   98146.082630   \n",
       "sample142  79938.738108   25274.748078   33503.735825   \n",
       "sample143      0.000000  137987.686312   64115.490610   \n",
       "sample145      0.000000  122899.454628  119532.346282   \n",
       "sample156      0.000000  182420.266467  194421.599787   \n",
       "\n",
       "           Acetatifactor_dash_muris  Acidovorax  Acinetobacter  \\\n",
       "SampleID                                                         \n",
       "sample141                  0.000000         0.0            0.0   \n",
       "sample142                  0.000000         0.0            0.0   \n",
       "sample143                  0.000000         0.0            0.0   \n",
       "sample145                  0.000000         0.0            0.0   \n",
       "sample156               4800.533328         0.0            0.0   \n",
       "\n",
       "           Acinetobacter_dash_sp_dot__dash_CIP_dash_56_dot_2  \\\n",
       "SampleID                                                       \n",
       "sample141                                                0.0   \n",
       "sample142                                                0.0   \n",
       "sample143                                                0.0   \n",
       "sample145                                                0.0   \n",
       "sample156                                                0.0   \n",
       "\n",
       "           Acutalibacter_dash_muris  Adlercreutzia_dash_mucosicola  \\\n",
       "SampleID                                                             \n",
       "sample141                       0.0                            0.0   \n",
       "sample142                       0.0                            0.0   \n",
       "sample143                       0.0                            0.0   \n",
       "sample145                       0.0                            0.0   \n",
       "sample156                       0.0                            0.0   \n",
       "\n",
       "           Adlercreutzia_dash_muris  ...  \\\n",
       "SampleID                             ...   \n",
       "sample141                       0.0  ...   \n",
       "sample142                       0.0  ...   \n",
       "sample143                       0.0  ...   \n",
       "sample145                       0.0  ...   \n",
       "sample156                       0.0  ...   \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_ventriosum_dash_group  \\\n",
       "SampleID                                                                                   \n",
       "sample141                                                0.0                               \n",
       "sample142                                                0.0                               \n",
       "sample143                                                0.0                               \n",
       "sample145                                                0.0                               \n",
       "sample156                                                0.0                               \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group  \\\n",
       "SampleID                                                                                     \n",
       "sample141                                       49784.244812                                 \n",
       "sample142                                       18809.114849                                 \n",
       "sample143                                       58540.230557                                 \n",
       "sample145                                       69025.721092                                 \n",
       "sample156                                           0.000000                                 \n",
       "\n",
       "           _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_gnavus_dash_group  \\\n",
       "SampleID                                                                                \n",
       "sample141                                                0.0                            \n",
       "sample142                                                0.0                            \n",
       "sample143                                                0.0                            \n",
       "sample145                                                0.0                            \n",
       "sample156                                                0.0                            \n",
       "\n",
       "           _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "SampleID                                                                                 \n",
       "sample141                                                0.0                             \n",
       "sample142                                                0.0                             \n",
       "sample143                                                0.0                             \n",
       "sample145                                                0.0                             \n",
       "sample156                                                0.0                             \n",
       "\n",
       "           bacterium_dash_enrichment_dash_culture_dash_clone_dash_M153  \\\n",
       "SampleID                                                                 \n",
       "sample141                                                0.0             \n",
       "sample142                                                0.0             \n",
       "sample143                                                0.0             \n",
       "sample145                                                0.0             \n",
       "sample156                                                0.0             \n",
       "\n",
       "           gut_dash_metagenome  human_dash_gut_dash_metagenome  \\\n",
       "SampleID                                                         \n",
       "sample141        200559.386244                        0.000000   \n",
       "sample142         48198.356800                        0.000000   \n",
       "sample143        167257.801590                        0.000000   \n",
       "sample145        368698.363883                        0.000000   \n",
       "sample156         24002.666640                     2400.266664   \n",
       "\n",
       "           marine_dash_metagenome  metagenome  mouse_dash_gut_dash_metagenome  \n",
       "SampleID                                                                       \n",
       "sample141                0.000000         0.0                    2.076714e+05  \n",
       "sample142             1763.354517         0.0                    1.081524e+05  \n",
       "sample143                0.000000         0.0                    2.717939e+05  \n",
       "sample145                0.000000         0.0                    7.407638e+04  \n",
       "sample156                0.000000         0.0                    1.101722e+06  \n",
       "\n",
       "[5 rows x 304 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = pd.read_excel('../../../../../../../data/16S_absolute_abundance.xlsx', index_col=0)\n",
    "df_16S = df_16S.loc[df_meta.index]\n",
    "df_16S.columns = [c.replace('/','_slash_').replace(' ','_space_').replace('[','_leftsquarebracket_').replace(']','_rightsquarebracket_').replace('-','_dash_').replace('.','_dot_').replace('(','_leftroundbracket').replace(')','_rightroundbracket_') for c in df_16S.columns]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 40 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Faecalibaculum</th>\n",
       "      <th>Parasutterella</th>\n",
       "      <th>Bacteroides_dash_uniformis</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Alloprevotella</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>...</th>\n",
       "      <th>Parabacteroides</th>\n",
       "      <th>Lactobacillus</th>\n",
       "      <th>Ruminococcaceae</th>\n",
       "      <th>Bilophila</th>\n",
       "      <th>Gastranaerophilales</th>\n",
       "      <th>Anaerotruncus</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Desulfovibrio</th>\n",
       "      <th>Bifidobacterium_dash_choerinum</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample141</th>\n",
       "      <td>2.558057e+07</td>\n",
       "      <td>2.803564e+06</td>\n",
       "      <td>7.311172e+05</td>\n",
       "      <td>1.503484e+07</td>\n",
       "      <td>6.520314e+06</td>\n",
       "      <td>19913.697925</td>\n",
       "      <td>3.797827e+05</td>\n",
       "      <td>6.841778e+05</td>\n",
       "      <td>1.877577e+05</td>\n",
       "      <td>5.789196e+05</td>\n",
       "      <td>...</td>\n",
       "      <td>1.934474e+05</td>\n",
       "      <td>1.069650e+06</td>\n",
       "      <td>641505.554582</td>\n",
       "      <td>2.361196e+05</td>\n",
       "      <td>2.474988e+05</td>\n",
       "      <td>3.442225e+05</td>\n",
       "      <td>2.076714e+05</td>\n",
       "      <td>9956.848962</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>98146.082630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample142</th>\n",
       "      <td>1.221770e+07</td>\n",
       "      <td>1.240226e+05</td>\n",
       "      <td>1.518836e+06</td>\n",
       "      <td>8.628681e+06</td>\n",
       "      <td>1.957324e+06</td>\n",
       "      <td>17045.760332</td>\n",
       "      <td>2.697932e+05</td>\n",
       "      <td>1.537645e+06</td>\n",
       "      <td>5.207774e+05</td>\n",
       "      <td>9.469214e+05</td>\n",
       "      <td>...</td>\n",
       "      <td>9.874785e+04</td>\n",
       "      <td>6.060062e+05</td>\n",
       "      <td>432021.856685</td>\n",
       "      <td>3.174038e+05</td>\n",
       "      <td>3.996937e+04</td>\n",
       "      <td>1.610530e+05</td>\n",
       "      <td>1.081524e+05</td>\n",
       "      <td>3526.709034</td>\n",
       "      <td>4114.493873</td>\n",
       "      <td>38206.014537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample143</th>\n",
       "      <td>2.765050e+07</td>\n",
       "      <td>3.721486e+05</td>\n",
       "      <td>1.919005e+07</td>\n",
       "      <td>7.158634e+06</td>\n",
       "      <td>4.014187e+05</td>\n",
       "      <td>27876.300265</td>\n",
       "      <td>2.028001e+06</td>\n",
       "      <td>3.982826e+07</td>\n",
       "      <td>1.076025e+06</td>\n",
       "      <td>1.998731e+06</td>\n",
       "      <td>...</td>\n",
       "      <td>7.668770e+06</td>\n",
       "      <td>5.658889e+05</td>\n",
       "      <td>682969.356493</td>\n",
       "      <td>9.617324e+04</td>\n",
       "      <td>2.313733e+05</td>\n",
       "      <td>9.700952e+05</td>\n",
       "      <td>2.717939e+05</td>\n",
       "      <td>25088.670239</td>\n",
       "      <td>33451.560318</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample145</th>\n",
       "      <td>4.461250e+07</td>\n",
       "      <td>6.279657e+05</td>\n",
       "      <td>3.589337e+06</td>\n",
       "      <td>1.075791e+07</td>\n",
       "      <td>7.912705e+04</td>\n",
       "      <td>40405.300152</td>\n",
       "      <td>3.833453e+06</td>\n",
       "      <td>2.133737e+07</td>\n",
       "      <td>1.880530e+06</td>\n",
       "      <td>2.452938e+06</td>\n",
       "      <td>...</td>\n",
       "      <td>4.865472e+05</td>\n",
       "      <td>3.198753e+04</td>\n",
       "      <td>590927.514717</td>\n",
       "      <td>1.144817e+05</td>\n",
       "      <td>1.188589e+06</td>\n",
       "      <td>5.825097e+05</td>\n",
       "      <td>7.407638e+04</td>\n",
       "      <td>8417.770865</td>\n",
       "      <td>43772.408498</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample156</th>\n",
       "      <td>6.293019e+07</td>\n",
       "      <td>2.520280e+05</td>\n",
       "      <td>6.328303e+07</td>\n",
       "      <td>5.762800e+07</td>\n",
       "      <td>8.175308e+06</td>\n",
       "      <td>67207.466593</td>\n",
       "      <td>1.785798e+06</td>\n",
       "      <td>1.848205e+05</td>\n",
       "      <td>8.160907e+04</td>\n",
       "      <td>1.022754e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>4.560507e+04</td>\n",
       "      <td>4.536504e+05</td>\n",
       "      <td>228025.333084</td>\n",
       "      <td>1.644183e+06</td>\n",
       "      <td>1.680187e+04</td>\n",
       "      <td>1.032115e+06</td>\n",
       "      <td>1.101722e+06</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14401.599984</td>\n",
       "      <td>48005.333281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Muribaculaceae  Akkermansia_dash_muciniphila  \\\n",
       "SampleID                                                  \n",
       "sample141    2.558057e+07                  2.803564e+06   \n",
       "sample142    1.221770e+07                  1.240226e+05   \n",
       "sample143    2.765050e+07                  3.721486e+05   \n",
       "sample145    4.461250e+07                  6.279657e+05   \n",
       "sample156    6.293019e+07                  2.520280e+05   \n",
       "\n",
       "           Bacteroides_dash_acidifaciens  Faecalibaculum  Parasutterella  \\\n",
       "SampleID                                                                   \n",
       "sample141                   7.311172e+05    1.503484e+07    6.520314e+06   \n",
       "sample142                   1.518836e+06    8.628681e+06    1.957324e+06   \n",
       "sample143                   1.919005e+07    7.158634e+06    4.014187e+05   \n",
       "sample145                   3.589337e+06    1.075791e+07    7.912705e+04   \n",
       "sample156                   6.328303e+07    5.762800e+07    8.175308e+06   \n",
       "\n",
       "           Bacteroides_dash_uniformis   Bacteroides  Alloprevotella  \\\n",
       "SampleID                                                              \n",
       "sample141                19913.697925  3.797827e+05    6.841778e+05   \n",
       "sample142                17045.760332  2.697932e+05    1.537645e+06   \n",
       "sample143                27876.300265  2.028001e+06    3.982826e+07   \n",
       "sample145                40405.300152  3.833453e+06    2.133737e+07   \n",
       "sample156                67207.466593  1.785798e+06    1.848205e+05   \n",
       "\n",
       "           Clostridia_dash_UCG_dash_014  \\\n",
       "SampleID                                  \n",
       "sample141                  1.877577e+05   \n",
       "sample142                  5.207774e+05   \n",
       "sample143                  1.076025e+06   \n",
       "sample145                  1.880530e+06   \n",
       "sample156                  8.160907e+04   \n",
       "\n",
       "           Lachnospiraceae_dash_NK4A136_dash_group  ...  Parabacteroides  \\\n",
       "SampleID                                            ...                    \n",
       "sample141                             5.789196e+05  ...     1.934474e+05   \n",
       "sample142                             9.469214e+05  ...     9.874785e+04   \n",
       "sample143                             1.998731e+06  ...     7.668770e+06   \n",
       "sample145                             2.452938e+06  ...     4.865472e+05   \n",
       "sample156                             1.022754e+07  ...     4.560507e+04   \n",
       "\n",
       "           Lactobacillus  Ruminococcaceae     Bilophila  Gastranaerophilales  \\\n",
       "SampleID                                                                       \n",
       "sample141   1.069650e+06    641505.554582  2.361196e+05         2.474988e+05   \n",
       "sample142   6.060062e+05    432021.856685  3.174038e+05         3.996937e+04   \n",
       "sample143   5.658889e+05    682969.356493  9.617324e+04         2.313733e+05   \n",
       "sample145   3.198753e+04    590927.514717  1.144817e+05         1.188589e+06   \n",
       "sample156   4.536504e+05    228025.333084  1.644183e+06         1.680187e+04   \n",
       "\n",
       "           Anaerotruncus  mouse_dash_gut_dash_metagenome  \\\n",
       "SampleID                                                   \n",
       "sample141   3.442225e+05                    2.076714e+05   \n",
       "sample142   1.610530e+05                    1.081524e+05   \n",
       "sample143   9.700952e+05                    2.717939e+05   \n",
       "sample145   5.825097e+05                    7.407638e+04   \n",
       "sample156   1.032115e+06                    1.101722e+06   \n",
       "\n",
       "           Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  Desulfovibrio  \\\n",
       "SampleID                                                                       \n",
       "sample141                                        9956.848962        0.000000   \n",
       "sample142                                        3526.709034     4114.493873   \n",
       "sample143                                       25088.670239    33451.560318   \n",
       "sample145                                        8417.770865    43772.408498   \n",
       "sample156                                           0.000000    14401.599984   \n",
       "\n",
       "           Bifidobacterium_dash_choerinum  \n",
       "SampleID                                   \n",
       "sample141                    98146.082630  \n",
       "sample142                    38206.014537  \n",
       "sample143                        0.000000  \n",
       "sample145                        0.000000  \n",
       "sample156                    48005.333281  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 40\n",
    "df_16S_T = df_16S.T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_16S_topX.to_csv('independent_variables_top40.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8219972118277203"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sample with minimum absolute abundance coverage\n",
    "(df_16S_topX.sum(axis=1)/df_16S.sum(axis=1)).min()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate derivative"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Acetate</th>\n",
       "      <th>Butyrate</th>\n",
       "      <th>Propionate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample141</th>\n",
       "      <td>5.880802</td>\n",
       "      <td>4.401419</td>\n",
       "      <td>12.404128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample142</th>\n",
       "      <td>18.463945</td>\n",
       "      <td>6.679364</td>\n",
       "      <td>18.569295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample143</th>\n",
       "      <td>23.205175</td>\n",
       "      <td>10.299538</td>\n",
       "      <td>27.076697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample145</th>\n",
       "      <td>8.669110</td>\n",
       "      <td>4.911881</td>\n",
       "      <td>1.435425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample156</th>\n",
       "      <td>12.145962</td>\n",
       "      <td>3.952820</td>\n",
       "      <td>6.676380</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Acetate   Butyrate  Propionate\n",
       "SampleID                                   \n",
       "sample141   5.880802   4.401419   12.404128\n",
       "sample142  18.463945   6.679364   18.569295\n",
       "sample143  23.205175  10.299538   27.076697\n",
       "sample145   8.669110   4.911881    1.435425\n",
       "sample156  12.145962   3.952820    6.676380"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_scfa_meta = pd.merge(df_meta, df_scfa, left_index=True, right_index=True, how='inner')\n",
    "df_scfa_deriv = deepcopy(df_scfa_meta)\n",
    "for curr_mice in set(df_scfa_deriv.MiceID):\n",
    "    curr_df = df_scfa_meta[df_scfa_meta.MiceID==curr_mice].sort_values(by='Day')\n",
    "    for scfa in ['Acetate','Butyrate','Propionate']:\n",
    "        xdata = np.array(curr_df['Day'])\n",
    "        ydata = np.array(curr_df[scfa])\n",
    "        cs = CubicSpline(xdata, ydata)\n",
    "        csd1 = cs.derivative(nu=1)\n",
    "        ydata_d1 = csd1(xdata)\n",
    "        df_scfa_deriv.loc[df_scfa_deriv.MiceID==curr_mice, scfa] = ydata_d1\n",
    "    \n",
    "df_scfa_deriv = df_scfa_deriv[['Acetate','Butyrate','Propionate']]\n",
    "df_scfa_deriv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_scfa_deriv.to_csv('response_variables_scfa_deriv.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Faecalibaculum</th>\n",
       "      <th>Parasutterella</th>\n",
       "      <th>Bacteroides_dash_uniformis</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Alloprevotella</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>...</th>\n",
       "      <th>Parabacteroides</th>\n",
       "      <th>Lactobacillus</th>\n",
       "      <th>Ruminococcaceae</th>\n",
       "      <th>Bilophila</th>\n",
       "      <th>Gastranaerophilales</th>\n",
       "      <th>Anaerotruncus</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Desulfovibrio</th>\n",
       "      <th>Bifidobacterium_dash_choerinum</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample141</th>\n",
       "      <td>0.068811</td>\n",
       "      <td>0.007541</td>\n",
       "      <td>0.001967</td>\n",
       "      <td>0.040443</td>\n",
       "      <td>0.017539</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.001022</td>\n",
       "      <td>0.001840</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>0.001557</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000520</td>\n",
       "      <td>0.002877</td>\n",
       "      <td>0.001726</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000666</td>\n",
       "      <td>0.000926</td>\n",
       "      <td>0.000559</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample142</th>\n",
       "      <td>0.032865</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>0.004086</td>\n",
       "      <td>0.023211</td>\n",
       "      <td>0.005265</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000726</td>\n",
       "      <td>0.004136</td>\n",
       "      <td>0.001401</td>\n",
       "      <td>0.002547</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000266</td>\n",
       "      <td>0.001630</td>\n",
       "      <td>0.001162</td>\n",
       "      <td>0.000854</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000433</td>\n",
       "      <td>0.000291</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.000103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample143</th>\n",
       "      <td>0.074379</td>\n",
       "      <td>0.001001</td>\n",
       "      <td>0.051620</td>\n",
       "      <td>0.019256</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>0.005455</td>\n",
       "      <td>0.107136</td>\n",
       "      <td>0.002894</td>\n",
       "      <td>0.005376</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020629</td>\n",
       "      <td>0.001522</td>\n",
       "      <td>0.001837</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0.000622</td>\n",
       "      <td>0.002610</td>\n",
       "      <td>0.000731</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample145</th>\n",
       "      <td>0.120006</td>\n",
       "      <td>0.001689</td>\n",
       "      <td>0.009655</td>\n",
       "      <td>0.028938</td>\n",
       "      <td>0.000213</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.010312</td>\n",
       "      <td>0.057397</td>\n",
       "      <td>0.005059</td>\n",
       "      <td>0.006598</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001309</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.001590</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.003197</td>\n",
       "      <td>0.001567</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample156</th>\n",
       "      <td>0.169279</td>\n",
       "      <td>0.000678</td>\n",
       "      <td>0.170228</td>\n",
       "      <td>0.155017</td>\n",
       "      <td>0.021991</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.004804</td>\n",
       "      <td>0.000497</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.027512</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.001220</td>\n",
       "      <td>0.000613</td>\n",
       "      <td>0.004423</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.002776</td>\n",
       "      <td>0.002964</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000129</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Muribaculaceae  Akkermansia_dash_muciniphila  \\\n",
       "SampleID                                                  \n",
       "sample141        0.068811                      0.007541   \n",
       "sample142        0.032865                      0.000334   \n",
       "sample143        0.074379                      0.001001   \n",
       "sample145        0.120006                      0.001689   \n",
       "sample156        0.169279                      0.000678   \n",
       "\n",
       "           Bacteroides_dash_acidifaciens  Faecalibaculum  Parasutterella  \\\n",
       "SampleID                                                                   \n",
       "sample141                       0.001967        0.040443        0.017539   \n",
       "sample142                       0.004086        0.023211        0.005265   \n",
       "sample143                       0.051620        0.019256        0.001080   \n",
       "sample145                       0.009655        0.028938        0.000213   \n",
       "sample156                       0.170228        0.155017        0.021991   \n",
       "\n",
       "           Bacteroides_dash_uniformis  Bacteroides  Alloprevotella  \\\n",
       "SampleID                                                             \n",
       "sample141                    0.000054     0.001022        0.001840   \n",
       "sample142                    0.000046     0.000726        0.004136   \n",
       "sample143                    0.000075     0.005455        0.107136   \n",
       "sample145                    0.000109     0.010312        0.057397   \n",
       "sample156                    0.000181     0.004804        0.000497   \n",
       "\n",
       "           Clostridia_dash_UCG_dash_014  \\\n",
       "SampleID                                  \n",
       "sample141                      0.000505   \n",
       "sample142                      0.001401   \n",
       "sample143                      0.002894   \n",
       "sample145                      0.005059   \n",
       "sample156                      0.000220   \n",
       "\n",
       "           Lachnospiraceae_dash_NK4A136_dash_group  ...  Parabacteroides  \\\n",
       "SampleID                                            ...                    \n",
       "sample141                                 0.001557  ...         0.000520   \n",
       "sample142                                 0.002547  ...         0.000266   \n",
       "sample143                                 0.005376  ...         0.020629   \n",
       "sample145                                 0.006598  ...         0.001309   \n",
       "sample156                                 0.027512  ...         0.000123   \n",
       "\n",
       "           Lactobacillus  Ruminococcaceae  Bilophila  Gastranaerophilales  \\\n",
       "SampleID                                                                    \n",
       "sample141       0.002877         0.001726   0.000635             0.000666   \n",
       "sample142       0.001630         0.001162   0.000854             0.000108   \n",
       "sample143       0.001522         0.001837   0.000259             0.000622   \n",
       "sample145       0.000086         0.001590   0.000308             0.003197   \n",
       "sample156       0.001220         0.000613   0.004423             0.000045   \n",
       "\n",
       "           Anaerotruncus  mouse_dash_gut_dash_metagenome  \\\n",
       "SampleID                                                   \n",
       "sample141       0.000926                        0.000559   \n",
       "sample142       0.000433                        0.000291   \n",
       "sample143       0.002610                        0.000731   \n",
       "sample145       0.001567                        0.000199   \n",
       "sample156       0.002776                        0.002964   \n",
       "\n",
       "           Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  Desulfovibrio  \\\n",
       "SampleID                                                                       \n",
       "sample141                                           0.000027        0.000000   \n",
       "sample142                                           0.000009        0.000011   \n",
       "sample143                                           0.000067        0.000090   \n",
       "sample145                                           0.000023        0.000118   \n",
       "sample156                                           0.000000        0.000039   \n",
       "\n",
       "           Bifidobacterium_dash_choerinum  \n",
       "SampleID                                   \n",
       "sample141                        0.000264  \n",
       "sample142                        0.000103  \n",
       "sample143                        0.000000  \n",
       "sample145                        0.000000  \n",
       "sample156                        0.000129  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# normalize bacteria data using \n",
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX = df_16S_topX/df_16S_topX.max().max()\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simple Correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SCFA</th>\n",
       "      <th>Variable</th>\n",
       "      <th>Pearson</th>\n",
       "      <th>Spearman</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acetate</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>-0.252873</td>\n",
       "      <td>-0.209436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acetate</td>\n",
       "      <td>Akkermansia_dash_muciniphila</td>\n",
       "      <td>-0.129561</td>\n",
       "      <td>-0.064612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acetate</td>\n",
       "      <td>Bacteroides_dash_acidifaciens</td>\n",
       "      <td>0.006933</td>\n",
       "      <td>0.026021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acetate</td>\n",
       "      <td>Faecalibaculum</td>\n",
       "      <td>-0.204028</td>\n",
       "      <td>-0.108495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acetate</td>\n",
       "      <td>Parasutterella</td>\n",
       "      <td>-0.197488</td>\n",
       "      <td>-0.214131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      SCFA                       Variable   Pearson  Spearman\n",
       "0  Acetate                 Muribaculaceae -0.252873 -0.209436\n",
       "1  Acetate   Akkermansia_dash_muciniphila -0.129561 -0.064612\n",
       "2  Acetate  Bacteroides_dash_acidifaciens  0.006933  0.026021\n",
       "3  Acetate                 Faecalibaculum -0.204028 -0.108495\n",
       "4  Acetate                 Parasutterella -0.197488 -0.214131"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "for scfa in ['Acetate','Butyrate','Propionate']:\n",
    "    for t in bacterial_taxa:\n",
    "        lines.append([scfa,\n",
    "                     t,\n",
    "                     pearsonr(df_scfa_deriv[scfa],df_16S_topX[t])[0],\n",
    "                     spearmanr(df_scfa_deriv[scfa],df_16S_topX[t])[0]]\n",
    "                    )\n",
    "df_corr = pd.DataFrame(lines, columns=['SCFA','Variable','Pearson','Spearman'])\n",
    "df_corr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_corr.to_excel('correlation_top40.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Elastic net"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rows = []\n",
    "for scfa in ['Acetate','Butyrate','Propionate']:\n",
    "    dydt = np.asarray(list(df_scfa_deriv[scfa]))\n",
    "    X = np.asarray(df_16S_topX.values)\n",
    "    l1_ratio = [1e-4, .1, .3, .5, .7, .9, .95, .99, 1]\n",
    "    reg = ElasticNetCV(\n",
    "        eps=1e-4,\n",
    "        n_alphas=10000,\n",
    "        cv=5,\n",
    "        random_state=0,\n",
    "        max_iter=100000,\n",
    "        tol=1e-6,\n",
    "        l1_ratio=l1_ratio\n",
    "    ).fit(X, dydt)\n",
    "    clf = ElasticNet(l1_ratio=reg.l1_ratio_, \n",
    "                     alpha=reg.alpha_, \n",
    "                     random_state=0,\n",
    "                     max_iter=100000,\n",
    "                     tol=1e-6,\n",
    "                    ).fit(X,dydt)\n",
    "    rows.append([scfa, reg.alpha_, reg.l1_ratio_, clf.score(X,dydt)]+ list(clf.coef_))\n",
    "df_elasticnet = pd.DataFrame(rows, columns=['SCFA','BestAlpha','L1Ratio','R2']+list(df_16S_topX.columns)).set_index('SCFA')\n",
    "df_elasticnet = df_elasticnet.loc[:, (df_elasticnet != 0).any(axis=0)]\n",
    "df_elasticnet.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_elasticnet = df_elasticnet.stack().reset_index()\n",
    "df_elasticnet = df_elasticnet[~df_elasticnet.level_1.isin(['BestAlpha','L1Ratio','R2'])]\n",
    "df_elasticnet.columns=['SCFA','Variable','Coefficient']\n",
    "df_elasticnet = df_elasticnet.reset_index(drop=True)\n",
    "df_elasticnet.to_excel('elasticnet_summary_top40.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Random forest regression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## parameter optimization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of trees in random forest\n",
    "n_estimators = [int(x) for x in np.linspace(start = 200, stop = 2000, num = 5)]\n",
    "# Number of features to consider at every split\n",
    "max_features = ['auto', 'sqrt']\n",
    "# Maximum number of levels in tree\n",
    "max_depth = [int(x) for x in np.linspace(10, 30, num = 3)]\n",
    "max_depth.append(None)\n",
    "# Minimum number of samples required to split a node\n",
    "min_samples_split = [2, 5]\n",
    "# Minimum number of samples required at each leaf node\n",
    "min_samples_leaf = [1, 2, 4]\n",
    "# Method of selecting samples for training each tree\n",
    "bootstrap = [True, False]\n",
    "\n",
    "# Create the random grid\n",
    "random_grid = {'n_estimators': n_estimators,\n",
    "               'max_features': max_features,\n",
    "               'max_depth': max_depth,\n",
    "               'min_samples_split': min_samples_split,\n",
    "               'min_samples_leaf': min_samples_leaf,\n",
    "               'bootstrap': bootstrap}\n",
    "\n",
    "lines = []\n",
    "for scfa in ['Acetate','Butyrate','Propionate']:\n",
    "    dydt = np.asarray(list(df_scfa_deriv[scfa]))\n",
    "    X = np.asarray(df_16S_topX.values)\n",
    "\n",
    "    # grid search\n",
    "    rf = RandomForestRegressor()\n",
    "    rf_random = GridSearchCV(\n",
    "        estimator = rf, \n",
    "        param_grid = random_grid, \n",
    "        cv = 5, \n",
    "        verbose=2, \n",
    "        n_jobs = -1)\n",
    "\n",
    "    # fit the random search model\n",
    "    rf_random.fit(X, dydt)\n",
    "    \n",
    "    # write to table\n",
    "    lines.append([scfa,\n",
    "           rf_random.best_params_['n_estimators'],\n",
    "           rf_random.best_params_['max_features'],\n",
    "           rf_random.best_params_['max_depth'],\n",
    "           rf_random.best_params_['min_samples_split'],\n",
    "           rf_random.best_params_['min_samples_leaf'],\n",
    "           rf_random.best_params_['bootstrap']\n",
    "           ])\n",
    "\n",
    "df_opt_paras = pd.DataFrame(lines, columns=['SCFA','n_estimators','max_features','max_depth','min_samples_split','min_samples_leaf','bootstrap'])\n",
    "df_opt_paras.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_opt_paras.to_excel('optimal_rf_hyperparameters_top40.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run RF using best parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_opt_paras = pd.read_excel('optimal_rf_hyperparameters_top40.xlsx', index_col=0)\n",
    "df_opt_paras.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rows = []\n",
    "for scfa in ['Acetate','Butyrate','Propionate']:\n",
    "    dydt = np.asarray(list(df_scfa_deriv[scfa]))\n",
    "    X = np.asarray(df_16S_topX.values)\n",
    "    regr = RandomForestRegressor(\n",
    "        random_state=0,\n",
    "        bootstrap=list(df_opt_paras.loc[df_opt_paras.SCFA==scfa,'bootstrap'])[0],\n",
    "        max_depth=None if str(list(df_opt_paras.loc[df_opt_paras.SCFA==scfa,'max_depth'])[0])=='nan' else list(df_opt_paras.loc[df_opt_paras.SCFA==scfa,'max_depth'])[0],\n",
    "        max_features=list(df_opt_paras.loc[df_opt_paras.SCFA==scfa,'max_features'])[0],\n",
    "        min_samples_leaf=list(df_opt_paras.loc[df_opt_paras.SCFA==scfa,'min_samples_leaf'])[0],\n",
    "        min_samples_split=list(df_opt_paras.loc[df_opt_paras.SCFA==scfa,'min_samples_split'])[0],\n",
    "        n_estimators=list(df_opt_paras.loc[df_opt_paras.SCFA==scfa,'n_estimators'])[0],\n",
    "        n_jobs=-1\n",
    "    )\n",
    "    clf = regr.fit(X, dydt)\n",
    "    rows.append([scfa, clf.score(X, dydt)]+ list(clf.feature_importances_))\n",
    "df_rf = pd.DataFrame(rows, columns=['SCFA','R2']+list(df_16S_topX.columns))\n",
    "df_rf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rf = df_rf.set_index('SCFA').stack().reset_index()\n",
    "df_rf = df_rf[~df_rf.level_1.isin(['R2'])]\n",
    "df_rf.columns=['SCFA','Variable','Gini']\n",
    "df_rf = df_rf.reset_index(drop=True)\n",
    "df_rf.to_excel('rf_importance_top40.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rf = pd.read_excel('rf_importance_top40.xlsx', index_col=0).set_index('SCFA')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20, 10), nrows=1, ncols=3, constrained_layout=True)\n",
    "for k,scfa in enumerate(['Acetate','Butyrate','Propionate']):\n",
    "    curr_df = df_rf.loc[scfa].reset_index()\n",
    "    curr_df.columns = ['SCFA','Taxon','Importance']\n",
    "    curr_df = curr_df.sort_values('Importance',ascending=False)\n",
    "    _ = sns.barplot(y='Taxon', x='Importance',  data=curr_df, orient='h', ax=ax[k], palette='Set1')\n",
    "    ax[k].legend(loc='lower right')\n",
    "    ax[k].set_xlim(0,0.12)\n",
    "    ax[k].set_xticks([0,0.06,0.12])\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Combine results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_corr = pd.read_excel('correlation_top40.xlsx', index_col=0)\n",
    "df_corr = df_corr.drop('Spearman', axis=1)\n",
    "df_corr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_elasnet = pd.read_excel('elasticnet_summary_top40.xlsx', index_col=0)\n",
    "df_elasnet.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rf = pd.read_excel('rf_importance_top40.xlsx', index_col=0)\n",
    "df_rf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_joined = pd.merge(df_rf, df_corr, left_on=['SCFA','Variable'], right_on=['SCFA','Variable'], how='inner')\n",
    "df_joined = pd.merge(df_rf, df_elasnet, left_on=['SCFA','Variable'], right_on=['SCFA','Variable'], how='inner')\n",
    "df_joined.columns = ['SCFA','Taxon','RandomForest','ElasticNet']\n",
    "df_joined['Taxon'] = [t.replace('_dash_','-').replace('_dot_','.') for t in df_joined['Taxon']]\n",
    "df_joined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10, 6), nrows=1, ncols=3, constrained_layout=True, sharex=True)\n",
    "cbar_ax = fig.add_axes([0.94, .45, .015, .2])\n",
    "for k,scfa in enumerate(['Acetate','Butyrate','Propionate']):\n",
    "    curr_df = df_joined[df_joined.SCFA==scfa].set_index('Taxon')\n",
    "    curr_df = curr_df.drop('SCFA', axis=1)\n",
    "    curr_df = curr_df.sort_values(by='RandomForest', ascending=False)\n",
    "    curr_df = curr_df.iloc[0:10,:] \n",
    "        \n",
    "    # normalize to -1 and 1\n",
    "    for c in ['ElasticNet','RandomForest']:\n",
    "        if c!='RandomForest':\n",
    "            curr_df.loc[curr_df[c]<0,c] = curr_df.loc[curr_df[c]<0,c]/np.abs(curr_df.loc[curr_df[c]<0,c].min())\n",
    "        curr_df.loc[curr_df[c]>=0,c] = curr_df.loc[curr_df[c]>=0,c]/curr_df.loc[curr_df[c]>=0,c].max()\n",
    "\n",
    "    # plot    \n",
    "    _ = sns.heatmap(curr_df.values, ax=ax[k], cmap='RdBu_r', vmin=-1, vmax=1, cbar = False if k!=2 else True, cbar_ax = None if k!=2 else cbar_ax, linewidths=0.25, linecolor='black')\n",
    "    ax[k].set_xticklabels(['RandomForest','ElasticNet'], rotation=90, fontsize=12)\n",
    "    ax[k].set_title(scfa, fontsize=15)\n",
    "    ax[k].set_yticklabels(curr_df.index, rotation=0, fontsize=12)\n",
    "    ax[k].tick_params(axis=u'both', which=u'both',length=0)\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "fig.savefig(\"scfa_producer_top10.pdf\", format=\"pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
