{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from scipy.interpolate import CubicSpline\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import warnings\n",
    "import itertools\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "import statsmodels.stats.api as sms\n",
    "import scipy.stats.kde as kde"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read 16S data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2</th>\n",
       "      <th>ASF356</th>\n",
       "      <th>Acetatifactor</th>\n",
       "      <th>Acetatifactor_dash_muris</th>\n",
       "      <th>Acidovorax</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Acinetobacter_dash_sp_dot__dash_CIP_dash_56_dot_2</th>\n",
       "      <th>Acutalibacter_dash_muris</th>\n",
       "      <th>Adlercreutzia_dash_mucosicola</th>\n",
       "      <th>Adlercreutzia_dash_muris</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_ventriosum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_gnavus_dash_group</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_M153</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>marine_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>0.0</td>\n",
       "      <td>108301.701888</td>\n",
       "      <td>7101.750943</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>63028.039623</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>165115.709436</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.322701e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>0.0</td>\n",
       "      <td>405068.034693</td>\n",
       "      <td>50064.588558</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>80103.341692</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>106501.033841</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.003112e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>0.0</td>\n",
       "      <td>36169.603950</td>\n",
       "      <td>45332.570284</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>221840.237562</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.866352e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>0.0</td>\n",
       "      <td>484302.054587</td>\n",
       "      <td>10150.043060</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>105850.449056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>919303.900024</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.714537e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>0.0</td>\n",
       "      <td>91489.970533</td>\n",
       "      <td>42374.302141</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12519.680178</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>581683.602124</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.335310e+05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 304 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            A2         ASF356  Acetatifactor  Acetatifactor_dash_muris  \\\n",
       "SampleID                                                                 \n",
       "sample131  0.0  108301.701888    7101.750943                       0.0   \n",
       "sample132  0.0  405068.034693   50064.588558                       0.0   \n",
       "sample133  0.0   36169.603950   45332.570284                       0.0   \n",
       "sample134  0.0  484302.054587   10150.043060                       0.0   \n",
       "sample135  0.0   91489.970533   42374.302141                       0.0   \n",
       "\n",
       "           Acidovorax  Acinetobacter  \\\n",
       "SampleID                               \n",
       "sample131         0.0            0.0   \n",
       "sample132         0.0            0.0   \n",
       "sample133         0.0            0.0   \n",
       "sample134         0.0            0.0   \n",
       "sample135         0.0            0.0   \n",
       "\n",
       "           Acinetobacter_dash_sp_dot__dash_CIP_dash_56_dot_2  \\\n",
       "SampleID                                                       \n",
       "sample131                                                0.0   \n",
       "sample132                                                0.0   \n",
       "sample133                                                0.0   \n",
       "sample134                                                0.0   \n",
       "sample135                                                0.0   \n",
       "\n",
       "           Acutalibacter_dash_muris  Adlercreutzia_dash_mucosicola  \\\n",
       "SampleID                                                             \n",
       "sample131                       0.0                       0.000000   \n",
       "sample132                       0.0                       0.000000   \n",
       "sample133                       0.0                       0.000000   \n",
       "sample134                       0.0                       0.000000   \n",
       "sample135                       0.0                   12519.680178   \n",
       "\n",
       "           Adlercreutzia_dash_muris  ...  \\\n",
       "SampleID                             ...   \n",
       "sample131                       0.0  ...   \n",
       "sample132                       0.0  ...   \n",
       "sample133                       0.0  ...   \n",
       "sample134                       0.0  ...   \n",
       "sample135                       0.0  ...   \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_ventriosum_dash_group  \\\n",
       "SampleID                                                                                   \n",
       "sample131                                                0.0                               \n",
       "sample132                                                0.0                               \n",
       "sample133                                                0.0                               \n",
       "sample134                                                0.0                               \n",
       "sample135                                                0.0                               \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group  \\\n",
       "SampleID                                                                                     \n",
       "sample131                                       63028.039623                                 \n",
       "sample132                                       80103.341692                                 \n",
       "sample133                                           0.000000                                 \n",
       "sample134                                      105850.449056                                 \n",
       "sample135                                           0.000000                                 \n",
       "\n",
       "           _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_gnavus_dash_group  \\\n",
       "SampleID                                                                                \n",
       "sample131                                                0.0                            \n",
       "sample132                                                0.0                            \n",
       "sample133                                                0.0                            \n",
       "sample134                                                0.0                            \n",
       "sample135                                                0.0                            \n",
       "\n",
       "           _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "SampleID                                                                                 \n",
       "sample131                                                0.0                             \n",
       "sample132                                                0.0                             \n",
       "sample133                                                0.0                             \n",
       "sample134                                                0.0                             \n",
       "sample135                                                0.0                             \n",
       "\n",
       "           bacterium_dash_enrichment_dash_culture_dash_clone_dash_M153  \\\n",
       "SampleID                                                                 \n",
       "sample131                                                0.0             \n",
       "sample132                                                0.0             \n",
       "sample133                                                0.0             \n",
       "sample134                                                0.0             \n",
       "sample135                                                0.0             \n",
       "\n",
       "           gut_dash_metagenome  human_dash_gut_dash_metagenome  \\\n",
       "SampleID                                                         \n",
       "sample131        165115.709436                             0.0   \n",
       "sample132        106501.033841                             0.0   \n",
       "sample133        221840.237562                             0.0   \n",
       "sample134        919303.900024                             0.0   \n",
       "sample135        581683.602124                             0.0   \n",
       "\n",
       "           marine_dash_metagenome  metagenome  mouse_dash_gut_dash_metagenome  \n",
       "SampleID                                                                       \n",
       "sample131                     0.0         0.0                    1.322701e+05  \n",
       "sample132                     0.0         0.0                    1.003112e+06  \n",
       "sample133                     0.0         0.0                    1.866352e+05  \n",
       "sample134                     0.0         0.0                    8.714537e+05  \n",
       "sample135                     0.0         0.0                    5.335310e+05  \n",
       "\n",
       "[5 rows x 304 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = pd.read_excel('../../../../data/16S_absolute_abundance.xlsx', index_col=0)\n",
    "df_16S.columns = [c.replace('/','_slash_').replace(' ','_space_').replace('[','_leftsquarebracket_').replace(']','_rightsquarebracket_').replace('-','_dash_').replace('.','_dot_').replace('(','_leftroundbracket').replace(')','_rightroundbracket_') for c in df_16S.columns]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add pseudocount for zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2</th>\n",
       "      <th>ASF356</th>\n",
       "      <th>Acetatifactor</th>\n",
       "      <th>Acetatifactor_dash_muris</th>\n",
       "      <th>Acidovorax</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Acinetobacter_dash_sp_dot__dash_CIP_dash_56_dot_2</th>\n",
       "      <th>Acutalibacter_dash_muris</th>\n",
       "      <th>Adlercreutzia_dash_mucosicola</th>\n",
       "      <th>Adlercreutzia_dash_muris</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_ventriosum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_gnavus_dash_group</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_M153</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>marine_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>887.718868</td>\n",
       "      <td>108301.701888</td>\n",
       "      <td>7101.750943</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>...</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>63028.039623</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>165115.709436</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>1.322701e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>910.265247</td>\n",
       "      <td>405068.034693</td>\n",
       "      <td>50064.588558</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>...</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>80103.341692</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>106501.033841</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>1.003112e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>482.261386</td>\n",
       "      <td>36169.603950</td>\n",
       "      <td>45332.570284</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>...</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>221840.237562</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>1.866352e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>2900.012303</td>\n",
       "      <td>484302.054587</td>\n",
       "      <td>10150.043060</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>...</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>105850.449056</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>919303.900024</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>8.714537e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>1926.104643</td>\n",
       "      <td>91489.970533</td>\n",
       "      <td>42374.302141</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>12519.680178</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>...</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>581683.602124</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>5.335310e+05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 304 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    A2         ASF356  Acetatifactor  \\\n",
       "SampleID                                               \n",
       "sample131   887.718868  108301.701888    7101.750943   \n",
       "sample132   910.265247  405068.034693   50064.588558   \n",
       "sample133   482.261386   36169.603950   45332.570284   \n",
       "sample134  2900.012303  484302.054587   10150.043060   \n",
       "sample135  1926.104643   91489.970533   42374.302141   \n",
       "\n",
       "           Acetatifactor_dash_muris   Acidovorax  Acinetobacter  \\\n",
       "SampleID                                                          \n",
       "sample131                887.718868   887.718868     887.718868   \n",
       "sample132                910.265247   910.265247     910.265247   \n",
       "sample133                482.261386   482.261386     482.261386   \n",
       "sample134               2900.012303  2900.012303    2900.012303   \n",
       "sample135               1926.104643  1926.104643    1926.104643   \n",
       "\n",
       "           Acinetobacter_dash_sp_dot__dash_CIP_dash_56_dot_2  \\\n",
       "SampleID                                                       \n",
       "sample131                                         887.718868   \n",
       "sample132                                         910.265247   \n",
       "sample133                                         482.261386   \n",
       "sample134                                        2900.012303   \n",
       "sample135                                        1926.104643   \n",
       "\n",
       "           Acutalibacter_dash_muris  Adlercreutzia_dash_mucosicola  \\\n",
       "SampleID                                                             \n",
       "sample131                887.718868                     887.718868   \n",
       "sample132                910.265247                     910.265247   \n",
       "sample133                482.261386                     482.261386   \n",
       "sample134               2900.012303                    2900.012303   \n",
       "sample135               1926.104643                   12519.680178   \n",
       "\n",
       "           Adlercreutzia_dash_muris  ...  \\\n",
       "SampleID                             ...   \n",
       "sample131                887.718868  ...   \n",
       "sample132                910.265247  ...   \n",
       "sample133                482.261386  ...   \n",
       "sample134               2900.012303  ...   \n",
       "sample135               1926.104643  ...   \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_ventriosum_dash_group  \\\n",
       "SampleID                                                                                   \n",
       "sample131                                         887.718868                               \n",
       "sample132                                         910.265247                               \n",
       "sample133                                         482.261386                               \n",
       "sample134                                        2900.012303                               \n",
       "sample135                                        1926.104643                               \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group  \\\n",
       "SampleID                                                                                     \n",
       "sample131                                       63028.039623                                 \n",
       "sample132                                       80103.341692                                 \n",
       "sample133                                         482.261386                                 \n",
       "sample134                                      105850.449056                                 \n",
       "sample135                                        1926.104643                                 \n",
       "\n",
       "           _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_gnavus_dash_group  \\\n",
       "SampleID                                                                                \n",
       "sample131                                         887.718868                            \n",
       "sample132                                         910.265247                            \n",
       "sample133                                         482.261386                            \n",
       "sample134                                        2900.012303                            \n",
       "sample135                                        1926.104643                            \n",
       "\n",
       "           _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "SampleID                                                                                 \n",
       "sample131                                         887.718868                             \n",
       "sample132                                         910.265247                             \n",
       "sample133                                         482.261386                             \n",
       "sample134                                        2900.012303                             \n",
       "sample135                                        1926.104643                             \n",
       "\n",
       "           bacterium_dash_enrichment_dash_culture_dash_clone_dash_M153  \\\n",
       "SampleID                                                                 \n",
       "sample131                                         887.718868             \n",
       "sample132                                         910.265247             \n",
       "sample133                                         482.261386             \n",
       "sample134                                        2900.012303             \n",
       "sample135                                        1926.104643             \n",
       "\n",
       "           gut_dash_metagenome  human_dash_gut_dash_metagenome  \\\n",
       "SampleID                                                         \n",
       "sample131        165115.709436                      887.718868   \n",
       "sample132        106501.033841                      910.265247   \n",
       "sample133        221840.237562                      482.261386   \n",
       "sample134        919303.900024                     2900.012303   \n",
       "sample135        581683.602124                     1926.104643   \n",
       "\n",
       "           marine_dash_metagenome   metagenome  mouse_dash_gut_dash_metagenome  \n",
       "SampleID                                                                        \n",
       "sample131              887.718868   887.718868                    1.322701e+05  \n",
       "sample132              910.265247   910.265247                    1.003112e+06  \n",
       "sample133              482.261386   482.261386                    1.866352e+05  \n",
       "sample134             2900.012303  2900.012303                    8.714537e+05  \n",
       "sample135             1926.104643  1926.104643                    5.335310e+05  \n",
       "\n",
       "[5 rows x 304 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample_id in df_16S.index:\n",
    "    sample = np.array(df_16S.loc[sample_id])\n",
    "    minval = np.min(sample[np.nonzero(sample)]) # minimum non-zero value\n",
    "    sample[sample==0] = minval\n",
    "    df_16S.loc[sample_id] = sample\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata for inulin and control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MiceID</th>\n",
       "      <th>Vendor</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>BC1</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>BC2</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>BC3</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>BC4</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>BC5</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          MiceID   Vendor  Day  Diet\n",
       "SampleID                            \n",
       "sample131    BC1  Beijing    0     0\n",
       "sample132    BC2  Beijing    0     0\n",
       "sample133    BC3  Beijing    0     0\n",
       "sample134    BC4  Beijing    0     0\n",
       "sample135    BC5  Beijing    0     0"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_excel('../../../../data/meta_data.xlsx', index_col=0)\n",
    "df_meta = df_meta[(df_meta.Diet=='Inulin') | (df_meta.Diet=='Control')]\n",
    "df_meta['Diet'] = [1 if g=='Inulin' else 0 for g in df_meta['Diet']]\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 20 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Faecalibaculum</th>\n",
       "      <th>Parasutterella</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Alloprevotella</th>\n",
       "      <th>Bacteroides_dash_uniformis</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Parabacteroides_dash_goldsteinii</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>Desulfovibrionaceae</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>Lachnospiraceae_dash_bacterium_dash_28_dash_4</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Helicobacter_dash_ganmani</th>\n",
       "      <th>Escherichia_dash_Shigella</th>\n",
       "      <th>Rikenellaceae_dash_RC9_dash_gut_dash_group</th>\n",
       "      <th>Mucispirillum_dash_schaedleri</th>\n",
       "      <th>Clostridia_dash_vadinBB60_dash_group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>2.178906e+07</td>\n",
       "      <td>1.682227e+06</td>\n",
       "      <td>3.771030e+06</td>\n",
       "      <td>2.175266e+07</td>\n",
       "      <td>2.213083e+06</td>\n",
       "      <td>3.826068e+05</td>\n",
       "      <td>9.717858e+06</td>\n",
       "      <td>69242.071699</td>\n",
       "      <td>3.380433e+06</td>\n",
       "      <td>103863.107548</td>\n",
       "      <td>2.719971e+06</td>\n",
       "      <td>1.819824e+05</td>\n",
       "      <td>170442.022643</td>\n",
       "      <td>1.012000e+05</td>\n",
       "      <td>6.249541e+05</td>\n",
       "      <td>5326.313208</td>\n",
       "      <td>32845.598114</td>\n",
       "      <td>6.779509e+06</td>\n",
       "      <td>2.008020e+06</td>\n",
       "      <td>470491.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>2.539822e+07</td>\n",
       "      <td>4.455748e+06</td>\n",
       "      <td>2.610641e+06</td>\n",
       "      <td>1.732963e+07</td>\n",
       "      <td>3.591907e+06</td>\n",
       "      <td>2.645231e+06</td>\n",
       "      <td>2.849130e+05</td>\n",
       "      <td>100129.177115</td>\n",
       "      <td>4.563160e+06</td>\n",
       "      <td>59167.241023</td>\n",
       "      <td>3.721164e+06</td>\n",
       "      <td>9.912789e+05</td>\n",
       "      <td>71910.954474</td>\n",
       "      <td>7.898372e+06</td>\n",
       "      <td>3.932346e+06</td>\n",
       "      <td>11833.448205</td>\n",
       "      <td>18205.304930</td>\n",
       "      <td>3.418046e+06</td>\n",
       "      <td>1.345281e+07</td>\n",
       "      <td>142911.643701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>1.379895e+07</td>\n",
       "      <td>1.339240e+06</td>\n",
       "      <td>3.729327e+06</td>\n",
       "      <td>1.250166e+07</td>\n",
       "      <td>4.753168e+06</td>\n",
       "      <td>7.802989e+05</td>\n",
       "      <td>1.040238e+06</td>\n",
       "      <td>49672.922758</td>\n",
       "      <td>7.508810e+05</td>\n",
       "      <td>46297.093056</td>\n",
       "      <td>7.354486e+05</td>\n",
       "      <td>1.504656e+05</td>\n",
       "      <td>55460.059390</td>\n",
       "      <td>2.083369e+05</td>\n",
       "      <td>5.290407e+05</td>\n",
       "      <td>5304.875246</td>\n",
       "      <td>9645.227720</td>\n",
       "      <td>5.057957e+06</td>\n",
       "      <td>4.054854e+06</td>\n",
       "      <td>175060.883119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>3.867746e+07</td>\n",
       "      <td>1.024574e+07</td>\n",
       "      <td>2.262880e+07</td>\n",
       "      <td>2.484296e+07</td>\n",
       "      <td>4.982221e+06</td>\n",
       "      <td>4.157168e+06</td>\n",
       "      <td>3.694616e+06</td>\n",
       "      <td>126150.535177</td>\n",
       "      <td>8.172235e+06</td>\n",
       "      <td>184150.781235</td>\n",
       "      <td>4.120917e+06</td>\n",
       "      <td>2.321460e+06</td>\n",
       "      <td>69600.295270</td>\n",
       "      <td>1.566007e+05</td>\n",
       "      <td>5.088072e+06</td>\n",
       "      <td>20300.086120</td>\n",
       "      <td>40600.172241</td>\n",
       "      <td>2.312180e+07</td>\n",
       "      <td>5.086622e+06</td>\n",
       "      <td>194300.824295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>1.994192e+07</td>\n",
       "      <td>4.775776e+06</td>\n",
       "      <td>4.727624e+06</td>\n",
       "      <td>1.935254e+07</td>\n",
       "      <td>3.164590e+06</td>\n",
       "      <td>1.040097e+06</td>\n",
       "      <td>1.814391e+06</td>\n",
       "      <td>25039.360356</td>\n",
       "      <td>1.251005e+06</td>\n",
       "      <td>72228.924105</td>\n",
       "      <td>1.094990e+06</td>\n",
       "      <td>2.725438e+05</td>\n",
       "      <td>533530.986054</td>\n",
       "      <td>6.067230e+04</td>\n",
       "      <td>5.441246e+05</td>\n",
       "      <td>6741.366250</td>\n",
       "      <td>33706.831249</td>\n",
       "      <td>1.064558e+07</td>\n",
       "      <td>7.617744e+05</td>\n",
       "      <td>45263.459106</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Muribaculaceae  Bacteroides_dash_acidifaciens  \\\n",
       "SampleID                                                   \n",
       "sample131    2.178906e+07                   1.682227e+06   \n",
       "sample132    2.539822e+07                   4.455748e+06   \n",
       "sample133    1.379895e+07                   1.339240e+06   \n",
       "sample134    3.867746e+07                   1.024574e+07   \n",
       "sample135    1.994192e+07                   4.775776e+06   \n",
       "\n",
       "           Akkermansia_dash_muciniphila  Faecalibaculum  Parasutterella  \\\n",
       "SampleID                                                                  \n",
       "sample131                  3.771030e+06    2.175266e+07    2.213083e+06   \n",
       "sample132                  2.610641e+06    1.732963e+07    3.591907e+06   \n",
       "sample133                  3.729327e+06    1.250166e+07    4.753168e+06   \n",
       "sample134                  2.262880e+07    2.484296e+07    4.982221e+06   \n",
       "sample135                  4.727624e+06    1.935254e+07    3.164590e+06   \n",
       "\n",
       "            Bacteroides  Alloprevotella  Bacteroides_dash_uniformis  \\\n",
       "SampleID                                                              \n",
       "sample131  3.826068e+05    9.717858e+06                69242.071699   \n",
       "sample132  2.645231e+06    2.849130e+05               100129.177115   \n",
       "sample133  7.802989e+05    1.040238e+06                49672.922758   \n",
       "sample134  4.157168e+06    3.694616e+06               126150.535177   \n",
       "sample135  1.040097e+06    1.814391e+06                25039.360356   \n",
       "\n",
       "           Lachnospiraceae  Parabacteroides_dash_goldsteinii  \\\n",
       "SampleID                                                       \n",
       "sample131     3.380433e+06                     103863.107548   \n",
       "sample132     4.563160e+06                      59167.241023   \n",
       "sample133     7.508810e+05                      46297.093056   \n",
       "sample134     8.172235e+06                     184150.781235   \n",
       "sample135     1.251005e+06                      72228.924105   \n",
       "\n",
       "           Lachnospiraceae_dash_NK4A136_dash_group  Desulfovibrionaceae  \\\n",
       "SampleID                                                                  \n",
       "sample131                             2.719971e+06         1.819824e+05   \n",
       "sample132                             3.721164e+06         9.912789e+05   \n",
       "sample133                             7.354486e+05         1.504656e+05   \n",
       "sample134                             4.120917e+06         2.321460e+06   \n",
       "sample135                             1.094990e+06         2.725438e+05   \n",
       "\n",
       "           Clostridia_dash_UCG_dash_014  \\\n",
       "SampleID                                  \n",
       "sample131                 170442.022643   \n",
       "sample132                  71910.954474   \n",
       "sample133                  55460.059390   \n",
       "sample134                  69600.295270   \n",
       "sample135                 533530.986054   \n",
       "\n",
       "           Lachnospiraceae_dash_bacterium_dash_28_dash_4  Oscillospiraceae  \\\n",
       "SampleID                                                                     \n",
       "sample131                                   1.012000e+05      6.249541e+05   \n",
       "sample132                                   7.898372e+06      3.932346e+06   \n",
       "sample133                                   2.083369e+05      5.290407e+05   \n",
       "sample134                                   1.566007e+05      5.088072e+06   \n",
       "sample135                                   6.067230e+04      5.441246e+05   \n",
       "\n",
       "           Helicobacter_dash_ganmani  Escherichia_dash_Shigella  \\\n",
       "SampleID                                                          \n",
       "sample131                5326.313208               32845.598114   \n",
       "sample132               11833.448205               18205.304930   \n",
       "sample133                5304.875246                9645.227720   \n",
       "sample134               20300.086120               40600.172241   \n",
       "sample135                6741.366250               33706.831249   \n",
       "\n",
       "           Rikenellaceae_dash_RC9_dash_gut_dash_group  \\\n",
       "SampleID                                                \n",
       "sample131                                6.779509e+06   \n",
       "sample132                                3.418046e+06   \n",
       "sample133                                5.057957e+06   \n",
       "sample134                                2.312180e+07   \n",
       "sample135                                1.064558e+07   \n",
       "\n",
       "           Mucispirillum_dash_schaedleri  Clostridia_dash_vadinBB60_dash_group  \n",
       "SampleID                                                                        \n",
       "sample131                   2.008020e+06                         470491.000005  \n",
       "sample132                   1.345281e+07                         142911.643701  \n",
       "sample133                   4.054854e+06                         175060.883119  \n",
       "sample134                   5.086622e+06                         194300.824295  \n",
       "sample135                   7.617744e+05                          45263.459106  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 20\n",
    "df_16S_T = df_16S.loc[df_meta.index].T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MiceID</th>\n",
       "      <th>Vendor</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Faecalibaculum</th>\n",
       "      <th>Parasutterella</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>...</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>Desulfovibrionaceae</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>Lachnospiraceae_dash_bacterium_dash_28_dash_4</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Helicobacter_dash_ganmani</th>\n",
       "      <th>Escherichia_dash_Shigella</th>\n",
       "      <th>Rikenellaceae_dash_RC9_dash_gut_dash_group</th>\n",
       "      <th>Mucispirillum_dash_schaedleri</th>\n",
       "      <th>Clostridia_dash_vadinBB60_dash_group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>BC1</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.058612</td>\n",
       "      <td>0.004525</td>\n",
       "      <td>0.010144</td>\n",
       "      <td>0.058514</td>\n",
       "      <td>0.005953</td>\n",
       "      <td>0.001029</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007317</td>\n",
       "      <td>0.000490</td>\n",
       "      <td>0.000458</td>\n",
       "      <td>0.000272</td>\n",
       "      <td>0.001681</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.018237</td>\n",
       "      <td>0.005401</td>\n",
       "      <td>0.001266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>BC2</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.068320</td>\n",
       "      <td>0.011986</td>\n",
       "      <td>0.007023</td>\n",
       "      <td>0.046616</td>\n",
       "      <td>0.009662</td>\n",
       "      <td>0.007116</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010010</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.000193</td>\n",
       "      <td>0.021246</td>\n",
       "      <td>0.010578</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.009194</td>\n",
       "      <td>0.036187</td>\n",
       "      <td>0.000384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>BC3</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.037119</td>\n",
       "      <td>0.003602</td>\n",
       "      <td>0.010032</td>\n",
       "      <td>0.033629</td>\n",
       "      <td>0.012786</td>\n",
       "      <td>0.002099</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>0.000405</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>0.000560</td>\n",
       "      <td>0.001423</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.013606</td>\n",
       "      <td>0.010907</td>\n",
       "      <td>0.000471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>BC4</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.104041</td>\n",
       "      <td>0.027561</td>\n",
       "      <td>0.060870</td>\n",
       "      <td>0.066826</td>\n",
       "      <td>0.013402</td>\n",
       "      <td>0.011183</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011085</td>\n",
       "      <td>0.006245</td>\n",
       "      <td>0.000187</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.013687</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.062197</td>\n",
       "      <td>0.013683</td>\n",
       "      <td>0.000523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>BC5</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.053643</td>\n",
       "      <td>0.012847</td>\n",
       "      <td>0.012717</td>\n",
       "      <td>0.052057</td>\n",
       "      <td>0.008513</td>\n",
       "      <td>0.002798</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002945</td>\n",
       "      <td>0.000733</td>\n",
       "      <td>0.001435</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>0.001464</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>0.028636</td>\n",
       "      <td>0.002049</td>\n",
       "      <td>0.000122</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          MiceID   Vendor  Day  Diet  Muribaculaceae  \\\n",
       "SampleID                                               \n",
       "sample131    BC1  Beijing    0     0        0.058612   \n",
       "sample132    BC2  Beijing    0     0        0.068320   \n",
       "sample133    BC3  Beijing    0     0        0.037119   \n",
       "sample134    BC4  Beijing    0     0        0.104041   \n",
       "sample135    BC5  Beijing    0     0        0.053643   \n",
       "\n",
       "           Bacteroides_dash_acidifaciens  Akkermansia_dash_muciniphila  \\\n",
       "SampleID                                                                 \n",
       "sample131                       0.004525                      0.010144   \n",
       "sample132                       0.011986                      0.007023   \n",
       "sample133                       0.003602                      0.010032   \n",
       "sample134                       0.027561                      0.060870   \n",
       "sample135                       0.012847                      0.012717   \n",
       "\n",
       "           Faecalibaculum  Parasutterella  Bacteroides  ...  \\\n",
       "SampleID                                                ...   \n",
       "sample131        0.058514        0.005953     0.001029  ...   \n",
       "sample132        0.046616        0.009662     0.007116  ...   \n",
       "sample133        0.033629        0.012786     0.002099  ...   \n",
       "sample134        0.066826        0.013402     0.011183  ...   \n",
       "sample135        0.052057        0.008513     0.002798  ...   \n",
       "\n",
       "           Lachnospiraceae_dash_NK4A136_dash_group  Desulfovibrionaceae  \\\n",
       "SampleID                                                                  \n",
       "sample131                                 0.007317             0.000490   \n",
       "sample132                                 0.010010             0.002666   \n",
       "sample133                                 0.001978             0.000405   \n",
       "sample134                                 0.011085             0.006245   \n",
       "sample135                                 0.002945             0.000733   \n",
       "\n",
       "           Clostridia_dash_UCG_dash_014  \\\n",
       "SampleID                                  \n",
       "sample131                      0.000458   \n",
       "sample132                      0.000193   \n",
       "sample133                      0.000149   \n",
       "sample134                      0.000187   \n",
       "sample135                      0.001435   \n",
       "\n",
       "           Lachnospiraceae_dash_bacterium_dash_28_dash_4  Oscillospiraceae  \\\n",
       "SampleID                                                                     \n",
       "sample131                                       0.000272          0.001681   \n",
       "sample132                                       0.021246          0.010578   \n",
       "sample133                                       0.000560          0.001423   \n",
       "sample134                                       0.000421          0.013687   \n",
       "sample135                                       0.000163          0.001464   \n",
       "\n",
       "           Helicobacter_dash_ganmani  Escherichia_dash_Shigella  \\\n",
       "SampleID                                                          \n",
       "sample131                   0.000014                   0.000088   \n",
       "sample132                   0.000032                   0.000049   \n",
       "sample133                   0.000014                   0.000026   \n",
       "sample134                   0.000055                   0.000109   \n",
       "sample135                   0.000018                   0.000091   \n",
       "\n",
       "           Rikenellaceae_dash_RC9_dash_gut_dash_group  \\\n",
       "SampleID                                                \n",
       "sample131                                    0.018237   \n",
       "sample132                                    0.009194   \n",
       "sample133                                    0.013606   \n",
       "sample134                                    0.062197   \n",
       "sample135                                    0.028636   \n",
       "\n",
       "           Mucispirillum_dash_schaedleri  Clostridia_dash_vadinBB60_dash_group  \n",
       "SampleID                                                                        \n",
       "sample131                       0.005401                              0.001266  \n",
       "sample132                       0.036187                              0.000384  \n",
       "sample133                       0.010907                              0.000471  \n",
       "sample134                       0.013683                              0.000523  \n",
       "sample135                       0.002049                              0.000122  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX_w_meta = pd.merge(df_meta, df_16S_topX/df_16S_topX.max().max(), left_index=True, right_index=True, how='inner')\n",
    "df_16S_topX_w_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate log-derivatives (dlog(B)/dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MiceID</th>\n",
       "      <th>Vendor</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Faecalibaculum</th>\n",
       "      <th>Parasutterella</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>...</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>Desulfovibrionaceae</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>Lachnospiraceae_dash_bacterium_dash_28_dash_4</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Helicobacter_dash_ganmani</th>\n",
       "      <th>Escherichia_dash_Shigella</th>\n",
       "      <th>Rikenellaceae_dash_RC9_dash_gut_dash_group</th>\n",
       "      <th>Mucispirillum_dash_schaedleri</th>\n",
       "      <th>Clostridia_dash_vadinBB60_dash_group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>BC1</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.035769</td>\n",
       "      <td>-0.001958</td>\n",
       "      <td>-0.015449</td>\n",
       "      <td>-0.122249</td>\n",
       "      <td>-0.014135</td>\n",
       "      <td>0.044030</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.365448e-03</td>\n",
       "      <td>0.080151</td>\n",
       "      <td>0.002214</td>\n",
       "      <td>-0.006143</td>\n",
       "      <td>0.008517</td>\n",
       "      <td>0.003323</td>\n",
       "      <td>0.001625</td>\n",
       "      <td>0.009386</td>\n",
       "      <td>-0.003710</td>\n",
       "      <td>-0.000439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>BC2</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.065024</td>\n",
       "      <td>-0.017470</td>\n",
       "      <td>-0.011927</td>\n",
       "      <td>-0.093016</td>\n",
       "      <td>-0.018327</td>\n",
       "      <td>-0.002971</td>\n",
       "      <td>...</td>\n",
       "      <td>8.422057e-03</td>\n",
       "      <td>0.127021</td>\n",
       "      <td>0.000410</td>\n",
       "      <td>-0.030942</td>\n",
       "      <td>-0.003068</td>\n",
       "      <td>0.045131</td>\n",
       "      <td>0.000424</td>\n",
       "      <td>0.005349</td>\n",
       "      <td>-0.044840</td>\n",
       "      <td>0.000987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>BC3</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.003791</td>\n",
       "      <td>-0.010972</td>\n",
       "      <td>-0.010358</td>\n",
       "      <td>-0.071602</td>\n",
       "      <td>-0.031663</td>\n",
       "      <td>0.033417</td>\n",
       "      <td>...</td>\n",
       "      <td>7.933539e-07</td>\n",
       "      <td>0.036681</td>\n",
       "      <td>0.004060</td>\n",
       "      <td>-0.000361</td>\n",
       "      <td>0.008703</td>\n",
       "      <td>0.009558</td>\n",
       "      <td>0.003965</td>\n",
       "      <td>0.004809</td>\n",
       "      <td>-0.016021</td>\n",
       "      <td>0.010510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>BC4</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.010212</td>\n",
       "      <td>-0.037511</td>\n",
       "      <td>-0.094830</td>\n",
       "      <td>-0.071193</td>\n",
       "      <td>-0.010250</td>\n",
       "      <td>-0.006809</td>\n",
       "      <td>...</td>\n",
       "      <td>2.616074e-02</td>\n",
       "      <td>0.003099</td>\n",
       "      <td>0.001159</td>\n",
       "      <td>0.000587</td>\n",
       "      <td>0.013388</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>-0.000006</td>\n",
       "      <td>-0.020317</td>\n",
       "      <td>-0.012572</td>\n",
       "      <td>0.004581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>BC5</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.071871</td>\n",
       "      <td>-0.004578</td>\n",
       "      <td>-0.016798</td>\n",
       "      <td>-0.029205</td>\n",
       "      <td>-0.000383</td>\n",
       "      <td>0.019338</td>\n",
       "      <td>...</td>\n",
       "      <td>2.071616e-02</td>\n",
       "      <td>0.008314</td>\n",
       "      <td>-0.001196</td>\n",
       "      <td>0.002778</td>\n",
       "      <td>0.004650</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>-0.012629</td>\n",
       "      <td>0.003524</td>\n",
       "      <td>0.002131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          MiceID   Vendor  Day  Diet  Muribaculaceae  \\\n",
       "SampleID                                               \n",
       "sample131    BC1  Beijing    0     0       -0.035769   \n",
       "sample132    BC2  Beijing    0     0       -0.065024   \n",
       "sample133    BC3  Beijing    0     0       -0.003791   \n",
       "sample134    BC4  Beijing    0     0       -0.010212   \n",
       "sample135    BC5  Beijing    0     0        0.071871   \n",
       "\n",
       "           Bacteroides_dash_acidifaciens  Akkermansia_dash_muciniphila  \\\n",
       "SampleID                                                                 \n",
       "sample131                      -0.001958                     -0.015449   \n",
       "sample132                      -0.017470                     -0.011927   \n",
       "sample133                      -0.010972                     -0.010358   \n",
       "sample134                      -0.037511                     -0.094830   \n",
       "sample135                      -0.004578                     -0.016798   \n",
       "\n",
       "           Faecalibaculum  Parasutterella  Bacteroides  ...  \\\n",
       "SampleID                                                ...   \n",
       "sample131       -0.122249       -0.014135     0.044030  ...   \n",
       "sample132       -0.093016       -0.018327    -0.002971  ...   \n",
       "sample133       -0.071602       -0.031663     0.033417  ...   \n",
       "sample134       -0.071193       -0.010250    -0.006809  ...   \n",
       "sample135       -0.029205       -0.000383     0.019338  ...   \n",
       "\n",
       "           Lachnospiraceae_dash_NK4A136_dash_group  Desulfovibrionaceae  \\\n",
       "SampleID                                                                  \n",
       "sample131                            -2.365448e-03             0.080151   \n",
       "sample132                             8.422057e-03             0.127021   \n",
       "sample133                             7.933539e-07             0.036681   \n",
       "sample134                             2.616074e-02             0.003099   \n",
       "sample135                             2.071616e-02             0.008314   \n",
       "\n",
       "           Clostridia_dash_UCG_dash_014  \\\n",
       "SampleID                                  \n",
       "sample131                      0.002214   \n",
       "sample132                      0.000410   \n",
       "sample133                      0.004060   \n",
       "sample134                      0.001159   \n",
       "sample135                     -0.001196   \n",
       "\n",
       "           Lachnospiraceae_dash_bacterium_dash_28_dash_4  Oscillospiraceae  \\\n",
       "SampleID                                                                     \n",
       "sample131                                      -0.006143          0.008517   \n",
       "sample132                                      -0.030942         -0.003068   \n",
       "sample133                                      -0.000361          0.008703   \n",
       "sample134                                       0.000587          0.013388   \n",
       "sample135                                       0.002778          0.004650   \n",
       "\n",
       "           Helicobacter_dash_ganmani  Escherichia_dash_Shigella  \\\n",
       "SampleID                                                          \n",
       "sample131                   0.003323                   0.001625   \n",
       "sample132                   0.045131                   0.000424   \n",
       "sample133                   0.009558                   0.003965   \n",
       "sample134                   0.000030                  -0.000006   \n",
       "sample135                   0.000011                  -0.000011   \n",
       "\n",
       "           Rikenellaceae_dash_RC9_dash_gut_dash_group  \\\n",
       "SampleID                                                \n",
       "sample131                                    0.009386   \n",
       "sample132                                    0.005349   \n",
       "sample133                                    0.004809   \n",
       "sample134                                   -0.020317   \n",
       "sample135                                   -0.012629   \n",
       "\n",
       "           Mucispirillum_dash_schaedleri  Clostridia_dash_vadinBB60_dash_group  \n",
       "SampleID                                                                        \n",
       "sample131                      -0.003710                             -0.000439  \n",
       "sample132                      -0.044840                              0.000987  \n",
       "sample133                      -0.016021                              0.010510  \n",
       "sample134                      -0.012572                              0.004581  \n",
       "sample135                       0.003524                              0.002131  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_deriv = deepcopy(df_16S_topX_w_meta)\n",
    "for curr_mice in set(df_deriv.MiceID):\n",
    "    curr_df = df_deriv[df_deriv.MiceID==curr_mice]\n",
    "    for taxon in bacterial_taxa:\n",
    "        xdata = np.array(curr_df['Day'])\n",
    "        ydata = np.array(curr_df[taxon])\n",
    "        cs = CubicSpline(xdata, ydata)\n",
    "        csd1 = cs.derivative(nu=1)\n",
    "        ydata_d1 = csd1(xdata)\n",
    "        df_deriv.loc[df_deriv.MiceID==curr_mice, taxon] = ydata_d1\n",
    "df_deriv.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct regression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-0.83447306, -1.47426183, -0.13514241, ..., -0.05649319,\n",
       "       -0.0596165 , -0.05797334])"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ymat should be samples by bacteria\n",
    "Ymat = df_deriv[bacterial_taxa].values\n",
    "Ymat = Ymat.flatten(order='F')\n",
    "Ymat = StandardScaler().fit_transform(Ymat.reshape(-1,1)).reshape(1,-1)[0] # standardize\n",
    "Ymat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6740"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Ymat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.00000000e+00, 0.00000000e+00, 5.86115481e-02, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 0.00000000e+00, 6.83200226e-02, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 0.00000000e+00, 3.71185148e-02, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       ...,\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        1.05807481e-04, 2.58294734e-04, 1.02695497e-04],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        1.15959266e-04, 2.21611042e-04, 6.69986871e-05],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        2.33952716e-05, 1.84699513e-05, 7.38798051e-06]])"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Xmat = np.zeros(shape=(topX*len(df_deriv.index), (topX+2)*topX))\n",
    "for k in np.arange(topX):\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)] = 1\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+1] = df_deriv.Diet.values\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+2:(k+1)*(topX+2)] = df_16S_topX_w_meta[bacterial_taxa].values\n",
    "Xmat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write stan program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "json_str = '{\\n\"N\" : %d,\\n'%(len(Ymat))\n",
    "\n",
    "# y variable\n",
    "json_str += '\\\"dlogX\\\" : [%s],\\n'%(','.join(list(Ymat.astype(str))))\n",
    "\n",
    "# x variable\n",
    "for k1,c1 in enumerate(bacterial_taxa):\n",
    "    # growth rate\n",
    "    json_str += '\\\"growth_rate_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)].astype(str))))\n",
    "    # diet response\n",
    "    json_str += '\\\"inulin_response_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)+1].astype(str))))\n",
    "    # bacterial interactions\n",
    "    for k2,c2 in enumerate(bacterial_taxa):\n",
    "        v = list(Xmat[:,k1*(topX+2)+2+k2].astype(str))\n",
    "        json_str += '\\\"pairwise_interaction_%s_%s\\\" : [%s]'%(c1,c2,','.join(v))\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            json_str += '\\n}'\n",
    "        else:\n",
    "            json_str += ',\\n'\n",
    "#print(json_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.data.json\", \"w\")\n",
    "text_file.write(\"%s\" % json_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# data block\n",
    "model_str = 'data {\\n'\n",
    "model_str += '\\tint<lower=0> N;\\n'\n",
    "model_str += '\\tvector[N] dlogX;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\tvector[N] growth_rate_%s;\\n'%(c1)\n",
    "    model_str += '\\tvector[N] inulin_response_%s;\\n'%(c1)\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tvector[N] pairwise_interaction_%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'\n",
    "\n",
    "# parameter block\n",
    "model_str += 'parameters {\\n\\treal<lower=0,upper=1> sigma;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\treal alpha__%s;\\n'%(c1) # growth rate\n",
    "    model_str += '\\treal epsilon__%s;\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\treal beta__%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'       \n",
    "        \n",
    "# model block\n",
    "model_str += 'model {\\n\\tsigma ~ uniform(0,1);\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\talpha__%s ~ normal(0,1);\\n'%(c1) # growth rate\n",
    "    model_str += '\\tepsilon__%s ~ normal(0,1);\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tbeta__%s_%s ~ normal(0,1);\\n'%(c1,c2)\n",
    "model_str += '\\tdlogX ~ normal('\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += 'alpha__%s*growth_rate_%s+'%(c1,c1) # growth rate\n",
    "    model_str += 'epsilon__%s*inulin_response_%s+'%(c1,c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s'%(c1,c2,c1,c2)\n",
    "        else:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s+'%(c1,c2,c1,c2)\n",
    "model_str += ', sigma);\\n}'\n",
    "#print(model_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.stan\", \"w\")\n",
    "text_file.write(\"%s\" % model_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot stan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hpd_grid(sample, alpha=0.05, roundto=2):\n",
    "    \"\"\"Calculate highest posterior density (HPD) of array for given alpha. \n",
    "    The HPD is the minimum width Bayesian credible interval (BCI). \n",
    "    The function works for multimodal distributions, returning more than one mode\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    sample : Numpy array or python list\n",
    "        An array containing MCMC samples\n",
    "    alpha : float\n",
    "        Desired probability of type I error (defaults to 0.05)\n",
    "    roundto: integer\n",
    "        Number of digits after the decimal point for the results\n",
    "    Returns\n",
    "    ----------\n",
    "    hpd: array with the lower \n",
    "          \n",
    "    \"\"\"\n",
    "    sample = np.asarray(sample)\n",
    "    sample = sample[~np.isnan(sample)]\n",
    "    # get upper and lower bounds\n",
    "    l = np.min(sample)\n",
    "    u = np.max(sample)\n",
    "    density = kde.gaussian_kde(sample)\n",
    "    x = np.linspace(l, u, 2000)\n",
    "    y = density.evaluate(x)\n",
    "    #y = density.evaluate(x, l, u) waitting for PR to be accepted\n",
    "    xy_zipped = zip(x, y/np.sum(y))\n",
    "    xy = sorted(xy_zipped, key=lambda x: x[1], reverse=True)\n",
    "    xy_cum_sum = 0\n",
    "    hdv = []\n",
    "    for val in xy:\n",
    "        xy_cum_sum += val[1]\n",
    "        hdv.append(val[0])\n",
    "        if xy_cum_sum >= (1-alpha):\n",
    "            break\n",
    "    hdv.sort()\n",
    "    diff = (u-l)/20  # differences of 5%\n",
    "    hpd = []\n",
    "    hpd.append(round(min(hdv), roundto))\n",
    "    for i in range(1, len(hdv)):\n",
    "        if hdv[i]-hdv[i-1] >= diff:\n",
    "            hpd.append(round(hdv[i-1], roundto))\n",
    "            hpd.append(round(hdv[i], roundto))\n",
    "    hpd.append(round(max(hdv), roundto))\n",
    "    ite = iter(hpd)\n",
    "    hpd = list(zip(ite, ite))\n",
    "    modes = []\n",
    "    for value in hpd:\n",
    "        x_hpd = x[(x > value[0]) & (x < value[1])]\n",
    "        y_hpd = y[(x > value[0]) & (x < value[1])]\n",
    "        modes.append(round(x_hpd[np.argmax(y_hpd)], roundto))\n",
    "    return hpd, x, y, modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"output_%d.csv\"%(i) for i in np.arange(1,4)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>0.15</td>\n",
       "      <td>-0.005</td>\n",
       "      <td>-0.044859</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides_dash_acidifaciens</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.145</td>\n",
       "      <td>1.835619</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Akkermansia_dash_muciniphila</td>\n",
       "      <td>-0.27</td>\n",
       "      <td>0.04</td>\n",
       "      <td>-0.115</td>\n",
       "      <td>-1.416896</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Faecalibaculum</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-0.040</td>\n",
       "      <td>-0.531562</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Parasutterella</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-0.040</td>\n",
       "      <td>-0.540205</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Type                           Taxa  Left  Right  Middle  \\\n",
       "0  basal_growth_rate                 Muribaculaceae -0.16   0.15  -0.005   \n",
       "1  basal_growth_rate  Bacteroides_dash_acidifaciens -0.01   0.30   0.145   \n",
       "2  basal_growth_rate   Akkermansia_dash_muciniphila -0.27   0.04  -0.115   \n",
       "3  basal_growth_rate                 Faecalibaculum -0.20   0.12  -0.040   \n",
       "4  basal_growth_rate                 Parasutterella -0.20   0.12  -0.040   \n",
       "\n",
       "        SNR  Significant  \n",
       "0 -0.044859        False  \n",
       "1  1.835619        False  \n",
       "2 -1.416896        False  \n",
       "3 -0.531562        False  \n",
       "4 -0.540205        False  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "    \n",
    "# basal growth rate\n",
    "for c in bacterial_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# inulin response\n",
    "for c in bacterial_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['inulin_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "\n",
    "# pairwise interactions\n",
    "for c1 in bacterial_taxa:\n",
    "    for c2 in bacterial_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stan_output_summary.to_excel('bayesian_regression_summary_top20.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>1.08</td>\n",
       "      <td>1.62</td>\n",
       "      <td>1.350</td>\n",
       "      <td>9.958040</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Bacteroides_dash_acidifaciens</td>\n",
       "      <td>1.89</td>\n",
       "      <td>2.42</td>\n",
       "      <td>2.155</td>\n",
       "      <td>15.846923</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Faecalibaculum</td>\n",
       "      <td>0.48</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.745</td>\n",
       "      <td>5.506163</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Parasutterella</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.410</td>\n",
       "      <td>3.039780</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.310</td>\n",
       "      <td>2.265775</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Muribaculaceae')</td>\n",
       "      <td>-1.59</td>\n",
       "      <td>-0.29</td>\n",
       "      <td>-0.940</td>\n",
       "      <td>-2.864951</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Bacteroides_dash_acidifaci...</td>\n",
       "      <td>0.23</td>\n",
       "      <td>2.31</td>\n",
       "      <td>1.270</td>\n",
       "      <td>2.392931</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Desulfovibrionaceae')</td>\n",
       "      <td>0.27</td>\n",
       "      <td>3.71</td>\n",
       "      <td>1.990</td>\n",
       "      <td>2.280134</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Clostridia_dash_UCG_dash_0...</td>\n",
       "      <td>-3.74</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>-1.900</td>\n",
       "      <td>-2.016358</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Muribaculac...</td>\n",
       "      <td>-3.81</td>\n",
       "      <td>-2.50</td>\n",
       "      <td>-3.155</td>\n",
       "      <td>-9.603366</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Akkermansia...</td>\n",
       "      <td>-4.57</td>\n",
       "      <td>-1.77</td>\n",
       "      <td>-3.170</td>\n",
       "      <td>-4.472396</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Faecalibacu...</td>\n",
       "      <td>-4.31</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>-2.655</td>\n",
       "      <td>-3.148145</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Parasuttere...</td>\n",
       "      <td>-4.26</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>-2.415</td>\n",
       "      <td>-2.600989</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Bacteroides')</td>\n",
       "      <td>-5.89</td>\n",
       "      <td>-2.35</td>\n",
       "      <td>-4.120</td>\n",
       "      <td>-4.574359</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Akkermansia_dash_muciniphila', 'Bacteroides_...</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2.09</td>\n",
       "      <td>1.050</td>\n",
       "      <td>1.970520</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Faecalibaculum', 'Muribaculaceae')</td>\n",
       "      <td>-1.77</td>\n",
       "      <td>-0.47</td>\n",
       "      <td>-1.120</td>\n",
       "      <td>-3.398489</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Escherichia_dash_Shigella', 'Escherichia_das...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>3.69</td>\n",
       "      <td>2.145</td>\n",
       "      <td>2.731059</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                               Taxa  \\\n",
       "20        inulin_response                                     Muribaculaceae   \n",
       "21        inulin_response                      Bacteroides_dash_acidifaciens   \n",
       "23        inulin_response                                     Faecalibaculum   \n",
       "24        inulin_response                                     Parasutterella   \n",
       "25        inulin_response                                        Bacteroides   \n",
       "40   pairwise_interaction               ('Muribaculaceae', 'Muribaculaceae')   \n",
       "41   pairwise_interaction  ('Muribaculaceae', 'Bacteroides_dash_acidifaci...   \n",
       "51   pairwise_interaction          ('Muribaculaceae', 'Desulfovibrionaceae')   \n",
       "52   pairwise_interaction  ('Muribaculaceae', 'Clostridia_dash_UCG_dash_0...   \n",
       "60   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Muribaculac...   \n",
       "62   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Akkermansia...   \n",
       "63   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Faecalibacu...   \n",
       "64   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Parasuttere...   \n",
       "65   pairwise_interaction   ('Bacteroides_dash_acidifaciens', 'Bacteroides')   \n",
       "81   pairwise_interaction  ('Akkermansia_dash_muciniphila', 'Bacteroides_...   \n",
       "100  pairwise_interaction               ('Faecalibaculum', 'Muribaculaceae')   \n",
       "376  pairwise_interaction  ('Escherichia_dash_Shigella', 'Escherichia_das...   \n",
       "\n",
       "     Left  Right  Middle        SNR  Significant  \n",
       "20   1.08   1.62   1.350   9.958040         True  \n",
       "21   1.89   2.42   2.155  15.846923         True  \n",
       "23   0.48   1.01   0.745   5.506163         True  \n",
       "24   0.14   0.68   0.410   3.039780         True  \n",
       "25   0.04   0.58   0.310   2.265775         True  \n",
       "40  -1.59  -0.29  -0.940  -2.864951         True  \n",
       "41   0.23   2.31   1.270   2.392931         True  \n",
       "51   0.27   3.71   1.990   2.280134         True  \n",
       "52  -3.74  -0.06  -1.900  -2.016358         True  \n",
       "60  -3.81  -2.50  -3.155  -9.603366         True  \n",
       "62  -4.57  -1.77  -3.170  -4.472396         True  \n",
       "63  -4.31  -1.00  -2.655  -3.148145         True  \n",
       "64  -4.26  -0.57  -2.415  -2.600989         True  \n",
       "65  -5.89  -2.35  -4.120  -4.574359         True  \n",
       "81   0.01   2.09   1.050   1.970520         True  \n",
       "100 -1.77  -0.47  -1.120  -3.398489         True  \n",
       "376  0.60   3.69   2.145   2.731059         True  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stan_output_summary = pd.read_excel('bayesian_regression_summary_top20.xlsx', index_col=0)\n",
    "df_stan_output_summary = df_stan_output_summary[df_stan_output_summary.Significant==True]\n",
    "df_stan_output_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
