{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from scipy.interpolate import CubicSpline\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import warnings\n",
    "import itertools\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "import statsmodels.stats.api as sms\n",
    "import scipy.stats.kde as kde"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read 16S data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Absiella_dash_argi</th>\n",
       "      <th>Absiella_dash_dolichum</th>\n",
       "      <th>Acetanaerobacterium</th>\n",
       "      <th>Acetoanaerobium</th>\n",
       "      <th>Acetobacterium</th>\n",
       "      <th>Acidaminococcus_dash_intestini</th>\n",
       "      <th>Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Actinobacillus</th>\n",
       "      <th>Actinobacteria</th>\n",
       "      <th>...</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70</th>\n",
       "      <th>marine_dash_group</th>\n",
       "      <th>marine_dash_sediment_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>rumen_dash_bacterium_dash_YS2</th>\n",
       "      <th>soil_dash_metagenome</th>\n",
       "      <th>swine_dash_fecal_dash_bacterium_dash_SD_dash_Pec5</th>\n",
       "      <th>vadinBE97</th>\n",
       "      <th>wallaby_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6444565</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.032972</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.009201</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443884</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034554</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007985</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444382</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.032344</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008121</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444917</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001194</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443930</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 533 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Absiella_dash_argi  Absiella_dash_dolichum  Acetanaerobacterium  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565                 0.0                     0.0                  0.0   \n",
       "SRR6443884                 0.0                     0.0                  0.0   \n",
       "SRR6444382                 0.0                     0.0                  0.0   \n",
       "SRR6444917                 0.0                     0.0                  0.0   \n",
       "SRR6443930                 0.0                     0.0                  0.0   \n",
       "\n",
       "            Acetoanaerobium  Acetobacterium  Acidaminococcus_dash_intestini  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565              0.0             0.0                             0.0   \n",
       "SRR6443884              0.0             0.0                             0.0   \n",
       "SRR6444382              0.0             0.0                             0.0   \n",
       "SRR6444917              0.0             0.0                             0.0   \n",
       "SRR6443930              0.0             0.0                             0.0   \n",
       "\n",
       "            Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828  \\\n",
       "SampleID                                                             \n",
       "SRR6444565                                                0.0        \n",
       "SRR6443884                                                0.0        \n",
       "SRR6444382                                                0.0        \n",
       "SRR6444917                                                0.0        \n",
       "SRR6443930                                                0.0        \n",
       "\n",
       "            Acinetobacter  Actinobacillus  Actinobacteria  ...  \\\n",
       "SampleID                                                   ...   \n",
       "SRR6444565            0.0             0.0             0.0  ...   \n",
       "SRR6443884            0.0             0.0             0.0  ...   \n",
       "SRR6444382            0.0             0.0             0.0  ...   \n",
       "SRR6444917            0.0             0.0             0.0  ...   \n",
       "SRR6443930            0.0             0.0             0.0  ...   \n",
       "\n",
       "            human_dash_gut_dash_metagenome  \\\n",
       "SampleID                                     \n",
       "SRR6444565                        0.032972   \n",
       "SRR6443884                        0.034554   \n",
       "SRR6444382                        0.032344   \n",
       "SRR6444917                        0.000000   \n",
       "SRR6443930                        0.000000   \n",
       "\n",
       "            iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70  \\\n",
       "SampleID                                                                                          \n",
       "SRR6444565                                                0.0                                     \n",
       "SRR6443884                                                0.0                                     \n",
       "SRR6444382                                                0.0                                     \n",
       "SRR6444917                                                0.0                                     \n",
       "SRR6443930                                                0.0                                     \n",
       "\n",
       "            marine_dash_group  marine_dash_sediment_dash_metagenome  \\\n",
       "SampleID                                                              \n",
       "SRR6444565                0.0                                   0.0   \n",
       "SRR6443884                0.0                                   0.0   \n",
       "SRR6444382                0.0                                   0.0   \n",
       "SRR6444917                0.0                                   0.0   \n",
       "SRR6443930                0.0                                   0.0   \n",
       "\n",
       "            metagenome  rumen_dash_bacterium_dash_YS2  soil_dash_metagenome  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565    0.009201                            0.0                   0.0   \n",
       "SRR6443884    0.007985                            0.0                   0.0   \n",
       "SRR6444382    0.008121                            0.0                   0.0   \n",
       "SRR6444917    0.001194                            0.0                   0.0   \n",
       "SRR6443930    0.000000                            0.0                   0.0   \n",
       "\n",
       "            swine_dash_fecal_dash_bacterium_dash_SD_dash_Pec5  vadinBE97  \\\n",
       "SampleID                                                                   \n",
       "SRR6444565                                                0.0        0.0   \n",
       "SRR6443884                                                0.0        0.0   \n",
       "SRR6444382                                                0.0        0.0   \n",
       "SRR6444917                                                0.0        0.0   \n",
       "SRR6443930                                                0.0        0.0   \n",
       "\n",
       "            wallaby_dash_gut_dash_metagenome  \n",
       "SampleID                                      \n",
       "SRR6444565                               0.0  \n",
       "SRR6443884                               0.0  \n",
       "SRR6444382                               0.0  \n",
       "SRR6444917                               0.0  \n",
       "SRR6443930                               0.0  \n",
       "\n",
       "[5 rows x 533 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = pd.read_excel('./16S_relative_abundance.xlsx', index_col=0)\n",
    "df_16S.columns = [c.replace('/','_slash_').replace(' ','_space_').replace('[','_leftsquarebracket_').replace(']','_rightsquarebracket_').replace('-','_dash_').replace('.','_dot_').replace('(','_leftroundbracket').replace(')','_rightroundbracket_') for c in df_16S.columns]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add pseudocount for zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Absiella_dash_argi</th>\n",
       "      <th>Absiella_dash_dolichum</th>\n",
       "      <th>Acetanaerobacterium</th>\n",
       "      <th>Acetoanaerobium</th>\n",
       "      <th>Acetobacterium</th>\n",
       "      <th>Acidaminococcus_dash_intestini</th>\n",
       "      <th>Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Actinobacillus</th>\n",
       "      <th>Actinobacteria</th>\n",
       "      <th>...</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70</th>\n",
       "      <th>marine_dash_group</th>\n",
       "      <th>marine_dash_sediment_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>rumen_dash_bacterium_dash_YS2</th>\n",
       "      <th>soil_dash_metagenome</th>\n",
       "      <th>swine_dash_fecal_dash_bacterium_dash_SD_dash_Pec5</th>\n",
       "      <th>vadinBE97</th>\n",
       "      <th>wallaby_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6444565</th>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>...</td>\n",
       "      <td>0.032972</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.009201</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443884</th>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034554</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.007985</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444382</th>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>...</td>\n",
       "      <td>0.032344</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.008121</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444917</th>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.001194</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443930</th>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 533 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Absiella_dash_argi  Absiella_dash_dolichum  Acetanaerobacterium  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565            0.000215                0.000215             0.000215   \n",
       "SRR6443884            0.000070                0.000070             0.000070   \n",
       "SRR6444382            0.000167                0.000167             0.000167   \n",
       "SRR6444917            0.000199                0.000199             0.000199   \n",
       "SRR6443930            0.000181                0.000181             0.000181   \n",
       "\n",
       "            Acetoanaerobium  Acetobacterium  Acidaminococcus_dash_intestini  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565         0.000215        0.000215                        0.000215   \n",
       "SRR6443884         0.000070        0.000070                        0.000070   \n",
       "SRR6444382         0.000167        0.000167                        0.000167   \n",
       "SRR6444917         0.000199        0.000199                        0.000199   \n",
       "SRR6443930         0.000181        0.000181                        0.000181   \n",
       "\n",
       "            Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828  \\\n",
       "SampleID                                                             \n",
       "SRR6444565                                           0.000215        \n",
       "SRR6443884                                           0.000070        \n",
       "SRR6444382                                           0.000167        \n",
       "SRR6444917                                           0.000199        \n",
       "SRR6443930                                           0.000181        \n",
       "\n",
       "            Acinetobacter  Actinobacillus  Actinobacteria  ...  \\\n",
       "SampleID                                                   ...   \n",
       "SRR6444565       0.000215        0.000215        0.000215  ...   \n",
       "SRR6443884       0.000070        0.000070        0.000070  ...   \n",
       "SRR6444382       0.000167        0.000167        0.000167  ...   \n",
       "SRR6444917       0.000199        0.000199        0.000199  ...   \n",
       "SRR6443930       0.000181        0.000181        0.000181  ...   \n",
       "\n",
       "            human_dash_gut_dash_metagenome  \\\n",
       "SampleID                                     \n",
       "SRR6444565                        0.032972   \n",
       "SRR6443884                        0.034554   \n",
       "SRR6444382                        0.032344   \n",
       "SRR6444917                        0.000199   \n",
       "SRR6443930                        0.000181   \n",
       "\n",
       "            iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70  \\\n",
       "SampleID                                                                                          \n",
       "SRR6444565                                           0.000215                                     \n",
       "SRR6443884                                           0.000070                                     \n",
       "SRR6444382                                           0.000167                                     \n",
       "SRR6444917                                           0.000199                                     \n",
       "SRR6443930                                           0.000181                                     \n",
       "\n",
       "            marine_dash_group  marine_dash_sediment_dash_metagenome  \\\n",
       "SampleID                                                              \n",
       "SRR6444565           0.000215                              0.000215   \n",
       "SRR6443884           0.000070                              0.000070   \n",
       "SRR6444382           0.000167                              0.000167   \n",
       "SRR6444917           0.000199                              0.000199   \n",
       "SRR6443930           0.000181                              0.000181   \n",
       "\n",
       "            metagenome  rumen_dash_bacterium_dash_YS2  soil_dash_metagenome  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565    0.009201                       0.000215              0.000215   \n",
       "SRR6443884    0.007985                       0.000070              0.000070   \n",
       "SRR6444382    0.008121                       0.000167              0.000167   \n",
       "SRR6444917    0.001194                       0.000199              0.000199   \n",
       "SRR6443930    0.000181                       0.000181              0.000181   \n",
       "\n",
       "            swine_dash_fecal_dash_bacterium_dash_SD_dash_Pec5  vadinBE97  \\\n",
       "SampleID                                                                   \n",
       "SRR6444565                                           0.000215   0.000215   \n",
       "SRR6443884                                           0.000070   0.000070   \n",
       "SRR6444382                                           0.000167   0.000167   \n",
       "SRR6444917                                           0.000199   0.000199   \n",
       "SRR6443930                                           0.000181   0.000181   \n",
       "\n",
       "            wallaby_dash_gut_dash_metagenome  \n",
       "SampleID                                      \n",
       "SRR6444565                          0.000215  \n",
       "SRR6443884                          0.000070  \n",
       "SRR6444382                          0.000167  \n",
       "SRR6444917                          0.000199  \n",
       "SRR6443930                          0.000181  \n",
       "\n",
       "[5 rows x 533 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample_id in df_16S.index:\n",
    "    sample = np.array(df_16S.loc[sample_id])\n",
    "    minval = np.min(sample[np.nonzero(sample)]) # minimum non-zero value\n",
    "    sample[sample==0] = minval\n",
    "    df_16S.loc[sample_id] = sample\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata for inulin and control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6444565</th>\n",
       "      <td>U222</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443884</th>\n",
       "      <td>U222</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444382</th>\n",
       "      <td>U222</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444917</th>\n",
       "      <td>U225</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443930</th>\n",
       "      <td>U225</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Subject_ID  status  Day  Diet\n",
       "#SampleID                               \n",
       "SRR6444565       U222  before    0     0\n",
       "SRR6443884       U222  before    0     0\n",
       "SRR6444382       U222  before    0     0\n",
       "SRR6444917       U225  before    0     0\n",
       "SRR6443930       U225  before    0     0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv('./metadata.txt', sep='\\t', index_col=0)\n",
    "df_meta = df_meta[['Group','Subject_ID','status']]\n",
    "df_meta = df_meta[(df_meta.Group=='inulin') | (df_meta.Group=='accessible')]\n",
    "df_meta['Day'] = [0 if s=='before' else 21 for s in df_meta.status]\n",
    "df_meta['Diet'] = [1 if g=='inulin' else 0 for g in df_meta['Group']]\n",
    "df_meta = df_meta.drop('Group', axis=1)\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 20 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Prevotella</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Parabacteroides</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>Parasutterella_dash_excrementihominis</th>\n",
       "      <th>Bacteroides_dash_plebeius_dash_DSM_dash_17135</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Lachnospiraceae_dash_UCG_dash_004</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Bacteroides_dash_coprocola</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6444565</th>\n",
       "      <td>0.197580</td>\n",
       "      <td>0.136827</td>\n",
       "      <td>0.024022</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.028962</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.037984</td>\n",
       "      <td>0.007446</td>\n",
       "      <td>0.009165</td>\n",
       "      <td>0.018580</td>\n",
       "      <td>0.006158</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.114631</td>\n",
       "      <td>0.006623</td>\n",
       "      <td>0.003508</td>\n",
       "      <td>0.036051</td>\n",
       "      <td>0.027852</td>\n",
       "      <td>0.001539</td>\n",
       "      <td>0.009201</td>\n",
       "      <td>0.000215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443884</th>\n",
       "      <td>0.243131</td>\n",
       "      <td>0.099128</td>\n",
       "      <td>0.017050</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.032008</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.037064</td>\n",
       "      <td>0.006764</td>\n",
       "      <td>0.004916</td>\n",
       "      <td>0.014191</td>\n",
       "      <td>0.005230</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.125837</td>\n",
       "      <td>0.005335</td>\n",
       "      <td>0.002789</td>\n",
       "      <td>0.034100</td>\n",
       "      <td>0.022873</td>\n",
       "      <td>0.003626</td>\n",
       "      <td>0.007985</td>\n",
       "      <td>0.000070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444382</th>\n",
       "      <td>0.214061</td>\n",
       "      <td>0.110743</td>\n",
       "      <td>0.028395</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.031565</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.037600</td>\n",
       "      <td>0.006981</td>\n",
       "      <td>0.007370</td>\n",
       "      <td>0.020525</td>\n",
       "      <td>0.007231</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.115360</td>\n",
       "      <td>0.006869</td>\n",
       "      <td>0.003560</td>\n",
       "      <td>0.032734</td>\n",
       "      <td>0.029730</td>\n",
       "      <td>0.004533</td>\n",
       "      <td>0.008121</td>\n",
       "      <td>0.000167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444917</th>\n",
       "      <td>0.578520</td>\n",
       "      <td>0.063804</td>\n",
       "      <td>0.117987</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.037393</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000630</td>\n",
       "      <td>0.003550</td>\n",
       "      <td>0.016358</td>\n",
       "      <td>0.014168</td>\n",
       "      <td>0.000299</td>\n",
       "      <td>0.026942</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.003782</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000630</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.001194</td>\n",
       "      <td>0.000199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443930</th>\n",
       "      <td>0.610679</td>\n",
       "      <td>0.055143</td>\n",
       "      <td>0.050196</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.079698</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.004404</td>\n",
       "      <td>0.008386</td>\n",
       "      <td>0.006154</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.023409</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Bacteroides  Faecalibacterium  Lachnospiraceae  \\\n",
       "#SampleID                                                    \n",
       "SRR6444565     0.197580          0.136827         0.024022   \n",
       "SRR6443884     0.243131          0.099128         0.017050   \n",
       "SRR6444382     0.214061          0.110743         0.028395   \n",
       "SRR6444917     0.578520          0.063804         0.117987   \n",
       "SRR6443930     0.610679          0.055143         0.050196   \n",
       "\n",
       "            Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_  \\\n",
       "#SampleID                                                                                           \n",
       "SRR6444565                                           0.000215                                       \n",
       "SRR6443884                                           0.000070                                       \n",
       "SRR6444382                                           0.000167                                       \n",
       "SRR6444917                                           0.000199                                       \n",
       "SRR6443930                                           0.000181                                       \n",
       "\n",
       "            Alistipes  Prevotella  \\\n",
       "#SampleID                           \n",
       "SRR6444565   0.028962    0.000215   \n",
       "SRR6443884   0.032008    0.000070   \n",
       "SRR6444382   0.031565    0.000167   \n",
       "SRR6444917   0.037393    0.000199   \n",
       "SRR6443930   0.079698    0.000181   \n",
       "\n",
       "            Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  \\\n",
       "#SampleID                                                        \n",
       "SRR6444565                                           0.037984    \n",
       "SRR6443884                                           0.037064    \n",
       "SRR6444382                                           0.037600    \n",
       "SRR6444917                                           0.000630    \n",
       "SRR6443930                                           0.000181    \n",
       "\n",
       "            Parabacteroides  Bifidobacterium   Blautia  Anaerostipes  \\\n",
       "#SampleID                                                              \n",
       "SRR6444565         0.007446         0.009165  0.018580      0.006158   \n",
       "SRR6443884         0.006764         0.004916  0.014191      0.005230   \n",
       "SRR6444382         0.006981         0.007370  0.020525      0.007231   \n",
       "SRR6444917         0.003550         0.016358  0.014168      0.000299   \n",
       "SRR6443930         0.004404         0.008386  0.006154      0.000181   \n",
       "\n",
       "            Parasutterella_dash_excrementihominis  \\\n",
       "#SampleID                                           \n",
       "SRR6444565                               0.000215   \n",
       "SRR6443884                               0.000070   \n",
       "SRR6444382                               0.000167   \n",
       "SRR6444917                               0.026942   \n",
       "SRR6443930                               0.023409   \n",
       "\n",
       "            Bacteroides_dash_plebeius_dash_DSM_dash_17135  Ruminococcus  \\\n",
       "#SampleID                                                                 \n",
       "SRR6444565                                       0.114631      0.006623   \n",
       "SRR6443884                                       0.125837      0.005335   \n",
       "SRR6444382                                       0.115360      0.006869   \n",
       "SRR6444917                                       0.000199      0.003782   \n",
       "SRR6443930                                       0.000181      0.000181   \n",
       "\n",
       "            Fusicatenibacter  gut_dash_metagenome  Roseburia  \\\n",
       "#SampleID                                                      \n",
       "SRR6444565          0.003508             0.036051   0.027852   \n",
       "SRR6443884          0.002789             0.034100   0.022873   \n",
       "SRR6444382          0.003560             0.032734   0.029730   \n",
       "SRR6444917          0.000199             0.000630   0.000199   \n",
       "SRR6443930          0.000181             0.000181   0.000181   \n",
       "\n",
       "            Lachnospiraceae_dash_UCG_dash_004  metagenome  \\\n",
       "#SampleID                                                   \n",
       "SRR6444565                           0.001539    0.009201   \n",
       "SRR6443884                           0.003626    0.007985   \n",
       "SRR6444382                           0.004533    0.008121   \n",
       "SRR6444917                           0.000199    0.001194   \n",
       "SRR6443930                           0.000181    0.000181   \n",
       "\n",
       "            Bacteroides_dash_coprocola  \n",
       "#SampleID                               \n",
       "SRR6444565                    0.000215  \n",
       "SRR6443884                    0.000070  \n",
       "SRR6444382                    0.000167  \n",
       "SRR6444917                    0.000199  \n",
       "SRR6443930                    0.000181  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 20\n",
    "df_16S_T = df_16S.loc[df_meta.index].T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Bacteroides',\n",
       " 'Faecalibacterium',\n",
       " 'Lachnospiraceae',\n",
       " 'Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_',\n",
       " 'Alistipes',\n",
       " 'Prevotella',\n",
       " 'Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57',\n",
       " 'Parabacteroides',\n",
       " 'Bifidobacterium',\n",
       " 'Blautia',\n",
       " 'Anaerostipes',\n",
       " 'Parasutterella_dash_excrementihominis',\n",
       " 'Bacteroides_dash_plebeius_dash_DSM_dash_17135',\n",
       " 'Ruminococcus',\n",
       " 'Fusicatenibacter',\n",
       " 'gut_dash_metagenome',\n",
       " 'Roseburia',\n",
       " 'Lachnospiraceae_dash_UCG_dash_004',\n",
       " 'metagenome',\n",
       " 'Bacteroides_dash_coprocola']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_16S_topX.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Prevotella</th>\n",
       "      <th>...</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>Parasutterella_dash_excrementihominis</th>\n",
       "      <th>Bacteroides_dash_plebeius_dash_DSM_dash_17135</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Lachnospiraceae_dash_UCG_dash_004</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Bacteroides_dash_coprocola</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U042</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.444946</td>\n",
       "      <td>0.100588</td>\n",
       "      <td>0.080670</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.027035</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003644</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.009113</td>\n",
       "      <td>0.022899</td>\n",
       "      <td>0.007670</td>\n",
       "      <td>0.025910</td>\n",
       "      <td>0.007428</td>\n",
       "      <td>0.011713</td>\n",
       "      <td>0.000128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U042</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.471585</td>\n",
       "      <td>0.037131</td>\n",
       "      <td>0.080455</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.038842</td>\n",
       "      <td>0.000406</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003687</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.010182</td>\n",
       "      <td>0.008189</td>\n",
       "      <td>0.016515</td>\n",
       "      <td>0.047139</td>\n",
       "      <td>0.008635</td>\n",
       "      <td>0.009576</td>\n",
       "      <td>0.000254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U068</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.045633</td>\n",
       "      <td>0.056045</td>\n",
       "      <td>0.043214</td>\n",
       "      <td>0.177269</td>\n",
       "      <td>0.000270</td>\n",
       "      <td>0.328534</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000425</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0.021873</td>\n",
       "      <td>0.004559</td>\n",
       "      <td>0.011077</td>\n",
       "      <td>0.002486</td>\n",
       "      <td>0.006962</td>\n",
       "      <td>0.011519</td>\n",
       "      <td>0.015651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U068</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.070620</td>\n",
       "      <td>0.093106</td>\n",
       "      <td>0.044719</td>\n",
       "      <td>0.120436</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>0.258661</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002214</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>0.008714</td>\n",
       "      <td>0.007876</td>\n",
       "      <td>0.017048</td>\n",
       "      <td>0.009354</td>\n",
       "      <td>0.004645</td>\n",
       "      <td>0.018583</td>\n",
       "      <td>0.004436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U075</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.227864</td>\n",
       "      <td>0.133095</td>\n",
       "      <td>0.049664</td>\n",
       "      <td>0.013882</td>\n",
       "      <td>0.017764</td>\n",
       "      <td>0.000785</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014899</td>\n",
       "      <td>0.016442</td>\n",
       "      <td>0.000179</td>\n",
       "      <td>0.014111</td>\n",
       "      <td>0.045612</td>\n",
       "      <td>0.011263</td>\n",
       "      <td>0.018615</td>\n",
       "      <td>0.002632</td>\n",
       "      <td>0.017789</td>\n",
       "      <td>0.000179</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Subject_ID  status  Day  Diet  Bacteroides  Faecalibacterium  \\\n",
       "0       U042  before    0     1     0.444946          0.100588   \n",
       "1       U042  during   21     1     0.471585          0.037131   \n",
       "2       U068  before    0     1     0.045633          0.056045   \n",
       "3       U068  during   21     1     0.070620          0.093106   \n",
       "4       U075  before    0     1     0.227864          0.133095   \n",
       "\n",
       "   Lachnospiraceae  \\\n",
       "0         0.080670   \n",
       "1         0.080455   \n",
       "2         0.043214   \n",
       "3         0.044719   \n",
       "4         0.049664   \n",
       "\n",
       "   Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_  \\\n",
       "0                                           0.000128                                       \n",
       "1                                           0.000254                                       \n",
       "2                                           0.177269                                       \n",
       "3                                           0.120436                                       \n",
       "4                                           0.013882                                       \n",
       "\n",
       "   Alistipes  Prevotella  ...  Anaerostipes  \\\n",
       "0   0.027035    0.000128  ...      0.003644   \n",
       "1   0.038842    0.000406  ...      0.003687   \n",
       "2   0.000270    0.328534  ...      0.000425   \n",
       "3   0.000256    0.258661  ...      0.002214   \n",
       "4   0.017764    0.000785  ...      0.014899   \n",
       "\n",
       "   Parasutterella_dash_excrementihominis  \\\n",
       "0                               0.000128   \n",
       "1                               0.000254   \n",
       "2                               0.000161   \n",
       "3                               0.000199   \n",
       "4                               0.016442   \n",
       "\n",
       "   Bacteroides_dash_plebeius_dash_DSM_dash_17135  Ruminococcus  \\\n",
       "0                                       0.000128      0.009113   \n",
       "1                                       0.000254      0.010182   \n",
       "2                                       0.000161      0.021873   \n",
       "3                                       0.000153      0.008714   \n",
       "4                                       0.000179      0.014111   \n",
       "\n",
       "   Fusicatenibacter  gut_dash_metagenome  Roseburia  \\\n",
       "0          0.022899             0.007670   0.025910   \n",
       "1          0.008189             0.016515   0.047139   \n",
       "2          0.004559             0.011077   0.002486   \n",
       "3          0.007876             0.017048   0.009354   \n",
       "4          0.045612             0.011263   0.018615   \n",
       "\n",
       "   Lachnospiraceae_dash_UCG_dash_004  metagenome  Bacteroides_dash_coprocola  \n",
       "0                           0.007428    0.011713                    0.000128  \n",
       "1                           0.008635    0.009576                    0.000254  \n",
       "2                           0.006962    0.011519                    0.015651  \n",
       "3                           0.004645    0.018583                    0.004436  \n",
       "4                           0.002632    0.017789                    0.000179  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX_w_meta = pd.merge(df_meta, df_16S_topX/df_16S_topX.max().max(), left_index=True, right_index=True, how='inner')\n",
    "df_16S_topX_w_meta = df_16S_topX_w_meta.groupby(['Subject_ID','status']).agg(np.mean).reset_index()\n",
    "df_16S_topX_w_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate log-derivatives (dlog(B)/dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Prevotella</th>\n",
       "      <th>...</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>Parasutterella_dash_excrementihominis</th>\n",
       "      <th>Bacteroides_dash_plebeius_dash_DSM_dash_17135</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Lachnospiraceae_dash_UCG_dash_004</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Bacteroides_dash_coprocola</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U042</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001269</td>\n",
       "      <td>-0.003022</td>\n",
       "      <td>-0.000010</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>5.622348e-04</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>5.970561e-06</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>-0.000700</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.001011</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>-0.000102</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U042</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001269</td>\n",
       "      <td>-0.003022</td>\n",
       "      <td>-0.000010</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>5.622348e-04</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>5.970561e-06</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>-0.000700</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.001011</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>-0.000102</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U068</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001190</td>\n",
       "      <td>0.001765</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>-0.002706</td>\n",
       "      <td>-7.087060e-07</td>\n",
       "      <td>-0.003327</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>-3.810299e-07</td>\n",
       "      <td>-0.000627</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>0.000284</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>-0.000110</td>\n",
       "      <td>0.000336</td>\n",
       "      <td>-0.000534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U068</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001190</td>\n",
       "      <td>0.001765</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>-0.002706</td>\n",
       "      <td>-7.087060e-07</td>\n",
       "      <td>-0.003327</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>-3.810299e-07</td>\n",
       "      <td>-0.000627</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>0.000284</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>-0.000110</td>\n",
       "      <td>0.000336</td>\n",
       "      <td>-0.000534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U075</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005197</td>\n",
       "      <td>-0.003111</td>\n",
       "      <td>-0.000158</td>\n",
       "      <td>-0.000459</td>\n",
       "      <td>6.566401e-04</td>\n",
       "      <td>-0.000024</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.000248</td>\n",
       "      <td>-4.237658e-06</td>\n",
       "      <td>0.000450</td>\n",
       "      <td>-0.000997</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>-0.000491</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>-0.000448</td>\n",
       "      <td>-0.000004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Subject_ID  status  Day  Diet  Bacteroides  Faecalibacterium  \\\n",
       "0       U042  before    0     1     0.001269         -0.003022   \n",
       "1       U042  during   21     1     0.001269         -0.003022   \n",
       "2       U068  before    0     1     0.001190          0.001765   \n",
       "3       U068  during   21     1     0.001190          0.001765   \n",
       "4       U075  before    0     1     0.005197         -0.003111   \n",
       "\n",
       "   Lachnospiraceae  \\\n",
       "0        -0.000010   \n",
       "1        -0.000010   \n",
       "2         0.000072   \n",
       "3         0.000072   \n",
       "4        -0.000158   \n",
       "\n",
       "   Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_  \\\n",
       "0                                           0.000006                                       \n",
       "1                                           0.000006                                       \n",
       "2                                          -0.002706                                       \n",
       "3                                          -0.002706                                       \n",
       "4                                          -0.000459                                       \n",
       "\n",
       "      Alistipes  Prevotella  ...  Anaerostipes  \\\n",
       "0  5.622348e-04    0.000013  ...      0.000002   \n",
       "1  5.622348e-04    0.000013  ...      0.000002   \n",
       "2 -7.087060e-07   -0.003327  ...      0.000085   \n",
       "3 -7.087060e-07   -0.003327  ...      0.000085   \n",
       "4  6.566401e-04   -0.000024  ...      0.000185   \n",
       "\n",
       "   Parasutterella_dash_excrementihominis  \\\n",
       "0                               0.000006   \n",
       "1                               0.000006   \n",
       "2                               0.000002   \n",
       "3                               0.000002   \n",
       "4                               0.000248   \n",
       "\n",
       "   Bacteroides_dash_plebeius_dash_DSM_dash_17135  Ruminococcus  \\\n",
       "0                                   5.970561e-06      0.000051   \n",
       "1                                   5.970561e-06      0.000051   \n",
       "2                                  -3.810299e-07     -0.000627   \n",
       "3                                  -3.810299e-07     -0.000627   \n",
       "4                                  -4.237658e-06      0.000450   \n",
       "\n",
       "   Fusicatenibacter  gut_dash_metagenome  Roseburia  \\\n",
       "0         -0.000700             0.000421   0.001011   \n",
       "1         -0.000700             0.000421   0.001011   \n",
       "2          0.000158             0.000284   0.000327   \n",
       "3          0.000158             0.000284   0.000327   \n",
       "4         -0.000997             0.000204  -0.000491   \n",
       "\n",
       "   Lachnospiraceae_dash_UCG_dash_004  metagenome  Bacteroides_dash_coprocola  \n",
       "0                           0.000057   -0.000102                    0.000006  \n",
       "1                           0.000057   -0.000102                    0.000006  \n",
       "2                          -0.000110    0.000336                   -0.000534  \n",
       "3                          -0.000110    0.000336                   -0.000534  \n",
       "4                           0.000112   -0.000448                   -0.000004  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_deriv = deepcopy(df_16S_topX_w_meta)\n",
    "for curr_sub in set(df_deriv.Subject_ID):\n",
    "    curr_df = df_deriv[df_deriv.Subject_ID==curr_sub].sort_values(by='Day')\n",
    "    for taxon in bacterial_taxa:\n",
    "        xdata = np.array(curr_df['Day'])\n",
    "        ydata = np.array(curr_df[taxon])\n",
    "        cs = CubicSpline(xdata, ydata)\n",
    "        csd1 = cs.derivative(nu=1)\n",
    "        ydata_d1 = csd1(xdata)\n",
    "        df_deriv.loc[df_deriv.Subject_ID==curr_sub, taxon] = ydata_d1\n",
    "df_deriv.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct regression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.89588099,  0.89588099,  0.83846764, ..., -2.43909054,\n",
       "       -0.01323803, -0.01323803])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ymat should be samples by bacteria\n",
    "Ymat = df_deriv[bacterial_taxa].values\n",
    "Ymat = Ymat.flatten(order='F')\n",
    "Ymat = StandardScaler().fit_transform(Ymat.reshape(-1,1)).reshape(1,-1)[0] # standardize\n",
    "Ymat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3560"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Ymat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.00000000e+00, 1.00000000e+00, 4.44946301e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 1.00000000e+00, 4.71585345e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 1.00000000e+00, 4.56331153e-02, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       ...,\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        9.46741789e-03, 6.74048698e-03, 6.98077419e-04],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        4.20216398e-02, 3.61930501e-03, 2.97519882e-04],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        4.50391700e-02, 4.41138219e-03, 7.73246735e-04]])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Xmat = np.zeros(shape=(topX*len(df_deriv.index), (topX+2)*topX))\n",
    "for k in np.arange(topX):\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)] = 1\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+1] = df_deriv.Diet.values\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+2:(k+1)*(topX+2)] = df_16S_topX_w_meta[bacterial_taxa].values\n",
    "Xmat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write stan program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "json_str = '{\\n\"N\" : %d,\\n'%(len(Ymat))\n",
    "\n",
    "# y variable\n",
    "json_str += '\\\"dlogX\\\" : [%s],\\n'%(','.join(list(Ymat.astype(str))))\n",
    "\n",
    "# x variable\n",
    "for k1,c1 in enumerate(bacterial_taxa):\n",
    "    # growth rate\n",
    "    json_str += '\\\"growth_rate_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)].astype(str))))\n",
    "    # diet response\n",
    "    json_str += '\\\"inulin_response_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)+1].astype(str))))\n",
    "    # bacterial interactions\n",
    "    for k2,c2 in enumerate(bacterial_taxa):\n",
    "        v = list(Xmat[:,k1*(topX+2)+2+k2].astype(str))\n",
    "        json_str += '\\\"pairwise_interaction_%s_%s\\\" : [%s]'%(c1,c2,','.join(v))\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            json_str += '\\n}'\n",
    "        else:\n",
    "            json_str += ',\\n'\n",
    "#print(json_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.data.json\", \"w\")\n",
    "text_file.write(\"%s\" % json_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# data block\n",
    "model_str = 'data {\\n'\n",
    "model_str += '\\tint<lower=0> N;\\n'\n",
    "model_str += '\\tvector[N] dlogX;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\tvector[N] growth_rate_%s;\\n'%(c1)\n",
    "    model_str += '\\tvector[N] inulin_response_%s;\\n'%(c1)\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tvector[N] pairwise_interaction_%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'\n",
    "\n",
    "# parameter block\n",
    "model_str += 'parameters {\\n\\treal<lower=0,upper=1> sigma;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\treal alpha__%s;\\n'%(c1) # growth rate\n",
    "    model_str += '\\treal epsilon__%s;\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\treal beta__%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'       \n",
    "        \n",
    "# model block\n",
    "model_str += 'model {\\n\\tsigma ~ uniform(0,1);\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\talpha__%s ~ normal(0,1);\\n'%(c1) # growth rate\n",
    "    model_str += '\\tepsilon__%s ~ normal(0,1);\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tbeta__%s_%s ~ normal(0,1);\\n'%(c1,c2)\n",
    "model_str += '\\tdlogX ~ normal('\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += 'alpha__%s*growth_rate_%s+'%(c1,c1) # growth rate\n",
    "    model_str += 'epsilon__%s*inulin_response_%s+'%(c1,c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s'%(c1,c2,c1,c2)\n",
    "        else:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s+'%(c1,c2,c1,c2)\n",
    "model_str += ', sigma);\\n}'\n",
    "#print(model_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.stan\", \"w\")\n",
    "text_file.write(\"%s\" % model_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot stan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hpd_grid(sample, alpha=0.05, roundto=2):\n",
    "    \"\"\"Calculate highest posterior density (HPD) of array for given alpha. \n",
    "    The HPD is the minimum width Bayesian credible interval (BCI). \n",
    "    The function works for multimodal distributions, returning more than one mode\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    sample : Numpy array or python list\n",
    "        An array containing MCMC samples\n",
    "    alpha : float\n",
    "        Desired probability of type I error (defaults to 0.05)\n",
    "    roundto: integer\n",
    "        Number of digits after the decimal point for the results\n",
    "    Returns\n",
    "    ----------\n",
    "    hpd: array with the lower \n",
    "          \n",
    "    \"\"\"\n",
    "    sample = np.asarray(sample)\n",
    "    sample = sample[~np.isnan(sample)]\n",
    "    # get upper and lower bounds\n",
    "    l = np.min(sample)\n",
    "    u = np.max(sample)\n",
    "    density = kde.gaussian_kde(sample)\n",
    "    x = np.linspace(l, u, 2000)\n",
    "    y = density.evaluate(x)\n",
    "    #y = density.evaluate(x, l, u) waitting for PR to be accepted\n",
    "    xy_zipped = zip(x, y/np.sum(y))\n",
    "    xy = sorted(xy_zipped, key=lambda x: x[1], reverse=True)\n",
    "    xy_cum_sum = 0\n",
    "    hdv = []\n",
    "    for val in xy:\n",
    "        xy_cum_sum += val[1]\n",
    "        hdv.append(val[0])\n",
    "        if xy_cum_sum >= (1-alpha):\n",
    "            break\n",
    "    hdv.sort()\n",
    "    diff = (u-l)/20  # differences of 5%\n",
    "    hpd = []\n",
    "    hpd.append(round(min(hdv), roundto))\n",
    "    for i in range(1, len(hdv)):\n",
    "        if hdv[i]-hdv[i-1] >= diff:\n",
    "            hpd.append(round(hdv[i-1], roundto))\n",
    "            hpd.append(round(hdv[i], roundto))\n",
    "    hpd.append(round(max(hdv), roundto))\n",
    "    ite = iter(hpd)\n",
    "    hpd = list(zip(ite, ite))\n",
    "    modes = []\n",
    "    for value in hpd:\n",
    "        x_hpd = x[(x > value[0]) & (x < value[1])]\n",
    "        y_hpd = y[(x > value[0]) & (x < value[1])]\n",
    "        modes.append(round(x_hpd[np.argmax(y_hpd)], roundto))\n",
    "    return hpd, x, y, modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"output_%d.csv\"%(i) for i in np.arange(1,4)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>0.15</td>\n",
       "      <td>-0.005</td>\n",
       "      <td>-0.044859</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides_dash_acidifaciens</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.145</td>\n",
       "      <td>1.835619</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Akkermansia_dash_muciniphila</td>\n",
       "      <td>-0.27</td>\n",
       "      <td>0.04</td>\n",
       "      <td>-0.115</td>\n",
       "      <td>-1.416896</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Faecalibaculum</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-0.040</td>\n",
       "      <td>-0.531562</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Parasutterella</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-0.040</td>\n",
       "      <td>-0.540205</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Type                           Taxa  Left  Right  Middle  \\\n",
       "0  basal_growth_rate                 Muribaculaceae -0.16   0.15  -0.005   \n",
       "1  basal_growth_rate  Bacteroides_dash_acidifaciens -0.01   0.30   0.145   \n",
       "2  basal_growth_rate   Akkermansia_dash_muciniphila -0.27   0.04  -0.115   \n",
       "3  basal_growth_rate                 Faecalibaculum -0.20   0.12  -0.040   \n",
       "4  basal_growth_rate                 Parasutterella -0.20   0.12  -0.040   \n",
       "\n",
       "        SNR  Significant  \n",
       "0 -0.044859        False  \n",
       "1  1.835619        False  \n",
       "2 -1.416896        False  \n",
       "3 -0.531562        False  \n",
       "4 -0.540205        False  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "    \n",
    "# basal growth rate\n",
    "for c in bacterial_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# inulin response\n",
    "for c in bacterial_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['inulin_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "\n",
    "# pairwise interactions\n",
    "for c1 in bacterial_taxa:\n",
    "    for c2 in bacterial_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stan_output_summary.to_excel('bayesian_regression_summary_top20.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>1.08</td>\n",
       "      <td>1.62</td>\n",
       "      <td>1.350</td>\n",
       "      <td>9.958040</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Bacteroides_dash_acidifaciens</td>\n",
       "      <td>1.89</td>\n",
       "      <td>2.42</td>\n",
       "      <td>2.155</td>\n",
       "      <td>15.846923</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Faecalibaculum</td>\n",
       "      <td>0.48</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.745</td>\n",
       "      <td>5.506163</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Parasutterella</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.410</td>\n",
       "      <td>3.039780</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.310</td>\n",
       "      <td>2.265775</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Muribaculaceae')</td>\n",
       "      <td>-1.59</td>\n",
       "      <td>-0.29</td>\n",
       "      <td>-0.940</td>\n",
       "      <td>-2.864951</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Bacteroides_dash_acidifaci...</td>\n",
       "      <td>0.23</td>\n",
       "      <td>2.31</td>\n",
       "      <td>1.270</td>\n",
       "      <td>2.392931</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Desulfovibrionaceae')</td>\n",
       "      <td>0.27</td>\n",
       "      <td>3.71</td>\n",
       "      <td>1.990</td>\n",
       "      <td>2.280134</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Muribaculaceae', 'Clostridia_dash_UCG_dash_0...</td>\n",
       "      <td>-3.74</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>-1.900</td>\n",
       "      <td>-2.016358</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Muribaculac...</td>\n",
       "      <td>-3.81</td>\n",
       "      <td>-2.50</td>\n",
       "      <td>-3.155</td>\n",
       "      <td>-9.603366</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Akkermansia...</td>\n",
       "      <td>-4.57</td>\n",
       "      <td>-1.77</td>\n",
       "      <td>-3.170</td>\n",
       "      <td>-4.472396</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Faecalibacu...</td>\n",
       "      <td>-4.31</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>-2.655</td>\n",
       "      <td>-3.148145</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Parasuttere...</td>\n",
       "      <td>-4.26</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>-2.415</td>\n",
       "      <td>-2.600989</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_acidifaciens', 'Bacteroides')</td>\n",
       "      <td>-5.89</td>\n",
       "      <td>-2.35</td>\n",
       "      <td>-4.120</td>\n",
       "      <td>-4.574359</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Akkermansia_dash_muciniphila', 'Bacteroides_...</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2.09</td>\n",
       "      <td>1.050</td>\n",
       "      <td>1.970520</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Faecalibaculum', 'Muribaculaceae')</td>\n",
       "      <td>-1.77</td>\n",
       "      <td>-0.47</td>\n",
       "      <td>-1.120</td>\n",
       "      <td>-3.398489</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Escherichia_dash_Shigella', 'Escherichia_das...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>3.69</td>\n",
       "      <td>2.145</td>\n",
       "      <td>2.731059</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                               Taxa  \\\n",
       "20        inulin_response                                     Muribaculaceae   \n",
       "21        inulin_response                      Bacteroides_dash_acidifaciens   \n",
       "23        inulin_response                                     Faecalibaculum   \n",
       "24        inulin_response                                     Parasutterella   \n",
       "25        inulin_response                                        Bacteroides   \n",
       "40   pairwise_interaction               ('Muribaculaceae', 'Muribaculaceae')   \n",
       "41   pairwise_interaction  ('Muribaculaceae', 'Bacteroides_dash_acidifaci...   \n",
       "51   pairwise_interaction          ('Muribaculaceae', 'Desulfovibrionaceae')   \n",
       "52   pairwise_interaction  ('Muribaculaceae', 'Clostridia_dash_UCG_dash_0...   \n",
       "60   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Muribaculac...   \n",
       "62   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Akkermansia...   \n",
       "63   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Faecalibacu...   \n",
       "64   pairwise_interaction  ('Bacteroides_dash_acidifaciens', 'Parasuttere...   \n",
       "65   pairwise_interaction   ('Bacteroides_dash_acidifaciens', 'Bacteroides')   \n",
       "81   pairwise_interaction  ('Akkermansia_dash_muciniphila', 'Bacteroides_...   \n",
       "100  pairwise_interaction               ('Faecalibaculum', 'Muribaculaceae')   \n",
       "376  pairwise_interaction  ('Escherichia_dash_Shigella', 'Escherichia_das...   \n",
       "\n",
       "     Left  Right  Middle        SNR  Significant  \n",
       "20   1.08   1.62   1.350   9.958040         True  \n",
       "21   1.89   2.42   2.155  15.846923         True  \n",
       "23   0.48   1.01   0.745   5.506163         True  \n",
       "24   0.14   0.68   0.410   3.039780         True  \n",
       "25   0.04   0.58   0.310   2.265775         True  \n",
       "40  -1.59  -0.29  -0.940  -2.864951         True  \n",
       "41   0.23   2.31   1.270   2.392931         True  \n",
       "51   0.27   3.71   1.990   2.280134         True  \n",
       "52  -3.74  -0.06  -1.900  -2.016358         True  \n",
       "60  -3.81  -2.50  -3.155  -9.603366         True  \n",
       "62  -4.57  -1.77  -3.170  -4.472396         True  \n",
       "63  -4.31  -1.00  -2.655  -3.148145         True  \n",
       "64  -4.26  -0.57  -2.415  -2.600989         True  \n",
       "65  -5.89  -2.35  -4.120  -4.574359         True  \n",
       "81   0.01   2.09   1.050   1.970520         True  \n",
       "100 -1.77  -0.47  -1.120  -3.398489         True  \n",
       "376  0.60   3.69   2.145   2.731059         True  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stan_output_summary = pd.read_excel('bayesian_regression_summary_top20.xlsx', index_col=0)\n",
    "df_stan_output_summary = df_stan_output_summary[df_stan_output_summary.Significant==True]\n",
    "df_stan_output_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
