{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from scipy.interpolate import CubicSpline\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import warnings\n",
    "import itertools\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "import statsmodels.stats.api as sms\n",
    "import scipy.stats.kde as kde"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read 16S data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Acetanaerobacterium</th>\n",
       "      <th>Acidaminococcus_dash_intestini</th>\n",
       "      <th>Akkermansia</th>\n",
       "      <th>Akkermansia_dash_sp_dot__dash_KLE1798</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Alistipes_dash_finegoldii_dash_DSM_dash_17242</th>\n",
       "      <th>Alistipes_dash_inops</th>\n",
       "      <th>Alistipes_dash_obesi</th>\n",
       "      <th>Alistipes_dash_shahii_dash_WAL_dash_8301</th>\n",
       "      <th>Alistipes_dash_sp_dot__dash_Marseille_dash_P5997</th>\n",
       "      <th>...</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c</th>\n",
       "      <th>butyrate_dash_producing_dash_bacterium_dash_L2_dash_10</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>methanogenic_dash_archaeon_dash_CH1270</th>\n",
       "      <th>rumen_dash_bacterium_dash_NK2A1</th>\n",
       "      <th>rumen_dash_bacterium_dash_YS2</th>\n",
       "      <th>vadinBE97</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.012382</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.018176</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>0.000873</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011868</td>\n",
       "      <td>0.003349</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008914</td>\n",
       "      <td>0.002265</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.012607</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.030136</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.022406</td>\n",
       "      <td>0.008317</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011350</td>\n",
       "      <td>0.002642</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003595</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.045532</td>\n",
       "      <td>0.005492</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002097</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011183</td>\n",
       "      <td>0.002546</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011083</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005820</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008059</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002361</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.045299</td>\n",
       "      <td>0.004396</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008628</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 314 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Acetanaerobacterium  Acidaminococcus_dash_intestini  Akkermansia  \\\n",
       "SampleID                                                                     \n",
       "sample1                   0.0                             0.0     0.012382   \n",
       "sample2                   0.0                             0.0     0.000000   \n",
       "sample3                   0.0                             0.0     0.030136   \n",
       "sample4                   0.0                             0.0     0.003595   \n",
       "sample5                   0.0                             0.0     0.005820   \n",
       "\n",
       "          Akkermansia_dash_sp_dot__dash_KLE1798  Alistipes  \\\n",
       "SampleID                                                     \n",
       "sample1                                     0.0   0.018176   \n",
       "sample2                                     0.0   0.011868   \n",
       "sample3                                     0.0   0.022406   \n",
       "sample4                                     0.0   0.045532   \n",
       "sample5                                     0.0   0.008059   \n",
       "\n",
       "          Alistipes_dash_finegoldii_dash_DSM_dash_17242  Alistipes_dash_inops  \\\n",
       "SampleID                                                                        \n",
       "sample1                                        0.000000                   0.0   \n",
       "sample2                                        0.003349                   0.0   \n",
       "sample3                                        0.008317                   0.0   \n",
       "sample4                                        0.005492                   0.0   \n",
       "sample5                                        0.000000                   0.0   \n",
       "\n",
       "          Alistipes_dash_obesi  Alistipes_dash_shahii_dash_WAL_dash_8301  \\\n",
       "SampleID                                                                   \n",
       "sample1               0.000000                                       0.0   \n",
       "sample2               0.000000                                       0.0   \n",
       "sample3               0.000000                                       0.0   \n",
       "sample4               0.002097                                       0.0   \n",
       "sample5               0.002361                                       0.0   \n",
       "\n",
       "          Alistipes_dash_sp_dot__dash_Marseille_dash_P5997  ...  \\\n",
       "SampleID                                                    ...   \n",
       "sample1                                                0.0  ...   \n",
       "sample2                                                0.0  ...   \n",
       "sample3                                                0.0  ...   \n",
       "sample4                                                0.0  ...   \n",
       "sample5                                                0.0  ...   \n",
       "\n",
       "          bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c  \\\n",
       "SampleID                                                                \n",
       "sample1                                                 0.0             \n",
       "sample2                                                 0.0             \n",
       "sample3                                                 0.0             \n",
       "sample4                                                 0.0             \n",
       "sample5                                                 0.0             \n",
       "\n",
       "          butyrate_dash_producing_dash_bacterium_dash_L2_dash_10  \\\n",
       "SampleID                                                           \n",
       "sample1                                                 0.0        \n",
       "sample2                                                 0.0        \n",
       "sample3                                                 0.0        \n",
       "sample4                                                 0.0        \n",
       "sample5                                                 0.0        \n",
       "\n",
       "          gut_dash_metagenome  human_dash_gut_dash_metagenome  \\\n",
       "SampleID                                                        \n",
       "sample1              0.005000                        0.000873   \n",
       "sample2              0.008914                        0.002265   \n",
       "sample3              0.011350                        0.002642   \n",
       "sample4              0.011183                        0.002546   \n",
       "sample5              0.045299                        0.004396   \n",
       "\n",
       "          iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70  \\\n",
       "SampleID                                                                                        \n",
       "sample1                                                 0.0                                     \n",
       "sample2                                                 0.0                                     \n",
       "sample3                                                 0.0                                     \n",
       "sample4                                                 0.0                                     \n",
       "sample5                                                 0.0                                     \n",
       "\n",
       "          metagenome  methanogenic_dash_archaeon_dash_CH1270  \\\n",
       "SampleID                                                       \n",
       "sample1     0.005715                                     0.0   \n",
       "sample2     0.012607                                     0.0   \n",
       "sample3     0.019520                                     0.0   \n",
       "sample4     0.011083                                     0.0   \n",
       "sample5     0.008628                                     0.0   \n",
       "\n",
       "          rumen_dash_bacterium_dash_NK2A1  rumen_dash_bacterium_dash_YS2  \\\n",
       "SampleID                                                                   \n",
       "sample1                               0.0                            0.0   \n",
       "sample2                               0.0                            0.0   \n",
       "sample3                               0.0                            0.0   \n",
       "sample4                               0.0                            0.0   \n",
       "sample5                               0.0                            0.0   \n",
       "\n",
       "          vadinBE97  \n",
       "SampleID             \n",
       "sample1         0.0  \n",
       "sample2         0.0  \n",
       "sample3         0.0  \n",
       "sample4         0.0  \n",
       "sample5         0.0  \n",
       "\n",
       "[5 rows x 314 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = pd.read_excel('./16S_relative_abundance.xlsx', index_col=0)\n",
    "df_16S.columns = [c.replace('/','_slash_').replace(' ','_space_').replace('[','_leftsquarebracket_').replace(']','_rightsquarebracket_').replace('-','_dash_').replace('.','_dot_').replace('(','_leftroundbracket').replace(')','_rightroundbracket_') for c in df_16S.columns]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add pseudocount for zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Acetanaerobacterium</th>\n",
       "      <th>Acidaminococcus_dash_intestini</th>\n",
       "      <th>Akkermansia</th>\n",
       "      <th>Akkermansia_dash_sp_dot__dash_KLE1798</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Alistipes_dash_finegoldii_dash_DSM_dash_17242</th>\n",
       "      <th>Alistipes_dash_inops</th>\n",
       "      <th>Alistipes_dash_obesi</th>\n",
       "      <th>Alistipes_dash_shahii_dash_WAL_dash_8301</th>\n",
       "      <th>Alistipes_dash_sp_dot__dash_Marseille_dash_P5997</th>\n",
       "      <th>...</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c</th>\n",
       "      <th>butyrate_dash_producing_dash_bacterium_dash_L2_dash_10</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>methanogenic_dash_archaeon_dash_CH1270</th>\n",
       "      <th>rumen_dash_bacterium_dash_NK2A1</th>\n",
       "      <th>rumen_dash_bacterium_dash_YS2</th>\n",
       "      <th>vadinBE97</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.012382</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.018176</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>0.000873</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.005715</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample2</th>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.011868</td>\n",
       "      <td>0.003349</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.008914</td>\n",
       "      <td>0.002265</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.012607</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.000148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample3</th>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.030136</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.022406</td>\n",
       "      <td>0.008317</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.011350</td>\n",
       "      <td>0.002642</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.019520</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.000196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample4</th>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.003595</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.045532</td>\n",
       "      <td>0.005492</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.002097</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.011183</td>\n",
       "      <td>0.002546</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.011083</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.000449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample5</th>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.005820</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.008059</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.002361</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.045299</td>\n",
       "      <td>0.004396</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.008628</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.000244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 314 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Acetanaerobacterium  Acidaminococcus_dash_intestini  Akkermansia  \\\n",
       "SampleID                                                                     \n",
       "sample1              0.000635                        0.000635     0.012382   \n",
       "sample2              0.000148                        0.000148     0.000148   \n",
       "sample3              0.000196                        0.000196     0.030136   \n",
       "sample4              0.000449                        0.000449     0.003595   \n",
       "sample5              0.000244                        0.000244     0.005820   \n",
       "\n",
       "          Akkermansia_dash_sp_dot__dash_KLE1798  Alistipes  \\\n",
       "SampleID                                                     \n",
       "sample1                                0.000635   0.018176   \n",
       "sample2                                0.000148   0.011868   \n",
       "sample3                                0.000196   0.022406   \n",
       "sample4                                0.000449   0.045532   \n",
       "sample5                                0.000244   0.008059   \n",
       "\n",
       "          Alistipes_dash_finegoldii_dash_DSM_dash_17242  Alistipes_dash_inops  \\\n",
       "SampleID                                                                        \n",
       "sample1                                        0.000635              0.000635   \n",
       "sample2                                        0.003349              0.000148   \n",
       "sample3                                        0.008317              0.000196   \n",
       "sample4                                        0.005492              0.000449   \n",
       "sample5                                        0.000244              0.000244   \n",
       "\n",
       "          Alistipes_dash_obesi  Alistipes_dash_shahii_dash_WAL_dash_8301  \\\n",
       "SampleID                                                                   \n",
       "sample1               0.000635                                  0.000635   \n",
       "sample2               0.000148                                  0.000148   \n",
       "sample3               0.000196                                  0.000196   \n",
       "sample4               0.002097                                  0.000449   \n",
       "sample5               0.002361                                  0.000244   \n",
       "\n",
       "          Alistipes_dash_sp_dot__dash_Marseille_dash_P5997  ...  \\\n",
       "SampleID                                                    ...   \n",
       "sample1                                           0.000635  ...   \n",
       "sample2                                           0.000148  ...   \n",
       "sample3                                           0.000196  ...   \n",
       "sample4                                           0.000449  ...   \n",
       "sample5                                           0.000244  ...   \n",
       "\n",
       "          bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c  \\\n",
       "SampleID                                                                \n",
       "sample1                                            0.000635             \n",
       "sample2                                            0.000148             \n",
       "sample3                                            0.000196             \n",
       "sample4                                            0.000449             \n",
       "sample5                                            0.000244             \n",
       "\n",
       "          butyrate_dash_producing_dash_bacterium_dash_L2_dash_10  \\\n",
       "SampleID                                                           \n",
       "sample1                                            0.000635        \n",
       "sample2                                            0.000148        \n",
       "sample3                                            0.000196        \n",
       "sample4                                            0.000449        \n",
       "sample5                                            0.000244        \n",
       "\n",
       "          gut_dash_metagenome  human_dash_gut_dash_metagenome  \\\n",
       "SampleID                                                        \n",
       "sample1              0.005000                        0.000873   \n",
       "sample2              0.008914                        0.002265   \n",
       "sample3              0.011350                        0.002642   \n",
       "sample4              0.011183                        0.002546   \n",
       "sample5              0.045299                        0.004396   \n",
       "\n",
       "          iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70  \\\n",
       "SampleID                                                                                        \n",
       "sample1                                            0.000635                                     \n",
       "sample2                                            0.000148                                     \n",
       "sample3                                            0.000196                                     \n",
       "sample4                                            0.000449                                     \n",
       "sample5                                            0.000244                                     \n",
       "\n",
       "          metagenome  methanogenic_dash_archaeon_dash_CH1270  \\\n",
       "SampleID                                                       \n",
       "sample1     0.005715                                0.000635   \n",
       "sample2     0.012607                                0.000148   \n",
       "sample3     0.019520                                0.000196   \n",
       "sample4     0.011083                                0.000449   \n",
       "sample5     0.008628                                0.000244   \n",
       "\n",
       "          rumen_dash_bacterium_dash_NK2A1  rumen_dash_bacterium_dash_YS2  \\\n",
       "SampleID                                                                   \n",
       "sample1                          0.000635                       0.000635   \n",
       "sample2                          0.000148                       0.000148   \n",
       "sample3                          0.000196                       0.000196   \n",
       "sample4                          0.000449                       0.000449   \n",
       "sample5                          0.000244                       0.000244   \n",
       "\n",
       "          vadinBE97  \n",
       "SampleID             \n",
       "sample1    0.000635  \n",
       "sample2    0.000148  \n",
       "sample3    0.000196  \n",
       "sample4    0.000449  \n",
       "sample5    0.000244  \n",
       "\n",
       "[5 rows x 314 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample_id in df_16S.index:\n",
    "    sample = np.array(df_16S.loc[sample_id])\n",
    "    minval = np.min(sample[np.nonzero(sample)]) # minimum non-zero value\n",
    "    sample[sample==0] = minval\n",
    "    df_16S.loc[sample_id] = sample\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata for maize starch and control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'inulin-maltodextrin', 'maltodextrin-inulin'}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df_meta.treatmentgroup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Day</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>1_A</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample2</th>\n",
       "      <td>1_B</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample3</th>\n",
       "      <td>1_A</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample4</th>\n",
       "      <td>1_B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample5</th>\n",
       "      <td>3_A</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Subject  Diet  Day\n",
       "#SampleID                   \n",
       "sample1       1_A     1   21\n",
       "sample2       1_B     0   21\n",
       "sample3       1_A     1    0\n",
       "sample4       1_B     0    0\n",
       "sample5       3_A     0   21"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv('./metadata-filtered.txt', sep='\\t', index_col=0)\n",
    "df_meta = df_meta.drop('#q2:types')\n",
    "df_meta = df_meta[df_meta.excluded=='FALSE']\n",
    "df_meta = df_meta[['subject_id','inulin','timepoint','treatmentgroup']]\n",
    "\n",
    "# remove patients that have odd number of samples\n",
    "subject_to_remove = []\n",
    "for sid in set(df_meta.subject_id):\n",
    "    _df = df_meta[df_meta.subject_id==sid]\n",
    "    if len(_df) % 2 != 0:\n",
    "        subject_to_remove.append(sid)\n",
    "df_meta = df_meta[~df_meta.subject_id.isin(subject_to_remove)]\n",
    "\n",
    "# runin1->intake1, runin2->intake2\n",
    "for sid in set(df_meta.subject_id):\n",
    "    df_meta.loc[(df_meta.subject_id==sid)&((df_meta.timepoint=='RunIn1')|(df_meta.timepoint=='Intake1')),'subject_id'] = sid+'_A'\n",
    "    df_meta.loc[(df_meta.subject_id==sid)&((df_meta.timepoint=='RunIn2')|(df_meta.timepoint=='Intake2')),'subject_id'] = sid+'_B'\n",
    "    \n",
    "for sid in set(df_meta.subject_id):\n",
    "    df_meta.loc[(df_meta.subject_id==sid)&((df_meta.timepoint=='RunIn1')|(df_meta.timepoint=='RunIn2')),'timepoint'] = 0\n",
    "    df_meta.loc[(df_meta.subject_id==sid)&((df_meta.timepoint=='Intake1')|(df_meta.timepoint=='Intake2')),'timepoint'] = 21\n",
    "    \n",
    "for sid in set(df_meta.subject_id):\n",
    "    treatment = list(set(list(df_meta.loc[df_meta.subject_id==sid,'treatmentgroup'])))\n",
    "    assert len(treatment)==1\n",
    "    if sid.endswith('_A'):\n",
    "        if treatment[0].split('-')[0]=='inulin':\n",
    "            df_meta.loc[df_meta.subject_id==sid,'inulin']=True\n",
    "        else:\n",
    "            df_meta.loc[df_meta.subject_id==sid,'inulin']=False\n",
    "    else:\n",
    "        # ends with _B\n",
    "        if treatment[0].split('-')[1]=='inulin':\n",
    "            df_meta.loc[df_meta.subject_id==sid,'inulin']=True\n",
    "        else:\n",
    "            df_meta.loc[df_meta.subject_id==sid,'inulin']=False\n",
    "\n",
    "df_meta.timepoint = df_meta.timepoint.astype(int)\n",
    "df_meta = df_meta.rename({'timepoint':'Day', 'subject_id':'Subject', 'inulin':'Diet'}, axis=1)\n",
    "df_meta['Diet'] = df_meta['Diet'].astype(int)\n",
    "df_meta = df_meta.drop('treatmentgroup', axis=1)\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 20 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Coprococcus</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>Dorea</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Christensenellaceae_dash_R_dash_7_dash_group</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Subdoligranulum</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_DJF_VR52</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>0.016827</td>\n",
       "      <td>0.044130</td>\n",
       "      <td>0.059370</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>0.064767</td>\n",
       "      <td>0.018176</td>\n",
       "      <td>0.011429</td>\n",
       "      <td>0.006429</td>\n",
       "      <td>0.005715</td>\n",
       "      <td>0.008334</td>\n",
       "      <td>0.009207</td>\n",
       "      <td>0.023018</td>\n",
       "      <td>0.000873</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.003651</td>\n",
       "      <td>0.048734</td>\n",
       "      <td>0.019287</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.013334</td>\n",
       "      <td>0.013811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample2</th>\n",
       "      <td>0.038018</td>\n",
       "      <td>0.071161</td>\n",
       "      <td>0.062395</td>\n",
       "      <td>0.008914</td>\n",
       "      <td>0.029203</td>\n",
       "      <td>0.011868</td>\n",
       "      <td>0.015710</td>\n",
       "      <td>0.016104</td>\n",
       "      <td>0.012607</td>\n",
       "      <td>0.002512</td>\n",
       "      <td>0.013641</td>\n",
       "      <td>0.003989</td>\n",
       "      <td>0.002265</td>\n",
       "      <td>0.019403</td>\n",
       "      <td>0.006008</td>\n",
       "      <td>0.022161</td>\n",
       "      <td>0.029105</td>\n",
       "      <td>0.008027</td>\n",
       "      <td>0.011376</td>\n",
       "      <td>0.021078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample3</th>\n",
       "      <td>0.094369</td>\n",
       "      <td>0.054645</td>\n",
       "      <td>0.090015</td>\n",
       "      <td>0.011350</td>\n",
       "      <td>0.033805</td>\n",
       "      <td>0.022406</td>\n",
       "      <td>0.013453</td>\n",
       "      <td>0.012915</td>\n",
       "      <td>0.019520</td>\n",
       "      <td>0.006800</td>\n",
       "      <td>0.020009</td>\n",
       "      <td>0.018003</td>\n",
       "      <td>0.002642</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.006458</td>\n",
       "      <td>0.045546</td>\n",
       "      <td>0.031114</td>\n",
       "      <td>0.002984</td>\n",
       "      <td>0.018933</td>\n",
       "      <td>0.019764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample4</th>\n",
       "      <td>0.137344</td>\n",
       "      <td>0.043784</td>\n",
       "      <td>0.059910</td>\n",
       "      <td>0.011183</td>\n",
       "      <td>0.062157</td>\n",
       "      <td>0.045532</td>\n",
       "      <td>0.016525</td>\n",
       "      <td>0.010634</td>\n",
       "      <td>0.011083</td>\n",
       "      <td>0.008238</td>\n",
       "      <td>0.019970</td>\n",
       "      <td>0.006291</td>\n",
       "      <td>0.002546</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.009685</td>\n",
       "      <td>0.072691</td>\n",
       "      <td>0.022866</td>\n",
       "      <td>0.003145</td>\n",
       "      <td>0.020170</td>\n",
       "      <td>0.008088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample5</th>\n",
       "      <td>0.172894</td>\n",
       "      <td>0.112617</td>\n",
       "      <td>0.038665</td>\n",
       "      <td>0.045299</td>\n",
       "      <td>0.017135</td>\n",
       "      <td>0.008059</td>\n",
       "      <td>0.013675</td>\n",
       "      <td>0.072527</td>\n",
       "      <td>0.008628</td>\n",
       "      <td>0.011152</td>\n",
       "      <td>0.008384</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.004396</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.075824</td>\n",
       "      <td>0.005698</td>\n",
       "      <td>0.008506</td>\n",
       "      <td>0.011559</td>\n",
       "      <td>0.006553</td>\n",
       "      <td>0.006349</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Bacteroides  Faecalibacterium  Lachnospiraceae  \\\n",
       "#SampleID                                                   \n",
       "sample1       0.016827          0.044130         0.059370   \n",
       "sample2       0.038018          0.071161         0.062395   \n",
       "sample3       0.094369          0.054645         0.090015   \n",
       "sample4       0.137344          0.043784         0.059910   \n",
       "sample5       0.172894          0.112617         0.038665   \n",
       "\n",
       "           gut_dash_metagenome  \\\n",
       "#SampleID                        \n",
       "sample1               0.005000   \n",
       "sample2               0.008914   \n",
       "sample3               0.011350   \n",
       "sample4               0.011183   \n",
       "sample5               0.045299   \n",
       "\n",
       "           Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  Alistipes  \\\n",
       "#SampleID                                                                  \n",
       "sample1                                             0.064767    0.018176   \n",
       "sample2                                             0.029203    0.011868   \n",
       "sample3                                             0.033805    0.022406   \n",
       "sample4                                             0.062157    0.045532   \n",
       "sample5                                             0.017135    0.008059   \n",
       "\n",
       "            Blautia  Coprococcus  metagenome  Anaerostipes     Dorea  \\\n",
       "#SampleID                                                              \n",
       "sample1    0.011429     0.006429    0.005715      0.008334  0.009207   \n",
       "sample2    0.015710     0.016104    0.012607      0.002512  0.013641   \n",
       "sample3    0.013453     0.012915    0.019520      0.006800  0.020009   \n",
       "sample4    0.016525     0.010634    0.011083      0.008238  0.019970   \n",
       "sample5    0.013675     0.072527    0.008628      0.011152  0.008384   \n",
       "\n",
       "           Bifidobacterium  human_dash_gut_dash_metagenome  Roseburia  \\\n",
       "#SampleID                                                               \n",
       "sample1           0.023018                        0.000873   0.000635   \n",
       "sample2           0.003989                        0.002265   0.019403   \n",
       "sample3           0.018003                        0.002642   0.000196   \n",
       "sample4           0.006291                        0.002546   0.000449   \n",
       "sample5           0.000244                        0.004396   0.000244   \n",
       "\n",
       "           Christensenellaceae_dash_R_dash_7_dash_group  Ruminococcus  \\\n",
       "#SampleID                                                               \n",
       "sample1                                        0.003651      0.048734   \n",
       "sample2                                        0.006008      0.022161   \n",
       "sample3                                        0.006458      0.045546   \n",
       "sample4                                        0.009685      0.072691   \n",
       "sample5                                        0.075824      0.005698   \n",
       "\n",
       "           Subdoligranulum  Fusicatenibacter  \\\n",
       "#SampleID                                      \n",
       "sample1           0.019287          0.000635   \n",
       "sample2           0.029105          0.008027   \n",
       "sample3           0.031114          0.002984   \n",
       "sample4           0.022866          0.003145   \n",
       "sample5           0.008506          0.011559   \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group  \\\n",
       "#SampleID                                                                              \n",
       "sample1                                             0.013334                           \n",
       "sample2                                             0.011376                           \n",
       "sample3                                             0.018933                           \n",
       "sample4                                             0.020170                           \n",
       "sample5                                             0.006553                           \n",
       "\n",
       "           Ruminococcus_dash_sp_dot__dash_DJF_VR52  \n",
       "#SampleID                                           \n",
       "sample1                                   0.013811  \n",
       "sample2                                   0.021078  \n",
       "sample3                                   0.019764  \n",
       "sample4                                   0.008088  \n",
       "sample5                                   0.006349  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 20\n",
    "df_16S_T = df_16S.loc[df_meta.index].T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Bacteroides',\n",
       " 'Faecalibacterium',\n",
       " 'Lachnospiraceae',\n",
       " 'gut_dash_metagenome',\n",
       " 'Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57',\n",
       " 'Alistipes',\n",
       " 'Blautia',\n",
       " 'Coprococcus',\n",
       " 'metagenome',\n",
       " 'Anaerostipes',\n",
       " 'Dorea',\n",
       " 'Bifidobacterium',\n",
       " 'human_dash_gut_dash_metagenome',\n",
       " 'Roseburia',\n",
       " 'Christensenellaceae_dash_R_dash_7_dash_group',\n",
       " 'Ruminococcus',\n",
       " 'Subdoligranulum',\n",
       " 'Fusicatenibacter',\n",
       " '_leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group',\n",
       " 'Ruminococcus_dash_sp_dot__dash_DJF_VR52']"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_16S_topX.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Day</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>...</th>\n",
       "      <th>Dorea</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Christensenellaceae_dash_R_dash_7_dash_group</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Subdoligranulum</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_DJF_VR52</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>1_A</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>0.024823</td>\n",
       "      <td>0.065101</td>\n",
       "      <td>0.087582</td>\n",
       "      <td>0.007377</td>\n",
       "      <td>0.095544</td>\n",
       "      <td>0.026813</td>\n",
       "      <td>0.016861</td>\n",
       "      <td>...</td>\n",
       "      <td>0.013582</td>\n",
       "      <td>0.033956</td>\n",
       "      <td>0.001288</td>\n",
       "      <td>0.000937</td>\n",
       "      <td>0.005386</td>\n",
       "      <td>0.071892</td>\n",
       "      <td>0.028453</td>\n",
       "      <td>0.000937</td>\n",
       "      <td>0.019671</td>\n",
       "      <td>0.020373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample2</th>\n",
       "      <td>1_B</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>0.056085</td>\n",
       "      <td>0.104977</td>\n",
       "      <td>0.092046</td>\n",
       "      <td>0.013149</td>\n",
       "      <td>0.043081</td>\n",
       "      <td>0.017508</td>\n",
       "      <td>0.023175</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020124</td>\n",
       "      <td>0.005885</td>\n",
       "      <td>0.003342</td>\n",
       "      <td>0.028623</td>\n",
       "      <td>0.008863</td>\n",
       "      <td>0.032692</td>\n",
       "      <td>0.042935</td>\n",
       "      <td>0.011842</td>\n",
       "      <td>0.016782</td>\n",
       "      <td>0.031094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample3</th>\n",
       "      <td>1_A</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.139213</td>\n",
       "      <td>0.080612</td>\n",
       "      <td>0.132790</td>\n",
       "      <td>0.016743</td>\n",
       "      <td>0.049869</td>\n",
       "      <td>0.033053</td>\n",
       "      <td>0.019846</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029517</td>\n",
       "      <td>0.026558</td>\n",
       "      <td>0.003897</td>\n",
       "      <td>0.000289</td>\n",
       "      <td>0.009526</td>\n",
       "      <td>0.067189</td>\n",
       "      <td>0.045899</td>\n",
       "      <td>0.004402</td>\n",
       "      <td>0.027929</td>\n",
       "      <td>0.029156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample4</th>\n",
       "      <td>1_B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.202610</td>\n",
       "      <td>0.064591</td>\n",
       "      <td>0.088379</td>\n",
       "      <td>0.016497</td>\n",
       "      <td>0.091694</td>\n",
       "      <td>0.067168</td>\n",
       "      <td>0.024378</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029460</td>\n",
       "      <td>0.009280</td>\n",
       "      <td>0.003756</td>\n",
       "      <td>0.000663</td>\n",
       "      <td>0.014288</td>\n",
       "      <td>0.107234</td>\n",
       "      <td>0.033731</td>\n",
       "      <td>0.004640</td>\n",
       "      <td>0.029754</td>\n",
       "      <td>0.011931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample5</th>\n",
       "      <td>3_A</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>0.255053</td>\n",
       "      <td>0.166133</td>\n",
       "      <td>0.057039</td>\n",
       "      <td>0.066825</td>\n",
       "      <td>0.025277</td>\n",
       "      <td>0.011888</td>\n",
       "      <td>0.020174</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012368</td>\n",
       "      <td>0.000360</td>\n",
       "      <td>0.006484</td>\n",
       "      <td>0.000360</td>\n",
       "      <td>0.111856</td>\n",
       "      <td>0.008406</td>\n",
       "      <td>0.012548</td>\n",
       "      <td>0.017052</td>\n",
       "      <td>0.009667</td>\n",
       "      <td>0.009366</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Subject  Diet  Day  Bacteroides  Faecalibacterium  Lachnospiraceae  \\\n",
       "#SampleID                                                                      \n",
       "sample1       1_A     1   21     0.024823          0.065101         0.087582   \n",
       "sample2       1_B     0   21     0.056085          0.104977         0.092046   \n",
       "sample3       1_A     1    0     0.139213          0.080612         0.132790   \n",
       "sample4       1_B     0    0     0.202610          0.064591         0.088379   \n",
       "sample5       3_A     0   21     0.255053          0.166133         0.057039   \n",
       "\n",
       "           gut_dash_metagenome  \\\n",
       "#SampleID                        \n",
       "sample1               0.007377   \n",
       "sample2               0.013149   \n",
       "sample3               0.016743   \n",
       "sample4               0.016497   \n",
       "sample5               0.066825   \n",
       "\n",
       "           Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  Alistipes  \\\n",
       "#SampleID                                                                  \n",
       "sample1                                             0.095544    0.026813   \n",
       "sample2                                             0.043081    0.017508   \n",
       "sample3                                             0.049869    0.033053   \n",
       "sample4                                             0.091694    0.067168   \n",
       "sample5                                             0.025277    0.011888   \n",
       "\n",
       "            Blautia  ...     Dorea  Bifidobacterium  \\\n",
       "#SampleID            ...                              \n",
       "sample1    0.016861  ...  0.013582         0.033956   \n",
       "sample2    0.023175  ...  0.020124         0.005885   \n",
       "sample3    0.019846  ...  0.029517         0.026558   \n",
       "sample4    0.024378  ...  0.029460         0.009280   \n",
       "sample5    0.020174  ...  0.012368         0.000360   \n",
       "\n",
       "           human_dash_gut_dash_metagenome  Roseburia  \\\n",
       "#SampleID                                              \n",
       "sample1                          0.001288   0.000937   \n",
       "sample2                          0.003342   0.028623   \n",
       "sample3                          0.003897   0.000289   \n",
       "sample4                          0.003756   0.000663   \n",
       "sample5                          0.006484   0.000360   \n",
       "\n",
       "           Christensenellaceae_dash_R_dash_7_dash_group  Ruminococcus  \\\n",
       "#SampleID                                                               \n",
       "sample1                                        0.005386      0.071892   \n",
       "sample2                                        0.008863      0.032692   \n",
       "sample3                                        0.009526      0.067189   \n",
       "sample4                                        0.014288      0.107234   \n",
       "sample5                                        0.111856      0.008406   \n",
       "\n",
       "           Subdoligranulum  Fusicatenibacter  \\\n",
       "#SampleID                                      \n",
       "sample1           0.028453          0.000937   \n",
       "sample2           0.042935          0.011842   \n",
       "sample3           0.045899          0.004402   \n",
       "sample4           0.033731          0.004640   \n",
       "sample5           0.012548          0.017052   \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group  \\\n",
       "#SampleID                                                                              \n",
       "sample1                                             0.019671                           \n",
       "sample2                                             0.016782                           \n",
       "sample3                                             0.027929                           \n",
       "sample4                                             0.029754                           \n",
       "sample5                                             0.009667                           \n",
       "\n",
       "           Ruminococcus_dash_sp_dot__dash_DJF_VR52  \n",
       "#SampleID                                           \n",
       "sample1                                   0.020373  \n",
       "sample2                                   0.031094  \n",
       "sample3                                   0.029156  \n",
       "sample4                                   0.011931  \n",
       "sample5                                   0.009366  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX_w_meta = pd.merge(df_meta, df_16S_topX/df_16S_topX.max().max(), left_index=True, right_index=True, how='inner')\n",
    "df_16S_topX_w_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate log-derivatives (dlog(B)/dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Day</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>...</th>\n",
       "      <th>Dorea</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Christensenellaceae_dash_R_dash_7_dash_group</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Subdoligranulum</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_DJF_VR52</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>1_A</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>-0.005447</td>\n",
       "      <td>-0.000739</td>\n",
       "      <td>-0.002153</td>\n",
       "      <td>-0.000446</td>\n",
       "      <td>0.002175</td>\n",
       "      <td>-0.000297</td>\n",
       "      <td>-0.000142</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000759</td>\n",
       "      <td>0.000352</td>\n",
       "      <td>-0.000124</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>-0.000197</td>\n",
       "      <td>0.000224</td>\n",
       "      <td>-0.000831</td>\n",
       "      <td>-0.000165</td>\n",
       "      <td>-0.000393</td>\n",
       "      <td>-0.000418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample2</th>\n",
       "      <td>1_B</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>-0.006977</td>\n",
       "      <td>0.001923</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>-0.000159</td>\n",
       "      <td>-0.002315</td>\n",
       "      <td>-0.002365</td>\n",
       "      <td>-0.000057</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000445</td>\n",
       "      <td>-0.000162</td>\n",
       "      <td>-0.000020</td>\n",
       "      <td>0.001331</td>\n",
       "      <td>-0.000258</td>\n",
       "      <td>-0.003550</td>\n",
       "      <td>0.000438</td>\n",
       "      <td>0.000343</td>\n",
       "      <td>-0.000618</td>\n",
       "      <td>0.000912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample3</th>\n",
       "      <td>1_A</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.005447</td>\n",
       "      <td>-0.000739</td>\n",
       "      <td>-0.002153</td>\n",
       "      <td>-0.000446</td>\n",
       "      <td>0.002175</td>\n",
       "      <td>-0.000297</td>\n",
       "      <td>-0.000142</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000759</td>\n",
       "      <td>0.000352</td>\n",
       "      <td>-0.000124</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>-0.000197</td>\n",
       "      <td>0.000224</td>\n",
       "      <td>-0.000831</td>\n",
       "      <td>-0.000165</td>\n",
       "      <td>-0.000393</td>\n",
       "      <td>-0.000418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample4</th>\n",
       "      <td>1_B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.006977</td>\n",
       "      <td>0.001923</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>-0.000159</td>\n",
       "      <td>-0.002315</td>\n",
       "      <td>-0.002365</td>\n",
       "      <td>-0.000057</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000445</td>\n",
       "      <td>-0.000162</td>\n",
       "      <td>-0.000020</td>\n",
       "      <td>0.001331</td>\n",
       "      <td>-0.000258</td>\n",
       "      <td>-0.003550</td>\n",
       "      <td>0.000438</td>\n",
       "      <td>0.000343</td>\n",
       "      <td>-0.000618</td>\n",
       "      <td>0.000912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample5</th>\n",
       "      <td>3_A</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>0.000615</td>\n",
       "      <td>-0.000337</td>\n",
       "      <td>-0.000245</td>\n",
       "      <td>0.000335</td>\n",
       "      <td>-0.001594</td>\n",
       "      <td>0.000134</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000269</td>\n",
       "      <td>-0.000233</td>\n",
       "      <td>-0.000536</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.001363</td>\n",
       "      <td>0.000268</td>\n",
       "      <td>-0.000352</td>\n",
       "      <td>-0.000302</td>\n",
       "      <td>-0.000610</td>\n",
       "      <td>-0.000948</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Subject  Diet  Day  Bacteroides  Faecalibacterium  Lachnospiraceae  \\\n",
       "#SampleID                                                                      \n",
       "sample1       1_A     1   21    -0.005447         -0.000739        -0.002153   \n",
       "sample2       1_B     0   21    -0.006977          0.001923         0.000175   \n",
       "sample3       1_A     1    0    -0.005447         -0.000739        -0.002153   \n",
       "sample4       1_B     0    0    -0.006977          0.001923         0.000175   \n",
       "sample5       3_A     0   21     0.000615         -0.000337        -0.000245   \n",
       "\n",
       "           gut_dash_metagenome  \\\n",
       "#SampleID                        \n",
       "sample1              -0.000446   \n",
       "sample2              -0.000159   \n",
       "sample3              -0.000446   \n",
       "sample4              -0.000159   \n",
       "sample5               0.000335   \n",
       "\n",
       "           Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  Alistipes  \\\n",
       "#SampleID                                                                  \n",
       "sample1                                             0.002175   -0.000297   \n",
       "sample2                                            -0.002315   -0.002365   \n",
       "sample3                                             0.002175   -0.000297   \n",
       "sample4                                            -0.002315   -0.002365   \n",
       "sample5                                            -0.001594    0.000134   \n",
       "\n",
       "            Blautia  ...     Dorea  Bifidobacterium  \\\n",
       "#SampleID            ...                              \n",
       "sample1   -0.000142  ... -0.000759         0.000352   \n",
       "sample2   -0.000057  ... -0.000445        -0.000162   \n",
       "sample3   -0.000142  ... -0.000759         0.000352   \n",
       "sample4   -0.000057  ... -0.000445        -0.000162   \n",
       "sample5    0.000260  ... -0.000269        -0.000233   \n",
       "\n",
       "           human_dash_gut_dash_metagenome  Roseburia  \\\n",
       "#SampleID                                              \n",
       "sample1                         -0.000124   0.000031   \n",
       "sample2                         -0.000020   0.001331   \n",
       "sample3                         -0.000124   0.000031   \n",
       "sample4                         -0.000020   0.001331   \n",
       "sample5                         -0.000536   0.000011   \n",
       "\n",
       "           Christensenellaceae_dash_R_dash_7_dash_group  Ruminococcus  \\\n",
       "#SampleID                                                               \n",
       "sample1                                       -0.000197      0.000224   \n",
       "sample2                                       -0.000258     -0.003550   \n",
       "sample3                                       -0.000197      0.000224   \n",
       "sample4                                       -0.000258     -0.003550   \n",
       "sample5                                        0.001363      0.000268   \n",
       "\n",
       "           Subdoligranulum  Fusicatenibacter  \\\n",
       "#SampleID                                      \n",
       "sample1          -0.000831         -0.000165   \n",
       "sample2           0.000438          0.000343   \n",
       "sample3          -0.000831         -0.000165   \n",
       "sample4           0.000438          0.000343   \n",
       "sample5          -0.000352         -0.000302   \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group  \\\n",
       "#SampleID                                                                              \n",
       "sample1                                            -0.000393                           \n",
       "sample2                                            -0.000618                           \n",
       "sample3                                            -0.000393                           \n",
       "sample4                                            -0.000618                           \n",
       "sample5                                            -0.000610                           \n",
       "\n",
       "           Ruminococcus_dash_sp_dot__dash_DJF_VR52  \n",
       "#SampleID                                           \n",
       "sample1                                  -0.000418  \n",
       "sample2                                   0.000912  \n",
       "sample3                                  -0.000418  \n",
       "sample4                                   0.000912  \n",
       "sample5                                  -0.000948  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_deriv = deepcopy(df_16S_topX_w_meta)\n",
    "for curr_sub in set(df_deriv.Subject):\n",
    "    curr_df = df_deriv[df_deriv.Subject==curr_sub].sort_values(by='Day')\n",
    "    for taxon in bacterial_taxa:\n",
    "        xdata = np.array(curr_df['Day'])\n",
    "        ydata = np.array(curr_df[taxon])\n",
    "        cs = CubicSpline(xdata, ydata)\n",
    "        csd1 = cs.derivative(nu=1)\n",
    "        ydata_d1 = csd1(xdata)\n",
    "        df_deriv.loc[df_deriv.Subject==curr_sub, taxon] = ydata_d1\n",
    "df_deriv.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct regression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-2.38964576, -3.05821695, -2.38964576, ...,  0.09548217,\n",
       "       -0.15072508,  0.09548217])"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ymat should be samples by bacteria\n",
    "Ymat = df_deriv[bacterial_taxa].values\n",
    "Ymat = Ymat.flatten(order='F')\n",
    "Ymat = StandardScaler().fit_transform(Ymat.reshape(-1,1)).reshape(1,-1)[0] # standardize\n",
    "Ymat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3240"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Ymat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.        , 1.        , 0.02482278, ..., 0.        , 0.        ,\n",
       "        0.        ],\n",
       "       [1.        , 0.        , 0.05608459, ..., 0.        , 0.        ,\n",
       "        0.        ],\n",
       "       [1.        , 1.        , 0.1392133 , ..., 0.        , 0.        ,\n",
       "        0.        ],\n",
       "       ...,\n",
       "       [0.        , 0.        , 0.        , ..., 0.02161112, 0.00993983,\n",
       "        0.01975141],\n",
       "       [0.        , 0.        , 0.        , ..., 0.01140879, 0.01369055,\n",
       "        0.01538074],\n",
       "       [0.        , 0.        , 0.        , ..., 0.03487752, 0.00555013,\n",
       "        0.01469523]])"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Xmat = np.zeros(shape=(topX*len(df_deriv.index), (topX+2)*topX))\n",
    "for k in np.arange(topX):\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)] = 1\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+1] = df_deriv.Diet.values\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+2:(k+1)*(topX+2)] = df_16S_topX_w_meta[bacterial_taxa].values\n",
    "Xmat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write stan program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "json_str = '{\\n\"N\" : %d,\\n'%(len(Ymat))\n",
    "\n",
    "# y variable\n",
    "json_str += '\\\"dlogX\\\" : [%s],\\n'%(','.join(list(Ymat.astype(str))))\n",
    "\n",
    "# x variable\n",
    "for k1,c1 in enumerate(bacterial_taxa):\n",
    "    # growth rate\n",
    "    json_str += '\\\"growth_rate_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)].astype(str))))\n",
    "    # diet response\n",
    "    json_str += '\\\"rs_response_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)+1].astype(str))))\n",
    "    # bacterial interactions\n",
    "    for k2,c2 in enumerate(bacterial_taxa):\n",
    "        v = list(Xmat[:,k1*(topX+2)+2+k2].astype(str))\n",
    "        json_str += '\\\"pairwise_interaction_%s_%s\\\" : [%s]'%(c1,c2,','.join(v))\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            json_str += '\\n}'\n",
    "        else:\n",
    "            json_str += ',\\n'\n",
    "#print(json_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.data.json\", \"w\")\n",
    "text_file.write(\"%s\" % json_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# data block\n",
    "model_str = 'data {\\n'\n",
    "model_str += '\\tint<lower=0> N;\\n'\n",
    "model_str += '\\tvector[N] dlogX;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\tvector[N] growth_rate_%s;\\n'%(c1)\n",
    "    model_str += '\\tvector[N] rs_response_%s;\\n'%(c1)\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tvector[N] pairwise_interaction_%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'\n",
    "\n",
    "# parameter block\n",
    "model_str += 'parameters {\\n\\treal<lower=0,upper=1> sigma;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\treal alpha__%s;\\n'%(c1) # growth rate\n",
    "    model_str += '\\treal epsilon__%s;\\n'%(c1) # rs response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\treal beta__%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'       \n",
    "        \n",
    "# model block\n",
    "model_str += 'model {\\n\\tsigma ~ uniform(0,1);\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\talpha__%s ~ normal(0,1);\\n'%(c1) # growth rate\n",
    "    model_str += '\\tepsilon__%s ~ normal(0,1);\\n'%(c1) # rs response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tbeta__%s_%s ~ normal(0,1);\\n'%(c1,c2)\n",
    "model_str += '\\tdlogX ~ normal('\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += 'alpha__%s*growth_rate_%s+'%(c1,c1) # growth rate\n",
    "    model_str += 'epsilon__%s*rs_response_%s+'%(c1,c1) # rs response\n",
    "    for c2 in bacterial_taxa:\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s'%(c1,c2,c1,c2)\n",
    "        else:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s+'%(c1,c2,c1,c2)\n",
    "model_str += ', sigma);\\n}'\n",
    "#print(model_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.stan\", \"w\")\n",
    "text_file.write(\"%s\" % model_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot stan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hpd_grid(sample, alpha=0.05, roundto=2):\n",
    "    \"\"\"Calculate highest posterior density (HPD) of array for given alpha. \n",
    "    The HPD is the minimum width Bayesian credible interval (BCI). \n",
    "    The function works for multimodal distributions, returning more than one mode\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    sample : Numpy array or python list\n",
    "        An array containing MCMC samples\n",
    "    alpha : float\n",
    "        Desired probability of type I error (defaults to 0.05)\n",
    "    roundto: integer\n",
    "        Number of digits after the decimal point for the results\n",
    "    Returns\n",
    "    ----------\n",
    "    hpd: array with the lower \n",
    "          \n",
    "    \"\"\"\n",
    "    sample = np.asarray(sample)\n",
    "    sample = sample[~np.isnan(sample)]\n",
    "    # get upper and lower bounds\n",
    "    l = np.min(sample)\n",
    "    u = np.max(sample)\n",
    "    density = kde.gaussian_kde(sample)\n",
    "    x = np.linspace(l, u, 2000)\n",
    "    y = density.evaluate(x)\n",
    "    #y = density.evaluate(x, l, u) waitting for PR to be accepted\n",
    "    xy_zipped = zip(x, y/np.sum(y))\n",
    "    xy = sorted(xy_zipped, key=lambda x: x[1], reverse=True)\n",
    "    xy_cum_sum = 0\n",
    "    hdv = []\n",
    "    for val in xy:\n",
    "        xy_cum_sum += val[1]\n",
    "        hdv.append(val[0])\n",
    "        if xy_cum_sum >= (1-alpha):\n",
    "            break\n",
    "    hdv.sort()\n",
    "    diff = (u-l)/20  # differences of 5%\n",
    "    hpd = []\n",
    "    hpd.append(round(min(hdv), roundto))\n",
    "    for i in range(1, len(hdv)):\n",
    "        if hdv[i]-hdv[i-1] >= diff:\n",
    "            hpd.append(round(hdv[i-1], roundto))\n",
    "            hpd.append(round(hdv[i], roundto))\n",
    "    hpd.append(round(max(hdv), roundto))\n",
    "    ite = iter(hpd)\n",
    "    hpd = list(zip(ite, ite))\n",
    "    modes = []\n",
    "    for value in hpd:\n",
    "        x_hpd = x[(x > value[0]) & (x < value[1])]\n",
    "        y_hpd = y[(x > value[0]) & (x < value[1])]\n",
    "        modes.append(round(x_hpd[np.argmax(y_hpd)], roundto))\n",
    "    return hpd, x, y, modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"output_%d.csv\"%(i) for i in np.arange(1,4)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>-0.42</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.159749</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "      <td>-0.60</td>\n",
       "      <td>0.32</td>\n",
       "      <td>-0.140</td>\n",
       "      <td>-0.605190</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>-0.66</td>\n",
       "      <td>0.26</td>\n",
       "      <td>-0.200</td>\n",
       "      <td>-0.856932</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>gut_dash_metagenome</td>\n",
       "      <td>-0.14</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.320</td>\n",
       "      <td>1.414249</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...</td>\n",
       "      <td>-0.34</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.115</td>\n",
       "      <td>0.486693</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Type                                               Taxa  Left  \\\n",
       "0  basal_growth_rate                                        Bacteroides -0.42   \n",
       "1  basal_growth_rate                                   Faecalibacterium -0.60   \n",
       "2  basal_growth_rate                                    Lachnospiraceae -0.66   \n",
       "3  basal_growth_rate                                gut_dash_metagenome -0.14   \n",
       "4  basal_growth_rate  Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das... -0.34   \n",
       "\n",
       "   Right  Middle       SNR  Significant  \n",
       "0   0.50   0.040  0.159749        False  \n",
       "1   0.32  -0.140 -0.605190        False  \n",
       "2   0.26  -0.200 -0.856932        False  \n",
       "3   0.78   0.320  1.414249        False  \n",
       "4   0.57   0.115  0.486693        False  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "    \n",
    "# basal growth rate\n",
    "for c in bacterial_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# inulin response\n",
    "for c in bacterial_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['rs_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "\n",
    "# pairwise interactions\n",
    "for c1 in bacterial_taxa:\n",
    "    for c2 in bacterial_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stan_output_summary.to_excel('bayesian_regression_summary_top20.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.405</td>\n",
       "      <td>2.684754</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...</td>\n",
       "      <td>-0.64</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>-0.345</td>\n",
       "      <td>-2.313357</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Anaerostipes</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.400</td>\n",
       "      <td>2.642284</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Bifidobacterium</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.485</td>\n",
       "      <td>3.229673</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Bacteroides')</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.270</td>\n",
       "      <td>3.432835</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Faecalibacterium')</td>\n",
       "      <td>-7.65</td>\n",
       "      <td>-4.91</td>\n",
       "      <td>-6.280</td>\n",
       "      <td>-9.042536</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'gut_dash_metagenome')</td>\n",
       "      <td>0.02</td>\n",
       "      <td>3.46</td>\n",
       "      <td>1.740</td>\n",
       "      <td>1.989138</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Ruminococcus_dash_sp_dot__das...</td>\n",
       "      <td>0.51</td>\n",
       "      <td>4.05</td>\n",
       "      <td>2.280</td>\n",
       "      <td>2.536649</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Blautia')</td>\n",
       "      <td>-3.70</td>\n",
       "      <td>-0.08</td>\n",
       "      <td>-1.890</td>\n",
       "      <td>-2.059112</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Faecalibacterium', 'Faecalibacterium')</td>\n",
       "      <td>1.53</td>\n",
       "      <td>4.27</td>\n",
       "      <td>2.900</td>\n",
       "      <td>4.187724</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bifidobacterium', 'Bifidobacterium')</td>\n",
       "      <td>0.31</td>\n",
       "      <td>3.93</td>\n",
       "      <td>2.120</td>\n",
       "      <td>2.333634</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                               Taxa  \\\n",
       "22            rs_response                                    Lachnospiraceae   \n",
       "24            rs_response  Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...   \n",
       "29            rs_response                                       Anaerostipes   \n",
       "31            rs_response                                    Bifidobacterium   \n",
       "40   pairwise_interaction                     ('Bacteroides', 'Bacteroides')   \n",
       "41   pairwise_interaction                ('Bacteroides', 'Faecalibacterium')   \n",
       "43   pairwise_interaction             ('Bacteroides', 'gut_dash_metagenome')   \n",
       "44   pairwise_interaction  ('Bacteroides', 'Ruminococcus_dash_sp_dot__das...   \n",
       "46   pairwise_interaction                         ('Bacteroides', 'Blautia')   \n",
       "61   pairwise_interaction           ('Faecalibacterium', 'Faecalibacterium')   \n",
       "271  pairwise_interaction             ('Bifidobacterium', 'Bifidobacterium')   \n",
       "\n",
       "     Left  Right  Middle       SNR  Significant  \n",
       "22   0.11   0.70   0.405  2.684754         True  \n",
       "24  -0.64  -0.05  -0.345 -2.313357         True  \n",
       "29   0.10   0.70   0.400  2.642284         True  \n",
       "31   0.19   0.78   0.485  3.229673         True  \n",
       "40   0.54   2.00   1.270  3.432835         True  \n",
       "41  -7.65  -4.91  -6.280 -9.042536         True  \n",
       "43   0.02   3.46   1.740  1.989138         True  \n",
       "44   0.51   4.05   2.280  2.536649         True  \n",
       "46  -3.70  -0.08  -1.890 -2.059112         True  \n",
       "61   1.53   4.27   2.900  4.187724         True  \n",
       "271  0.31   3.93   2.120  2.333634         True  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stan_output_summary = pd.read_excel('bayesian_regression_summary_top20.xlsx', index_col=0)\n",
    "df_stan_output_summary = df_stan_output_summary[df_stan_output_summary.Significant==True]\n",
    "df_stan_output_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
