{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from scipy.interpolate import CubicSpline\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import warnings\n",
    "import itertools\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "import statsmodels.stats.api as sms\n",
    "import scipy.stats.kde as kde"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read 16S data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Absiella_dash_argi</th>\n",
       "      <th>Absiella_dash_dolichum</th>\n",
       "      <th>Acetanaerobacterium</th>\n",
       "      <th>Acetoanaerobium</th>\n",
       "      <th>Acetobacterium</th>\n",
       "      <th>Acidaminococcus_dash_intestini</th>\n",
       "      <th>Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Actinobacillus</th>\n",
       "      <th>Actinobacteria</th>\n",
       "      <th>...</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70</th>\n",
       "      <th>marine_dash_group</th>\n",
       "      <th>marine_dash_sediment_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>rumen_dash_bacterium_dash_YS2</th>\n",
       "      <th>soil_dash_metagenome</th>\n",
       "      <th>swine_dash_fecal_dash_bacterium_dash_SD_dash_Pec5</th>\n",
       "      <th>vadinBE97</th>\n",
       "      <th>wallaby_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6444565</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.032972</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.009201</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443884</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034554</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007985</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444382</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.032344</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008121</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444917</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001194</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443930</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 533 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Absiella_dash_argi  Absiella_dash_dolichum  Acetanaerobacterium  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565                 0.0                     0.0                  0.0   \n",
       "SRR6443884                 0.0                     0.0                  0.0   \n",
       "SRR6444382                 0.0                     0.0                  0.0   \n",
       "SRR6444917                 0.0                     0.0                  0.0   \n",
       "SRR6443930                 0.0                     0.0                  0.0   \n",
       "\n",
       "            Acetoanaerobium  Acetobacterium  Acidaminococcus_dash_intestini  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565              0.0             0.0                             0.0   \n",
       "SRR6443884              0.0             0.0                             0.0   \n",
       "SRR6444382              0.0             0.0                             0.0   \n",
       "SRR6444917              0.0             0.0                             0.0   \n",
       "SRR6443930              0.0             0.0                             0.0   \n",
       "\n",
       "            Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828  \\\n",
       "SampleID                                                             \n",
       "SRR6444565                                                0.0        \n",
       "SRR6443884                                                0.0        \n",
       "SRR6444382                                                0.0        \n",
       "SRR6444917                                                0.0        \n",
       "SRR6443930                                                0.0        \n",
       "\n",
       "            Acinetobacter  Actinobacillus  Actinobacteria  ...  \\\n",
       "SampleID                                                   ...   \n",
       "SRR6444565            0.0             0.0             0.0  ...   \n",
       "SRR6443884            0.0             0.0             0.0  ...   \n",
       "SRR6444382            0.0             0.0             0.0  ...   \n",
       "SRR6444917            0.0             0.0             0.0  ...   \n",
       "SRR6443930            0.0             0.0             0.0  ...   \n",
       "\n",
       "            human_dash_gut_dash_metagenome  \\\n",
       "SampleID                                     \n",
       "SRR6444565                        0.032972   \n",
       "SRR6443884                        0.034554   \n",
       "SRR6444382                        0.032344   \n",
       "SRR6444917                        0.000000   \n",
       "SRR6443930                        0.000000   \n",
       "\n",
       "            iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70  \\\n",
       "SampleID                                                                                          \n",
       "SRR6444565                                                0.0                                     \n",
       "SRR6443884                                                0.0                                     \n",
       "SRR6444382                                                0.0                                     \n",
       "SRR6444917                                                0.0                                     \n",
       "SRR6443930                                                0.0                                     \n",
       "\n",
       "            marine_dash_group  marine_dash_sediment_dash_metagenome  \\\n",
       "SampleID                                                              \n",
       "SRR6444565                0.0                                   0.0   \n",
       "SRR6443884                0.0                                   0.0   \n",
       "SRR6444382                0.0                                   0.0   \n",
       "SRR6444917                0.0                                   0.0   \n",
       "SRR6443930                0.0                                   0.0   \n",
       "\n",
       "            metagenome  rumen_dash_bacterium_dash_YS2  soil_dash_metagenome  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565    0.009201                            0.0                   0.0   \n",
       "SRR6443884    0.007985                            0.0                   0.0   \n",
       "SRR6444382    0.008121                            0.0                   0.0   \n",
       "SRR6444917    0.001194                            0.0                   0.0   \n",
       "SRR6443930    0.000000                            0.0                   0.0   \n",
       "\n",
       "            swine_dash_fecal_dash_bacterium_dash_SD_dash_Pec5  vadinBE97  \\\n",
       "SampleID                                                                   \n",
       "SRR6444565                                                0.0        0.0   \n",
       "SRR6443884                                                0.0        0.0   \n",
       "SRR6444382                                                0.0        0.0   \n",
       "SRR6444917                                                0.0        0.0   \n",
       "SRR6443930                                                0.0        0.0   \n",
       "\n",
       "            wallaby_dash_gut_dash_metagenome  \n",
       "SampleID                                      \n",
       "SRR6444565                               0.0  \n",
       "SRR6443884                               0.0  \n",
       "SRR6444382                               0.0  \n",
       "SRR6444917                               0.0  \n",
       "SRR6443930                               0.0  \n",
       "\n",
       "[5 rows x 533 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = pd.read_excel('./16S_relative_abundance.xlsx', index_col=0)\n",
    "df_16S.columns = [c.replace('/','_slash_').replace(' ','_space_').replace('[','_leftsquarebracket_').replace(']','_rightsquarebracket_').replace('-','_dash_').replace('.','_dot_').replace('(','_leftroundbracket').replace(')','_rightroundbracket_') for c in df_16S.columns]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add pseudocount for zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Absiella_dash_argi</th>\n",
       "      <th>Absiella_dash_dolichum</th>\n",
       "      <th>Acetanaerobacterium</th>\n",
       "      <th>Acetoanaerobium</th>\n",
       "      <th>Acetobacterium</th>\n",
       "      <th>Acidaminococcus_dash_intestini</th>\n",
       "      <th>Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Actinobacillus</th>\n",
       "      <th>Actinobacteria</th>\n",
       "      <th>...</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70</th>\n",
       "      <th>marine_dash_group</th>\n",
       "      <th>marine_dash_sediment_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>rumen_dash_bacterium_dash_YS2</th>\n",
       "      <th>soil_dash_metagenome</th>\n",
       "      <th>swine_dash_fecal_dash_bacterium_dash_SD_dash_Pec5</th>\n",
       "      <th>vadinBE97</th>\n",
       "      <th>wallaby_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6444565</th>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>...</td>\n",
       "      <td>0.032972</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.009201</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443884</th>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034554</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.007985</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444382</th>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>...</td>\n",
       "      <td>0.032344</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.008121</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444917</th>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.001194</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443930</th>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 533 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Absiella_dash_argi  Absiella_dash_dolichum  Acetanaerobacterium  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565            0.000215                0.000215             0.000215   \n",
       "SRR6443884            0.000070                0.000070             0.000070   \n",
       "SRR6444382            0.000167                0.000167             0.000167   \n",
       "SRR6444917            0.000199                0.000199             0.000199   \n",
       "SRR6443930            0.000181                0.000181             0.000181   \n",
       "\n",
       "            Acetoanaerobium  Acetobacterium  Acidaminococcus_dash_intestini  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565         0.000215        0.000215                        0.000215   \n",
       "SRR6443884         0.000070        0.000070                        0.000070   \n",
       "SRR6444382         0.000167        0.000167                        0.000167   \n",
       "SRR6444917         0.000199        0.000199                        0.000199   \n",
       "SRR6443930         0.000181        0.000181                        0.000181   \n",
       "\n",
       "            Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828  \\\n",
       "SampleID                                                             \n",
       "SRR6444565                                           0.000215        \n",
       "SRR6443884                                           0.000070        \n",
       "SRR6444382                                           0.000167        \n",
       "SRR6444917                                           0.000199        \n",
       "SRR6443930                                           0.000181        \n",
       "\n",
       "            Acinetobacter  Actinobacillus  Actinobacteria  ...  \\\n",
       "SampleID                                                   ...   \n",
       "SRR6444565       0.000215        0.000215        0.000215  ...   \n",
       "SRR6443884       0.000070        0.000070        0.000070  ...   \n",
       "SRR6444382       0.000167        0.000167        0.000167  ...   \n",
       "SRR6444917       0.000199        0.000199        0.000199  ...   \n",
       "SRR6443930       0.000181        0.000181        0.000181  ...   \n",
       "\n",
       "            human_dash_gut_dash_metagenome  \\\n",
       "SampleID                                     \n",
       "SRR6444565                        0.032972   \n",
       "SRR6443884                        0.034554   \n",
       "SRR6444382                        0.032344   \n",
       "SRR6444917                        0.000199   \n",
       "SRR6443930                        0.000181   \n",
       "\n",
       "            iron_dash_reducing_dash_bacterium_dash_enrichment_dash_culture_dash_clone_dash_HN70  \\\n",
       "SampleID                                                                                          \n",
       "SRR6444565                                           0.000215                                     \n",
       "SRR6443884                                           0.000070                                     \n",
       "SRR6444382                                           0.000167                                     \n",
       "SRR6444917                                           0.000199                                     \n",
       "SRR6443930                                           0.000181                                     \n",
       "\n",
       "            marine_dash_group  marine_dash_sediment_dash_metagenome  \\\n",
       "SampleID                                                              \n",
       "SRR6444565           0.000215                              0.000215   \n",
       "SRR6443884           0.000070                              0.000070   \n",
       "SRR6444382           0.000167                              0.000167   \n",
       "SRR6444917           0.000199                              0.000199   \n",
       "SRR6443930           0.000181                              0.000181   \n",
       "\n",
       "            metagenome  rumen_dash_bacterium_dash_YS2  soil_dash_metagenome  \\\n",
       "SampleID                                                                      \n",
       "SRR6444565    0.009201                       0.000215              0.000215   \n",
       "SRR6443884    0.007985                       0.000070              0.000070   \n",
       "SRR6444382    0.008121                       0.000167              0.000167   \n",
       "SRR6444917    0.001194                       0.000199              0.000199   \n",
       "SRR6443930    0.000181                       0.000181              0.000181   \n",
       "\n",
       "            swine_dash_fecal_dash_bacterium_dash_SD_dash_Pec5  vadinBE97  \\\n",
       "SampleID                                                                   \n",
       "SRR6444565                                           0.000215   0.000215   \n",
       "SRR6443884                                           0.000070   0.000070   \n",
       "SRR6444382                                           0.000167   0.000167   \n",
       "SRR6444917                                           0.000199   0.000199   \n",
       "SRR6443930                                           0.000181   0.000181   \n",
       "\n",
       "            wallaby_dash_gut_dash_metagenome  \n",
       "SampleID                                      \n",
       "SRR6444565                          0.000215  \n",
       "SRR6443884                          0.000070  \n",
       "SRR6444382                          0.000167  \n",
       "SRR6444917                          0.000199  \n",
       "SRR6443930                          0.000181  \n",
       "\n",
       "[5 rows x 533 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample_id in df_16S.index:\n",
    "    sample = np.array(df_16S.loc[sample_id])\n",
    "    minval = np.min(sample[np.nonzero(sample)]) # minimum non-zero value\n",
    "    sample[sample==0] = minval\n",
    "    df_16S.loc[sample_id] = sample\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata for maize starch and control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6444565</th>\n",
       "      <td>U222</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443884</th>\n",
       "      <td>U222</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444382</th>\n",
       "      <td>U222</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444917</th>\n",
       "      <td>U225</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443930</th>\n",
       "      <td>U225</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Subject_ID  status  Day  Diet\n",
       "#SampleID                               \n",
       "SRR6444565       U222  before    0     0\n",
       "SRR6443884       U222  before    0     0\n",
       "SRR6444382       U222  before    0     0\n",
       "SRR6444917       U225  before    0     0\n",
       "SRR6443930       U225  before    0     0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv('./metadata.txt', sep='\\t', index_col=0)\n",
    "df_meta = df_meta[['Group','Subject_ID','status']]\n",
    "df_meta = df_meta[(df_meta.Group=='himaize') | (df_meta.Group=='accessible')]\n",
    "df_meta['Day'] = [0 if s=='before' else 21 for s in df_meta.status]\n",
    "df_meta['Diet'] = [1 if g=='himaize' else 0 for g in df_meta['Group']]\n",
    "df_meta = df_meta.drop('Group', axis=1)\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 20 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Prevotella</th>\n",
       "      <th>Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Parabacteroides</th>\n",
       "      <th>Bacteroides_dash_plebeius_dash_DSM_dash_17135</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Parasutterella_dash_excrementihominis</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>Bacteroides_dash_coprocola</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Lachnospira</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>Subdoligranulum</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6444565</th>\n",
       "      <td>0.197580</td>\n",
       "      <td>0.136827</td>\n",
       "      <td>0.024022</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.037984</td>\n",
       "      <td>0.028962</td>\n",
       "      <td>0.007446</td>\n",
       "      <td>0.114631</td>\n",
       "      <td>0.018580</td>\n",
       "      <td>0.036051</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.027852</td>\n",
       "      <td>0.006623</td>\n",
       "      <td>0.006158</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.009201</td>\n",
       "      <td>0.010847</td>\n",
       "      <td>0.003508</td>\n",
       "      <td>0.005907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443884</th>\n",
       "      <td>0.243131</td>\n",
       "      <td>0.099128</td>\n",
       "      <td>0.017050</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.037064</td>\n",
       "      <td>0.032008</td>\n",
       "      <td>0.006764</td>\n",
       "      <td>0.125837</td>\n",
       "      <td>0.014191</td>\n",
       "      <td>0.034100</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.022873</td>\n",
       "      <td>0.005335</td>\n",
       "      <td>0.005230</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.007985</td>\n",
       "      <td>0.013982</td>\n",
       "      <td>0.002789</td>\n",
       "      <td>0.004079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444382</th>\n",
       "      <td>0.214061</td>\n",
       "      <td>0.110743</td>\n",
       "      <td>0.028395</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.037600</td>\n",
       "      <td>0.031565</td>\n",
       "      <td>0.006981</td>\n",
       "      <td>0.115360</td>\n",
       "      <td>0.020525</td>\n",
       "      <td>0.032734</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.029730</td>\n",
       "      <td>0.006869</td>\n",
       "      <td>0.007231</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.008121</td>\n",
       "      <td>0.011263</td>\n",
       "      <td>0.003560</td>\n",
       "      <td>0.005006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444917</th>\n",
       "      <td>0.578520</td>\n",
       "      <td>0.063804</td>\n",
       "      <td>0.117987</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000630</td>\n",
       "      <td>0.037393</td>\n",
       "      <td>0.003550</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.014168</td>\n",
       "      <td>0.000630</td>\n",
       "      <td>0.026942</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.003782</td>\n",
       "      <td>0.000299</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.001194</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443930</th>\n",
       "      <td>0.610679</td>\n",
       "      <td>0.055143</td>\n",
       "      <td>0.050196</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.079698</td>\n",
       "      <td>0.004404</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.006154</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.023409</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Bacteroides  Faecalibacterium  Lachnospiraceae  Prevotella  \\\n",
       "#SampleID                                                                \n",
       "SRR6444565     0.197580          0.136827         0.024022    0.000215   \n",
       "SRR6443884     0.243131          0.099128         0.017050    0.000070   \n",
       "SRR6444382     0.214061          0.110743         0.028395    0.000167   \n",
       "SRR6444917     0.578520          0.063804         0.117987    0.000199   \n",
       "SRR6443930     0.610679          0.055143         0.050196    0.000181   \n",
       "\n",
       "            Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_  \\\n",
       "#SampleID                                                                                           \n",
       "SRR6444565                                           0.000215                                       \n",
       "SRR6443884                                           0.000070                                       \n",
       "SRR6444382                                           0.000167                                       \n",
       "SRR6444917                                           0.000199                                       \n",
       "SRR6443930                                           0.000181                                       \n",
       "\n",
       "            Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  Alistipes  \\\n",
       "#SampleID                                                                   \n",
       "SRR6444565                                           0.037984    0.028962   \n",
       "SRR6443884                                           0.037064    0.032008   \n",
       "SRR6444382                                           0.037600    0.031565   \n",
       "SRR6444917                                           0.000630    0.037393   \n",
       "SRR6443930                                           0.000181    0.079698   \n",
       "\n",
       "            Parabacteroides  Bacteroides_dash_plebeius_dash_DSM_dash_17135  \\\n",
       "#SampleID                                                                    \n",
       "SRR6444565         0.007446                                       0.114631   \n",
       "SRR6443884         0.006764                                       0.125837   \n",
       "SRR6444382         0.006981                                       0.115360   \n",
       "SRR6444917         0.003550                                       0.000199   \n",
       "SRR6443930         0.004404                                       0.000181   \n",
       "\n",
       "             Blautia  gut_dash_metagenome  \\\n",
       "#SampleID                                   \n",
       "SRR6444565  0.018580             0.036051   \n",
       "SRR6443884  0.014191             0.034100   \n",
       "SRR6444382  0.020525             0.032734   \n",
       "SRR6444917  0.014168             0.000630   \n",
       "SRR6443930  0.006154             0.000181   \n",
       "\n",
       "            Parasutterella_dash_excrementihominis  Roseburia  Ruminococcus  \\\n",
       "#SampleID                                                                    \n",
       "SRR6444565                               0.000215   0.027852      0.006623   \n",
       "SRR6443884                               0.000070   0.022873      0.005335   \n",
       "SRR6444382                               0.000167   0.029730      0.006869   \n",
       "SRR6444917                               0.026942   0.000199      0.003782   \n",
       "SRR6443930                               0.023409   0.000181      0.000181   \n",
       "\n",
       "            Anaerostipes  Bacteroides_dash_coprocola  metagenome  Lachnospira  \\\n",
       "#SampleID                                                                       \n",
       "SRR6444565      0.006158                    0.000215    0.009201     0.010847   \n",
       "SRR6443884      0.005230                    0.000070    0.007985     0.013982   \n",
       "SRR6444382      0.007231                    0.000167    0.008121     0.011263   \n",
       "SRR6444917      0.000299                    0.000199    0.001194     0.000199   \n",
       "SRR6443930      0.000181                    0.000181    0.000181     0.000181   \n",
       "\n",
       "            Fusicatenibacter  Subdoligranulum  \n",
       "#SampleID                                      \n",
       "SRR6444565          0.003508         0.005907  \n",
       "SRR6443884          0.002789         0.004079  \n",
       "SRR6444382          0.003560         0.005006  \n",
       "SRR6444917          0.000199         0.000199  \n",
       "SRR6443930          0.000181         0.000181  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 20\n",
    "df_16S_T = df_16S.loc[df_meta.index].T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Bacteroides',\n",
       " 'Faecalibacterium',\n",
       " 'Lachnospiraceae',\n",
       " 'Prevotella',\n",
       " 'Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_',\n",
       " 'Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57',\n",
       " 'Alistipes',\n",
       " 'Parabacteroides',\n",
       " 'Bacteroides_dash_plebeius_dash_DSM_dash_17135',\n",
       " 'Blautia',\n",
       " 'gut_dash_metagenome',\n",
       " 'Parasutterella_dash_excrementihominis',\n",
       " 'Roseburia',\n",
       " 'Ruminococcus',\n",
       " 'Anaerostipes',\n",
       " 'Bacteroides_dash_coprocola',\n",
       " 'metagenome',\n",
       " 'Lachnospira',\n",
       " 'Fusicatenibacter',\n",
       " 'Subdoligranulum']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_16S_topX.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Prevotella</th>\n",
       "      <th>Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>...</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Parasutterella_dash_excrementihominis</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>Bacteroides_dash_coprocola</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Lachnospira</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>Subdoligranulum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U043</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.566560</td>\n",
       "      <td>0.181597</td>\n",
       "      <td>0.014493</td>\n",
       "      <td>0.000411</td>\n",
       "      <td>0.000411</td>\n",
       "      <td>0.048429</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003573</td>\n",
       "      <td>0.000411</td>\n",
       "      <td>0.030010</td>\n",
       "      <td>0.000411</td>\n",
       "      <td>0.015531</td>\n",
       "      <td>0.000411</td>\n",
       "      <td>0.000411</td>\n",
       "      <td>0.000411</td>\n",
       "      <td>0.000411</td>\n",
       "      <td>0.000411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U043</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.531151</td>\n",
       "      <td>0.161992</td>\n",
       "      <td>0.016003</td>\n",
       "      <td>0.004825</td>\n",
       "      <td>0.002562</td>\n",
       "      <td>0.068180</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002982</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.054666</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.010017</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.000241</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.000185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U044</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.067995</td>\n",
       "      <td>0.090265</td>\n",
       "      <td>0.034407</td>\n",
       "      <td>0.146068</td>\n",
       "      <td>0.008834</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>...</td>\n",
       "      <td>0.076503</td>\n",
       "      <td>0.004303</td>\n",
       "      <td>0.003725</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.013594</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.003754</td>\n",
       "      <td>0.003344</td>\n",
       "      <td>0.012889</td>\n",
       "      <td>0.015563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U044</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.053179</td>\n",
       "      <td>0.064275</td>\n",
       "      <td>0.078139</td>\n",
       "      <td>0.153087</td>\n",
       "      <td>0.010942</td>\n",
       "      <td>0.000443</td>\n",
       "      <td>...</td>\n",
       "      <td>0.094488</td>\n",
       "      <td>0.004591</td>\n",
       "      <td>0.003187</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>0.012505</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>0.012696</td>\n",
       "      <td>0.003020</td>\n",
       "      <td>0.009586</td>\n",
       "      <td>0.013215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U045</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.094715</td>\n",
       "      <td>0.093115</td>\n",
       "      <td>0.057557</td>\n",
       "      <td>0.136424</td>\n",
       "      <td>0.132611</td>\n",
       "      <td>0.054321</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006494</td>\n",
       "      <td>0.000201</td>\n",
       "      <td>0.000201</td>\n",
       "      <td>0.063839</td>\n",
       "      <td>0.025264</td>\n",
       "      <td>0.000201</td>\n",
       "      <td>0.034321</td>\n",
       "      <td>0.000201</td>\n",
       "      <td>0.021834</td>\n",
       "      <td>0.003986</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Subject_ID  status  Day  Diet  Bacteroides  Faecalibacterium  \\\n",
       "0       U043  before    0     1     0.566560          0.181597   \n",
       "1       U043  during   21     1     0.531151          0.161992   \n",
       "2       U044  before    0     1     0.067995          0.090265   \n",
       "3       U044  during   21     1     0.053179          0.064275   \n",
       "4       U045  before    0     1     0.094715          0.093115   \n",
       "\n",
       "   Lachnospiraceae  Prevotella  \\\n",
       "0         0.014493    0.000411   \n",
       "1         0.016003    0.004825   \n",
       "2         0.034407    0.146068   \n",
       "3         0.078139    0.153087   \n",
       "4         0.057557    0.136424   \n",
       "\n",
       "   Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_  \\\n",
       "0                                           0.000411                                       \n",
       "1                                           0.002562                                       \n",
       "2                                           0.008834                                       \n",
       "3                                           0.010942                                       \n",
       "4                                           0.132611                                       \n",
       "\n",
       "   Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  ...  \\\n",
       "0                                           0.048429   ...   \n",
       "1                                           0.068180   ...   \n",
       "2                                           0.000273   ...   \n",
       "3                                           0.000443   ...   \n",
       "4                                           0.054321   ...   \n",
       "\n",
       "   gut_dash_metagenome  Parasutterella_dash_excrementihominis  Roseburia  \\\n",
       "0             0.003573                               0.000411   0.030010   \n",
       "1             0.002982                               0.000185   0.054666   \n",
       "2             0.076503                               0.004303   0.003725   \n",
       "3             0.094488                               0.004591   0.003187   \n",
       "4             0.006494                               0.000201   0.000201   \n",
       "\n",
       "   Ruminococcus  Anaerostipes  Bacteroides_dash_coprocola  metagenome  \\\n",
       "0      0.000411      0.015531                    0.000411    0.000411   \n",
       "1      0.000185      0.010017                    0.000185    0.000241   \n",
       "2      0.000220      0.013594                    0.000220    0.003754   \n",
       "3      0.000151      0.012505                    0.000151    0.012696   \n",
       "4      0.063839      0.025264                    0.000201    0.034321   \n",
       "\n",
       "   Lachnospira  Fusicatenibacter  Subdoligranulum  \n",
       "0     0.000411          0.000411         0.000411  \n",
       "1     0.000185          0.000185         0.000185  \n",
       "2     0.003344          0.012889         0.015563  \n",
       "3     0.003020          0.009586         0.013215  \n",
       "4     0.000201          0.021834         0.003986  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX_w_meta = pd.merge(df_meta, df_16S_topX/df_16S_topX.max().max(), left_index=True, right_index=True, how='inner')\n",
    "df_16S_topX_w_meta = df_16S_topX_w_meta.groupby(['Subject_ID','status']).agg(np.mean).reset_index()\n",
    "df_16S_topX_w_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate log-derivatives (dlog(B)/dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Prevotella</th>\n",
       "      <th>Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>...</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Parasutterella_dash_excrementihominis</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>Bacteroides_dash_coprocola</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Lachnospira</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>Subdoligranulum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U043</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.001686</td>\n",
       "      <td>-0.000934</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000210</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.000941</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000028</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>-0.000263</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>-0.000008</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>-0.000011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U043</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.001686</td>\n",
       "      <td>-0.000934</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000210</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.000941</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000028</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>-0.000263</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>-0.000008</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>-0.000011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U044</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.000706</td>\n",
       "      <td>-0.001238</td>\n",
       "      <td>0.002082</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000856</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>-0.000026</td>\n",
       "      <td>-0.000003</td>\n",
       "      <td>-0.000052</td>\n",
       "      <td>-0.000003</td>\n",
       "      <td>0.000426</td>\n",
       "      <td>-0.000015</td>\n",
       "      <td>-0.000157</td>\n",
       "      <td>-0.000112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U044</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.000706</td>\n",
       "      <td>-0.001238</td>\n",
       "      <td>0.002082</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000856</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>-0.000026</td>\n",
       "      <td>-0.000003</td>\n",
       "      <td>-0.000052</td>\n",
       "      <td>-0.000003</td>\n",
       "      <td>0.000426</td>\n",
       "      <td>-0.000015</td>\n",
       "      <td>-0.000157</td>\n",
       "      <td>-0.000112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U045</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.000367</td>\n",
       "      <td>0.000556</td>\n",
       "      <td>-0.000868</td>\n",
       "      <td>0.000506</td>\n",
       "      <td>0.000669</td>\n",
       "      <td>0.001538</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000109</td>\n",
       "      <td>-0.000004</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.000737</td>\n",
       "      <td>-0.000446</td>\n",
       "      <td>-0.000004</td>\n",
       "      <td>-0.000621</td>\n",
       "      <td>-0.000004</td>\n",
       "      <td>-0.000488</td>\n",
       "      <td>-0.000043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Subject_ID  status  Day  Diet  Bacteroides  Faecalibacterium  \\\n",
       "0       U043  before    0     1    -0.001686         -0.000934   \n",
       "1       U043  during   21     1    -0.001686         -0.000934   \n",
       "2       U044  before    0     1    -0.000706         -0.001238   \n",
       "3       U044  during   21     1    -0.000706         -0.001238   \n",
       "4       U045  before    0     1    -0.000367          0.000556   \n",
       "\n",
       "   Lachnospiraceae  Prevotella  \\\n",
       "0         0.000072    0.000210   \n",
       "1         0.000072    0.000210   \n",
       "2         0.002082    0.000334   \n",
       "3         0.002082    0.000334   \n",
       "4        -0.000868    0.000506   \n",
       "\n",
       "   Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_  \\\n",
       "0                                           0.000102                                       \n",
       "1                                           0.000102                                       \n",
       "2                                           0.000100                                       \n",
       "3                                           0.000100                                       \n",
       "4                                           0.000669                                       \n",
       "\n",
       "   Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  ...  \\\n",
       "0                                           0.000941   ...   \n",
       "1                                           0.000941   ...   \n",
       "2                                           0.000008   ...   \n",
       "3                                           0.000008   ...   \n",
       "4                                           0.001538   ...   \n",
       "\n",
       "   gut_dash_metagenome  Parasutterella_dash_excrementihominis  Roseburia  \\\n",
       "0            -0.000028                              -0.000011   0.001174   \n",
       "1            -0.000028                              -0.000011   0.001174   \n",
       "2             0.000856                               0.000014  -0.000026   \n",
       "3             0.000856                               0.000014  -0.000026   \n",
       "4            -0.000109                              -0.000004   0.000067   \n",
       "\n",
       "   Ruminococcus  Anaerostipes  Bacteroides_dash_coprocola  metagenome  \\\n",
       "0     -0.000011     -0.000263                   -0.000011   -0.000008   \n",
       "1     -0.000011     -0.000263                   -0.000011   -0.000008   \n",
       "2     -0.000003     -0.000052                   -0.000003    0.000426   \n",
       "3     -0.000003     -0.000052                   -0.000003    0.000426   \n",
       "4      0.000737     -0.000446                   -0.000004   -0.000621   \n",
       "\n",
       "   Lachnospira  Fusicatenibacter  Subdoligranulum  \n",
       "0    -0.000011         -0.000011        -0.000011  \n",
       "1    -0.000011         -0.000011        -0.000011  \n",
       "2    -0.000015         -0.000157        -0.000112  \n",
       "3    -0.000015         -0.000157        -0.000112  \n",
       "4    -0.000004         -0.000488        -0.000043  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_deriv = deepcopy(df_16S_topX_w_meta)\n",
    "for curr_sub in set(df_deriv.Subject_ID):\n",
    "    curr_df = df_deriv[df_deriv.Subject_ID==curr_sub].sort_values(by='Day')\n",
    "    for taxon in bacterial_taxa:\n",
    "        xdata = np.array(curr_df['Day'])\n",
    "        ydata = np.array(curr_df[taxon])\n",
    "        cs = CubicSpline(xdata, ydata)\n",
    "        csd1 = cs.derivative(nu=1)\n",
    "        ydata_d1 = csd1(xdata)\n",
    "        df_deriv.loc[df_deriv.Subject_ID==curr_sub, taxon] = ydata_d1\n",
    "df_deriv.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct regression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-1.24868729, -1.24868729, -0.5259887 , ...,  0.12887521,\n",
       "        0.35415503,  0.35415503])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ymat should be samples by bacteria\n",
    "Ymat = df_deriv[bacterial_taxa].values\n",
    "Ymat = Ymat.flatten(order='F')\n",
    "Ymat = StandardScaler().fit_transform(Ymat.reshape(-1,1)).reshape(1,-1)[0] # standardize\n",
    "Ymat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3280"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Ymat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.00000000e+00, 1.00000000e+00, 5.66560062e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 1.00000000e+00, 5.31150643e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 1.00000000e+00, 6.79946621e-02, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       ...,\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        1.29988210e-03, 4.69189295e-03, 3.10771634e-02],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        2.84350475e-03, 5.87330579e-04, 1.78306210e-02],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        4.29800958e-03, 4.90891772e-03, 2.80945526e-02]])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Xmat = np.zeros(shape=(topX*len(df_deriv.index), (topX+2)*topX))\n",
    "for k in np.arange(topX):\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)] = 1\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+1] = df_deriv.Diet.values\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+2:(k+1)*(topX+2)] = df_16S_topX_w_meta[bacterial_taxa].values\n",
    "Xmat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write stan program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "json_str = '{\\n\"N\" : %d,\\n'%(len(Ymat))\n",
    "\n",
    "# y variable\n",
    "json_str += '\\\"dlogX\\\" : [%s],\\n'%(','.join(list(Ymat.astype(str))))\n",
    "\n",
    "# x variable\n",
    "for k1,c1 in enumerate(bacterial_taxa):\n",
    "    # growth rate\n",
    "    json_str += '\\\"growth_rate_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)].astype(str))))\n",
    "    # diet response\n",
    "    json_str += '\\\"inulin_response_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)+1].astype(str))))\n",
    "    # bacterial interactions\n",
    "    for k2,c2 in enumerate(bacterial_taxa):\n",
    "        v = list(Xmat[:,k1*(topX+2)+2+k2].astype(str))\n",
    "        json_str += '\\\"pairwise_interaction_%s_%s\\\" : [%s]'%(c1,c2,','.join(v))\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            json_str += '\\n}'\n",
    "        else:\n",
    "            json_str += ',\\n'\n",
    "#print(json_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa_2/mice_scfa_2.data.json\", \"w\")\n",
    "text_file.write(\"%s\" % json_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# data block\n",
    "model_str = 'data {\\n'\n",
    "model_str += '\\tint<lower=0> N;\\n'\n",
    "model_str += '\\tvector[N] dlogX;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\tvector[N] growth_rate_%s;\\n'%(c1)\n",
    "    model_str += '\\tvector[N] inulin_response_%s;\\n'%(c1)\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tvector[N] pairwise_interaction_%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'\n",
    "\n",
    "# parameter block\n",
    "model_str += 'parameters {\\n\\treal<lower=0,upper=1> sigma;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\treal alpha__%s;\\n'%(c1) # growth rate\n",
    "    model_str += '\\treal epsilon__%s;\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\treal beta__%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'       \n",
    "        \n",
    "# model block\n",
    "model_str += 'model {\\n\\tsigma ~ uniform(0,1);\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\talpha__%s ~ normal(0,1);\\n'%(c1) # growth rate\n",
    "    model_str += '\\tepsilon__%s ~ normal(0,1);\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tbeta__%s_%s ~ normal(0,1);\\n'%(c1,c2)\n",
    "model_str += '\\tdlogX ~ normal('\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += 'alpha__%s*growth_rate_%s+'%(c1,c1) # growth rate\n",
    "    model_str += 'epsilon__%s*inulin_response_%s+'%(c1,c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s'%(c1,c2,c1,c2)\n",
    "        else:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s+'%(c1,c2,c1,c2)\n",
    "model_str += ', sigma);\\n}'\n",
    "#print(model_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa_2/mice_scfa_2.stan\", \"w\")\n",
    "text_file.write(\"%s\" % model_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot stan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hpd_grid(sample, alpha=0.05, roundto=2):\n",
    "    \"\"\"Calculate highest posterior density (HPD) of array for given alpha. \n",
    "    The HPD is the minimum width Bayesian credible interval (BCI). \n",
    "    The function works for multimodal distributions, returning more than one mode\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    sample : Numpy array or python list\n",
    "        An array containing MCMC samples\n",
    "    alpha : float\n",
    "        Desired probability of type I error (defaults to 0.05)\n",
    "    roundto: integer\n",
    "        Number of digits after the decimal point for the results\n",
    "    Returns\n",
    "    ----------\n",
    "    hpd: array with the lower \n",
    "          \n",
    "    \"\"\"\n",
    "    sample = np.asarray(sample)\n",
    "    sample = sample[~np.isnan(sample)]\n",
    "    # get upper and lower bounds\n",
    "    l = np.min(sample)\n",
    "    u = np.max(sample)\n",
    "    density = kde.gaussian_kde(sample)\n",
    "    x = np.linspace(l, u, 2000)\n",
    "    y = density.evaluate(x)\n",
    "    #y = density.evaluate(x, l, u) waitting for PR to be accepted\n",
    "    xy_zipped = zip(x, y/np.sum(y))\n",
    "    xy = sorted(xy_zipped, key=lambda x: x[1], reverse=True)\n",
    "    xy_cum_sum = 0\n",
    "    hdv = []\n",
    "    for val in xy:\n",
    "        xy_cum_sum += val[1]\n",
    "        hdv.append(val[0])\n",
    "        if xy_cum_sum >= (1-alpha):\n",
    "            break\n",
    "    hdv.sort()\n",
    "    diff = (u-l)/20  # differences of 5%\n",
    "    hpd = []\n",
    "    hpd.append(round(min(hdv), roundto))\n",
    "    for i in range(1, len(hdv)):\n",
    "        if hdv[i]-hdv[i-1] >= diff:\n",
    "            hpd.append(round(hdv[i-1], roundto))\n",
    "            hpd.append(round(hdv[i], roundto))\n",
    "    hpd.append(round(max(hdv), roundto))\n",
    "    ite = iter(hpd)\n",
    "    hpd = list(zip(ite, ite))\n",
    "    modes = []\n",
    "    for value in hpd:\n",
    "        x_hpd = x[(x > value[0]) & (x < value[1])]\n",
    "        y_hpd = y[(x > value[0]) & (x < value[1])]\n",
    "        modes.append(round(x_hpd[np.argmax(y_hpd)], roundto))\n",
    "    return hpd, x, y, modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"output_%d_rs.csv\"%(i) for i in np.arange(1,4)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>-1.14</td>\n",
       "      <td>-0.11</td>\n",
       "      <td>-0.625</td>\n",
       "      <td>-2.409779</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "      <td>-0.66</td>\n",
       "      <td>0.36</td>\n",
       "      <td>-0.150</td>\n",
       "      <td>-0.596851</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>-0.42</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.360303</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>-0.39</td>\n",
       "      <td>0.63</td>\n",
       "      <td>0.120</td>\n",
       "      <td>0.454670</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Trichuris_dash_trichiura_dash__leftroundbracke...</td>\n",
       "      <td>-0.68</td>\n",
       "      <td>0.33</td>\n",
       "      <td>-0.175</td>\n",
       "      <td>-0.646623</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Type                                               Taxa  Left  \\\n",
       "0  basal_growth_rate                                        Bacteroides -1.14   \n",
       "1  basal_growth_rate                                   Faecalibacterium -0.66   \n",
       "2  basal_growth_rate                                    Lachnospiraceae -0.42   \n",
       "3  basal_growth_rate                                         Prevotella -0.39   \n",
       "4  basal_growth_rate  Trichuris_dash_trichiura_dash__leftroundbracke... -0.68   \n",
       "\n",
       "   Right  Middle       SNR  Significant  \n",
       "0  -0.11  -0.625 -2.409779         True  \n",
       "1   0.36  -0.150 -0.596851        False  \n",
       "2   0.61   0.095  0.360303        False  \n",
       "3   0.63   0.120  0.454670        False  \n",
       "4   0.33  -0.175 -0.646623        False  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "    \n",
    "# basal growth rate\n",
    "for c in bacterial_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# inulin response\n",
    "for c in bacterial_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['rs_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "\n",
    "# pairwise interactions\n",
    "for c1 in bacterial_taxa:\n",
    "    for c2 in bacterial_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stan_output_summary.to_excel('bayesian_regression_summary_top20_rs.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>-1.14</td>\n",
       "      <td>-0.11</td>\n",
       "      <td>-0.625</td>\n",
       "      <td>-2.409779</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...</td>\n",
       "      <td>0.10</td>\n",
       "      <td>1.13</td>\n",
       "      <td>0.615</td>\n",
       "      <td>2.367750</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides_dash_plebeius_dash_DSM_dash_17135</td>\n",
       "      <td>-1.05</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>-0.535</td>\n",
       "      <td>-2.071025</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.515</td>\n",
       "      <td>3.448608</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Bacteroides')</td>\n",
       "      <td>1.07</td>\n",
       "      <td>2.51</td>\n",
       "      <td>1.790</td>\n",
       "      <td>4.897480</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Faecalibacterium')</td>\n",
       "      <td>0.51</td>\n",
       "      <td>3.99</td>\n",
       "      <td>2.250</td>\n",
       "      <td>2.558270</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Lachnospiraceae')</td>\n",
       "      <td>-4.60</td>\n",
       "      <td>-1.04</td>\n",
       "      <td>-2.820</td>\n",
       "      <td>-3.107074</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Prevotella')</td>\n",
       "      <td>-2.47</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>-1.335</td>\n",
       "      <td>-2.315486</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Trichuris_dash_trichiura_dash...</td>\n",
       "      <td>1.42</td>\n",
       "      <td>3.85</td>\n",
       "      <td>2.635</td>\n",
       "      <td>4.293403</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Ruminococcus_dash_sp_dot__das...</td>\n",
       "      <td>-4.96</td>\n",
       "      <td>-1.38</td>\n",
       "      <td>-3.170</td>\n",
       "      <td>-3.496771</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Bacteroides_dash_plebeius_das...</td>\n",
       "      <td>0.42</td>\n",
       "      <td>3.61</td>\n",
       "      <td>2.015</td>\n",
       "      <td>2.507289</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Parasutterella_dash_excrement...</td>\n",
       "      <td>-3.87</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>-1.950</td>\n",
       "      <td>-2.028178</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Bacteroides_dash_coprocola')</td>\n",
       "      <td>-3.28</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>-1.645</td>\n",
       "      <td>-1.969291</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Faecalibacterium', 'Bacteroides_dash_plebeiu...</td>\n",
       "      <td>0.15</td>\n",
       "      <td>3.39</td>\n",
       "      <td>1.770</td>\n",
       "      <td>2.159146</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Prevotella', 'Prevotella')</td>\n",
       "      <td>5.02</td>\n",
       "      <td>7.30</td>\n",
       "      <td>6.160</td>\n",
       "      <td>10.663421</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Prevotella', 'Trichuris_dash_trichiura_dash_...</td>\n",
       "      <td>-3.58</td>\n",
       "      <td>-1.14</td>\n",
       "      <td>-2.360</td>\n",
       "      <td>-3.792889</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Trichuris_dash_trichiura_dash__leftroundbrac...</td>\n",
       "      <td>-3.33</td>\n",
       "      <td>-0.88</td>\n",
       "      <td>-2.105</td>\n",
       "      <td>-3.374894</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Trichuris_dash_trichiura_dash__leftroundbrac...</td>\n",
       "      <td>-3.77</td>\n",
       "      <td>-0.55</td>\n",
       "      <td>-2.160</td>\n",
       "      <td>-2.640688</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_d...</td>\n",
       "      <td>-1.90</td>\n",
       "      <td>-0.43</td>\n",
       "      <td>-1.165</td>\n",
       "      <td>-3.131135</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_d...</td>\n",
       "      <td>-2.85</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>-1.710</td>\n",
       "      <td>-2.992284</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_d...</td>\n",
       "      <td>0.35</td>\n",
       "      <td>3.94</td>\n",
       "      <td>2.145</td>\n",
       "      <td>2.358529</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_plebeius_dash_DSM_dash_1713...</td>\n",
       "      <td>-5.43</td>\n",
       "      <td>-2.23</td>\n",
       "      <td>-3.830</td>\n",
       "      <td>-4.727823</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_coprocola', 'Bacteroides_da...</td>\n",
       "      <td>-4.96</td>\n",
       "      <td>-1.76</td>\n",
       "      <td>-3.360</td>\n",
       "      <td>-4.136909</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides_dash_coprocola', 'Bacteroides_da...</td>\n",
       "      <td>1.92</td>\n",
       "      <td>5.23</td>\n",
       "      <td>3.575</td>\n",
       "      <td>4.245876</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                               Taxa  \\\n",
       "0       basal_growth_rate                                        Bacteroides   \n",
       "5       basal_growth_rate  Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...   \n",
       "8       basal_growth_rate      Bacteroides_dash_plebeius_dash_DSM_dash_17135   \n",
       "25            rs_response  Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...   \n",
       "40   pairwise_interaction                     ('Bacteroides', 'Bacteroides')   \n",
       "41   pairwise_interaction                ('Bacteroides', 'Faecalibacterium')   \n",
       "42   pairwise_interaction                 ('Bacteroides', 'Lachnospiraceae')   \n",
       "43   pairwise_interaction                      ('Bacteroides', 'Prevotella')   \n",
       "44   pairwise_interaction  ('Bacteroides', 'Trichuris_dash_trichiura_dash...   \n",
       "45   pairwise_interaction  ('Bacteroides', 'Ruminococcus_dash_sp_dot__das...   \n",
       "48   pairwise_interaction  ('Bacteroides', 'Bacteroides_dash_plebeius_das...   \n",
       "51   pairwise_interaction  ('Bacteroides', 'Parasutterella_dash_excrement...   \n",
       "55   pairwise_interaction      ('Bacteroides', 'Bacteroides_dash_coprocola')   \n",
       "68   pairwise_interaction  ('Faecalibacterium', 'Bacteroides_dash_plebeiu...   \n",
       "103  pairwise_interaction                       ('Prevotella', 'Prevotella')   \n",
       "104  pairwise_interaction  ('Prevotella', 'Trichuris_dash_trichiura_dash_...   \n",
       "124  pairwise_interaction  ('Trichuris_dash_trichiura_dash__leftroundbrac...   \n",
       "128  pairwise_interaction  ('Trichuris_dash_trichiura_dash__leftroundbrac...   \n",
       "140  pairwise_interaction  ('Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_d...   \n",
       "143  pairwise_interaction  ('Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_d...   \n",
       "145  pairwise_interaction  ('Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_d...   \n",
       "208  pairwise_interaction  ('Bacteroides_dash_plebeius_dash_DSM_dash_1713...   \n",
       "348  pairwise_interaction  ('Bacteroides_dash_coprocola', 'Bacteroides_da...   \n",
       "355  pairwise_interaction  ('Bacteroides_dash_coprocola', 'Bacteroides_da...   \n",
       "\n",
       "     Left  Right  Middle        SNR  Significant  \n",
       "0   -1.14  -0.11  -0.625  -2.409779         True  \n",
       "5    0.10   1.13   0.615   2.367750         True  \n",
       "8   -1.05  -0.02  -0.535  -2.071025         True  \n",
       "25   0.22   0.81   0.515   3.448608         True  \n",
       "40   1.07   2.51   1.790   4.897480         True  \n",
       "41   0.51   3.99   2.250   2.558270         True  \n",
       "42  -4.60  -1.04  -2.820  -3.107074         True  \n",
       "43  -2.47  -0.20  -1.335  -2.315486         True  \n",
       "44   1.42   3.85   2.635   4.293403         True  \n",
       "45  -4.96  -1.38  -3.170  -3.496771         True  \n",
       "48   0.42   3.61   2.015   2.507289         True  \n",
       "51  -3.87  -0.03  -1.950  -2.028178         True  \n",
       "55  -3.28  -0.01  -1.645  -1.969291         True  \n",
       "68   0.15   3.39   1.770   2.159146         True  \n",
       "103  5.02   7.30   6.160  10.663421         True  \n",
       "104 -3.58  -1.14  -2.360  -3.792889         True  \n",
       "124 -3.33  -0.88  -2.105  -3.374894         True  \n",
       "128 -3.77  -0.55  -2.160  -2.640688         True  \n",
       "140 -1.90  -0.43  -1.165  -3.131135         True  \n",
       "143 -2.85  -0.57  -1.710  -2.992284         True  \n",
       "145  0.35   3.94   2.145   2.358529         True  \n",
       "208 -5.43  -2.23  -3.830  -4.727823         True  \n",
       "348 -4.96  -1.76  -3.360  -4.136909         True  \n",
       "355  1.92   5.23   3.575   4.245876         True  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stan_output_summary = pd.read_excel('bayesian_regression_summary_top20_rs.xlsx', index_col=0)\n",
    "df_stan_output_summary = df_stan_output_summary[df_stan_output_summary.Significant==True]\n",
    "df_stan_output_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
