{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.interpolate import CubicSpline\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "%run -i '../../utils.py'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bac = pd.read_csv('16S_relative_abundance_lowest_taxonomy.csv', index_col=0)\n",
    "df_meta = pd.read_csv('metadata.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate stan program files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_bacteria_taxa = generate_stan_files_for_fiber_respones(\n",
    "    df_bac, # 16S data (relative or absolute), rows are samples, columns are taxa\n",
    "    df_meta, # meta data, rows are samples, columns are SubjectID, Day, and Dose\n",
    "    prefix='creswell_2020_lowest_taxonomy', # prefix of stan file name\n",
    "    topN=20, # select the topN taxa to run bayesian regression\n",
    "    stan_path='/Users/liaoc/Documents/cmdstan-2.25.0/projects/microbiome_fiber_response_LD'\n",
    ") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract stan output and show significant fiber responses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"creswell_2020_output_%d.csv\"%(i) for i in [1,2,3]])\n",
    "\n",
    "lines = []\n",
    "# basal growth rate\n",
    "for c in selected_bacteria_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# fiber response\n",
    "for c in selected_bacteria_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['fiber_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# pairwise interactions\n",
    "for c1 in selected_bacteria_taxa:\n",
    "    for c2 in selected_bacteria_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.to_excel('bayesian_regression_summary.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Lachnospiraceae, Akkermansia_muciniphila)</td>\n",
       "      <td>0.11</td>\n",
       "      <td>1.63</td>\n",
       "      <td>0.870</td>\n",
       "      <td>2.270559</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Lachnospiraceae, Roseburia_intestinalis)</td>\n",
       "      <td>0.59</td>\n",
       "      <td>3.42</td>\n",
       "      <td>2.005</td>\n",
       "      <td>2.780318</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Faecalibacterium_prausnitzii, Gemmiger_formic...</td>\n",
       "      <td>0.20</td>\n",
       "      <td>3.54</td>\n",
       "      <td>1.870</td>\n",
       "      <td>2.234029</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Prevotella_copri, Anaerostipes_hadrus)</td>\n",
       "      <td>-2.41</td>\n",
       "      <td>-0.63</td>\n",
       "      <td>-1.520</td>\n",
       "      <td>-3.372511</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Prevotella_copri, Bifidobacterium_adolescentis)</td>\n",
       "      <td>0.28</td>\n",
       "      <td>1.87</td>\n",
       "      <td>1.075</td>\n",
       "      <td>2.674802</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Prevotella_copri, Prevotella_copri)</td>\n",
       "      <td>0.17</td>\n",
       "      <td>1.18</td>\n",
       "      <td>0.675</td>\n",
       "      <td>2.628095</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Roseburia_faecis, Roseburia_faecis)</td>\n",
       "      <td>0.11</td>\n",
       "      <td>2.22</td>\n",
       "      <td>1.165</td>\n",
       "      <td>2.202735</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Akkermansia_muciniphila, Anaerostipes_hadrus)</td>\n",
       "      <td>-2.08</td>\n",
       "      <td>-0.30</td>\n",
       "      <td>-1.190</td>\n",
       "      <td>-2.648681</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Akkermansia_muciniphila, Bifidobacterium_adol...</td>\n",
       "      <td>-1.59</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>-0.805</td>\n",
       "      <td>-2.011412</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Akkermansia_muciniphila, Akkermansia_muciniph...</td>\n",
       "      <td>-2.49</td>\n",
       "      <td>-0.97</td>\n",
       "      <td>-1.730</td>\n",
       "      <td>-4.480156</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                               Taxa  \\\n",
       "50   pairwise_interaction         (Lachnospiraceae, Akkermansia_muciniphila)   \n",
       "58   pairwise_interaction          (Lachnospiraceae, Roseburia_intestinalis)   \n",
       "75   pairwise_interaction  (Faecalibacterium_prausnitzii, Gemmiger_formic...   \n",
       "164  pairwise_interaction            (Prevotella_copri, Anaerostipes_hadrus)   \n",
       "165  pairwise_interaction   (Prevotella_copri, Bifidobacterium_adolescentis)   \n",
       "166  pairwise_interaction               (Prevotella_copri, Prevotella_copri)   \n",
       "229  pairwise_interaction               (Roseburia_faecis, Roseburia_faecis)   \n",
       "244  pairwise_interaction     (Akkermansia_muciniphila, Anaerostipes_hadrus)   \n",
       "245  pairwise_interaction  (Akkermansia_muciniphila, Bifidobacterium_adol...   \n",
       "250  pairwise_interaction  (Akkermansia_muciniphila, Akkermansia_muciniph...   \n",
       "\n",
       "     Left  Right  Middle       SNR  Significant  \n",
       "50   0.11   1.63   0.870  2.270559         True  \n",
       "58   0.59   3.42   2.005  2.780318         True  \n",
       "75   0.20   3.54   1.870  2.234029         True  \n",
       "164 -2.41  -0.63  -1.520 -3.372511         True  \n",
       "165  0.28   1.87   1.075  2.674802         True  \n",
       "166  0.17   1.18   0.675  2.628095         True  \n",
       "229  0.11   2.22   1.165  2.202735         True  \n",
       "244 -2.08  -0.30  -1.190 -2.648681         True  \n",
       "245 -1.59  -0.02  -0.805 -2.011412         True  \n",
       "250 -2.49  -0.97  -1.730 -4.480156         True  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select only significant ones\n",
    "df_stan_output_summary[df_stan_output_summary.Significant==True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
