{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.interpolate import CubicSpline\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "%run -i '../../utils.py'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bac = pd.read_csv('16S_relative_abundance_lowest_taxonomy.csv', index_col=0)\n",
    "df_meta = pd.read_csv('metadata.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate stan program files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_bacteria_taxa = generate_stan_files_for_fiber_respones(\n",
    "    df_bac, # 16S data (relative or absolute), rows are samples, columns are taxa\n",
    "    df_meta, # meta data, rows are samples, columns are SubjectID, Day, and Dose\n",
    "    prefix='creswell_2020_lowest_taxonomy', # prefix of stan file name\n",
    "    topN=20, # select the topN taxa to run bayesian regression\n",
    "    stan_path='/Users/liaoc/Documents/cmdstan-2.25.0/projects/microbiome_fiber_response_LD'\n",
    ") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract stan output and show significant fiber responses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"creswell_2020_lowest_taxonomy_output_%d.csv\"%(i) for i in [1,2,3,4,5,6,7,8,9]])\n",
    "\n",
    "lines = []\n",
    "# basal growth rate\n",
    "for c in selected_bacteria_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# fiber response\n",
    "for c in selected_bacteria_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['fiber_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# pairwise interactions\n",
    "for c1 in selected_bacteria_taxa:\n",
    "    for c2 in selected_bacteria_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.to_excel('bayesian_regression_summary_lowest_taxonomy.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Anaerostipes_hadrus</td>\n",
       "      <td>-0.54</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>-0.275</td>\n",
       "      <td>-2.046499</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.450</td>\n",
       "      <td>3.271727</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>fiber_response</td>\n",
       "      <td>Anaerostipes_hadrus</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.035</td>\n",
       "      <td>2.616084</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>fiber_response</td>\n",
       "      <td>Prevotella_copri</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>-0.035</td>\n",
       "      <td>-2.553260</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>fiber_response</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>-0.035</td>\n",
       "      <td>-3.069801</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Dorea_longicatena, Akkermansia_muciniphila)</td>\n",
       "      <td>0.07</td>\n",
       "      <td>1.57</td>\n",
       "      <td>0.820</td>\n",
       "      <td>2.139535</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Anaerostipes_hadrus, Anaerostipes_hadrus)</td>\n",
       "      <td>0.33</td>\n",
       "      <td>2.07</td>\n",
       "      <td>1.200</td>\n",
       "      <td>2.706596</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Anaerostipes_hadrus, Akkermansia_muciniphila)</td>\n",
       "      <td>2.38</td>\n",
       "      <td>3.89</td>\n",
       "      <td>3.135</td>\n",
       "      <td>8.252109</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Bifidobacterium_adolescentis, Akkermansia_muc...</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.66</td>\n",
       "      <td>0.910</td>\n",
       "      <td>2.423329</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Prevotella_copri, Prevotella_copri)</td>\n",
       "      <td>0.92</td>\n",
       "      <td>1.92</td>\n",
       "      <td>1.420</td>\n",
       "      <td>5.584872</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Prevotella_copri, Roseburia_intestinalis)</td>\n",
       "      <td>0.59</td>\n",
       "      <td>3.40</td>\n",
       "      <td>1.995</td>\n",
       "      <td>2.773700</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Roseburia_faecis, Roseburia_faecis)</td>\n",
       "      <td>0.12</td>\n",
       "      <td>2.25</td>\n",
       "      <td>1.185</td>\n",
       "      <td>2.189039</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Akkermansia_muciniphila, Anaerostipes_hadrus)</td>\n",
       "      <td>-3.28</td>\n",
       "      <td>-1.52</td>\n",
       "      <td>-2.400</td>\n",
       "      <td>-5.390934</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Akkermansia_muciniphila, Bifidobacterium_adol...</td>\n",
       "      <td>0.13</td>\n",
       "      <td>1.71</td>\n",
       "      <td>0.920</td>\n",
       "      <td>2.265066</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Akkermansia_muciniphila, Akkermansia_muciniph...</td>\n",
       "      <td>-8.37</td>\n",
       "      <td>-6.87</td>\n",
       "      <td>-7.620</td>\n",
       "      <td>-19.903939</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Roseburia_intestinalis, Roseburia_intestinalis)</td>\n",
       "      <td>-3.14</td>\n",
       "      <td>-0.37</td>\n",
       "      <td>-1.755</td>\n",
       "      <td>-2.537861</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                               Taxa  \\\n",
       "4       basal_growth_rate                                Anaerostipes_hadrus   \n",
       "10      basal_growth_rate                            Akkermansia_muciniphila   \n",
       "24         fiber_response                                Anaerostipes_hadrus   \n",
       "26         fiber_response                                   Prevotella_copri   \n",
       "30         fiber_response                            Akkermansia_muciniphila   \n",
       "90   pairwise_interaction       (Dorea_longicatena, Akkermansia_muciniphila)   \n",
       "124  pairwise_interaction         (Anaerostipes_hadrus, Anaerostipes_hadrus)   \n",
       "130  pairwise_interaction     (Anaerostipes_hadrus, Akkermansia_muciniphila)   \n",
       "150  pairwise_interaction  (Bifidobacterium_adolescentis, Akkermansia_muc...   \n",
       "166  pairwise_interaction               (Prevotella_copri, Prevotella_copri)   \n",
       "178  pairwise_interaction         (Prevotella_copri, Roseburia_intestinalis)   \n",
       "229  pairwise_interaction               (Roseburia_faecis, Roseburia_faecis)   \n",
       "244  pairwise_interaction     (Akkermansia_muciniphila, Anaerostipes_hadrus)   \n",
       "245  pairwise_interaction  (Akkermansia_muciniphila, Bifidobacterium_adol...   \n",
       "250  pairwise_interaction  (Akkermansia_muciniphila, Akkermansia_muciniph...   \n",
       "418  pairwise_interaction   (Roseburia_intestinalis, Roseburia_intestinalis)   \n",
       "\n",
       "     Left  Right  Middle        SNR  Significant  \n",
       "4   -0.54  -0.01  -0.275  -2.046499         True  \n",
       "10   0.18   0.72   0.450   3.271727         True  \n",
       "24   0.01   0.06   0.035   2.616084         True  \n",
       "26  -0.06  -0.01  -0.035  -2.553260         True  \n",
       "30  -0.06  -0.01  -0.035  -3.069801         True  \n",
       "90   0.07   1.57   0.820   2.139535         True  \n",
       "124  0.33   2.07   1.200   2.706596         True  \n",
       "130  2.38   3.89   3.135   8.252109         True  \n",
       "150  0.16   1.66   0.910   2.423329         True  \n",
       "166  0.92   1.92   1.420   5.584872         True  \n",
       "178  0.59   3.40   1.995   2.773700         True  \n",
       "229  0.12   2.25   1.185   2.189039         True  \n",
       "244 -3.28  -1.52  -2.400  -5.390934         True  \n",
       "245  0.13   1.71   0.920   2.265066         True  \n",
       "250 -8.37  -6.87  -7.620 -19.903939         True  \n",
       "418 -3.14  -0.37  -1.755  -2.537861         True  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select only significant ones\n",
    "df_stan_output_summary[df_stan_output_summary.Significant==True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
