{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from scipy.interpolate import CubicSpline\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import warnings\n",
    "import itertools\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "import statsmodels.stats.api as sms\n",
    "import scipy.stats.kde as kde"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read 16S data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abiotrophia</th>\n",
       "      <th>Acetanaerobacterium</th>\n",
       "      <th>Acidaminococcus</th>\n",
       "      <th>Actinidia_dash_chinensis</th>\n",
       "      <th>Actinomyces</th>\n",
       "      <th>Actinomyces_dash_massiliensis</th>\n",
       "      <th>Adlercreutzia</th>\n",
       "      <th>Agathobacter</th>\n",
       "      <th>Akkermansia</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>bacterium_dash_YE57</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_BBMC_dash_9</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_Ecwsrb026</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c</th>\n",
       "      <th>butyrate_dash_producing_dash_bacterium_dash_L2_dash_10</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>vadinBE97</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6185161</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004546</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.013533</td>\n",
       "      <td>0.002041</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003486</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.004507</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185162</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001597</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002067</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>0.000869</td>\n",
       "      <td>0.034805</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185163</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.018110</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007531</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005917</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026358</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185164</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002307</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000773</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000177</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008329</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.034066</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185165</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.052867</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002736</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083844</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 469 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Abiotrophia  Acetanaerobacterium  Acidaminococcus  \\\n",
       "SampleID                                                        \n",
       "SRR6185161          0.0                  0.0         0.004546   \n",
       "SRR6185162          0.0                  0.0         0.000000   \n",
       "SRR6185163          0.0                  0.0         0.000000   \n",
       "SRR6185164          0.0                  0.0         0.000000   \n",
       "SRR6185165          0.0                  0.0         0.000000   \n",
       "\n",
       "            Actinidia_dash_chinensis  Actinomyces  \\\n",
       "SampleID                                            \n",
       "SRR6185161                       0.0     0.000000   \n",
       "SRR6185162                       0.0     0.000000   \n",
       "SRR6185163                       0.0     0.000000   \n",
       "SRR6185164                       0.0     0.000114   \n",
       "SRR6185165                       0.0     0.000000   \n",
       "\n",
       "            Actinomyces_dash_massiliensis  Adlercreutzia  Agathobacter  \\\n",
       "SampleID                                                                 \n",
       "SRR6185161                            0.0            0.0      0.013533   \n",
       "SRR6185162                            0.0            0.0      0.001597   \n",
       "SRR6185163                            0.0            0.0      0.018110   \n",
       "SRR6185164                            0.0            0.0      0.002307   \n",
       "SRR6185165                            0.0            0.0      0.052867   \n",
       "\n",
       "            Akkermansia  Akkermansia_dash_muciniphila  ...  \\\n",
       "SampleID                                               ...   \n",
       "SRR6185161     0.002041                           0.0  ...   \n",
       "SRR6185162     0.000000                           0.0  ...   \n",
       "SRR6185163     0.000000                           0.0  ...   \n",
       "SRR6185164     0.000000                           0.0  ...   \n",
       "SRR6185165     0.000000                           0.0  ...   \n",
       "\n",
       "            _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "SampleID                                                                                  \n",
       "SRR6185161                                           0.003009                             \n",
       "SRR6185162                                           0.002067                             \n",
       "SRR6185163                                           0.007531                             \n",
       "SRR6185164                                           0.000773                             \n",
       "SRR6185165                                           0.020375                             \n",
       "\n",
       "            bacterium_dash_YE57  \\\n",
       "SampleID                          \n",
       "SRR6185161                  0.0   \n",
       "SRR6185162                  0.0   \n",
       "SRR6185163                  0.0   \n",
       "SRR6185164                  0.0   \n",
       "SRR6185165                  0.0   \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_BBMC_dash_9  \\\n",
       "SampleID                                                                         \n",
       "SRR6185161                                                0.0                    \n",
       "SRR6185162                                                0.0                    \n",
       "SRR6185163                                                0.0                    \n",
       "SRR6185164                                                0.0                    \n",
       "SRR6185165                                                0.0                    \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_Ecwsrb026  \\\n",
       "SampleID                                                                       \n",
       "SRR6185161                                           0.000000                  \n",
       "SRR6185162                                           0.000000                  \n",
       "SRR6185163                                           0.000000                  \n",
       "SRR6185164                                           0.000177                  \n",
       "SRR6185165                                           0.000220                  \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c  \\\n",
       "SampleID                                                                  \n",
       "SRR6185161                                                0.0             \n",
       "SRR6185162                                                0.0             \n",
       "SRR6185163                                                0.0             \n",
       "SRR6185164                                                0.0             \n",
       "SRR6185165                                                0.0             \n",
       "\n",
       "            butyrate_dash_producing_dash_bacterium_dash_L2_dash_10  \\\n",
       "SampleID                                                             \n",
       "SRR6185161                                                0.0        \n",
       "SRR6185162                                                0.0        \n",
       "SRR6185163                                                0.0        \n",
       "SRR6185164                                                0.0        \n",
       "SRR6185165                                                0.0        \n",
       "\n",
       "            gut_dash_metagenome  human_dash_gut_dash_metagenome  metagenome  \\\n",
       "SampleID                                                                      \n",
       "SRR6185161             0.003486                        0.000066    0.004507   \n",
       "SRR6185162             0.001174                        0.000869    0.034805   \n",
       "SRR6185163             0.005917                        0.000000    0.026358   \n",
       "SRR6185164             0.008329                        0.000254    0.034066   \n",
       "SRR6185165             0.002736                        0.000000    0.083844   \n",
       "\n",
       "            vadinBE97  \n",
       "SampleID               \n",
       "SRR6185161        0.0  \n",
       "SRR6185162        0.0  \n",
       "SRR6185163        0.0  \n",
       "SRR6185164        0.0  \n",
       "SRR6185165        0.0  \n",
       "\n",
       "[5 rows x 469 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = pd.read_excel('./16S_relative_abundance.xlsx', index_col=0)\n",
    "df_16S.columns = [c.replace('/','_slash_').replace(' ','_space_').replace('[','_leftsquarebracket_').replace(']','_rightsquarebracket_').replace('-','_dash_').replace('.','_dot_').replace('(','_leftroundbracket').replace(')','_rightroundbracket_') for c in df_16S.columns]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add pseudocount for zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abiotrophia</th>\n",
       "      <th>Acetanaerobacterium</th>\n",
       "      <th>Acidaminococcus</th>\n",
       "      <th>Actinidia_dash_chinensis</th>\n",
       "      <th>Actinomyces</th>\n",
       "      <th>Actinomyces_dash_massiliensis</th>\n",
       "      <th>Adlercreutzia</th>\n",
       "      <th>Agathobacter</th>\n",
       "      <th>Akkermansia</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>bacterium_dash_YE57</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_BBMC_dash_9</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_Ecwsrb026</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c</th>\n",
       "      <th>butyrate_dash_producing_dash_bacterium_dash_L2_dash_10</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>vadinBE97</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6185161</th>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.004546</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.013533</td>\n",
       "      <td>0.002041</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003009</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.003486</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.004507</td>\n",
       "      <td>0.000013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185162</th>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.001597</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002067</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>0.000869</td>\n",
       "      <td>0.034805</td>\n",
       "      <td>0.000117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185163</th>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.018110</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007531</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.005917</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.026358</td>\n",
       "      <td>0.000717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185164</th>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.002307</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000773</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000177</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.008329</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.034066</td>\n",
       "      <td>0.000025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185165</th>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.052867</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020375</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.002736</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.083844</td>\n",
       "      <td>0.000073</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 469 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Abiotrophia  Acetanaerobacterium  Acidaminococcus  \\\n",
       "SampleID                                                        \n",
       "SRR6185161     0.000013             0.000013         0.004546   \n",
       "SRR6185162     0.000117             0.000117         0.000117   \n",
       "SRR6185163     0.000717             0.000717         0.000717   \n",
       "SRR6185164     0.000025             0.000025         0.000025   \n",
       "SRR6185165     0.000073             0.000073         0.000073   \n",
       "\n",
       "            Actinidia_dash_chinensis  Actinomyces  \\\n",
       "SampleID                                            \n",
       "SRR6185161                  0.000013     0.000013   \n",
       "SRR6185162                  0.000117     0.000117   \n",
       "SRR6185163                  0.000717     0.000717   \n",
       "SRR6185164                  0.000025     0.000114   \n",
       "SRR6185165                  0.000073     0.000073   \n",
       "\n",
       "            Actinomyces_dash_massiliensis  Adlercreutzia  Agathobacter  \\\n",
       "SampleID                                                                 \n",
       "SRR6185161                       0.000013       0.000013      0.013533   \n",
       "SRR6185162                       0.000117       0.000117      0.001597   \n",
       "SRR6185163                       0.000717       0.000717      0.018110   \n",
       "SRR6185164                       0.000025       0.000025      0.002307   \n",
       "SRR6185165                       0.000073       0.000073      0.052867   \n",
       "\n",
       "            Akkermansia  Akkermansia_dash_muciniphila  ...  \\\n",
       "SampleID                                               ...   \n",
       "SRR6185161     0.002041                      0.000013  ...   \n",
       "SRR6185162     0.000117                      0.000117  ...   \n",
       "SRR6185163     0.000717                      0.000717  ...   \n",
       "SRR6185164     0.000025                      0.000025  ...   \n",
       "SRR6185165     0.000073                      0.000073  ...   \n",
       "\n",
       "            _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "SampleID                                                                                  \n",
       "SRR6185161                                           0.003009                             \n",
       "SRR6185162                                           0.002067                             \n",
       "SRR6185163                                           0.007531                             \n",
       "SRR6185164                                           0.000773                             \n",
       "SRR6185165                                           0.020375                             \n",
       "\n",
       "            bacterium_dash_YE57  \\\n",
       "SampleID                          \n",
       "SRR6185161             0.000013   \n",
       "SRR6185162             0.000117   \n",
       "SRR6185163             0.000717   \n",
       "SRR6185164             0.000025   \n",
       "SRR6185165             0.000073   \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_BBMC_dash_9  \\\n",
       "SampleID                                                                         \n",
       "SRR6185161                                           0.000013                    \n",
       "SRR6185162                                           0.000117                    \n",
       "SRR6185163                                           0.000717                    \n",
       "SRR6185164                                           0.000025                    \n",
       "SRR6185165                                           0.000073                    \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_Ecwsrb026  \\\n",
       "SampleID                                                                       \n",
       "SRR6185161                                           0.000013                  \n",
       "SRR6185162                                           0.000117                  \n",
       "SRR6185163                                           0.000717                  \n",
       "SRR6185164                                           0.000177                  \n",
       "SRR6185165                                           0.000220                  \n",
       "\n",
       "            bacterium_dash_enrichment_dash_culture_dash_clone_dash_RB7c  \\\n",
       "SampleID                                                                  \n",
       "SRR6185161                                           0.000013             \n",
       "SRR6185162                                           0.000117             \n",
       "SRR6185163                                           0.000717             \n",
       "SRR6185164                                           0.000025             \n",
       "SRR6185165                                           0.000073             \n",
       "\n",
       "            butyrate_dash_producing_dash_bacterium_dash_L2_dash_10  \\\n",
       "SampleID                                                             \n",
       "SRR6185161                                           0.000013        \n",
       "SRR6185162                                           0.000117        \n",
       "SRR6185163                                           0.000717        \n",
       "SRR6185164                                           0.000025        \n",
       "SRR6185165                                           0.000073        \n",
       "\n",
       "            gut_dash_metagenome  human_dash_gut_dash_metagenome  metagenome  \\\n",
       "SampleID                                                                      \n",
       "SRR6185161             0.003486                        0.000066    0.004507   \n",
       "SRR6185162             0.001174                        0.000869    0.034805   \n",
       "SRR6185163             0.005917                        0.000717    0.026358   \n",
       "SRR6185164             0.008329                        0.000254    0.034066   \n",
       "SRR6185165             0.002736                        0.000073    0.083844   \n",
       "\n",
       "            vadinBE97  \n",
       "SampleID               \n",
       "SRR6185161   0.000013  \n",
       "SRR6185162   0.000117  \n",
       "SRR6185163   0.000717  \n",
       "SRR6185164   0.000025  \n",
       "SRR6185165   0.000073  \n",
       "\n",
       "[5 rows x 469 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample_id in df_16S.index:\n",
    "    sample = np.array(df_16S.loc[sample_id])\n",
    "    minval = np.min(sample[np.nonzero(sample)]) # minimum non-zero value\n",
    "    sample[sample==0] = minval\n",
    "    df_16S.loc[sample_id] = sample\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata for inulin and control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>dose</th>\n",
       "      <th>Day</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6185259</th>\n",
       "      <td>Prebiotic</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>post</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185177</th>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>post</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185262</th>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT04</td>\n",
       "      <td>post</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185178</th>\n",
       "      <td>Prebiotic</td>\n",
       "      <td>ADAPT04</td>\n",
       "      <td>post</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185264</th>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT06</td>\n",
       "      <td>post</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Group Subject_ID status  dose  Day\n",
       "#SampleID                                         \n",
       "SRR6185259  Prebiotic    ADAPT03   post    16   21\n",
       "SRR6185177    Placebo    ADAPT03   post    16   21\n",
       "SRR6185262    Placebo    ADAPT04   post    16   21\n",
       "SRR6185178  Prebiotic    ADAPT04   post    16   21\n",
       "SRR6185264    Placebo    ADAPT06   post    16   21"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv('metadata.txt', sep='\\t', index_col=0)\n",
    "df_meta = df_meta[['Group','Subject_ID','status','dose']]\n",
    "df_meta['Day'] = [0 if s=='pre' else 21 for s in df_meta.status]\n",
    "df_meta['Diet'] = [1 if g=='Prebiotic' else 0 for g in df_meta['Group']]\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 20 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>Agathobacter</th>\n",
       "      <th>Prevotella</th>\n",
       "      <th>Subdoligranulum</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Christensenellaceae_dash_R_dash_7_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group</th>\n",
       "      <th>Coprococcus</th>\n",
       "      <th>Dorea</th>\n",
       "      <th>UCG_dash_002</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6185224</th>\n",
       "      <td>0.094739</td>\n",
       "      <td>0.034865</td>\n",
       "      <td>0.003189</td>\n",
       "      <td>0.044047</td>\n",
       "      <td>0.025218</td>\n",
       "      <td>0.007017</td>\n",
       "      <td>0.000326</td>\n",
       "      <td>0.039055</td>\n",
       "      <td>0.037309</td>\n",
       "      <td>0.004085</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.062539</td>\n",
       "      <td>0.029233</td>\n",
       "      <td>0.012754</td>\n",
       "      <td>0.013441</td>\n",
       "      <td>0.014570</td>\n",
       "      <td>0.029559</td>\n",
       "      <td>0.038717</td>\n",
       "      <td>0.034097</td>\n",
       "      <td>0.030629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185259</th>\n",
       "      <td>0.066649</td>\n",
       "      <td>0.040161</td>\n",
       "      <td>0.101466</td>\n",
       "      <td>0.069750</td>\n",
       "      <td>0.076858</td>\n",
       "      <td>0.008313</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.049645</td>\n",
       "      <td>0.030776</td>\n",
       "      <td>0.025531</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.051409</td>\n",
       "      <td>0.042289</td>\n",
       "      <td>0.014102</td>\n",
       "      <td>0.014662</td>\n",
       "      <td>0.007208</td>\n",
       "      <td>0.014679</td>\n",
       "      <td>0.021788</td>\n",
       "      <td>0.018819</td>\n",
       "      <td>0.023107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185212</th>\n",
       "      <td>0.068807</td>\n",
       "      <td>0.015762</td>\n",
       "      <td>0.012607</td>\n",
       "      <td>0.049940</td>\n",
       "      <td>0.020162</td>\n",
       "      <td>0.000855</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.065287</td>\n",
       "      <td>0.012306</td>\n",
       "      <td>0.020250</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.086455</td>\n",
       "      <td>0.047375</td>\n",
       "      <td>0.009842</td>\n",
       "      <td>0.012318</td>\n",
       "      <td>0.017246</td>\n",
       "      <td>0.039318</td>\n",
       "      <td>0.031864</td>\n",
       "      <td>0.060234</td>\n",
       "      <td>0.015071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185177</th>\n",
       "      <td>0.060489</td>\n",
       "      <td>0.044513</td>\n",
       "      <td>0.003528</td>\n",
       "      <td>0.050959</td>\n",
       "      <td>0.019494</td>\n",
       "      <td>0.001221</td>\n",
       "      <td>0.000155</td>\n",
       "      <td>0.060468</td>\n",
       "      <td>0.016390</td>\n",
       "      <td>0.034870</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.079993</td>\n",
       "      <td>0.059433</td>\n",
       "      <td>0.007729</td>\n",
       "      <td>0.005691</td>\n",
       "      <td>0.013896</td>\n",
       "      <td>0.030896</td>\n",
       "      <td>0.042888</td>\n",
       "      <td>0.032914</td>\n",
       "      <td>0.023953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185278</th>\n",
       "      <td>0.164667</td>\n",
       "      <td>0.074385</td>\n",
       "      <td>0.004485</td>\n",
       "      <td>0.039542</td>\n",
       "      <td>0.027734</td>\n",
       "      <td>0.083783</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.039328</td>\n",
       "      <td>0.037355</td>\n",
       "      <td>0.035342</td>\n",
       "      <td>0.134308</td>\n",
       "      <td>0.023239</td>\n",
       "      <td>0.019161</td>\n",
       "      <td>0.051655</td>\n",
       "      <td>0.014096</td>\n",
       "      <td>0.005390</td>\n",
       "      <td>0.005055</td>\n",
       "      <td>0.001688</td>\n",
       "      <td>0.004739</td>\n",
       "      <td>0.000203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Blautia  Faecalibacterium  Bifidobacterium  metagenome  \\\n",
       "#SampleID                                                             \n",
       "SRR6185224  0.094739          0.034865         0.003189    0.044047   \n",
       "SRR6185259  0.066649          0.040161         0.101466    0.069750   \n",
       "SRR6185212  0.068807          0.015762         0.012607    0.049940   \n",
       "SRR6185177  0.060489          0.044513         0.003528    0.050959   \n",
       "SRR6185278  0.164667          0.074385         0.004485    0.039542   \n",
       "\n",
       "            Anaerostipes  Agathobacter  Prevotella  Subdoligranulum  \\\n",
       "#SampleID                                                             \n",
       "SRR6185224      0.025218      0.007017    0.000326         0.039055   \n",
       "SRR6185259      0.076858      0.008313    0.000033         0.049645   \n",
       "SRR6185212      0.020162      0.000855    0.000025         0.065287   \n",
       "SRR6185177      0.019494      0.001221    0.000155         0.060468   \n",
       "SRR6185278      0.027734      0.083783    0.000041         0.039328   \n",
       "\n",
       "            Bacteroides  Fusicatenibacter  \\\n",
       "#SampleID                                   \n",
       "SRR6185224     0.037309          0.004085   \n",
       "SRR6185259     0.030776          0.025531   \n",
       "SRR6185212     0.012306          0.020250   \n",
       "SRR6185177     0.016390          0.034870   \n",
       "SRR6185278     0.037355          0.035342   \n",
       "\n",
       "            Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  \\\n",
       "#SampleID                                                        \n",
       "SRR6185224                                           0.000047    \n",
       "SRR6185259                                           0.000033    \n",
       "SRR6185212                                           0.000025    \n",
       "SRR6185177                                           0.000031    \n",
       "SRR6185278                                           0.134308    \n",
       "\n",
       "            Christensenellaceae_dash_R_dash_7_dash_group  \\\n",
       "#SampleID                                                  \n",
       "SRR6185224                                      0.062539   \n",
       "SRR6185259                                      0.051409   \n",
       "SRR6185212                                      0.086455   \n",
       "SRR6185177                                      0.079993   \n",
       "SRR6185278                                      0.023239   \n",
       "\n",
       "            _leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group  \\\n",
       "#SampleID                                                                               \n",
       "SRR6185224                                           0.029233                           \n",
       "SRR6185259                                           0.042289                           \n",
       "SRR6185212                                           0.047375                           \n",
       "SRR6185177                                           0.059433                           \n",
       "SRR6185278                                           0.019161                           \n",
       "\n",
       "            Coprococcus     Dorea  UCG_dash_002  Lachnospiraceae  \\\n",
       "#SampleID                                                          \n",
       "SRR6185224     0.012754  0.013441      0.014570         0.029559   \n",
       "SRR6185259     0.014102  0.014662      0.007208         0.014679   \n",
       "SRR6185212     0.009842  0.012318      0.017246         0.039318   \n",
       "SRR6185177     0.007729  0.005691      0.013896         0.030896   \n",
       "SRR6185278     0.051655  0.014096      0.005390         0.005055   \n",
       "\n",
       "            gut_dash_metagenome  Clostridia_dash_UCG_dash_014  \\\n",
       "#SampleID                                                       \n",
       "SRR6185224             0.038717                      0.034097   \n",
       "SRR6185259             0.021788                      0.018819   \n",
       "SRR6185212             0.031864                      0.060234   \n",
       "SRR6185177             0.042888                      0.032914   \n",
       "SRR6185278             0.001688                      0.004739   \n",
       "\n",
       "            _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \n",
       "#SampleID                                                                                \n",
       "SRR6185224                                           0.030629                            \n",
       "SRR6185259                                           0.023107                            \n",
       "SRR6185212                                           0.015071                            \n",
       "SRR6185177                                           0.023953                            \n",
       "SRR6185278                                           0.000203                            "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 20\n",
    "df_16S_T = df_16S.loc[df_meta.index].T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Blautia',\n",
       " 'Faecalibacterium',\n",
       " 'Bifidobacterium',\n",
       " 'metagenome',\n",
       " 'Anaerostipes',\n",
       " 'Agathobacter',\n",
       " 'Prevotella',\n",
       " 'Subdoligranulum',\n",
       " 'Bacteroides',\n",
       " 'Fusicatenibacter',\n",
       " 'Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57',\n",
       " 'Christensenellaceae_dash_R_dash_7_dash_group',\n",
       " '_leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group',\n",
       " 'Coprococcus',\n",
       " 'Dorea',\n",
       " 'UCG_dash_002',\n",
       " 'Lachnospiraceae',\n",
       " 'gut_dash_metagenome',\n",
       " 'Clostridia_dash_UCG_dash_014',\n",
       " '_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_16S_topX.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>...</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Christensenellaceae_dash_R_dash_7_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group</th>\n",
       "      <th>Coprococcus</th>\n",
       "      <th>Dorea</th>\n",
       "      <th>UCG_dash_002</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6185224</th>\n",
       "      <td>Prebiotic</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>pre</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.147887</td>\n",
       "      <td>0.054425</td>\n",
       "      <td>0.004977</td>\n",
       "      <td>0.068757</td>\n",
       "      <td>0.039365</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.097623</td>\n",
       "      <td>0.045632</td>\n",
       "      <td>0.019910</td>\n",
       "      <td>0.020981</td>\n",
       "      <td>0.022744</td>\n",
       "      <td>0.046141</td>\n",
       "      <td>0.060437</td>\n",
       "      <td>0.053226</td>\n",
       "      <td>0.047812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185259</th>\n",
       "      <td>Prebiotic</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>post</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104039</td>\n",
       "      <td>0.062691</td>\n",
       "      <td>0.158389</td>\n",
       "      <td>0.108880</td>\n",
       "      <td>0.119976</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.080250</td>\n",
       "      <td>0.066013</td>\n",
       "      <td>0.022013</td>\n",
       "      <td>0.022888</td>\n",
       "      <td>0.011251</td>\n",
       "      <td>0.022914</td>\n",
       "      <td>0.034010</td>\n",
       "      <td>0.029376</td>\n",
       "      <td>0.036070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185212</th>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>pre</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.107408</td>\n",
       "      <td>0.024605</td>\n",
       "      <td>0.019680</td>\n",
       "      <td>0.077956</td>\n",
       "      <td>0.031473</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.134956</td>\n",
       "      <td>0.073953</td>\n",
       "      <td>0.015364</td>\n",
       "      <td>0.019229</td>\n",
       "      <td>0.026921</td>\n",
       "      <td>0.061376</td>\n",
       "      <td>0.049740</td>\n",
       "      <td>0.094026</td>\n",
       "      <td>0.023526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185177</th>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>post</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0.094423</td>\n",
       "      <td>0.069485</td>\n",
       "      <td>0.005508</td>\n",
       "      <td>0.079547</td>\n",
       "      <td>0.030430</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.124869</td>\n",
       "      <td>0.092776</td>\n",
       "      <td>0.012065</td>\n",
       "      <td>0.008883</td>\n",
       "      <td>0.021692</td>\n",
       "      <td>0.048229</td>\n",
       "      <td>0.066949</td>\n",
       "      <td>0.051379</td>\n",
       "      <td>0.037391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185278</th>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT04</td>\n",
       "      <td>pre</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.257045</td>\n",
       "      <td>0.116115</td>\n",
       "      <td>0.007001</td>\n",
       "      <td>0.061725</td>\n",
       "      <td>0.043293</td>\n",
       "      <td>...</td>\n",
       "      <td>0.209656</td>\n",
       "      <td>0.036276</td>\n",
       "      <td>0.029910</td>\n",
       "      <td>0.080633</td>\n",
       "      <td>0.022004</td>\n",
       "      <td>0.008414</td>\n",
       "      <td>0.007890</td>\n",
       "      <td>0.002635</td>\n",
       "      <td>0.007398</td>\n",
       "      <td>0.000318</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Group Subject_ID status  Day  Diet   Blautia  \\\n",
       "#SampleID                                                      \n",
       "SRR6185224  Prebiotic    ADAPT03    pre    0     1  0.147887   \n",
       "SRR6185259  Prebiotic    ADAPT03   post   21     1  0.104039   \n",
       "SRR6185212    Placebo    ADAPT03    pre    0     0  0.107408   \n",
       "SRR6185177    Placebo    ADAPT03   post   21     0  0.094423   \n",
       "SRR6185278    Placebo    ADAPT04    pre    0     0  0.257045   \n",
       "\n",
       "            Faecalibacterium  Bifidobacterium  metagenome  Anaerostipes  ...  \\\n",
       "#SampleID                                                                ...   \n",
       "SRR6185224          0.054425         0.004977    0.068757      0.039365  ...   \n",
       "SRR6185259          0.062691         0.158389    0.108880      0.119976  ...   \n",
       "SRR6185212          0.024605         0.019680    0.077956      0.031473  ...   \n",
       "SRR6185177          0.069485         0.005508    0.079547      0.030430  ...   \n",
       "SRR6185278          0.116115         0.007001    0.061725      0.043293  ...   \n",
       "\n",
       "            Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  \\\n",
       "#SampleID                                                        \n",
       "SRR6185224                                           0.000073    \n",
       "SRR6185259                                           0.000051    \n",
       "SRR6185212                                           0.000039    \n",
       "SRR6185177                                           0.000048    \n",
       "SRR6185278                                           0.209656    \n",
       "\n",
       "            Christensenellaceae_dash_R_dash_7_dash_group  \\\n",
       "#SampleID                                                  \n",
       "SRR6185224                                      0.097623   \n",
       "SRR6185259                                      0.080250   \n",
       "SRR6185212                                      0.134956   \n",
       "SRR6185177                                      0.124869   \n",
       "SRR6185278                                      0.036276   \n",
       "\n",
       "            _leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group  \\\n",
       "#SampleID                                                                               \n",
       "SRR6185224                                           0.045632                           \n",
       "SRR6185259                                           0.066013                           \n",
       "SRR6185212                                           0.073953                           \n",
       "SRR6185177                                           0.092776                           \n",
       "SRR6185278                                           0.029910                           \n",
       "\n",
       "            Coprococcus     Dorea  UCG_dash_002  Lachnospiraceae  \\\n",
       "#SampleID                                                          \n",
       "SRR6185224     0.019910  0.020981      0.022744         0.046141   \n",
       "SRR6185259     0.022013  0.022888      0.011251         0.022914   \n",
       "SRR6185212     0.015364  0.019229      0.026921         0.061376   \n",
       "SRR6185177     0.012065  0.008883      0.021692         0.048229   \n",
       "SRR6185278     0.080633  0.022004      0.008414         0.007890   \n",
       "\n",
       "            gut_dash_metagenome  Clostridia_dash_UCG_dash_014  \\\n",
       "#SampleID                                                       \n",
       "SRR6185224             0.060437                      0.053226   \n",
       "SRR6185259             0.034010                      0.029376   \n",
       "SRR6185212             0.049740                      0.094026   \n",
       "SRR6185177             0.066949                      0.051379   \n",
       "SRR6185278             0.002635                      0.007398   \n",
       "\n",
       "            _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \n",
       "#SampleID                                                                                \n",
       "SRR6185224                                           0.047812                            \n",
       "SRR6185259                                           0.036070                            \n",
       "SRR6185212                                           0.023526                            \n",
       "SRR6185177                                           0.037391                            \n",
       "SRR6185278                                           0.000318                            \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX_w_meta = pd.merge(df_meta, df_16S_topX/df_16S_topX.max().max(), left_index=True, right_index=True, how='inner')\n",
    "df_16S_topX_w_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate log-derivatives (dlog(B)/dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>...</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Christensenellaceae_dash_R_dash_7_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group</th>\n",
       "      <th>Coprococcus</th>\n",
       "      <th>Dorea</th>\n",
       "      <th>UCG_dash_002</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Clostridia_dash_UCG_dash_014</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6185224</th>\n",
       "      <td>Prebiotic</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>pre</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.002088</td>\n",
       "      <td>0.000394</td>\n",
       "      <td>0.007305</td>\n",
       "      <td>0.001911</td>\n",
       "      <td>0.003839</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.008151e-06</td>\n",
       "      <td>-0.000827</td>\n",
       "      <td>0.000970</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>-0.000547</td>\n",
       "      <td>-0.001106</td>\n",
       "      <td>-0.001258</td>\n",
       "      <td>-0.001136</td>\n",
       "      <td>-0.000559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185259</th>\n",
       "      <td>Prebiotic</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>post</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.002088</td>\n",
       "      <td>0.000394</td>\n",
       "      <td>0.007305</td>\n",
       "      <td>0.001911</td>\n",
       "      <td>0.003839</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.008151e-06</td>\n",
       "      <td>-0.000827</td>\n",
       "      <td>0.000970</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>-0.000547</td>\n",
       "      <td>-0.001106</td>\n",
       "      <td>-0.001258</td>\n",
       "      <td>-0.001136</td>\n",
       "      <td>-0.000559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185212</th>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>pre</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.000618</td>\n",
       "      <td>0.002137</td>\n",
       "      <td>-0.000675</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>-0.000050</td>\n",
       "      <td>...</td>\n",
       "      <td>4.386858e-07</td>\n",
       "      <td>-0.000480</td>\n",
       "      <td>0.000896</td>\n",
       "      <td>-0.000157</td>\n",
       "      <td>-0.000493</td>\n",
       "      <td>-0.000249</td>\n",
       "      <td>-0.000626</td>\n",
       "      <td>0.000819</td>\n",
       "      <td>-0.002031</td>\n",
       "      <td>0.000660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185177</th>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT03</td>\n",
       "      <td>post</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.000618</td>\n",
       "      <td>0.002137</td>\n",
       "      <td>-0.000675</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>-0.000050</td>\n",
       "      <td>...</td>\n",
       "      <td>4.386858e-07</td>\n",
       "      <td>-0.000480</td>\n",
       "      <td>0.000896</td>\n",
       "      <td>-0.000157</td>\n",
       "      <td>-0.000493</td>\n",
       "      <td>-0.000249</td>\n",
       "      <td>-0.000626</td>\n",
       "      <td>0.000819</td>\n",
       "      <td>-0.002031</td>\n",
       "      <td>0.000660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6185278</th>\n",
       "      <td>Placebo</td>\n",
       "      <td>ADAPT04</td>\n",
       "      <td>pre</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.002717</td>\n",
       "      <td>0.000875</td>\n",
       "      <td>0.002291</td>\n",
       "      <td>-0.001571</td>\n",
       "      <td>0.001320</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.784231e-03</td>\n",
       "      <td>-0.001420</td>\n",
       "      <td>0.000243</td>\n",
       "      <td>-0.002790</td>\n",
       "      <td>-0.000377</td>\n",
       "      <td>0.000395</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>-0.000346</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Group Subject_ID status  Day  Diet   Blautia  \\\n",
       "#SampleID                                                      \n",
       "SRR6185224  Prebiotic    ADAPT03    pre    0     1 -0.002088   \n",
       "SRR6185259  Prebiotic    ADAPT03   post   21     1 -0.002088   \n",
       "SRR6185212    Placebo    ADAPT03    pre    0     0 -0.000618   \n",
       "SRR6185177    Placebo    ADAPT03   post   21     0 -0.000618   \n",
       "SRR6185278    Placebo    ADAPT04    pre    0     0 -0.002717   \n",
       "\n",
       "            Faecalibacterium  Bifidobacterium  metagenome  Anaerostipes  ...  \\\n",
       "#SampleID                                                                ...   \n",
       "SRR6185224          0.000394         0.007305    0.001911      0.003839  ...   \n",
       "SRR6185259          0.000394         0.007305    0.001911      0.003839  ...   \n",
       "SRR6185212          0.002137        -0.000675    0.000076     -0.000050  ...   \n",
       "SRR6185177          0.002137        -0.000675    0.000076     -0.000050  ...   \n",
       "SRR6185278          0.000875         0.002291   -0.001571      0.001320  ...   \n",
       "\n",
       "            Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  \\\n",
       "#SampleID                                                        \n",
       "SRR6185224                                      -1.008151e-06    \n",
       "SRR6185259                                      -1.008151e-06    \n",
       "SRR6185212                                       4.386858e-07    \n",
       "SRR6185177                                       4.386858e-07    \n",
       "SRR6185278                                      -2.784231e-03    \n",
       "\n",
       "            Christensenellaceae_dash_R_dash_7_dash_group  \\\n",
       "#SampleID                                                  \n",
       "SRR6185224                                     -0.000827   \n",
       "SRR6185259                                     -0.000827   \n",
       "SRR6185212                                     -0.000480   \n",
       "SRR6185177                                     -0.000480   \n",
       "SRR6185278                                     -0.001420   \n",
       "\n",
       "            _leftsquarebracket_Eubacterium_rightsquarebracket__dash_hallii_dash_group  \\\n",
       "#SampleID                                                                               \n",
       "SRR6185224                                           0.000970                           \n",
       "SRR6185259                                           0.000970                           \n",
       "SRR6185212                                           0.000896                           \n",
       "SRR6185177                                           0.000896                           \n",
       "SRR6185278                                           0.000243                           \n",
       "\n",
       "            Coprococcus     Dorea  UCG_dash_002  Lachnospiraceae  \\\n",
       "#SampleID                                                          \n",
       "SRR6185224     0.000100  0.000091     -0.000547        -0.001106   \n",
       "SRR6185259     0.000100  0.000091     -0.000547        -0.001106   \n",
       "SRR6185212    -0.000157 -0.000493     -0.000249        -0.000626   \n",
       "SRR6185177    -0.000157 -0.000493     -0.000249        -0.000626   \n",
       "SRR6185278    -0.002790 -0.000377      0.000395         0.000018   \n",
       "\n",
       "            gut_dash_metagenome  Clostridia_dash_UCG_dash_014  \\\n",
       "#SampleID                                                       \n",
       "SRR6185224            -0.001258                     -0.001136   \n",
       "SRR6185259            -0.001258                     -0.001136   \n",
       "SRR6185212             0.000819                     -0.002031   \n",
       "SRR6185177             0.000819                     -0.002031   \n",
       "SRR6185278             0.000273                     -0.000346   \n",
       "\n",
       "            _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \n",
       "#SampleID                                                                                \n",
       "SRR6185224                                          -0.000559                            \n",
       "SRR6185259                                          -0.000559                            \n",
       "SRR6185212                                           0.000660                            \n",
       "SRR6185177                                           0.000660                            \n",
       "SRR6185278                                           0.000003                            \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_deriv = deepcopy(df_16S_topX_w_meta)\n",
    "for curr_sub in set(df_deriv.Subject_ID):\n",
    "    for group in ['Prebiotic','Placebo']:\n",
    "        curr_df = df_deriv[(df_deriv.Subject_ID==curr_sub) & (df_deriv.Group==group)].sort_values(by='Day')\n",
    "        if len(curr_df.index) == 0:\n",
    "            continue\n",
    "        for taxon in bacterial_taxa:\n",
    "            xdata = np.array(curr_df['Day'])\n",
    "            ydata = np.array(curr_df[taxon])\n",
    "            cs = CubicSpline(xdata, ydata)\n",
    "            csd1 = cs.derivative(nu=1)\n",
    "            ydata_d1 = csd1(xdata)\n",
    "            df_deriv.loc[(df_deriv.Subject_ID==curr_sub)&(df_deriv.Group==group), taxon] = ydata_d1\n",
    "df_deriv.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct regression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-0.87410934, -0.87410934, -0.28067247, ..., -0.14311646,\n",
       "       -0.31333864, -0.31333864])"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ymat should be samples by bacteria\n",
    "Ymat = df_deriv[bacterial_taxa].values\n",
    "Ymat = Ymat.flatten(order='F')\n",
    "Ymat = StandardScaler().fit_transform(Ymat.reshape(-1,1)).reshape(1,-1)[0] # standardize\n",
    "Ymat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2680"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Ymat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.00000000e+00, 1.00000000e+00, 1.47887459e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 1.00000000e+00, 1.04039337e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 0.00000000e+00, 1.07407674e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       ...,\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        1.46143086e-02, 1.27524508e-04, 1.50989017e-02],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        2.81549771e-02, 4.20851676e-05, 1.93802197e-02],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        5.44162667e-03, 2.06905957e-05, 4.69676522e-03]])"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Xmat = np.zeros(shape=(topX*len(df_deriv.index), (topX+2)*topX))\n",
    "for k in np.arange(topX):\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)] = 1\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+1] = df_deriv.Diet.values\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+2:(k+1)*(topX+2)] = df_16S_topX_w_meta[bacterial_taxa].values\n",
    "Xmat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write stan program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "json_str = '{\\n\"N\" : %d,\\n'%(len(Ymat))\n",
    "\n",
    "# y variable\n",
    "json_str += '\\\"dlogX\\\" : [%s],\\n'%(','.join(list(Ymat.astype(str))))\n",
    "\n",
    "# x variable\n",
    "for k1,c1 in enumerate(bacterial_taxa):\n",
    "    # growth rate\n",
    "    json_str += '\\\"growth_rate_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)].astype(str))))\n",
    "    # diet response\n",
    "    json_str += '\\\"inulin_response_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)+1].astype(str))))\n",
    "    # bacterial interactions\n",
    "    for k2,c2 in enumerate(bacterial_taxa):\n",
    "        v = list(Xmat[:,k1*(topX+2)+2+k2].astype(str))\n",
    "        json_str += '\\\"pairwise_interaction_%s_%s\\\" : [%s]'%(c1,c2,','.join(v))\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            json_str += '\\n}'\n",
    "        else:\n",
    "            json_str += ',\\n'\n",
    "#print(json_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.data.json\", \"w\")\n",
    "text_file.write(\"%s\" % json_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# data block\n",
    "model_str = 'data {\\n'\n",
    "model_str += '\\tint<lower=0> N;\\n'\n",
    "model_str += '\\tvector[N] dlogX;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\tvector[N] growth_rate_%s;\\n'%(c1)\n",
    "    model_str += '\\tvector[N] inulin_response_%s;\\n'%(c1)\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tvector[N] pairwise_interaction_%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'\n",
    "\n",
    "# parameter block\n",
    "model_str += 'parameters {\\n\\treal<lower=0,upper=1> sigma;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\treal alpha__%s;\\n'%(c1) # growth rate\n",
    "    model_str += '\\treal epsilon__%s;\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\treal beta__%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'       \n",
    "        \n",
    "# model block\n",
    "model_str += 'model {\\n\\tsigma ~ uniform(0,1);\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\talpha__%s ~ normal(0,1);\\n'%(c1) # growth rate\n",
    "    model_str += '\\tepsilon__%s ~ normal(0,1);\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tbeta__%s_%s ~ normal(0,1);\\n'%(c1,c2)\n",
    "model_str += '\\tdlogX ~ normal('\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += 'alpha__%s*growth_rate_%s+'%(c1,c1) # growth rate\n",
    "    model_str += 'epsilon__%s*inulin_response_%s+'%(c1,c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s'%(c1,c2,c1,c2)\n",
    "        else:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s+'%(c1,c2,c1,c2)\n",
    "model_str += ', sigma);\\n}'\n",
    "#print(model_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.stan\", \"w\")\n",
    "text_file.write(\"%s\" % model_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot stan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hpd_grid(sample, alpha=0.05, roundto=2):\n",
    "    \"\"\"Calculate highest posterior density (HPD) of array for given alpha. \n",
    "    The HPD is the minimum width Bayesian credible interval (BCI). \n",
    "    The function works for multimodal distributions, returning more than one mode\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    sample : Numpy array or python list\n",
    "        An array containing MCMC samples\n",
    "    alpha : float\n",
    "        Desired probability of type I error (defaults to 0.05)\n",
    "    roundto: integer\n",
    "        Number of digits after the decimal point for the results\n",
    "    Returns\n",
    "    ----------\n",
    "    hpd: array with the lower \n",
    "          \n",
    "    \"\"\"\n",
    "    sample = np.asarray(sample)\n",
    "    sample = sample[~np.isnan(sample)]\n",
    "    # get upper and lower bounds\n",
    "    l = np.min(sample)\n",
    "    u = np.max(sample)\n",
    "    density = kde.gaussian_kde(sample)\n",
    "    x = np.linspace(l, u, 2000)\n",
    "    y = density.evaluate(x)\n",
    "    #y = density.evaluate(x, l, u) waitting for PR to be accepted\n",
    "    xy_zipped = zip(x, y/np.sum(y))\n",
    "    xy = sorted(xy_zipped, key=lambda x: x[1], reverse=True)\n",
    "    xy_cum_sum = 0\n",
    "    hdv = []\n",
    "    for val in xy:\n",
    "        xy_cum_sum += val[1]\n",
    "        hdv.append(val[0])\n",
    "        if xy_cum_sum >= (1-alpha):\n",
    "            break\n",
    "    hdv.sort()\n",
    "    diff = (u-l)/20  # differences of 5%\n",
    "    hpd = []\n",
    "    hpd.append(round(min(hdv), roundto))\n",
    "    for i in range(1, len(hdv)):\n",
    "        if hdv[i]-hdv[i-1] >= diff:\n",
    "            hpd.append(round(hdv[i-1], roundto))\n",
    "            hpd.append(round(hdv[i], roundto))\n",
    "    hpd.append(round(max(hdv), roundto))\n",
    "    ite = iter(hpd)\n",
    "    hpd = list(zip(ite, ite))\n",
    "    modes = []\n",
    "    for value in hpd:\n",
    "        x_hpd = x[(x > value[0]) & (x < value[1])]\n",
    "        y_hpd = y[(x > value[0]) & (x < value[1])]\n",
    "        modes.append(round(x_hpd[np.argmax(y_hpd)], roundto))\n",
    "    return hpd, x, y, modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"output_%d.csv\"%(i) for i in np.arange(1,4)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>0.15</td>\n",
       "      <td>-0.005</td>\n",
       "      <td>-0.044859</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides_dash_acidifaciens</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.145</td>\n",
       "      <td>1.835619</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Akkermansia_dash_muciniphila</td>\n",
       "      <td>-0.27</td>\n",
       "      <td>0.04</td>\n",
       "      <td>-0.115</td>\n",
       "      <td>-1.416896</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Faecalibaculum</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-0.040</td>\n",
       "      <td>-0.531562</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Parasutterella</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-0.040</td>\n",
       "      <td>-0.540205</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Type                           Taxa  Left  Right  Middle  \\\n",
       "0  basal_growth_rate                 Muribaculaceae -0.16   0.15  -0.005   \n",
       "1  basal_growth_rate  Bacteroides_dash_acidifaciens -0.01   0.30   0.145   \n",
       "2  basal_growth_rate   Akkermansia_dash_muciniphila -0.27   0.04  -0.115   \n",
       "3  basal_growth_rate                 Faecalibaculum -0.20   0.12  -0.040   \n",
       "4  basal_growth_rate                 Parasutterella -0.20   0.12  -0.040   \n",
       "\n",
       "        SNR  Significant  \n",
       "0 -0.044859        False  \n",
       "1  1.835619        False  \n",
       "2 -1.416896        False  \n",
       "3 -0.531562        False  \n",
       "4 -0.540205        False  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "    \n",
    "# basal growth rate\n",
    "for c in bacterial_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# inulin response\n",
    "for c in bacterial_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['inulin_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "\n",
    "# pairwise interactions\n",
    "for c1 in bacterial_taxa:\n",
    "    for c2 in bacterial_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stan_output_summary.to_excel('bayesian_regression_summary_top20.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>-0.99</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>-0.505</td>\n",
       "      <td>-2.050357</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Bifidobacterium</td>\n",
       "      <td>1.57</td>\n",
       "      <td>2.19</td>\n",
       "      <td>1.880</td>\n",
       "      <td>12.046869</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Anaerostipes</td>\n",
       "      <td>0.94</td>\n",
       "      <td>1.56</td>\n",
       "      <td>1.250</td>\n",
       "      <td>7.995026</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.425</td>\n",
       "      <td>2.720541</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>-0.66</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>-0.350</td>\n",
       "      <td>-2.239965</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Coprococcus</td>\n",
       "      <td>-0.65</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>-0.340</td>\n",
       "      <td>-2.163012</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Blautia', 'Blautia')</td>\n",
       "      <td>-3.91</td>\n",
       "      <td>-1.19</td>\n",
       "      <td>-2.550</td>\n",
       "      <td>-3.696627</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Blautia', 'Faecalibacterium')</td>\n",
       "      <td>-3.30</td>\n",
       "      <td>-0.11</td>\n",
       "      <td>-1.705</td>\n",
       "      <td>-2.138517</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Blautia', 'Bifidobacterium')</td>\n",
       "      <td>-3.51</td>\n",
       "      <td>-0.87</td>\n",
       "      <td>-2.190</td>\n",
       "      <td>-3.272662</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bifidobacterium', 'Bifidobacterium')</td>\n",
       "      <td>1.35</td>\n",
       "      <td>3.97</td>\n",
       "      <td>2.660</td>\n",
       "      <td>3.995891</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Prevotella', 'Faecalibacterium')</td>\n",
       "      <td>0.23</td>\n",
       "      <td>3.40</td>\n",
       "      <td>1.815</td>\n",
       "      <td>2.260957</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Prevotella', 'Prevotella')</td>\n",
       "      <td>-6.48</td>\n",
       "      <td>-4.34</td>\n",
       "      <td>-5.410</td>\n",
       "      <td>-9.988830</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                    Taxa  Left  \\\n",
       "6       basal_growth_rate                              Prevotella -0.99   \n",
       "22        inulin_response                         Bifidobacterium  1.57   \n",
       "24        inulin_response                            Anaerostipes  0.94   \n",
       "26        inulin_response                              Prevotella  0.12   \n",
       "28        inulin_response                             Bacteroides -0.66   \n",
       "33        inulin_response                             Coprococcus -0.65   \n",
       "40   pairwise_interaction                  ('Blautia', 'Blautia') -3.91   \n",
       "41   pairwise_interaction         ('Blautia', 'Faecalibacterium') -3.30   \n",
       "42   pairwise_interaction          ('Blautia', 'Bifidobacterium') -3.51   \n",
       "82   pairwise_interaction  ('Bifidobacterium', 'Bifidobacterium')  1.35   \n",
       "161  pairwise_interaction      ('Prevotella', 'Faecalibacterium')  0.23   \n",
       "166  pairwise_interaction            ('Prevotella', 'Prevotella') -6.48   \n",
       "\n",
       "     Right  Middle        SNR  Significant  \n",
       "6    -0.02  -0.505  -2.050357         True  \n",
       "22    2.19   1.880  12.046869         True  \n",
       "24    1.56   1.250   7.995026         True  \n",
       "26    0.73   0.425   2.720541         True  \n",
       "28   -0.04  -0.350  -2.239965         True  \n",
       "33   -0.03  -0.340  -2.163012         True  \n",
       "40   -1.19  -2.550  -3.696627         True  \n",
       "41   -0.11  -1.705  -2.138517         True  \n",
       "42   -0.87  -2.190  -3.272662         True  \n",
       "82    3.97   2.660   3.995891         True  \n",
       "161   3.40   1.815   2.260957         True  \n",
       "166  -4.34  -5.410  -9.988830         True  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stan_output_summary = pd.read_excel('bayesian_regression_summary_top20.xlsx', index_col=0)\n",
    "df_stan_output_summary = df_stan_output_summary[df_stan_output_summary.Significant==True]\n",
    "df_stan_output_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
