{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.interpolate import CubicSpline\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "%run -i '../../utils.py'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bac = pd.read_excel('./16S_relative_abundance.xlsx', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata for inulin and control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_type</th>\n",
       "      <th>Host_Age</th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>Time_point_original</th>\n",
       "      <th>Time_point</th>\n",
       "      <th>semester</th>\n",
       "      <th>status</th>\n",
       "      <th>Group</th>\n",
       "      <th>dose</th>\n",
       "      <th>Acetate</th>\n",
       "      <th>Propionate</th>\n",
       "      <th>Butyrate</th>\n",
       "      <th>total_SCFA</th>\n",
       "      <th>Collection_Date</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6444565</th>\n",
       "      <td>feces</td>\n",
       "      <td>17-29 yo</td>\n",
       "      <td>U222</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>fall16</td>\n",
       "      <td>before</td>\n",
       "      <td>accessible</td>\n",
       "      <td>0</td>\n",
       "      <td>73.981150</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.548664</td>\n",
       "      <td>84.529814</td>\n",
       "      <td>2016/9/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443884</th>\n",
       "      <td>feces</td>\n",
       "      <td>17-29 yo</td>\n",
       "      <td>U222</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>fall16</td>\n",
       "      <td>before</td>\n",
       "      <td>accessible</td>\n",
       "      <td>0</td>\n",
       "      <td>71.237839</td>\n",
       "      <td>15.792592</td>\n",
       "      <td>18.146317</td>\n",
       "      <td>105.176748</td>\n",
       "      <td>2016/9/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444382</th>\n",
       "      <td>feces</td>\n",
       "      <td>17-29 yo</td>\n",
       "      <td>U222</td>\n",
       "      <td>3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>fall16</td>\n",
       "      <td>before</td>\n",
       "      <td>accessible</td>\n",
       "      <td>0</td>\n",
       "      <td>29.743311</td>\n",
       "      <td>5.881475</td>\n",
       "      <td>10.214394</td>\n",
       "      <td>45.839180</td>\n",
       "      <td>2016/9/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444917</th>\n",
       "      <td>feces</td>\n",
       "      <td>17-29 yo</td>\n",
       "      <td>U225</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>fall16</td>\n",
       "      <td>before</td>\n",
       "      <td>accessible</td>\n",
       "      <td>0</td>\n",
       "      <td>28.019063</td>\n",
       "      <td>13.531843</td>\n",
       "      <td>8.422327</td>\n",
       "      <td>49.973233</td>\n",
       "      <td>2016/9/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443930</th>\n",
       "      <td>feces</td>\n",
       "      <td>17-29 yo</td>\n",
       "      <td>U225</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>fall16</td>\n",
       "      <td>before</td>\n",
       "      <td>accessible</td>\n",
       "      <td>0</td>\n",
       "      <td>17.114913</td>\n",
       "      <td>8.984568</td>\n",
       "      <td>3.395816</td>\n",
       "      <td>29.495297</td>\n",
       "      <td>2016/9/1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample_type  Host_Age Subject_ID Time_point_original  Time_point  \\\n",
       "#SampleID                                                                     \n",
       "SRR6444565       feces  17-29 yo       U222                   1         1.0   \n",
       "SRR6443884       feces  17-29 yo       U222                   2         2.0   \n",
       "SRR6444382       feces  17-29 yo       U222                   3         3.0   \n",
       "SRR6444917       feces  17-29 yo       U225                   1         1.0   \n",
       "SRR6443930       feces  17-29 yo       U225                   2         2.0   \n",
       "\n",
       "           semester  status       Group  dose    Acetate  Propionate  \\\n",
       "#SampleID                                                              \n",
       "SRR6444565   fall16  before  accessible     0  73.981150         NaN   \n",
       "SRR6443884   fall16  before  accessible     0  71.237839   15.792592   \n",
       "SRR6444382   fall16  before  accessible     0  29.743311    5.881475   \n",
       "SRR6444917   fall16  before  accessible     0  28.019063   13.531843   \n",
       "SRR6443930   fall16  before  accessible     0  17.114913    8.984568   \n",
       "\n",
       "             Butyrate  total_SCFA Collection_Date  \n",
       "#SampleID                                          \n",
       "SRR6444565  10.548664   84.529814        2016/9/1  \n",
       "SRR6443884  18.146317  105.176748        2016/9/1  \n",
       "SRR6444382  10.214394   45.839180        2016/9/1  \n",
       "SRR6444917   8.422327   49.973233        2016/9/1  \n",
       "SRR6443930   3.395816   29.495297        2016/9/1  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv('metadata.txt', sep='\\t', index_col=0)\n",
    "#df_meta = df_meta[['Group','Subject_ID','status','dose']]\n",
    "# df_meta = df_meta[(df_meta.Group=='inulin') | (df_meta.Group=='accessible')]\n",
    "# df_meta['Day'] = [0 if s=='before' else 21 for s in df_meta.status]\n",
    "# df_meta['Diet'] = [1 if g=='inulin' else 0 for g in df_meta['Group']]\n",
    "# df_meta = df_meta.drop('Group', axis=1)\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 20 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Prevotella</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Parabacteroides</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>Parasutterella_dash_excrementihominis</th>\n",
       "      <th>Bacteroides_dash_plebeius_dash_DSM_dash_17135</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Lachnospiraceae_dash_UCG_dash_004</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Bacteroides_dash_coprocola</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR6444565</th>\n",
       "      <td>0.197580</td>\n",
       "      <td>0.136827</td>\n",
       "      <td>0.024022</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.028962</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.037984</td>\n",
       "      <td>0.007446</td>\n",
       "      <td>0.009165</td>\n",
       "      <td>0.018580</td>\n",
       "      <td>0.006158</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.114631</td>\n",
       "      <td>0.006623</td>\n",
       "      <td>0.003508</td>\n",
       "      <td>0.036051</td>\n",
       "      <td>0.027852</td>\n",
       "      <td>0.001539</td>\n",
       "      <td>0.009201</td>\n",
       "      <td>0.000215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443884</th>\n",
       "      <td>0.243131</td>\n",
       "      <td>0.099128</td>\n",
       "      <td>0.017050</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.032008</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.037064</td>\n",
       "      <td>0.006764</td>\n",
       "      <td>0.004916</td>\n",
       "      <td>0.014191</td>\n",
       "      <td>0.005230</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.125837</td>\n",
       "      <td>0.005335</td>\n",
       "      <td>0.002789</td>\n",
       "      <td>0.034100</td>\n",
       "      <td>0.022873</td>\n",
       "      <td>0.003626</td>\n",
       "      <td>0.007985</td>\n",
       "      <td>0.000070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444382</th>\n",
       "      <td>0.214061</td>\n",
       "      <td>0.110743</td>\n",
       "      <td>0.028395</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.031565</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.037600</td>\n",
       "      <td>0.006981</td>\n",
       "      <td>0.007370</td>\n",
       "      <td>0.020525</td>\n",
       "      <td>0.007231</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.115360</td>\n",
       "      <td>0.006869</td>\n",
       "      <td>0.003560</td>\n",
       "      <td>0.032734</td>\n",
       "      <td>0.029730</td>\n",
       "      <td>0.004533</td>\n",
       "      <td>0.008121</td>\n",
       "      <td>0.000167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6444917</th>\n",
       "      <td>0.578520</td>\n",
       "      <td>0.063804</td>\n",
       "      <td>0.117987</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.037393</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000630</td>\n",
       "      <td>0.003550</td>\n",
       "      <td>0.016358</td>\n",
       "      <td>0.014168</td>\n",
       "      <td>0.000299</td>\n",
       "      <td>0.026942</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.003782</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000630</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.001194</td>\n",
       "      <td>0.000199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6443930</th>\n",
       "      <td>0.610679</td>\n",
       "      <td>0.055143</td>\n",
       "      <td>0.050196</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.079698</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.004404</td>\n",
       "      <td>0.008386</td>\n",
       "      <td>0.006154</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.023409</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000181</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Bacteroides  Faecalibacterium  Lachnospiraceae  \\\n",
       "#SampleID                                                    \n",
       "SRR6444565     0.197580          0.136827         0.024022   \n",
       "SRR6443884     0.243131          0.099128         0.017050   \n",
       "SRR6444382     0.214061          0.110743         0.028395   \n",
       "SRR6444917     0.578520          0.063804         0.117987   \n",
       "SRR6443930     0.610679          0.055143         0.050196   \n",
       "\n",
       "            Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_  \\\n",
       "#SampleID                                                                                           \n",
       "SRR6444565                                           0.000215                                       \n",
       "SRR6443884                                           0.000070                                       \n",
       "SRR6444382                                           0.000167                                       \n",
       "SRR6444917                                           0.000199                                       \n",
       "SRR6443930                                           0.000181                                       \n",
       "\n",
       "            Alistipes  Prevotella  \\\n",
       "#SampleID                           \n",
       "SRR6444565   0.028962    0.000215   \n",
       "SRR6443884   0.032008    0.000070   \n",
       "SRR6444382   0.031565    0.000167   \n",
       "SRR6444917   0.037393    0.000199   \n",
       "SRR6443930   0.079698    0.000181   \n",
       "\n",
       "            Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  \\\n",
       "#SampleID                                                        \n",
       "SRR6444565                                           0.037984    \n",
       "SRR6443884                                           0.037064    \n",
       "SRR6444382                                           0.037600    \n",
       "SRR6444917                                           0.000630    \n",
       "SRR6443930                                           0.000181    \n",
       "\n",
       "            Parabacteroides  Bifidobacterium   Blautia  Anaerostipes  \\\n",
       "#SampleID                                                              \n",
       "SRR6444565         0.007446         0.009165  0.018580      0.006158   \n",
       "SRR6443884         0.006764         0.004916  0.014191      0.005230   \n",
       "SRR6444382         0.006981         0.007370  0.020525      0.007231   \n",
       "SRR6444917         0.003550         0.016358  0.014168      0.000299   \n",
       "SRR6443930         0.004404         0.008386  0.006154      0.000181   \n",
       "\n",
       "            Parasutterella_dash_excrementihominis  \\\n",
       "#SampleID                                           \n",
       "SRR6444565                               0.000215   \n",
       "SRR6443884                               0.000070   \n",
       "SRR6444382                               0.000167   \n",
       "SRR6444917                               0.026942   \n",
       "SRR6443930                               0.023409   \n",
       "\n",
       "            Bacteroides_dash_plebeius_dash_DSM_dash_17135  Ruminococcus  \\\n",
       "#SampleID                                                                 \n",
       "SRR6444565                                       0.114631      0.006623   \n",
       "SRR6443884                                       0.125837      0.005335   \n",
       "SRR6444382                                       0.115360      0.006869   \n",
       "SRR6444917                                       0.000199      0.003782   \n",
       "SRR6443930                                       0.000181      0.000181   \n",
       "\n",
       "            Fusicatenibacter  gut_dash_metagenome  Roseburia  \\\n",
       "#SampleID                                                      \n",
       "SRR6444565          0.003508             0.036051   0.027852   \n",
       "SRR6443884          0.002789             0.034100   0.022873   \n",
       "SRR6444382          0.003560             0.032734   0.029730   \n",
       "SRR6444917          0.000199             0.000630   0.000199   \n",
       "SRR6443930          0.000181             0.000181   0.000181   \n",
       "\n",
       "            Lachnospiraceae_dash_UCG_dash_004  metagenome  \\\n",
       "#SampleID                                                   \n",
       "SRR6444565                           0.001539    0.009201   \n",
       "SRR6443884                           0.003626    0.007985   \n",
       "SRR6444382                           0.004533    0.008121   \n",
       "SRR6444917                           0.000199    0.001194   \n",
       "SRR6443930                           0.000181    0.000181   \n",
       "\n",
       "            Bacteroides_dash_coprocola  \n",
       "#SampleID                               \n",
       "SRR6444565                    0.000215  \n",
       "SRR6443884                    0.000070  \n",
       "SRR6444382                    0.000167  \n",
       "SRR6444917                    0.000199  \n",
       "SRR6443930                    0.000181  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 20\n",
    "df_16S_T = df_16S.loc[df_meta.index].T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Bacteroides',\n",
       " 'Faecalibacterium',\n",
       " 'Lachnospiraceae',\n",
       " 'Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_',\n",
       " 'Alistipes',\n",
       " 'Prevotella',\n",
       " 'Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57',\n",
       " 'Parabacteroides',\n",
       " 'Bifidobacterium',\n",
       " 'Blautia',\n",
       " 'Anaerostipes',\n",
       " 'Parasutterella_dash_excrementihominis',\n",
       " 'Bacteroides_dash_plebeius_dash_DSM_dash_17135',\n",
       " 'Ruminococcus',\n",
       " 'Fusicatenibacter',\n",
       " 'gut_dash_metagenome',\n",
       " 'Roseburia',\n",
       " 'Lachnospiraceae_dash_UCG_dash_004',\n",
       " 'metagenome',\n",
       " 'Bacteroides_dash_coprocola']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_16S_topX.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Prevotella</th>\n",
       "      <th>...</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>Parasutterella_dash_excrementihominis</th>\n",
       "      <th>Bacteroides_dash_plebeius_dash_DSM_dash_17135</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Lachnospiraceae_dash_UCG_dash_004</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Bacteroides_dash_coprocola</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U042</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.444946</td>\n",
       "      <td>0.100588</td>\n",
       "      <td>0.080670</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.027035</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003644</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.009113</td>\n",
       "      <td>0.022899</td>\n",
       "      <td>0.007670</td>\n",
       "      <td>0.025910</td>\n",
       "      <td>0.007428</td>\n",
       "      <td>0.011713</td>\n",
       "      <td>0.000128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U042</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.471585</td>\n",
       "      <td>0.037131</td>\n",
       "      <td>0.080455</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.038842</td>\n",
       "      <td>0.000406</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003687</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.010182</td>\n",
       "      <td>0.008189</td>\n",
       "      <td>0.016515</td>\n",
       "      <td>0.047139</td>\n",
       "      <td>0.008635</td>\n",
       "      <td>0.009576</td>\n",
       "      <td>0.000254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U068</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.045633</td>\n",
       "      <td>0.056045</td>\n",
       "      <td>0.043214</td>\n",
       "      <td>0.177269</td>\n",
       "      <td>0.000270</td>\n",
       "      <td>0.328534</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000425</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0.021873</td>\n",
       "      <td>0.004559</td>\n",
       "      <td>0.011077</td>\n",
       "      <td>0.002486</td>\n",
       "      <td>0.006962</td>\n",
       "      <td>0.011519</td>\n",
       "      <td>0.015651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U068</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.070620</td>\n",
       "      <td>0.093106</td>\n",
       "      <td>0.044719</td>\n",
       "      <td>0.120436</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>0.258661</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002214</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>0.008714</td>\n",
       "      <td>0.007876</td>\n",
       "      <td>0.017048</td>\n",
       "      <td>0.009354</td>\n",
       "      <td>0.004645</td>\n",
       "      <td>0.018583</td>\n",
       "      <td>0.004436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U075</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.227864</td>\n",
       "      <td>0.133095</td>\n",
       "      <td>0.049664</td>\n",
       "      <td>0.013882</td>\n",
       "      <td>0.017764</td>\n",
       "      <td>0.000785</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014899</td>\n",
       "      <td>0.016442</td>\n",
       "      <td>0.000179</td>\n",
       "      <td>0.014111</td>\n",
       "      <td>0.045612</td>\n",
       "      <td>0.011263</td>\n",
       "      <td>0.018615</td>\n",
       "      <td>0.002632</td>\n",
       "      <td>0.017789</td>\n",
       "      <td>0.000179</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Subject_ID  status  Day  Diet  Bacteroides  Faecalibacterium  \\\n",
       "0       U042  before    0     1     0.444946          0.100588   \n",
       "1       U042  during   21     1     0.471585          0.037131   \n",
       "2       U068  before    0     1     0.045633          0.056045   \n",
       "3       U068  during   21     1     0.070620          0.093106   \n",
       "4       U075  before    0     1     0.227864          0.133095   \n",
       "\n",
       "   Lachnospiraceae  \\\n",
       "0         0.080670   \n",
       "1         0.080455   \n",
       "2         0.043214   \n",
       "3         0.044719   \n",
       "4         0.049664   \n",
       "\n",
       "   Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_  \\\n",
       "0                                           0.000128                                       \n",
       "1                                           0.000254                                       \n",
       "2                                           0.177269                                       \n",
       "3                                           0.120436                                       \n",
       "4                                           0.013882                                       \n",
       "\n",
       "   Alistipes  Prevotella  ...  Anaerostipes  \\\n",
       "0   0.027035    0.000128  ...      0.003644   \n",
       "1   0.038842    0.000406  ...      0.003687   \n",
       "2   0.000270    0.328534  ...      0.000425   \n",
       "3   0.000256    0.258661  ...      0.002214   \n",
       "4   0.017764    0.000785  ...      0.014899   \n",
       "\n",
       "   Parasutterella_dash_excrementihominis  \\\n",
       "0                               0.000128   \n",
       "1                               0.000254   \n",
       "2                               0.000161   \n",
       "3                               0.000199   \n",
       "4                               0.016442   \n",
       "\n",
       "   Bacteroides_dash_plebeius_dash_DSM_dash_17135  Ruminococcus  \\\n",
       "0                                       0.000128      0.009113   \n",
       "1                                       0.000254      0.010182   \n",
       "2                                       0.000161      0.021873   \n",
       "3                                       0.000153      0.008714   \n",
       "4                                       0.000179      0.014111   \n",
       "\n",
       "   Fusicatenibacter  gut_dash_metagenome  Roseburia  \\\n",
       "0          0.022899             0.007670   0.025910   \n",
       "1          0.008189             0.016515   0.047139   \n",
       "2          0.004559             0.011077   0.002486   \n",
       "3          0.007876             0.017048   0.009354   \n",
       "4          0.045612             0.011263   0.018615   \n",
       "\n",
       "   Lachnospiraceae_dash_UCG_dash_004  metagenome  Bacteroides_dash_coprocola  \n",
       "0                           0.007428    0.011713                    0.000128  \n",
       "1                           0.008635    0.009576                    0.000254  \n",
       "2                           0.006962    0.011519                    0.015651  \n",
       "3                           0.004645    0.018583                    0.004436  \n",
       "4                           0.002632    0.017789                    0.000179  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX_w_meta = pd.merge(df_meta, df_16S_topX/df_16S_topX.max().max(), left_index=True, right_index=True, how='inner')\n",
    "df_16S_topX_w_meta = df_16S_topX_w_meta.groupby(['Subject_ID','status']).agg(np.mean).reset_index()\n",
    "df_16S_topX_w_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate log-derivatives (dlog(B)/dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>status</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Prevotella</th>\n",
       "      <th>...</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>Parasutterella_dash_excrementihominis</th>\n",
       "      <th>Bacteroides_dash_plebeius_dash_DSM_dash_17135</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>Lachnospiraceae_dash_UCG_dash_004</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>Bacteroides_dash_coprocola</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U042</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001269</td>\n",
       "      <td>-0.003022</td>\n",
       "      <td>-0.000010</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>5.622348e-04</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>5.970561e-06</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>-0.000700</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.001011</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>-0.000102</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U042</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001269</td>\n",
       "      <td>-0.003022</td>\n",
       "      <td>-0.000010</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>5.622348e-04</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>5.970561e-06</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>-0.000700</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.001011</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>-0.000102</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U068</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001190</td>\n",
       "      <td>0.001765</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>-0.002706</td>\n",
       "      <td>-7.087060e-07</td>\n",
       "      <td>-0.003327</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>-3.810299e-07</td>\n",
       "      <td>-0.000627</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>0.000284</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>-0.000110</td>\n",
       "      <td>0.000336</td>\n",
       "      <td>-0.000534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U068</td>\n",
       "      <td>during</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001190</td>\n",
       "      <td>0.001765</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>-0.002706</td>\n",
       "      <td>-7.087060e-07</td>\n",
       "      <td>-0.003327</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>-3.810299e-07</td>\n",
       "      <td>-0.000627</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>0.000284</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>-0.000110</td>\n",
       "      <td>0.000336</td>\n",
       "      <td>-0.000534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U075</td>\n",
       "      <td>before</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005197</td>\n",
       "      <td>-0.003111</td>\n",
       "      <td>-0.000158</td>\n",
       "      <td>-0.000459</td>\n",
       "      <td>6.566401e-04</td>\n",
       "      <td>-0.000024</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.000248</td>\n",
       "      <td>-4.237658e-06</td>\n",
       "      <td>0.000450</td>\n",
       "      <td>-0.000997</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>-0.000491</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>-0.000448</td>\n",
       "      <td>-0.000004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Subject_ID  status  Day  Diet  Bacteroides  Faecalibacterium  \\\n",
       "0       U042  before    0     1     0.001269         -0.003022   \n",
       "1       U042  during   21     1     0.001269         -0.003022   \n",
       "2       U068  before    0     1     0.001190          0.001765   \n",
       "3       U068  during   21     1     0.001190          0.001765   \n",
       "4       U075  before    0     1     0.005197         -0.003111   \n",
       "\n",
       "   Lachnospiraceae  \\\n",
       "0        -0.000010   \n",
       "1        -0.000010   \n",
       "2         0.000072   \n",
       "3         0.000072   \n",
       "4        -0.000158   \n",
       "\n",
       "   Trichuris_dash_trichiura_dash__leftroundbrackethuman_dash_whipworm_rightroundbracket_  \\\n",
       "0                                           0.000006                                       \n",
       "1                                           0.000006                                       \n",
       "2                                          -0.002706                                       \n",
       "3                                          -0.002706                                       \n",
       "4                                          -0.000459                                       \n",
       "\n",
       "      Alistipes  Prevotella  ...  Anaerostipes  \\\n",
       "0  5.622348e-04    0.000013  ...      0.000002   \n",
       "1  5.622348e-04    0.000013  ...      0.000002   \n",
       "2 -7.087060e-07   -0.003327  ...      0.000085   \n",
       "3 -7.087060e-07   -0.003327  ...      0.000085   \n",
       "4  6.566401e-04   -0.000024  ...      0.000185   \n",
       "\n",
       "   Parasutterella_dash_excrementihominis  \\\n",
       "0                               0.000006   \n",
       "1                               0.000006   \n",
       "2                               0.000002   \n",
       "3                               0.000002   \n",
       "4                               0.000248   \n",
       "\n",
       "   Bacteroides_dash_plebeius_dash_DSM_dash_17135  Ruminococcus  \\\n",
       "0                                   5.970561e-06      0.000051   \n",
       "1                                   5.970561e-06      0.000051   \n",
       "2                                  -3.810299e-07     -0.000627   \n",
       "3                                  -3.810299e-07     -0.000627   \n",
       "4                                  -4.237658e-06      0.000450   \n",
       "\n",
       "   Fusicatenibacter  gut_dash_metagenome  Roseburia  \\\n",
       "0         -0.000700             0.000421   0.001011   \n",
       "1         -0.000700             0.000421   0.001011   \n",
       "2          0.000158             0.000284   0.000327   \n",
       "3          0.000158             0.000284   0.000327   \n",
       "4         -0.000997             0.000204  -0.000491   \n",
       "\n",
       "   Lachnospiraceae_dash_UCG_dash_004  metagenome  Bacteroides_dash_coprocola  \n",
       "0                           0.000057   -0.000102                    0.000006  \n",
       "1                           0.000057   -0.000102                    0.000006  \n",
       "2                          -0.000110    0.000336                   -0.000534  \n",
       "3                          -0.000110    0.000336                   -0.000534  \n",
       "4                           0.000112   -0.000448                   -0.000004  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_deriv = deepcopy(df_16S_topX_w_meta)\n",
    "for curr_sub in set(df_deriv.Subject_ID):\n",
    "    curr_df = df_deriv[df_deriv.Subject_ID==curr_sub].sort_values(by='Day')\n",
    "    for taxon in bacterial_taxa:\n",
    "        xdata = np.array(curr_df['Day'])\n",
    "        ydata = np.array(curr_df[taxon])\n",
    "        cs = CubicSpline(xdata, ydata)\n",
    "        csd1 = cs.derivative(nu=1)\n",
    "        ydata_d1 = csd1(xdata)\n",
    "        df_deriv.loc[df_deriv.Subject_ID==curr_sub, taxon] = ydata_d1\n",
    "df_deriv.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct regression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.89588099,  0.89588099,  0.83846764, ..., -2.43909054,\n",
       "       -0.01323803, -0.01323803])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ymat should be samples by bacteria\n",
    "Ymat = df_deriv[bacterial_taxa].values\n",
    "Ymat = Ymat.flatten(order='F')\n",
    "Ymat = StandardScaler().fit_transform(Ymat.reshape(-1,1)).reshape(1,-1)[0] # standardize\n",
    "Ymat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3560"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Ymat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.00000000e+00, 1.00000000e+00, 4.44946301e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 1.00000000e+00, 4.71585345e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 1.00000000e+00, 4.56331153e-02, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       ...,\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        9.46741789e-03, 6.74048698e-03, 6.98077419e-04],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        4.20216398e-02, 3.61930501e-03, 2.97519882e-04],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        4.50391700e-02, 4.41138219e-03, 7.73246735e-04]])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Xmat = np.zeros(shape=(topX*len(df_deriv.index), (topX+2)*topX))\n",
    "for k in np.arange(topX):\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)] = 1\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+1] = df_deriv.Diet.values\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+2:(k+1)*(topX+2)] = df_16S_topX_w_meta[bacterial_taxa].values\n",
    "Xmat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write stan program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "json_str = '{\\n\"N\" : %d,\\n'%(len(Ymat))\n",
    "\n",
    "# y variable\n",
    "json_str += '\\\"dlogX\\\" : [%s],\\n'%(','.join(list(Ymat.astype(str))))\n",
    "\n",
    "# x variable\n",
    "for k1,c1 in enumerate(bacterial_taxa):\n",
    "    # growth rate\n",
    "    json_str += '\\\"growth_rate_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)].astype(str))))\n",
    "    # diet response\n",
    "    json_str += '\\\"inulin_response_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)+1].astype(str))))\n",
    "    # bacterial interactions\n",
    "    for k2,c2 in enumerate(bacterial_taxa):\n",
    "        v = list(Xmat[:,k1*(topX+2)+2+k2].astype(str))\n",
    "        json_str += '\\\"pairwise_interaction_%s_%s\\\" : [%s]'%(c1,c2,','.join(v))\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            json_str += '\\n}'\n",
    "        else:\n",
    "            json_str += ',\\n'\n",
    "#print(json_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.data.json\", \"w\")\n",
    "text_file.write(\"%s\" % json_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# data block\n",
    "model_str = 'data {\\n'\n",
    "model_str += '\\tint<lower=0> N;\\n'\n",
    "model_str += '\\tvector[N] dlogX;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\tvector[N] growth_rate_%s;\\n'%(c1)\n",
    "    model_str += '\\tvector[N] inulin_response_%s;\\n'%(c1)\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tvector[N] pairwise_interaction_%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'\n",
    "\n",
    "# parameter block\n",
    "model_str += 'parameters {\\n\\treal<lower=0,upper=1> sigma;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\treal alpha__%s;\\n'%(c1) # growth rate\n",
    "    model_str += '\\treal epsilon__%s;\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\treal beta__%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'       \n",
    "        \n",
    "# model block\n",
    "model_str += 'model {\\n\\tsigma ~ uniform(0,1);\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\talpha__%s ~ normal(0,1);\\n'%(c1) # growth rate\n",
    "    model_str += '\\tepsilon__%s ~ normal(0,1);\\n'%(c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tbeta__%s_%s ~ normal(0,1);\\n'%(c1,c2)\n",
    "model_str += '\\tdlogX ~ normal('\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += 'alpha__%s*growth_rate_%s+'%(c1,c1) # growth rate\n",
    "    model_str += 'epsilon__%s*inulin_response_%s+'%(c1,c1) # inulin response\n",
    "    for c2 in bacterial_taxa:\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s'%(c1,c2,c1,c2)\n",
    "        else:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s+'%(c1,c2,c1,c2)\n",
    "model_str += ', sigma);\\n}'\n",
    "#print(model_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.stan\", \"w\")\n",
    "text_file.write(\"%s\" % model_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot stan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hpd_grid(sample, alpha=0.05, roundto=2):\n",
    "    \"\"\"Calculate highest posterior density (HPD) of array for given alpha. \n",
    "    The HPD is the minimum width Bayesian credible interval (BCI). \n",
    "    The function works for multimodal distributions, returning more than one mode\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    sample : Numpy array or python list\n",
    "        An array containing MCMC samples\n",
    "    alpha : float\n",
    "        Desired probability of type I error (defaults to 0.05)\n",
    "    roundto: integer\n",
    "        Number of digits after the decimal point for the results\n",
    "    Returns\n",
    "    ----------\n",
    "    hpd: array with the lower \n",
    "          \n",
    "    \"\"\"\n",
    "    sample = np.asarray(sample)\n",
    "    sample = sample[~np.isnan(sample)]\n",
    "    # get upper and lower bounds\n",
    "    l = np.min(sample)\n",
    "    u = np.max(sample)\n",
    "    density = kde.gaussian_kde(sample)\n",
    "    x = np.linspace(l, u, 2000)\n",
    "    y = density.evaluate(x)\n",
    "    #y = density.evaluate(x, l, u) waitting for PR to be accepted\n",
    "    xy_zipped = zip(x, y/np.sum(y))\n",
    "    xy = sorted(xy_zipped, key=lambda x: x[1], reverse=True)\n",
    "    xy_cum_sum = 0\n",
    "    hdv = []\n",
    "    for val in xy:\n",
    "        xy_cum_sum += val[1]\n",
    "        hdv.append(val[0])\n",
    "        if xy_cum_sum >= (1-alpha):\n",
    "            break\n",
    "    hdv.sort()\n",
    "    diff = (u-l)/20  # differences of 5%\n",
    "    hpd = []\n",
    "    hpd.append(round(min(hdv), roundto))\n",
    "    for i in range(1, len(hdv)):\n",
    "        if hdv[i]-hdv[i-1] >= diff:\n",
    "            hpd.append(round(hdv[i-1], roundto))\n",
    "            hpd.append(round(hdv[i], roundto))\n",
    "    hpd.append(round(max(hdv), roundto))\n",
    "    ite = iter(hpd)\n",
    "    hpd = list(zip(ite, ite))\n",
    "    modes = []\n",
    "    for value in hpd:\n",
    "        x_hpd = x[(x > value[0]) & (x < value[1])]\n",
    "        y_hpd = y[(x > value[0]) & (x < value[1])]\n",
    "        modes.append(round(x_hpd[np.argmax(y_hpd)], roundto))\n",
    "    return hpd, x, y, modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"output_%d_inulin.csv\"%(i) for i in np.arange(1,4)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>-0.36</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.150</td>\n",
       "      <td>0.584667</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "      <td>-0.92</td>\n",
       "      <td>0.09</td>\n",
       "      <td>-0.415</td>\n",
       "      <td>-1.623619</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>-0.72</td>\n",
       "      <td>0.31</td>\n",
       "      <td>-0.205</td>\n",
       "      <td>-0.761524</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Trichuris_dash_trichiura_dash__leftroundbracke...</td>\n",
       "      <td>-0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.067067</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Alistipes</td>\n",
       "      <td>-0.35</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.155</td>\n",
       "      <td>0.600150</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Type                                               Taxa  Left  \\\n",
       "0  basal_growth_rate                                        Bacteroides -0.36   \n",
       "1  basal_growth_rate                                   Faecalibacterium -0.92   \n",
       "2  basal_growth_rate                                    Lachnospiraceae -0.72   \n",
       "3  basal_growth_rate  Trichuris_dash_trichiura_dash__leftroundbracke... -0.48   \n",
       "4  basal_growth_rate                                          Alistipes -0.35   \n",
       "\n",
       "   Right  Middle       SNR  Significant  \n",
       "0   0.66   0.150  0.584667        False  \n",
       "1   0.09  -0.415 -1.623619        False  \n",
       "2   0.31  -0.205 -0.761524        False  \n",
       "3   0.52   0.020  0.067067        False  \n",
       "4   0.66   0.155  0.600150        False  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "    \n",
    "# basal growth rate\n",
    "for c in bacterial_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# inulin response\n",
    "for c in bacterial_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['inulin_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "\n",
    "# pairwise interactions\n",
    "for c1 in bacterial_taxa:\n",
    "    for c2 in bacterial_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stan_output_summary.to_excel('bayesian_regression_summary_top20_inulin.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bifidobacterium</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.06</td>\n",
       "      <td>0.545</td>\n",
       "      <td>2.116647</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.555</td>\n",
       "      <td>3.799384</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Trichuris_dash_trichiura_dash__leftroundbracke...</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.415</td>\n",
       "      <td>2.871528</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...</td>\n",
       "      <td>-0.64</td>\n",
       "      <td>-0.08</td>\n",
       "      <td>-0.360</td>\n",
       "      <td>-2.507589</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Bifidobacterium</td>\n",
       "      <td>0.48</td>\n",
       "      <td>1.05</td>\n",
       "      <td>0.765</td>\n",
       "      <td>5.313565</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>inulin_response</td>\n",
       "      <td>Anaerostipes</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.71</td>\n",
       "      <td>0.420</td>\n",
       "      <td>2.886729</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Lachnospiraceae')</td>\n",
       "      <td>-4.73</td>\n",
       "      <td>-1.15</td>\n",
       "      <td>-2.940</td>\n",
       "      <td>-3.240980</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Trichuris_dash_trichiura_dash...</td>\n",
       "      <td>-3.53</td>\n",
       "      <td>-0.67</td>\n",
       "      <td>-2.100</td>\n",
       "      <td>-2.880639</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Bifidobacterium')</td>\n",
       "      <td>-4.12</td>\n",
       "      <td>-0.62</td>\n",
       "      <td>-2.370</td>\n",
       "      <td>-2.633815</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Faecalibacterium', 'Faecalibacterium')</td>\n",
       "      <td>0.59</td>\n",
       "      <td>3.96</td>\n",
       "      <td>2.275</td>\n",
       "      <td>2.663886</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Trichuris_dash_trichiura_dash__leftroundbrac...</td>\n",
       "      <td>-3.45</td>\n",
       "      <td>-0.36</td>\n",
       "      <td>-1.905</td>\n",
       "      <td>-2.409717</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Prevotella', 'Prevotella')</td>\n",
       "      <td>0.73</td>\n",
       "      <td>3.76</td>\n",
       "      <td>2.245</td>\n",
       "      <td>2.915239</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bifidobacterium', 'Bacteroides')</td>\n",
       "      <td>-1.74</td>\n",
       "      <td>-0.17</td>\n",
       "      <td>-0.955</td>\n",
       "      <td>-2.416430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bifidobacterium', 'Bifidobacterium')</td>\n",
       "      <td>4.48</td>\n",
       "      <td>8.04</td>\n",
       "      <td>6.260</td>\n",
       "      <td>6.932907</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                               Taxa  \\\n",
       "8       basal_growth_rate                                    Bifidobacterium   \n",
       "21        inulin_response                                   Faecalibacterium   \n",
       "23        inulin_response  Trichuris_dash_trichiura_dash__leftroundbracke...   \n",
       "26        inulin_response  Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_das...   \n",
       "28        inulin_response                                    Bifidobacterium   \n",
       "30        inulin_response                                       Anaerostipes   \n",
       "42   pairwise_interaction                 ('Bacteroides', 'Lachnospiraceae')   \n",
       "43   pairwise_interaction  ('Bacteroides', 'Trichuris_dash_trichiura_dash...   \n",
       "48   pairwise_interaction                 ('Bacteroides', 'Bifidobacterium')   \n",
       "61   pairwise_interaction           ('Faecalibacterium', 'Faecalibacterium')   \n",
       "112  pairwise_interaction  ('Trichuris_dash_trichiura_dash__leftroundbrac...   \n",
       "145  pairwise_interaction                       ('Prevotella', 'Prevotella')   \n",
       "200  pairwise_interaction                 ('Bifidobacterium', 'Bacteroides')   \n",
       "208  pairwise_interaction             ('Bifidobacterium', 'Bifidobacterium')   \n",
       "\n",
       "     Left  Right  Middle       SNR  Significant  \n",
       "8    0.03   1.06   0.545  2.116647         True  \n",
       "21   0.26   0.85   0.555  3.799384         True  \n",
       "23   0.13   0.70   0.415  2.871528         True  \n",
       "26  -0.64  -0.08  -0.360 -2.507589         True  \n",
       "28   0.48   1.05   0.765  5.313565         True  \n",
       "30   0.13   0.71   0.420  2.886729         True  \n",
       "42  -4.73  -1.15  -2.940 -3.240980         True  \n",
       "43  -3.53  -0.67  -2.100 -2.880639         True  \n",
       "48  -4.12  -0.62  -2.370 -2.633815         True  \n",
       "61   0.59   3.96   2.275  2.663886         True  \n",
       "112 -3.45  -0.36  -1.905 -2.409717         True  \n",
       "145  0.73   3.76   2.245  2.915239         True  \n",
       "200 -1.74  -0.17  -0.955 -2.416430         True  \n",
       "208  4.48   8.04   6.260  6.932907         True  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stan_output_summary = pd.read_excel('bayesian_regression_summary_top20_inulin.xlsx', index_col=0)\n",
    "df_stan_output_summary = df_stan_output_summary[df_stan_output_summary.Significant==True]\n",
    "df_stan_output_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
