{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from scipy.interpolate import CubicSpline\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import warnings\n",
    "import itertools\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "import statsmodels.stats.api as sms\n",
    "import scipy.stats.kde as kde"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read 16S data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2</th>\n",
       "      <th>ASF356</th>\n",
       "      <th>Acetatifactor</th>\n",
       "      <th>Acetatifactor_dash_muris</th>\n",
       "      <th>Acidovorax</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Acinetobacter_dash_sp_dot__dash_CIP_dash_56_dot_2</th>\n",
       "      <th>Acutalibacter_dash_muris</th>\n",
       "      <th>Adlercreutzia_dash_mucosicola</th>\n",
       "      <th>Adlercreutzia_dash_muris</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_ventriosum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_gnavus_dash_group</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_M153</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>marine_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>0.0</td>\n",
       "      <td>108301.701888</td>\n",
       "      <td>7101.750943</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>63028.039623</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>165115.709436</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.322701e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>0.0</td>\n",
       "      <td>405068.034693</td>\n",
       "      <td>50064.588558</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>80103.341692</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>106501.033841</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.003112e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>0.0</td>\n",
       "      <td>36169.603950</td>\n",
       "      <td>45332.570284</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>221840.237562</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.866352e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>0.0</td>\n",
       "      <td>484302.054587</td>\n",
       "      <td>10150.043060</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>105850.449056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>919303.900024</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.714537e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>0.0</td>\n",
       "      <td>91489.970533</td>\n",
       "      <td>42374.302141</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12519.680178</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>581683.602124</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.335310e+05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 304 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            A2         ASF356  Acetatifactor  Acetatifactor_dash_muris  \\\n",
       "SampleID                                                                 \n",
       "sample131  0.0  108301.701888    7101.750943                       0.0   \n",
       "sample132  0.0  405068.034693   50064.588558                       0.0   \n",
       "sample133  0.0   36169.603950   45332.570284                       0.0   \n",
       "sample134  0.0  484302.054587   10150.043060                       0.0   \n",
       "sample135  0.0   91489.970533   42374.302141                       0.0   \n",
       "\n",
       "           Acidovorax  Acinetobacter  \\\n",
       "SampleID                               \n",
       "sample131         0.0            0.0   \n",
       "sample132         0.0            0.0   \n",
       "sample133         0.0            0.0   \n",
       "sample134         0.0            0.0   \n",
       "sample135         0.0            0.0   \n",
       "\n",
       "           Acinetobacter_dash_sp_dot__dash_CIP_dash_56_dot_2  \\\n",
       "SampleID                                                       \n",
       "sample131                                                0.0   \n",
       "sample132                                                0.0   \n",
       "sample133                                                0.0   \n",
       "sample134                                                0.0   \n",
       "sample135                                                0.0   \n",
       "\n",
       "           Acutalibacter_dash_muris  Adlercreutzia_dash_mucosicola  \\\n",
       "SampleID                                                             \n",
       "sample131                       0.0                       0.000000   \n",
       "sample132                       0.0                       0.000000   \n",
       "sample133                       0.0                       0.000000   \n",
       "sample134                       0.0                       0.000000   \n",
       "sample135                       0.0                   12519.680178   \n",
       "\n",
       "           Adlercreutzia_dash_muris  ...  \\\n",
       "SampleID                             ...   \n",
       "sample131                       0.0  ...   \n",
       "sample132                       0.0  ...   \n",
       "sample133                       0.0  ...   \n",
       "sample134                       0.0  ...   \n",
       "sample135                       0.0  ...   \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_ventriosum_dash_group  \\\n",
       "SampleID                                                                                   \n",
       "sample131                                                0.0                               \n",
       "sample132                                                0.0                               \n",
       "sample133                                                0.0                               \n",
       "sample134                                                0.0                               \n",
       "sample135                                                0.0                               \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group  \\\n",
       "SampleID                                                                                     \n",
       "sample131                                       63028.039623                                 \n",
       "sample132                                       80103.341692                                 \n",
       "sample133                                           0.000000                                 \n",
       "sample134                                      105850.449056                                 \n",
       "sample135                                           0.000000                                 \n",
       "\n",
       "           _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_gnavus_dash_group  \\\n",
       "SampleID                                                                                \n",
       "sample131                                                0.0                            \n",
       "sample132                                                0.0                            \n",
       "sample133                                                0.0                            \n",
       "sample134                                                0.0                            \n",
       "sample135                                                0.0                            \n",
       "\n",
       "           _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "SampleID                                                                                 \n",
       "sample131                                                0.0                             \n",
       "sample132                                                0.0                             \n",
       "sample133                                                0.0                             \n",
       "sample134                                                0.0                             \n",
       "sample135                                                0.0                             \n",
       "\n",
       "           bacterium_dash_enrichment_dash_culture_dash_clone_dash_M153  \\\n",
       "SampleID                                                                 \n",
       "sample131                                                0.0             \n",
       "sample132                                                0.0             \n",
       "sample133                                                0.0             \n",
       "sample134                                                0.0             \n",
       "sample135                                                0.0             \n",
       "\n",
       "           gut_dash_metagenome  human_dash_gut_dash_metagenome  \\\n",
       "SampleID                                                         \n",
       "sample131        165115.709436                             0.0   \n",
       "sample132        106501.033841                             0.0   \n",
       "sample133        221840.237562                             0.0   \n",
       "sample134        919303.900024                             0.0   \n",
       "sample135        581683.602124                             0.0   \n",
       "\n",
       "           marine_dash_metagenome  metagenome  mouse_dash_gut_dash_metagenome  \n",
       "SampleID                                                                       \n",
       "sample131                     0.0         0.0                    1.322701e+05  \n",
       "sample132                     0.0         0.0                    1.003112e+06  \n",
       "sample133                     0.0         0.0                    1.866352e+05  \n",
       "sample134                     0.0         0.0                    8.714537e+05  \n",
       "sample135                     0.0         0.0                    5.335310e+05  \n",
       "\n",
       "[5 rows x 304 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = pd.read_excel('../../../../../data/16S_absolute_abundance.xlsx', index_col=0)\n",
    "df_16S.columns = [c.replace('/','_slash_').replace(' ','_space_').replace('[','_leftsquarebracket_').replace(']','_rightsquarebracket_').replace('-','_dash_').replace('.','_dot_').replace('(','_leftroundbracket').replace(')','_rightroundbracket_') for c in df_16S.columns]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add pseudocount for zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2</th>\n",
       "      <th>ASF356</th>\n",
       "      <th>Acetatifactor</th>\n",
       "      <th>Acetatifactor_dash_muris</th>\n",
       "      <th>Acidovorax</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Acinetobacter_dash_sp_dot__dash_CIP_dash_56_dot_2</th>\n",
       "      <th>Acutalibacter_dash_muris</th>\n",
       "      <th>Adlercreutzia_dash_mucosicola</th>\n",
       "      <th>Adlercreutzia_dash_muris</th>\n",
       "      <th>...</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_ventriosum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_gnavus_dash_group</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>bacterium_dash_enrichment_dash_culture_dash_clone_dash_M153</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>marine_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>887.718868</td>\n",
       "      <td>108301.701888</td>\n",
       "      <td>7101.750943</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>...</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>63028.039623</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>165115.709436</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>887.718868</td>\n",
       "      <td>1.322701e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>910.265247</td>\n",
       "      <td>405068.034693</td>\n",
       "      <td>50064.588558</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>...</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>80103.341692</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>106501.033841</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>910.265247</td>\n",
       "      <td>1.003112e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>482.261386</td>\n",
       "      <td>36169.603950</td>\n",
       "      <td>45332.570284</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>...</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>221840.237562</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>482.261386</td>\n",
       "      <td>1.866352e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>2900.012303</td>\n",
       "      <td>484302.054587</td>\n",
       "      <td>10150.043060</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>...</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>105850.449056</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>919303.900024</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>2900.012303</td>\n",
       "      <td>8.714537e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>1926.104643</td>\n",
       "      <td>91489.970533</td>\n",
       "      <td>42374.302141</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>12519.680178</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>...</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>581683.602124</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>5.335310e+05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 304 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    A2         ASF356  Acetatifactor  \\\n",
       "SampleID                                               \n",
       "sample131   887.718868  108301.701888    7101.750943   \n",
       "sample132   910.265247  405068.034693   50064.588558   \n",
       "sample133   482.261386   36169.603950   45332.570284   \n",
       "sample134  2900.012303  484302.054587   10150.043060   \n",
       "sample135  1926.104643   91489.970533   42374.302141   \n",
       "\n",
       "           Acetatifactor_dash_muris   Acidovorax  Acinetobacter  \\\n",
       "SampleID                                                          \n",
       "sample131                887.718868   887.718868     887.718868   \n",
       "sample132                910.265247   910.265247     910.265247   \n",
       "sample133                482.261386   482.261386     482.261386   \n",
       "sample134               2900.012303  2900.012303    2900.012303   \n",
       "sample135               1926.104643  1926.104643    1926.104643   \n",
       "\n",
       "           Acinetobacter_dash_sp_dot__dash_CIP_dash_56_dot_2  \\\n",
       "SampleID                                                       \n",
       "sample131                                         887.718868   \n",
       "sample132                                         910.265247   \n",
       "sample133                                         482.261386   \n",
       "sample134                                        2900.012303   \n",
       "sample135                                        1926.104643   \n",
       "\n",
       "           Acutalibacter_dash_muris  Adlercreutzia_dash_mucosicola  \\\n",
       "SampleID                                                             \n",
       "sample131                887.718868                     887.718868   \n",
       "sample132                910.265247                     910.265247   \n",
       "sample133                482.261386                     482.261386   \n",
       "sample134               2900.012303                    2900.012303   \n",
       "sample135               1926.104643                   12519.680178   \n",
       "\n",
       "           Adlercreutzia_dash_muris  ...  \\\n",
       "SampleID                             ...   \n",
       "sample131                887.718868  ...   \n",
       "sample132                910.265247  ...   \n",
       "sample133                482.261386  ...   \n",
       "sample134               2900.012303  ...   \n",
       "sample135               1926.104643  ...   \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_ventriosum_dash_group  \\\n",
       "SampleID                                                                                   \n",
       "sample131                                         887.718868                               \n",
       "sample132                                         910.265247                               \n",
       "sample133                                         482.261386                               \n",
       "sample134                                        2900.012303                               \n",
       "sample135                                        1926.104643                               \n",
       "\n",
       "           _leftsquarebracket_Eubacterium_rightsquarebracket__dash_xylanophilum_dash_group  \\\n",
       "SampleID                                                                                     \n",
       "sample131                                       63028.039623                                 \n",
       "sample132                                       80103.341692                                 \n",
       "sample133                                         482.261386                                 \n",
       "sample134                                      105850.449056                                 \n",
       "sample135                                        1926.104643                                 \n",
       "\n",
       "           _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_gnavus_dash_group  \\\n",
       "SampleID                                                                                \n",
       "sample131                                         887.718868                            \n",
       "sample132                                         910.265247                            \n",
       "sample133                                         482.261386                            \n",
       "sample134                                        2900.012303                            \n",
       "sample135                                        1926.104643                            \n",
       "\n",
       "           _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "SampleID                                                                                 \n",
       "sample131                                         887.718868                             \n",
       "sample132                                         910.265247                             \n",
       "sample133                                         482.261386                             \n",
       "sample134                                        2900.012303                             \n",
       "sample135                                        1926.104643                             \n",
       "\n",
       "           bacterium_dash_enrichment_dash_culture_dash_clone_dash_M153  \\\n",
       "SampleID                                                                 \n",
       "sample131                                         887.718868             \n",
       "sample132                                         910.265247             \n",
       "sample133                                         482.261386             \n",
       "sample134                                        2900.012303             \n",
       "sample135                                        1926.104643             \n",
       "\n",
       "           gut_dash_metagenome  human_dash_gut_dash_metagenome  \\\n",
       "SampleID                                                         \n",
       "sample131        165115.709436                      887.718868   \n",
       "sample132        106501.033841                      910.265247   \n",
       "sample133        221840.237562                      482.261386   \n",
       "sample134        919303.900024                     2900.012303   \n",
       "sample135        581683.602124                     1926.104643   \n",
       "\n",
       "           marine_dash_metagenome   metagenome  mouse_dash_gut_dash_metagenome  \n",
       "SampleID                                                                        \n",
       "sample131              887.718868   887.718868                    1.322701e+05  \n",
       "sample132              910.265247   910.265247                    1.003112e+06  \n",
       "sample133              482.261386   482.261386                    1.866352e+05  \n",
       "sample134             2900.012303  2900.012303                    8.714537e+05  \n",
       "sample135             1926.104643  1926.104643                    5.335310e+05  \n",
       "\n",
       "[5 rows x 304 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample_id in df_16S.index:\n",
    "    sample = np.array(df_16S.loc[sample_id])\n",
    "    minval = np.min(sample[np.nonzero(sample)]) # minimum non-zero value\n",
    "    sample[sample==0] = minval\n",
    "    df_16S.loc[sample_id] = sample\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata for rs and control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MiceID</th>\n",
       "      <th>Vendor</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>BC1</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>BC2</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>BC3</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>BC4</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>BC5</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          MiceID   Vendor  Day  Diet\n",
       "SampleID                            \n",
       "sample131    BC1  Beijing    0     0\n",
       "sample132    BC2  Beijing    0     0\n",
       "sample133    BC3  Beijing    0     0\n",
       "sample134    BC4  Beijing    0     0\n",
       "sample135    BC5  Beijing    0     0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_excel('../../../../../data/meta_data.xlsx', index_col=0)\n",
    "df_meta = df_meta[(df_meta.Diet=='Resistant starch') | (df_meta.Diet=='Control')]\n",
    "df_meta['Diet'] = [1 if g=='Resistant starch' else 0 for g in df_meta['Diet']]\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 20 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Faecalibaculum</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>Parabacteroides_dash_goldsteinii</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Alloprevotella</th>\n",
       "      <th>Lachnospiraceae_dash_bacterium_dash_28_dash_4</th>\n",
       "      <th>Desulfovibrionaceae</th>\n",
       "      <th>Parasutterella</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Rikenellaceae_dash_RC9_dash_gut_dash_group</th>\n",
       "      <th>Lachnospiraceae_dash_bacterium_dash_10_dash_1</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Helicobacter_dash_ganmani</th>\n",
       "      <th>Mucispirillum_dash_schaedleri</th>\n",
       "      <th>Clostridia_dash_vadinBB60_dash_group</th>\n",
       "      <th>Alistipes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>2.178906e+07</td>\n",
       "      <td>2.175266e+07</td>\n",
       "      <td>1.682227e+06</td>\n",
       "      <td>2.719971e+06</td>\n",
       "      <td>103863.107548</td>\n",
       "      <td>3.826068e+05</td>\n",
       "      <td>3.380433e+06</td>\n",
       "      <td>9.717858e+06</td>\n",
       "      <td>1.012000e+05</td>\n",
       "      <td>1.819824e+05</td>\n",
       "      <td>2.213083e+06</td>\n",
       "      <td>3.771030e+06</td>\n",
       "      <td>6.249541e+05</td>\n",
       "      <td>6.779509e+06</td>\n",
       "      <td>674666.339629</td>\n",
       "      <td>7101.750943</td>\n",
       "      <td>5326.313208</td>\n",
       "      <td>2.008020e+06</td>\n",
       "      <td>470491.000005</td>\n",
       "      <td>1.064375e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>2.539822e+07</td>\n",
       "      <td>1.732963e+07</td>\n",
       "      <td>4.455748e+06</td>\n",
       "      <td>3.721164e+06</td>\n",
       "      <td>59167.241023</td>\n",
       "      <td>2.645231e+06</td>\n",
       "      <td>4.563160e+06</td>\n",
       "      <td>2.849130e+05</td>\n",
       "      <td>7.898372e+06</td>\n",
       "      <td>9.912789e+05</td>\n",
       "      <td>3.591907e+06</td>\n",
       "      <td>2.610641e+06</td>\n",
       "      <td>3.932346e+06</td>\n",
       "      <td>3.418046e+06</td>\n",
       "      <td>90116.259404</td>\n",
       "      <td>10012.917712</td>\n",
       "      <td>11833.448205</td>\n",
       "      <td>1.345281e+07</td>\n",
       "      <td>142911.643701</td>\n",
       "      <td>1.247063e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>1.379895e+07</td>\n",
       "      <td>1.250166e+07</td>\n",
       "      <td>1.339240e+06</td>\n",
       "      <td>7.354486e+05</td>\n",
       "      <td>46297.093056</td>\n",
       "      <td>7.802989e+05</td>\n",
       "      <td>7.508810e+05</td>\n",
       "      <td>1.040238e+06</td>\n",
       "      <td>2.083369e+05</td>\n",
       "      <td>1.504656e+05</td>\n",
       "      <td>4.753168e+06</td>\n",
       "      <td>3.729327e+06</td>\n",
       "      <td>5.290407e+05</td>\n",
       "      <td>5.057957e+06</td>\n",
       "      <td>21701.762370</td>\n",
       "      <td>4822.613860</td>\n",
       "      <td>5304.875246</td>\n",
       "      <td>4.054854e+06</td>\n",
       "      <td>175060.883119</td>\n",
       "      <td>1.064833e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>3.867746e+07</td>\n",
       "      <td>2.484296e+07</td>\n",
       "      <td>1.024574e+07</td>\n",
       "      <td>4.120917e+06</td>\n",
       "      <td>184150.781235</td>\n",
       "      <td>4.157168e+06</td>\n",
       "      <td>8.172235e+06</td>\n",
       "      <td>3.694616e+06</td>\n",
       "      <td>1.566007e+05</td>\n",
       "      <td>2.321460e+06</td>\n",
       "      <td>4.982221e+06</td>\n",
       "      <td>2.262880e+07</td>\n",
       "      <td>5.088072e+06</td>\n",
       "      <td>2.312180e+07</td>\n",
       "      <td>511852.171464</td>\n",
       "      <td>21750.092272</td>\n",
       "      <td>20300.086120</td>\n",
       "      <td>5.086622e+06</td>\n",
       "      <td>194300.824295</td>\n",
       "      <td>8.069284e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>1.994192e+07</td>\n",
       "      <td>1.935254e+07</td>\n",
       "      <td>4.775776e+06</td>\n",
       "      <td>1.094990e+06</td>\n",
       "      <td>72228.924105</td>\n",
       "      <td>1.040097e+06</td>\n",
       "      <td>1.251005e+06</td>\n",
       "      <td>1.814391e+06</td>\n",
       "      <td>6.067230e+04</td>\n",
       "      <td>2.725438e+05</td>\n",
       "      <td>3.164590e+06</td>\n",
       "      <td>4.727624e+06</td>\n",
       "      <td>5.441246e+05</td>\n",
       "      <td>1.064558e+07</td>\n",
       "      <td>42374.302141</td>\n",
       "      <td>1926.104643</td>\n",
       "      <td>6741.366250</td>\n",
       "      <td>7.617744e+05</td>\n",
       "      <td>45263.459106</td>\n",
       "      <td>2.844857e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Muribaculaceae  Faecalibaculum  Bacteroides_dash_acidifaciens  \\\n",
       "SampleID                                                                   \n",
       "sample131    2.178906e+07    2.175266e+07                   1.682227e+06   \n",
       "sample132    2.539822e+07    1.732963e+07                   4.455748e+06   \n",
       "sample133    1.379895e+07    1.250166e+07                   1.339240e+06   \n",
       "sample134    3.867746e+07    2.484296e+07                   1.024574e+07   \n",
       "sample135    1.994192e+07    1.935254e+07                   4.775776e+06   \n",
       "\n",
       "           Lachnospiraceae_dash_NK4A136_dash_group  \\\n",
       "SampleID                                             \n",
       "sample131                             2.719971e+06   \n",
       "sample132                             3.721164e+06   \n",
       "sample133                             7.354486e+05   \n",
       "sample134                             4.120917e+06   \n",
       "sample135                             1.094990e+06   \n",
       "\n",
       "           Parabacteroides_dash_goldsteinii   Bacteroides  Lachnospiraceae  \\\n",
       "SampleID                                                                     \n",
       "sample131                     103863.107548  3.826068e+05     3.380433e+06   \n",
       "sample132                      59167.241023  2.645231e+06     4.563160e+06   \n",
       "sample133                      46297.093056  7.802989e+05     7.508810e+05   \n",
       "sample134                     184150.781235  4.157168e+06     8.172235e+06   \n",
       "sample135                      72228.924105  1.040097e+06     1.251005e+06   \n",
       "\n",
       "           Alloprevotella  Lachnospiraceae_dash_bacterium_dash_28_dash_4  \\\n",
       "SampleID                                                                   \n",
       "sample131    9.717858e+06                                   1.012000e+05   \n",
       "sample132    2.849130e+05                                   7.898372e+06   \n",
       "sample133    1.040238e+06                                   2.083369e+05   \n",
       "sample134    3.694616e+06                                   1.566007e+05   \n",
       "sample135    1.814391e+06                                   6.067230e+04   \n",
       "\n",
       "           Desulfovibrionaceae  Parasutterella  Akkermansia_dash_muciniphila  \\\n",
       "SampleID                                                                       \n",
       "sample131         1.819824e+05    2.213083e+06                  3.771030e+06   \n",
       "sample132         9.912789e+05    3.591907e+06                  2.610641e+06   \n",
       "sample133         1.504656e+05    4.753168e+06                  3.729327e+06   \n",
       "sample134         2.321460e+06    4.982221e+06                  2.262880e+07   \n",
       "sample135         2.725438e+05    3.164590e+06                  4.727624e+06   \n",
       "\n",
       "           Oscillospiraceae  Rikenellaceae_dash_RC9_dash_gut_dash_group  \\\n",
       "SampleID                                                                  \n",
       "sample131      6.249541e+05                                6.779509e+06   \n",
       "sample132      3.932346e+06                                3.418046e+06   \n",
       "sample133      5.290407e+05                                5.057957e+06   \n",
       "sample134      5.088072e+06                                2.312180e+07   \n",
       "sample135      5.441246e+05                                1.064558e+07   \n",
       "\n",
       "           Lachnospiraceae_dash_bacterium_dash_10_dash_1  \\\n",
       "SampleID                                                   \n",
       "sample131                                  674666.339629   \n",
       "sample132                                   90116.259404   \n",
       "sample133                                   21701.762370   \n",
       "sample134                                  511852.171464   \n",
       "sample135                                   42374.302141   \n",
       "\n",
       "           Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  \\\n",
       "SampleID                                                        \n",
       "sample131                                        7101.750943    \n",
       "sample132                                       10012.917712    \n",
       "sample133                                        4822.613860    \n",
       "sample134                                       21750.092272    \n",
       "sample135                                        1926.104643    \n",
       "\n",
       "           Helicobacter_dash_ganmani  Mucispirillum_dash_schaedleri  \\\n",
       "SampleID                                                              \n",
       "sample131                5326.313208                   2.008020e+06   \n",
       "sample132               11833.448205                   1.345281e+07   \n",
       "sample133                5304.875246                   4.054854e+06   \n",
       "sample134               20300.086120                   5.086622e+06   \n",
       "sample135                6741.366250                   7.617744e+05   \n",
       "\n",
       "           Clostridia_dash_vadinBB60_dash_group     Alistipes  \n",
       "SampleID                                                       \n",
       "sample131                         470491.000005  1.064375e+06  \n",
       "sample132                         142911.643701  1.247063e+06  \n",
       "sample133                         175060.883119  1.064833e+06  \n",
       "sample134                         194300.824295  8.069284e+06  \n",
       "sample135                          45263.459106  2.844857e+06  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 20\n",
    "df_16S_T = df_16S.loc[df_meta.index].T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Muribaculaceae',\n",
       " 'Faecalibaculum',\n",
       " 'Bacteroides_dash_acidifaciens',\n",
       " 'Lachnospiraceae_dash_NK4A136_dash_group',\n",
       " 'Parabacteroides_dash_goldsteinii',\n",
       " 'Bacteroides',\n",
       " 'Lachnospiraceae',\n",
       " 'Alloprevotella',\n",
       " 'Lachnospiraceae_dash_bacterium_dash_28_dash_4',\n",
       " 'Desulfovibrionaceae',\n",
       " 'Parasutterella',\n",
       " 'Akkermansia_dash_muciniphila',\n",
       " 'Oscillospiraceae',\n",
       " 'Rikenellaceae_dash_RC9_dash_gut_dash_group',\n",
       " 'Lachnospiraceae_dash_bacterium_dash_10_dash_1',\n",
       " 'Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57',\n",
       " 'Helicobacter_dash_ganmani',\n",
       " 'Mucispirillum_dash_schaedleri',\n",
       " 'Clostridia_dash_vadinBB60_dash_group',\n",
       " 'Alistipes']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_16S_topX.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MiceID</th>\n",
       "      <th>Vendor</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Faecalibaculum</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>Parabacteroides_dash_goldsteinii</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>...</th>\n",
       "      <th>Parasutterella</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Rikenellaceae_dash_RC9_dash_gut_dash_group</th>\n",
       "      <th>Lachnospiraceae_dash_bacterium_dash_10_dash_1</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Helicobacter_dash_ganmani</th>\n",
       "      <th>Mucispirillum_dash_schaedleri</th>\n",
       "      <th>Clostridia_dash_vadinBB60_dash_group</th>\n",
       "      <th>Alistipes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>BC1</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.121890</td>\n",
       "      <td>0.121687</td>\n",
       "      <td>0.009411</td>\n",
       "      <td>0.015216</td>\n",
       "      <td>0.000581</td>\n",
       "      <td>0.002140</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012380</td>\n",
       "      <td>0.021096</td>\n",
       "      <td>0.003496</td>\n",
       "      <td>0.037925</td>\n",
       "      <td>0.003774</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>0.011233</td>\n",
       "      <td>0.002632</td>\n",
       "      <td>0.005954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>BC2</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.142080</td>\n",
       "      <td>0.096944</td>\n",
       "      <td>0.024926</td>\n",
       "      <td>0.020817</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>0.014798</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020094</td>\n",
       "      <td>0.014604</td>\n",
       "      <td>0.021998</td>\n",
       "      <td>0.019121</td>\n",
       "      <td>0.000504</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.075256</td>\n",
       "      <td>0.000799</td>\n",
       "      <td>0.006976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>BC3</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.077193</td>\n",
       "      <td>0.069936</td>\n",
       "      <td>0.007492</td>\n",
       "      <td>0.004114</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0.004365</td>\n",
       "      <td>...</td>\n",
       "      <td>0.026590</td>\n",
       "      <td>0.020862</td>\n",
       "      <td>0.002960</td>\n",
       "      <td>0.028295</td>\n",
       "      <td>0.000121</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>0.022683</td>\n",
       "      <td>0.000979</td>\n",
       "      <td>0.005957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>BC4</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.216366</td>\n",
       "      <td>0.138974</td>\n",
       "      <td>0.057316</td>\n",
       "      <td>0.023053</td>\n",
       "      <td>0.001030</td>\n",
       "      <td>0.023256</td>\n",
       "      <td>...</td>\n",
       "      <td>0.027871</td>\n",
       "      <td>0.126588</td>\n",
       "      <td>0.028463</td>\n",
       "      <td>0.129346</td>\n",
       "      <td>0.002863</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.028455</td>\n",
       "      <td>0.001087</td>\n",
       "      <td>0.045140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>BC5</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.111557</td>\n",
       "      <td>0.108260</td>\n",
       "      <td>0.026716</td>\n",
       "      <td>0.006125</td>\n",
       "      <td>0.000404</td>\n",
       "      <td>0.005818</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017703</td>\n",
       "      <td>0.026447</td>\n",
       "      <td>0.003044</td>\n",
       "      <td>0.059553</td>\n",
       "      <td>0.000237</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.004261</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.015914</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          MiceID   Vendor  Day  Diet  Muribaculaceae  Faecalibaculum  \\\n",
       "SampleID                                                               \n",
       "sample131    BC1  Beijing    0     0        0.121890        0.121687   \n",
       "sample132    BC2  Beijing    0     0        0.142080        0.096944   \n",
       "sample133    BC3  Beijing    0     0        0.077193        0.069936   \n",
       "sample134    BC4  Beijing    0     0        0.216366        0.138974   \n",
       "sample135    BC5  Beijing    0     0        0.111557        0.108260   \n",
       "\n",
       "           Bacteroides_dash_acidifaciens  \\\n",
       "SampleID                                   \n",
       "sample131                       0.009411   \n",
       "sample132                       0.024926   \n",
       "sample133                       0.007492   \n",
       "sample134                       0.057316   \n",
       "sample135                       0.026716   \n",
       "\n",
       "           Lachnospiraceae_dash_NK4A136_dash_group  \\\n",
       "SampleID                                             \n",
       "sample131                                 0.015216   \n",
       "sample132                                 0.020817   \n",
       "sample133                                 0.004114   \n",
       "sample134                                 0.023053   \n",
       "sample135                                 0.006125   \n",
       "\n",
       "           Parabacteroides_dash_goldsteinii  Bacteroides  ...  Parasutterella  \\\n",
       "SampleID                                                  ...                   \n",
       "sample131                          0.000581     0.002140  ...        0.012380   \n",
       "sample132                          0.000331     0.014798  ...        0.020094   \n",
       "sample133                          0.000259     0.004365  ...        0.026590   \n",
       "sample134                          0.001030     0.023256  ...        0.027871   \n",
       "sample135                          0.000404     0.005818  ...        0.017703   \n",
       "\n",
       "           Akkermansia_dash_muciniphila  Oscillospiraceae  \\\n",
       "SampleID                                                    \n",
       "sample131                      0.021096          0.003496   \n",
       "sample132                      0.014604          0.021998   \n",
       "sample133                      0.020862          0.002960   \n",
       "sample134                      0.126588          0.028463   \n",
       "sample135                      0.026447          0.003044   \n",
       "\n",
       "           Rikenellaceae_dash_RC9_dash_gut_dash_group  \\\n",
       "SampleID                                                \n",
       "sample131                                    0.037925   \n",
       "sample132                                    0.019121   \n",
       "sample133                                    0.028295   \n",
       "sample134                                    0.129346   \n",
       "sample135                                    0.059553   \n",
       "\n",
       "           Lachnospiraceae_dash_bacterium_dash_10_dash_1  \\\n",
       "SampleID                                                   \n",
       "sample131                                       0.003774   \n",
       "sample132                                       0.000504   \n",
       "sample133                                       0.000121   \n",
       "sample134                                       0.002863   \n",
       "sample135                                       0.000237   \n",
       "\n",
       "           Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  \\\n",
       "SampleID                                                        \n",
       "sample131                                           0.000040    \n",
       "sample132                                           0.000056    \n",
       "sample133                                           0.000027    \n",
       "sample134                                           0.000122    \n",
       "sample135                                           0.000011    \n",
       "\n",
       "           Helicobacter_dash_ganmani  Mucispirillum_dash_schaedleri  \\\n",
       "SampleID                                                              \n",
       "sample131                   0.000030                       0.011233   \n",
       "sample132                   0.000066                       0.075256   \n",
       "sample133                   0.000030                       0.022683   \n",
       "sample134                   0.000114                       0.028455   \n",
       "sample135                   0.000038                       0.004261   \n",
       "\n",
       "           Clostridia_dash_vadinBB60_dash_group  Alistipes  \n",
       "SampleID                                                    \n",
       "sample131                              0.002632   0.005954  \n",
       "sample132                              0.000799   0.006976  \n",
       "sample133                              0.000979   0.005957  \n",
       "sample134                              0.001087   0.045140  \n",
       "sample135                              0.000253   0.015914  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX_w_meta = pd.merge(df_meta, df_16S_topX/df_16S_topX.max().max(), left_index=True, right_index=True, how='inner')\n",
    "df_16S_topX_w_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate log-derivatives (dlog(B)/dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MiceID</th>\n",
       "      <th>Vendor</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Muribaculaceae</th>\n",
       "      <th>Faecalibaculum</th>\n",
       "      <th>Bacteroides_dash_acidifaciens</th>\n",
       "      <th>Lachnospiraceae_dash_NK4A136_dash_group</th>\n",
       "      <th>Parabacteroides_dash_goldsteinii</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>...</th>\n",
       "      <th>Parasutterella</th>\n",
       "      <th>Akkermansia_dash_muciniphila</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Rikenellaceae_dash_RC9_dash_gut_dash_group</th>\n",
       "      <th>Lachnospiraceae_dash_bacterium_dash_10_dash_1</th>\n",
       "      <th>Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57</th>\n",
       "      <th>Helicobacter_dash_ganmani</th>\n",
       "      <th>Mucispirillum_dash_schaedleri</th>\n",
       "      <th>Clostridia_dash_vadinBB60_dash_group</th>\n",
       "      <th>Alistipes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample131</th>\n",
       "      <td>BC1</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.074386</td>\n",
       "      <td>-0.254233</td>\n",
       "      <td>-0.004071</td>\n",
       "      <td>-0.004919</td>\n",
       "      <td>0.000152</td>\n",
       "      <td>0.091566</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.029396</td>\n",
       "      <td>-0.032129</td>\n",
       "      <td>0.017712</td>\n",
       "      <td>0.019519</td>\n",
       "      <td>-0.005393</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.006910</td>\n",
       "      <td>-0.007715</td>\n",
       "      <td>-0.000913</td>\n",
       "      <td>0.039701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample132</th>\n",
       "      <td>BC2</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.135226</td>\n",
       "      <td>-0.193439</td>\n",
       "      <td>-0.036332</td>\n",
       "      <td>0.017515</td>\n",
       "      <td>0.000315</td>\n",
       "      <td>-0.006180</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.038113</td>\n",
       "      <td>-0.024804</td>\n",
       "      <td>-0.006381</td>\n",
       "      <td>0.011125</td>\n",
       "      <td>0.000221</td>\n",
       "      <td>-0.000053</td>\n",
       "      <td>0.093857</td>\n",
       "      <td>-0.093251</td>\n",
       "      <td>0.002053</td>\n",
       "      <td>0.009724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample133</th>\n",
       "      <td>BC3</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.007883</td>\n",
       "      <td>-0.148905</td>\n",
       "      <td>-0.022818</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>0.069495</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.065846</td>\n",
       "      <td>-0.021541</td>\n",
       "      <td>0.018099</td>\n",
       "      <td>0.010002</td>\n",
       "      <td>-0.001413</td>\n",
       "      <td>-0.000018</td>\n",
       "      <td>0.019876</td>\n",
       "      <td>-0.033317</td>\n",
       "      <td>0.021856</td>\n",
       "      <td>0.063697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample134</th>\n",
       "      <td>BC4</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.021238</td>\n",
       "      <td>-0.148056</td>\n",
       "      <td>-0.078009</td>\n",
       "      <td>0.054405</td>\n",
       "      <td>-0.000708</td>\n",
       "      <td>-0.014161</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.021315</td>\n",
       "      <td>-0.197212</td>\n",
       "      <td>0.027841</td>\n",
       "      <td>-0.042253</td>\n",
       "      <td>-0.002511</td>\n",
       "      <td>-0.000173</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>-0.026145</td>\n",
       "      <td>0.009527</td>\n",
       "      <td>-0.045128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample135</th>\n",
       "      <td>BC5</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.149464</td>\n",
       "      <td>-0.060736</td>\n",
       "      <td>-0.009520</td>\n",
       "      <td>0.043082</td>\n",
       "      <td>0.000137</td>\n",
       "      <td>0.040215</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000796</td>\n",
       "      <td>-0.034933</td>\n",
       "      <td>0.009670</td>\n",
       "      <td>-0.026264</td>\n",
       "      <td>0.000484</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.007328</td>\n",
       "      <td>0.004431</td>\n",
       "      <td>0.022808</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          MiceID   Vendor  Day  Diet  Muribaculaceae  Faecalibaculum  \\\n",
       "SampleID                                                               \n",
       "sample131    BC1  Beijing    0     0       -0.074386       -0.254233   \n",
       "sample132    BC2  Beijing    0     0       -0.135226       -0.193439   \n",
       "sample133    BC3  Beijing    0     0       -0.007883       -0.148905   \n",
       "sample134    BC4  Beijing    0     0       -0.021238       -0.148056   \n",
       "sample135    BC5  Beijing    0     0        0.149464       -0.060736   \n",
       "\n",
       "           Bacteroides_dash_acidifaciens  \\\n",
       "SampleID                                   \n",
       "sample131                      -0.004071   \n",
       "sample132                      -0.036332   \n",
       "sample133                      -0.022818   \n",
       "sample134                      -0.078009   \n",
       "sample135                      -0.009520   \n",
       "\n",
       "           Lachnospiraceae_dash_NK4A136_dash_group  \\\n",
       "SampleID                                             \n",
       "sample131                                -0.004919   \n",
       "sample132                                 0.017515   \n",
       "sample133                                 0.000002   \n",
       "sample134                                 0.054405   \n",
       "sample135                                 0.043082   \n",
       "\n",
       "           Parabacteroides_dash_goldsteinii  Bacteroides  ...  Parasutterella  \\\n",
       "SampleID                                                  ...                   \n",
       "sample131                          0.000152     0.091566  ...       -0.029396   \n",
       "sample132                          0.000315    -0.006180  ...       -0.038113   \n",
       "sample133                          0.000183     0.069495  ...       -0.065846   \n",
       "sample134                         -0.000708    -0.014161  ...       -0.021315   \n",
       "sample135                          0.000137     0.040215  ...       -0.000796   \n",
       "\n",
       "           Akkermansia_dash_muciniphila  Oscillospiraceae  \\\n",
       "SampleID                                                    \n",
       "sample131                     -0.032129          0.017712   \n",
       "sample132                     -0.024804         -0.006381   \n",
       "sample133                     -0.021541          0.018099   \n",
       "sample134                     -0.197212          0.027841   \n",
       "sample135                     -0.034933          0.009670   \n",
       "\n",
       "           Rikenellaceae_dash_RC9_dash_gut_dash_group  \\\n",
       "SampleID                                                \n",
       "sample131                                    0.019519   \n",
       "sample132                                    0.011125   \n",
       "sample133                                    0.010002   \n",
       "sample134                                   -0.042253   \n",
       "sample135                                   -0.026264   \n",
       "\n",
       "           Lachnospiraceae_dash_bacterium_dash_10_dash_1  \\\n",
       "SampleID                                                   \n",
       "sample131                                      -0.005393   \n",
       "sample132                                       0.000221   \n",
       "sample133                                      -0.001413   \n",
       "sample134                                      -0.002511   \n",
       "sample135                                       0.000484   \n",
       "\n",
       "           Ruminococcus_dash_sp_dot__dash_N15_dot_MGS_dash_57  \\\n",
       "SampleID                                                        \n",
       "sample131                                           0.000088    \n",
       "sample132                                          -0.000053    \n",
       "sample133                                          -0.000018    \n",
       "sample134                                          -0.000173    \n",
       "sample135                                           0.000010    \n",
       "\n",
       "           Helicobacter_dash_ganmani  Mucispirillum_dash_schaedleri  \\\n",
       "SampleID                                                              \n",
       "sample131                   0.006910                      -0.007715   \n",
       "sample132                   0.093857                      -0.093251   \n",
       "sample133                   0.019876                      -0.033317   \n",
       "sample134                   0.000063                      -0.026145   \n",
       "sample135                   0.000023                       0.007328   \n",
       "\n",
       "           Clostridia_dash_vadinBB60_dash_group  Alistipes  \n",
       "SampleID                                                    \n",
       "sample131                             -0.000913   0.039701  \n",
       "sample132                              0.002053   0.009724  \n",
       "sample133                              0.021856   0.063697  \n",
       "sample134                              0.009527  -0.045128  \n",
       "sample135                              0.004431   0.022808  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_deriv = deepcopy(df_16S_topX_w_meta)\n",
    "for curr_mice in set(df_deriv.MiceID):\n",
    "    curr_df = df_deriv[df_deriv.MiceID==curr_mice]\n",
    "    for taxon in bacterial_taxa:\n",
    "        xdata = np.array(curr_df['Day'])\n",
    "        ydata = np.array(curr_df[taxon])\n",
    "        cs = CubicSpline(xdata, ydata)\n",
    "        csd1 = cs.derivative(nu=1)\n",
    "        ydata_d1 = csd1(xdata)\n",
    "        df_deriv.loc[df_deriv.MiceID==curr_mice, taxon] = ydata_d1\n",
    "df_deriv.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct regression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-2.44202305, -4.44927117, -0.24797056, ...,  0.09420514,\n",
       "        0.18872496,  0.10023815])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ymat should be samples by bacteria\n",
    "Ymat = df_deriv[bacterial_taxa].values\n",
    "Ymat = Ymat.flatten(order='F')\n",
    "Ymat = StandardScaler().fit_transform(Ymat.reshape(-1,1)).reshape(1,-1)[0] # standardize\n",
    "Ymat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7200"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Ymat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.00000000e+00, 0.00000000e+00, 1.21890330e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 0.00000000e+00, 1.42080364e-01, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       [1.00000000e+00, 0.00000000e+00, 7.71927743e-02, ...,\n",
       "        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],\n",
       "       ...,\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        8.52639403e-04, 3.11541320e-03, 7.79673146e-03],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        1.77657663e-04, 1.00598652e-02, 1.67109239e-02],\n",
       "       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,\n",
       "        8.25431847e-05, 2.60386228e-03, 8.34436558e-03]])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Xmat = np.zeros(shape=(topX*len(df_deriv.index), (topX+2)*topX))\n",
    "for k in np.arange(topX):\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)] = 1\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+1] = df_deriv.Diet.values\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+2:(k+1)*(topX+2)] = df_16S_topX_w_meta[bacterial_taxa].values\n",
    "Xmat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write stan program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "json_str = '{\\n\"N\" : %d,\\n'%(len(Ymat))\n",
    "\n",
    "# y variable\n",
    "json_str += '\\\"dlogX\\\" : [%s],\\n'%(','.join(list(Ymat.astype(str))))\n",
    "\n",
    "# x variable\n",
    "for k1,c1 in enumerate(bacterial_taxa):\n",
    "    # growth rate\n",
    "    json_str += '\\\"growth_rate_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)].astype(str))))\n",
    "    # diet response\n",
    "    json_str += '\\\"rs_response_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)+1].astype(str))))\n",
    "    # bacterial interactions\n",
    "    for k2,c2 in enumerate(bacterial_taxa):\n",
    "        v = list(Xmat[:,k1*(topX+2)+2+k2].astype(str))\n",
    "        json_str += '\\\"pairwise_interaction_%s_%s\\\" : [%s]'%(c1,c2,','.join(v))\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            json_str += '\\n}'\n",
    "        else:\n",
    "            json_str += ',\\n'\n",
    "#print(json_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.data.json\", \"w\")\n",
    "text_file.write(\"%s\" % json_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# data block\n",
    "model_str = 'data {\\n'\n",
    "model_str += '\\tint<lower=0> N;\\n'\n",
    "model_str += '\\tvector[N] dlogX;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\tvector[N] growth_rate_%s;\\n'%(c1)\n",
    "    model_str += '\\tvector[N] rs_response_%s;\\n'%(c1)\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tvector[N] pairwise_interaction_%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'\n",
    "\n",
    "# parameter block\n",
    "model_str += 'parameters {\\n\\treal<lower=0,upper=1> sigma;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\treal alpha__%s;\\n'%(c1) # growth rate\n",
    "    model_str += '\\treal epsilon__%s;\\n'%(c1) # resistance starch response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\treal beta__%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'       \n",
    "        \n",
    "# model block\n",
    "model_str += 'model {\\n\\tsigma ~ uniform(0,1);\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\talpha__%s ~ normal(0,1);\\n'%(c1) # growth rate\n",
    "    model_str += '\\tepsilon__%s ~ normal(0,1);\\n'%(c1) # rs response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tbeta__%s_%s ~ normal(0,1);\\n'%(c1,c2)\n",
    "model_str += '\\tdlogX ~ normal('\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += 'alpha__%s*growth_rate_%s+'%(c1,c1) # growth rate\n",
    "    model_str += 'epsilon__%s*rs_response_%s+'%(c1,c1) # rs response\n",
    "    for c2 in bacterial_taxa:\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s'%(c1,c2,c1,c2)\n",
    "        else:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s+'%(c1,c2,c1,c2)\n",
    "model_str += ', sigma);\\n}'\n",
    "#print(model_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa/mice_scfa.stan\", \"w\")\n",
    "text_file.write(\"%s\" % model_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot stan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hpd_grid(sample, alpha=0.05, roundto=2):\n",
    "    \"\"\"Calculate highest posterior density (HPD) of array for given alpha. \n",
    "    The HPD is the minimum width Bayesian credible interval (BCI). \n",
    "    The function works for multimodal distributions, returning more than one mode\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    sample : Numpy array or python list\n",
    "        An array containing MCMC samples\n",
    "    alpha : float\n",
    "        Desired probability of type I error (defaults to 0.05)\n",
    "    roundto: integer\n",
    "        Number of digits after the decimal point for the results\n",
    "    Returns\n",
    "    ----------\n",
    "    hpd: array with the lower \n",
    "          \n",
    "    \"\"\"\n",
    "    sample = np.asarray(sample)\n",
    "    sample = sample[~np.isnan(sample)]\n",
    "    # get upper and lower bounds\n",
    "    l = np.min(sample)\n",
    "    u = np.max(sample)\n",
    "    density = kde.gaussian_kde(sample)\n",
    "    x = np.linspace(l, u, 2000)\n",
    "    y = density.evaluate(x)\n",
    "    #y = density.evaluate(x, l, u) waitting for PR to be accepted\n",
    "    xy_zipped = zip(x, y/np.sum(y))\n",
    "    xy = sorted(xy_zipped, key=lambda x: x[1], reverse=True)\n",
    "    xy_cum_sum = 0\n",
    "    hdv = []\n",
    "    for val in xy:\n",
    "        xy_cum_sum += val[1]\n",
    "        hdv.append(val[0])\n",
    "        if xy_cum_sum >= (1-alpha):\n",
    "            break\n",
    "    hdv.sort()\n",
    "    diff = (u-l)/20  # differences of 5%\n",
    "    hpd = []\n",
    "    hpd.append(round(min(hdv), roundto))\n",
    "    for i in range(1, len(hdv)):\n",
    "        if hdv[i]-hdv[i-1] >= diff:\n",
    "            hpd.append(round(hdv[i-1], roundto))\n",
    "            hpd.append(round(hdv[i], roundto))\n",
    "    hpd.append(round(max(hdv), roundto))\n",
    "    ite = iter(hpd)\n",
    "    hpd = list(zip(ite, ite))\n",
    "    modes = []\n",
    "    for value in hpd:\n",
    "        x_hpd = x[(x > value[0]) & (x < value[1])]\n",
    "        y_hpd = y[(x > value[0]) & (x < value[1])]\n",
    "        modes.append(round(x_hpd[np.argmax(y_hpd)], roundto))\n",
    "    return hpd, x, y, modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"output_%d.csv\"%(i) for i in [1,2,3]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>-0.15</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.045</td>\n",
       "      <td>0.443050</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Faecalibaculum</td>\n",
       "      <td>-0.18</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.209583</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides_dash_acidifaciens</td>\n",
       "      <td>-0.34</td>\n",
       "      <td>0.04</td>\n",
       "      <td>-0.150</td>\n",
       "      <td>-1.542393</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Lachnospiraceae_dash_NK4A136_dash_group</td>\n",
       "      <td>-0.25</td>\n",
       "      <td>0.14</td>\n",
       "      <td>-0.055</td>\n",
       "      <td>-0.590692</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Parabacteroides_dash_goldsteinii</td>\n",
       "      <td>-0.36</td>\n",
       "      <td>0.03</td>\n",
       "      <td>-0.165</td>\n",
       "      <td>-1.638676</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Type                                     Taxa  Left  Right  \\\n",
       "0  basal_growth_rate                           Muribaculaceae -0.15   0.24   \n",
       "1  basal_growth_rate                           Faecalibaculum -0.18   0.22   \n",
       "2  basal_growth_rate            Bacteroides_dash_acidifaciens -0.34   0.04   \n",
       "3  basal_growth_rate  Lachnospiraceae_dash_NK4A136_dash_group -0.25   0.14   \n",
       "4  basal_growth_rate         Parabacteroides_dash_goldsteinii -0.36   0.03   \n",
       "\n",
       "   Middle       SNR  Significant  \n",
       "0   0.045  0.443050        False  \n",
       "1   0.020  0.209583        False  \n",
       "2  -0.150 -1.542393        False  \n",
       "3  -0.055 -0.590692        False  \n",
       "4  -0.165 -1.638676        False  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "    \n",
    "# basal growth rate\n",
    "for c in bacterial_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# rs response\n",
    "for c in bacterial_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['rs_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "\n",
    "# pairwise interactions\n",
    "for c1 in bacterial_taxa:\n",
    "    for c2 in bacterial_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stan_output_summary.to_excel('bayesian_regression_summary_top20.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Muribaculaceae</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.660</td>\n",
       "      <td>5.491318</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Faecalibaculum</td>\n",
       "      <td>0.85</td>\n",
       "      <td>1.32</td>\n",
       "      <td>1.085</td>\n",
       "      <td>8.986253</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Alloprevotella</td>\n",
       "      <td>-0.49</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>-0.255</td>\n",
       "      <td>-2.153641</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Desulfovibrionaceae</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.410</td>\n",
       "      <td>3.338960</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Muribaculaceae, Muribaculaceae)</td>\n",
       "      <td>2.88</td>\n",
       "      <td>4.31</td>\n",
       "      <td>3.595</td>\n",
       "      <td>9.897982</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Muribaculaceae, Faecalibaculum)</td>\n",
       "      <td>-4.73</td>\n",
       "      <td>-1.85</td>\n",
       "      <td>-3.290</td>\n",
       "      <td>-4.506180</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Muribaculaceae, Lachnospiraceae_dash_NK4A136_...</td>\n",
       "      <td>-5.85</td>\n",
       "      <td>-2.90</td>\n",
       "      <td>-4.375</td>\n",
       "      <td>-5.836662</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Muribaculaceae, Lachnospiraceae_dash_bacteriu...</td>\n",
       "      <td>-3.28</td>\n",
       "      <td>-0.39</td>\n",
       "      <td>-1.835</td>\n",
       "      <td>-2.509557</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Muribaculaceae, Desulfovibrionaceae)</td>\n",
       "      <td>-2.88</td>\n",
       "      <td>-0.11</td>\n",
       "      <td>-1.495</td>\n",
       "      <td>-2.125255</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Muribaculaceae, Lachnospiraceae_dash_bacteriu...</td>\n",
       "      <td>-3.75</td>\n",
       "      <td>-0.73</td>\n",
       "      <td>-2.240</td>\n",
       "      <td>-2.912772</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Muribaculaceae, Ruminococcus_dash_sp_dot__das...</td>\n",
       "      <td>-5.96</td>\n",
       "      <td>-2.93</td>\n",
       "      <td>-4.445</td>\n",
       "      <td>-5.779005</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Faecalibaculum, Muribaculaceae)</td>\n",
       "      <td>-1.59</td>\n",
       "      <td>-0.14</td>\n",
       "      <td>-0.865</td>\n",
       "      <td>-2.327044</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Faecalibaculum, Faecalibaculum)</td>\n",
       "      <td>-3.23</td>\n",
       "      <td>-0.34</td>\n",
       "      <td>-1.785</td>\n",
       "      <td>-2.407113</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Faecalibaculum, Lachnospiraceae_dash_NK4A136_...</td>\n",
       "      <td>-3.52</td>\n",
       "      <td>-0.61</td>\n",
       "      <td>-2.065</td>\n",
       "      <td>-2.831374</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Parabacteroides_dash_goldsteinii, Parabactero...</td>\n",
       "      <td>-3.57</td>\n",
       "      <td>-0.39</td>\n",
       "      <td>-1.980</td>\n",
       "      <td>-2.461324</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Bacteroides, Muribaculaceae)</td>\n",
       "      <td>-1.72</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>-2.728627</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Bacteroides, Bacteroides)</td>\n",
       "      <td>-3.47</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>-1.875</td>\n",
       "      <td>-2.327825</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Alloprevotella, Muribaculaceae)</td>\n",
       "      <td>1.06</td>\n",
       "      <td>2.51</td>\n",
       "      <td>1.785</td>\n",
       "      <td>4.878909</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Lachnospiraceae_dash_bacterium_dash_28_dash_4...</td>\n",
       "      <td>-2.95</td>\n",
       "      <td>-1.50</td>\n",
       "      <td>-2.225</td>\n",
       "      <td>-6.072116</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Lachnospiraceae_dash_bacterium_dash_28_dash_4...</td>\n",
       "      <td>0.25</td>\n",
       "      <td>3.28</td>\n",
       "      <td>1.765</td>\n",
       "      <td>2.292892</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Lachnospiraceae_dash_bacterium_dash_28_dash_4...</td>\n",
       "      <td>-3.26</td>\n",
       "      <td>-0.09</td>\n",
       "      <td>-1.675</td>\n",
       "      <td>-2.092139</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Desulfovibrionaceae, Muribaculaceae)</td>\n",
       "      <td>-1.73</td>\n",
       "      <td>-0.29</td>\n",
       "      <td>-1.010</td>\n",
       "      <td>-2.738673</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>(Lachnospiraceae_dash_bacterium_dash_10_dash_1...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3.63</td>\n",
       "      <td>2.125</td>\n",
       "      <td>2.786922</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                               Taxa  \\\n",
       "20            rs_response                                     Muribaculaceae   \n",
       "21            rs_response                                     Faecalibaculum   \n",
       "27            rs_response                                     Alloprevotella   \n",
       "29            rs_response                                Desulfovibrionaceae   \n",
       "40   pairwise_interaction                   (Muribaculaceae, Muribaculaceae)   \n",
       "41   pairwise_interaction                   (Muribaculaceae, Faecalibaculum)   \n",
       "43   pairwise_interaction  (Muribaculaceae, Lachnospiraceae_dash_NK4A136_...   \n",
       "48   pairwise_interaction  (Muribaculaceae, Lachnospiraceae_dash_bacteriu...   \n",
       "49   pairwise_interaction              (Muribaculaceae, Desulfovibrionaceae)   \n",
       "54   pairwise_interaction  (Muribaculaceae, Lachnospiraceae_dash_bacteriu...   \n",
       "55   pairwise_interaction  (Muribaculaceae, Ruminococcus_dash_sp_dot__das...   \n",
       "60   pairwise_interaction                   (Faecalibaculum, Muribaculaceae)   \n",
       "61   pairwise_interaction                   (Faecalibaculum, Faecalibaculum)   \n",
       "63   pairwise_interaction  (Faecalibaculum, Lachnospiraceae_dash_NK4A136_...   \n",
       "124  pairwise_interaction  (Parabacteroides_dash_goldsteinii, Parabactero...   \n",
       "140  pairwise_interaction                      (Bacteroides, Muribaculaceae)   \n",
       "145  pairwise_interaction                         (Bacteroides, Bacteroides)   \n",
       "180  pairwise_interaction                   (Alloprevotella, Muribaculaceae)   \n",
       "200  pairwise_interaction  (Lachnospiraceae_dash_bacterium_dash_28_dash_4...   \n",
       "202  pairwise_interaction  (Lachnospiraceae_dash_bacterium_dash_28_dash_4...   \n",
       "205  pairwise_interaction  (Lachnospiraceae_dash_bacterium_dash_28_dash_4...   \n",
       "220  pairwise_interaction              (Desulfovibrionaceae, Muribaculaceae)   \n",
       "334  pairwise_interaction  (Lachnospiraceae_dash_bacterium_dash_10_dash_1...   \n",
       "\n",
       "     Left  Right  Middle       SNR  Significant  \n",
       "20   0.42   0.90   0.660  5.491318         True  \n",
       "21   0.85   1.32   1.085  8.986253         True  \n",
       "27  -0.49  -0.02  -0.255 -2.153641         True  \n",
       "29   0.17   0.65   0.410  3.338960         True  \n",
       "40   2.88   4.31   3.595  9.897982         True  \n",
       "41  -4.73  -1.85  -3.290 -4.506180         True  \n",
       "43  -5.85  -2.90  -4.375 -5.836662         True  \n",
       "48  -3.28  -0.39  -1.835 -2.509557         True  \n",
       "49  -2.88  -0.11  -1.495 -2.125255         True  \n",
       "54  -3.75  -0.73  -2.240 -2.912772         True  \n",
       "55  -5.96  -2.93  -4.445 -5.779005         True  \n",
       "60  -1.59  -0.14  -0.865 -2.327044         True  \n",
       "61  -3.23  -0.34  -1.785 -2.407113         True  \n",
       "63  -3.52  -0.61  -2.065 -2.831374         True  \n",
       "124 -3.57  -0.39  -1.980 -2.461324         True  \n",
       "140 -1.72  -0.28  -1.000 -2.728627         True  \n",
       "145 -3.47  -0.28  -1.875 -2.327825         True  \n",
       "180  1.06   2.51   1.785  4.878909         True  \n",
       "200 -2.95  -1.50  -2.225 -6.072116         True  \n",
       "202  0.25   3.28   1.765  2.292892         True  \n",
       "205 -3.26  -0.09  -1.675 -2.092139         True  \n",
       "220 -1.73  -0.29  -1.010 -2.738673         True  \n",
       "334  0.62   3.63   2.125  2.786922         True  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_stan_output_summary = pd.read_excel('bayesian_regression_summary_top20.xlsx', index_col=0)\n",
    "df_stan_output_summary = df_stan_output_summary[df_stan_output_summary.Significant==True]\n",
    "df_stan_output_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
