{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from copy import deepcopy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from scipy.interpolate import CubicSpline\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import warnings\n",
    "import itertools\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import arviz as az\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "import statsmodels.stats.api as sms\n",
    "import scipy.stats.kde as kde"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read 16S data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abiotrophia_dash_defectiva</th>\n",
       "      <th>Absiella_dash_argi</th>\n",
       "      <th>Acholeplasmatales_dash_bacterium_dash_canine_dash_oral_dash_taxon_dash_316</th>\n",
       "      <th>Acidaminococcus</th>\n",
       "      <th>Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2764</th>\n",
       "      <th>Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Actinobacillus</th>\n",
       "      <th>Actinobacteria</th>\n",
       "      <th>Actinobacteriota</th>\n",
       "      <th>...</th>\n",
       "      <th>bacterium_dash_mpn_dash_isolate_dash_group_dash_16</th>\n",
       "      <th>bacterium_dash_oral_dash_clone_dash_ASCD05</th>\n",
       "      <th>butyrate_dash_producing_dash_bacterium_dash_SM4_slash_1</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "      <th>rumen_dash_bacterium_dash_NK4A78</th>\n",
       "      <th>swine_dash_fecal_dash_bacterium_dash_SD_dash_Xyl5</th>\n",
       "      <th>wallaby_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR10022259</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.039097</td>\n",
       "      <td>0.009150</td>\n",
       "      <td>0.009223</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022258</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.041739</td>\n",
       "      <td>0.018898</td>\n",
       "      <td>0.005560</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022120</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.091772</td>\n",
       "      <td>0.073355</td>\n",
       "      <td>0.030477</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022238</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.094728</td>\n",
       "      <td>0.054310</td>\n",
       "      <td>0.017197</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006318</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022162</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.094973</td>\n",
       "      <td>0.055338</td>\n",
       "      <td>0.009547</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.010040</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 426 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Abiotrophia_dash_defectiva  Absiella_dash_argi  \\\n",
       "SampleID                                                      \n",
       "SRR10022259                         0.0                 0.0   \n",
       "SRR10022258                         0.0                 0.0   \n",
       "SRR10022120                         0.0                 0.0   \n",
       "SRR10022238                         0.0                 0.0   \n",
       "SRR10022162                         0.0                 0.0   \n",
       "\n",
       "             Acholeplasmatales_dash_bacterium_dash_canine_dash_oral_dash_taxon_dash_316  \\\n",
       "SampleID                                                                                  \n",
       "SRR10022259                                                0.0                            \n",
       "SRR10022258                                                0.0                            \n",
       "SRR10022120                                                0.0                            \n",
       "SRR10022238                                                0.0                            \n",
       "SRR10022162                                                0.0                            \n",
       "\n",
       "             Acidaminococcus  \\\n",
       "SampleID                       \n",
       "SRR10022259              0.0   \n",
       "SRR10022258              0.0   \n",
       "SRR10022120              0.0   \n",
       "SRR10022238              0.0   \n",
       "SRR10022162              0.0   \n",
       "\n",
       "             Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2764  \\\n",
       "SampleID                                                              \n",
       "SRR10022259                                                0.0        \n",
       "SRR10022258                                                0.0        \n",
       "SRR10022120                                                0.0        \n",
       "SRR10022238                                                0.0        \n",
       "SRR10022162                                                0.0        \n",
       "\n",
       "             Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828  \\\n",
       "SampleID                                                              \n",
       "SRR10022259                                                0.0        \n",
       "SRR10022258                                                0.0        \n",
       "SRR10022120                                                0.0        \n",
       "SRR10022238                                                0.0        \n",
       "SRR10022162                                                0.0        \n",
       "\n",
       "             Acinetobacter  Actinobacillus  Actinobacteria  Actinobacteriota  \\\n",
       "SampleID                                                                       \n",
       "SRR10022259            0.0             0.0             0.0               0.0   \n",
       "SRR10022258            0.0             0.0             0.0               0.0   \n",
       "SRR10022120            0.0             0.0             0.0               0.0   \n",
       "SRR10022238            0.0             0.0             0.0               0.0   \n",
       "SRR10022162            0.0             0.0             0.0               0.0   \n",
       "\n",
       "             ...  bacterium_dash_mpn_dash_isolate_dash_group_dash_16  \\\n",
       "SampleID     ...                                                       \n",
       "SRR10022259  ...                                                0.0    \n",
       "SRR10022258  ...                                                0.0    \n",
       "SRR10022120  ...                                                0.0    \n",
       "SRR10022238  ...                                                0.0    \n",
       "SRR10022162  ...                                                0.0    \n",
       "\n",
       "             bacterium_dash_oral_dash_clone_dash_ASCD05  \\\n",
       "SampleID                                                  \n",
       "SRR10022259                                         0.0   \n",
       "SRR10022258                                         0.0   \n",
       "SRR10022120                                         0.0   \n",
       "SRR10022238                                         0.0   \n",
       "SRR10022162                                         0.0   \n",
       "\n",
       "             butyrate_dash_producing_dash_bacterium_dash_SM4_slash_1  \\\n",
       "SampleID                                                               \n",
       "SRR10022259                                                0.0         \n",
       "SRR10022258                                                0.0         \n",
       "SRR10022120                                                0.0         \n",
       "SRR10022238                                                0.0         \n",
       "SRR10022162                                                0.0         \n",
       "\n",
       "             gut_dash_metagenome  human_dash_gut_dash_metagenome  metagenome  \\\n",
       "SampleID                                                                       \n",
       "SRR10022259             0.039097                        0.009150    0.009223   \n",
       "SRR10022258             0.041739                        0.018898    0.005560   \n",
       "SRR10022120             0.091772                        0.073355    0.030477   \n",
       "SRR10022238             0.094728                        0.054310    0.017197   \n",
       "SRR10022162             0.094973                        0.055338    0.009547   \n",
       "\n",
       "             mouse_dash_gut_dash_metagenome  rumen_dash_bacterium_dash_NK4A78  \\\n",
       "SampleID                                                                        \n",
       "SRR10022259                             0.0                          0.000000   \n",
       "SRR10022258                             0.0                          0.000000   \n",
       "SRR10022120                             0.0                          0.000000   \n",
       "SRR10022238                             0.0                          0.006318   \n",
       "SRR10022162                             0.0                          0.000000   \n",
       "\n",
       "             swine_dash_fecal_dash_bacterium_dash_SD_dash_Xyl5  \\\n",
       "SampleID                                                         \n",
       "SRR10022259                                                0.0   \n",
       "SRR10022258                                                0.0   \n",
       "SRR10022120                                                0.0   \n",
       "SRR10022238                                                0.0   \n",
       "SRR10022162                                                0.0   \n",
       "\n",
       "             wallaby_dash_gut_dash_metagenome  \n",
       "SampleID                                       \n",
       "SRR10022259                          0.000000  \n",
       "SRR10022258                          0.001126  \n",
       "SRR10022120                          0.001085  \n",
       "SRR10022238                          0.000000  \n",
       "SRR10022162                          0.010040  \n",
       "\n",
       "[5 rows x 426 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_16S = pd.read_excel('./16S_relative_abundance.xlsx', index_col=0)\n",
    "df_16S.columns = [c.replace('/','_slash_').replace(' ','_space_').replace('[','_leftsquarebracket_').replace(']','_rightsquarebracket_').replace('-','_dash_').replace('.','_dot_').replace('(','_leftroundbracket').replace(')','_rightroundbracket_') for c in df_16S.columns]\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add pseudocount for zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abiotrophia_dash_defectiva</th>\n",
       "      <th>Absiella_dash_argi</th>\n",
       "      <th>Acholeplasmatales_dash_bacterium_dash_canine_dash_oral_dash_taxon_dash_316</th>\n",
       "      <th>Acidaminococcus</th>\n",
       "      <th>Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2764</th>\n",
       "      <th>Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828</th>\n",
       "      <th>Acinetobacter</th>\n",
       "      <th>Actinobacillus</th>\n",
       "      <th>Actinobacteria</th>\n",
       "      <th>Actinobacteriota</th>\n",
       "      <th>...</th>\n",
       "      <th>bacterium_dash_mpn_dash_isolate_dash_group_dash_16</th>\n",
       "      <th>bacterium_dash_oral_dash_clone_dash_ASCD05</th>\n",
       "      <th>butyrate_dash_producing_dash_bacterium_dash_SM4_slash_1</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>human_dash_gut_dash_metagenome</th>\n",
       "      <th>metagenome</th>\n",
       "      <th>mouse_dash_gut_dash_metagenome</th>\n",
       "      <th>rumen_dash_bacterium_dash_NK4A78</th>\n",
       "      <th>swine_dash_fecal_dash_bacterium_dash_SD_dash_Xyl5</th>\n",
       "      <th>wallaby_dash_gut_dash_metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR10022259</th>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.039097</td>\n",
       "      <td>0.009150</td>\n",
       "      <td>0.009223</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022258</th>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.041739</td>\n",
       "      <td>0.018898</td>\n",
       "      <td>0.005560</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.001126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022120</th>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.091772</td>\n",
       "      <td>0.073355</td>\n",
       "      <td>0.030477</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.001085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022238</th>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.094728</td>\n",
       "      <td>0.054310</td>\n",
       "      <td>0.017197</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.006318</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022162</th>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.094973</td>\n",
       "      <td>0.055338</td>\n",
       "      <td>0.009547</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.010040</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 426 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Abiotrophia_dash_defectiva  Absiella_dash_argi  \\\n",
       "SampleID                                                      \n",
       "SRR10022259                    0.000072            0.000072   \n",
       "SRR10022258                    0.000176            0.000176   \n",
       "SRR10022120                    0.000186            0.000186   \n",
       "SRR10022238                    0.000209            0.000209   \n",
       "SRR10022162                    0.000132            0.000132   \n",
       "\n",
       "             Acholeplasmatales_dash_bacterium_dash_canine_dash_oral_dash_taxon_dash_316  \\\n",
       "SampleID                                                                                  \n",
       "SRR10022259                                           0.000072                            \n",
       "SRR10022258                                           0.000176                            \n",
       "SRR10022120                                           0.000186                            \n",
       "SRR10022238                                           0.000209                            \n",
       "SRR10022162                                           0.000132                            \n",
       "\n",
       "             Acidaminococcus  \\\n",
       "SampleID                       \n",
       "SRR10022259         0.000072   \n",
       "SRR10022258         0.000176   \n",
       "SRR10022120         0.000186   \n",
       "SRR10022238         0.000209   \n",
       "SRR10022162         0.000132   \n",
       "\n",
       "             Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2764  \\\n",
       "SampleID                                                              \n",
       "SRR10022259                                           0.000072        \n",
       "SRR10022258                                           0.000176        \n",
       "SRR10022120                                           0.000186        \n",
       "SRR10022238                                           0.000209        \n",
       "SRR10022162                                           0.000132        \n",
       "\n",
       "             Acidaminococcus_dash_sp_dot__dash_Marseille_dash_P2828  \\\n",
       "SampleID                                                              \n",
       "SRR10022259                                           0.000072        \n",
       "SRR10022258                                           0.000176        \n",
       "SRR10022120                                           0.000186        \n",
       "SRR10022238                                           0.000209        \n",
       "SRR10022162                                           0.000132        \n",
       "\n",
       "             Acinetobacter  Actinobacillus  Actinobacteria  Actinobacteriota  \\\n",
       "SampleID                                                                       \n",
       "SRR10022259       0.000072        0.000072        0.000072          0.000072   \n",
       "SRR10022258       0.000176        0.000176        0.000176          0.000176   \n",
       "SRR10022120       0.000186        0.000186        0.000186          0.000186   \n",
       "SRR10022238       0.000209        0.000209        0.000209          0.000209   \n",
       "SRR10022162       0.000132        0.000132        0.000132          0.000132   \n",
       "\n",
       "             ...  bacterium_dash_mpn_dash_isolate_dash_group_dash_16  \\\n",
       "SampleID     ...                                                       \n",
       "SRR10022259  ...                                           0.000072    \n",
       "SRR10022258  ...                                           0.000176    \n",
       "SRR10022120  ...                                           0.000186    \n",
       "SRR10022238  ...                                           0.000209    \n",
       "SRR10022162  ...                                           0.000132    \n",
       "\n",
       "             bacterium_dash_oral_dash_clone_dash_ASCD05  \\\n",
       "SampleID                                                  \n",
       "SRR10022259                                    0.000072   \n",
       "SRR10022258                                    0.000176   \n",
       "SRR10022120                                    0.000186   \n",
       "SRR10022238                                    0.000209   \n",
       "SRR10022162                                    0.000132   \n",
       "\n",
       "             butyrate_dash_producing_dash_bacterium_dash_SM4_slash_1  \\\n",
       "SampleID                                                               \n",
       "SRR10022259                                           0.000072         \n",
       "SRR10022258                                           0.000176         \n",
       "SRR10022120                                           0.000186         \n",
       "SRR10022238                                           0.000209         \n",
       "SRR10022162                                           0.000132         \n",
       "\n",
       "             gut_dash_metagenome  human_dash_gut_dash_metagenome  metagenome  \\\n",
       "SampleID                                                                       \n",
       "SRR10022259             0.039097                        0.009150    0.009223   \n",
       "SRR10022258             0.041739                        0.018898    0.005560   \n",
       "SRR10022120             0.091772                        0.073355    0.030477   \n",
       "SRR10022238             0.094728                        0.054310    0.017197   \n",
       "SRR10022162             0.094973                        0.055338    0.009547   \n",
       "\n",
       "             mouse_dash_gut_dash_metagenome  rumen_dash_bacterium_dash_NK4A78  \\\n",
       "SampleID                                                                        \n",
       "SRR10022259                        0.000072                          0.000072   \n",
       "SRR10022258                        0.000176                          0.000176   \n",
       "SRR10022120                        0.000186                          0.000186   \n",
       "SRR10022238                        0.000209                          0.006318   \n",
       "SRR10022162                        0.000132                          0.000132   \n",
       "\n",
       "             swine_dash_fecal_dash_bacterium_dash_SD_dash_Xyl5  \\\n",
       "SampleID                                                         \n",
       "SRR10022259                                           0.000072   \n",
       "SRR10022258                                           0.000176   \n",
       "SRR10022120                                           0.000186   \n",
       "SRR10022238                                           0.000209   \n",
       "SRR10022162                                           0.000132   \n",
       "\n",
       "             wallaby_dash_gut_dash_metagenome  \n",
       "SampleID                                       \n",
       "SRR10022259                          0.000072  \n",
       "SRR10022258                          0.001126  \n",
       "SRR10022120                          0.001085  \n",
       "SRR10022238                          0.000209  \n",
       "SRR10022162                          0.010040  \n",
       "\n",
       "[5 rows x 426 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample_id in df_16S.index:\n",
    "    sample = np.array(df_16S.loc[sample_id])\n",
    "    minval = np.min(sample[np.nonzero(sample)]) # minimum non-zero value\n",
    "    sample[sample==0] = minval\n",
    "    df_16S.loc[sample_id] = sample\n",
    "df_16S.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read metadata for maize starch and control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject</th>\n",
       "      <th>Dose</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR10022259</th>\n",
       "      <td>F01</td>\n",
       "      <td>baseline</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022258</th>\n",
       "      <td>F01</td>\n",
       "      <td>10g</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022120</th>\n",
       "      <td>F01</td>\n",
       "      <td>20g</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022238</th>\n",
       "      <td>F01</td>\n",
       "      <td>35g</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022162</th>\n",
       "      <td>F01</td>\n",
       "      <td>50g</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Subject      Dose  Day  Diet\n",
       "sample-id                               \n",
       "SRR10022259     F01  baseline    0     0\n",
       "SRR10022258     F01       10g    1     0\n",
       "SRR10022120     F01       20g    2     0\n",
       "SRR10022238     F01       35g    3     0\n",
       "SRR10022162     F01       50g    4     0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv('./metadata.tsv', sep='\\t', index_col=0)\n",
    "df_meta = df_meta.drop('#q2:types')\n",
    "df_meta.Subject = [s.strip() for s in df_meta.Subject]\n",
    "df_meta = df_meta[['Treatment','Subject','Dose','week']]\n",
    "df_meta = df_meta[(df_meta.Treatment=='Maize') | (df_meta.Treatment=='Control')]\n",
    "df_meta.week = df_meta.week.astype(int)\n",
    "df_meta = df_meta.rename({'week':'Day'}, axis=1)\n",
    "df_meta['Diet'] = [1 if g=='Maize' else 0 for g in df_meta['Treatment']]\n",
    "df_meta = df_meta.drop('Treatment', axis=1)\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the 20 most abundant taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Agathobacter</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>Prevotella</th>\n",
       "      <th>Faecalibacterium_dash_sp_dot_</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Blautia_dash_sp_dot__dash_GD9</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Bifidobacterium_dash_faecale</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>Streptococcus</th>\n",
       "      <th>Dorea_dash_longicatena</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR10022259</th>\n",
       "      <td>0.249412</td>\n",
       "      <td>0.018771</td>\n",
       "      <td>0.047380</td>\n",
       "      <td>0.030200</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.039097</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.043257</td>\n",
       "      <td>0.030164</td>\n",
       "      <td>0.010886</td>\n",
       "      <td>0.031249</td>\n",
       "      <td>0.028934</td>\n",
       "      <td>0.027524</td>\n",
       "      <td>0.001881</td>\n",
       "      <td>0.013635</td>\n",
       "      <td>0.028030</td>\n",
       "      <td>0.003364</td>\n",
       "      <td>0.020616</td>\n",
       "      <td>0.023220</td>\n",
       "      <td>0.009223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022258</th>\n",
       "      <td>0.080345</td>\n",
       "      <td>0.015379</td>\n",
       "      <td>0.102939</td>\n",
       "      <td>0.030547</td>\n",
       "      <td>0.012036</td>\n",
       "      <td>0.041739</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.007672</td>\n",
       "      <td>0.031357</td>\n",
       "      <td>0.021749</td>\n",
       "      <td>0.019778</td>\n",
       "      <td>0.018335</td>\n",
       "      <td>0.009889</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.042231</td>\n",
       "      <td>0.013373</td>\n",
       "      <td>0.074362</td>\n",
       "      <td>0.013197</td>\n",
       "      <td>0.018371</td>\n",
       "      <td>0.005560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022120</th>\n",
       "      <td>0.097941</td>\n",
       "      <td>0.010634</td>\n",
       "      <td>0.087400</td>\n",
       "      <td>0.023191</td>\n",
       "      <td>0.006635</td>\n",
       "      <td>0.091772</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.008309</td>\n",
       "      <td>0.040460</td>\n",
       "      <td>0.003689</td>\n",
       "      <td>0.028276</td>\n",
       "      <td>0.014107</td>\n",
       "      <td>0.019656</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.033701</td>\n",
       "      <td>0.008619</td>\n",
       "      <td>0.017362</td>\n",
       "      <td>0.011037</td>\n",
       "      <td>0.004837</td>\n",
       "      <td>0.030477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022238</th>\n",
       "      <td>0.089289</td>\n",
       "      <td>0.027699</td>\n",
       "      <td>0.084142</td>\n",
       "      <td>0.044017</td>\n",
       "      <td>0.003598</td>\n",
       "      <td>0.094728</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.004895</td>\n",
       "      <td>0.025732</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.036444</td>\n",
       "      <td>0.002385</td>\n",
       "      <td>0.006151</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.029289</td>\n",
       "      <td>0.008452</td>\n",
       "      <td>0.021297</td>\n",
       "      <td>0.010167</td>\n",
       "      <td>0.013933</td>\n",
       "      <td>0.017197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022162</th>\n",
       "      <td>0.111137</td>\n",
       "      <td>0.033216</td>\n",
       "      <td>0.060506</td>\n",
       "      <td>0.020641</td>\n",
       "      <td>0.011160</td>\n",
       "      <td>0.094973</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.005991</td>\n",
       "      <td>0.025974</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.014188</td>\n",
       "      <td>0.010238</td>\n",
       "      <td>0.004280</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.032854</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.025447</td>\n",
       "      <td>0.009152</td>\n",
       "      <td>0.013892</td>\n",
       "      <td>0.009547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Bacteroides  Agathobacter  Faecalibacterium   Blautia  \\\n",
       "sample-id                                                            \n",
       "SRR10022259     0.249412      0.018771          0.047380  0.030200   \n",
       "SRR10022258     0.080345      0.015379          0.102939  0.030547   \n",
       "SRR10022120     0.097941      0.010634          0.087400  0.023191   \n",
       "SRR10022238     0.089289      0.027699          0.084142  0.044017   \n",
       "SRR10022162     0.111137      0.033216          0.060506  0.020641   \n",
       "\n",
       "             Bifidobacterium  gut_dash_metagenome  Prevotella  \\\n",
       "sample-id                                                       \n",
       "SRR10022259         0.000072             0.039097    0.000072   \n",
       "SRR10022258         0.012036             0.041739    0.000176   \n",
       "SRR10022120         0.006635             0.091772    0.000186   \n",
       "SRR10022238         0.003598             0.094728    0.000209   \n",
       "SRR10022162         0.011160             0.094973    0.000132   \n",
       "\n",
       "             Faecalibacterium_dash_sp_dot_  Alistipes  \\\n",
       "sample-id                                               \n",
       "SRR10022259                       0.043257   0.030164   \n",
       "SRR10022258                       0.007672   0.031357   \n",
       "SRR10022120                       0.008309   0.040460   \n",
       "SRR10022238                       0.004895   0.025732   \n",
       "SRR10022162                       0.005991   0.025974   \n",
       "\n",
       "             Blautia_dash_sp_dot__dash_GD9  Fusicatenibacter  Ruminococcus  \\\n",
       "sample-id                                                                    \n",
       "SRR10022259                       0.010886          0.031249      0.028934   \n",
       "SRR10022258                       0.021749          0.019778      0.018335   \n",
       "SRR10022120                       0.003689          0.028276      0.014107   \n",
       "SRR10022238                       0.000209          0.036444      0.002385   \n",
       "SRR10022162                       0.000132          0.014188      0.010238   \n",
       "\n",
       "             Lachnospiraceae  Bifidobacterium_dash_faecale  Anaerostipes  \\\n",
       "sample-id                                                                  \n",
       "SRR10022259         0.027524                      0.001881      0.013635   \n",
       "SRR10022258         0.009889                      0.000176      0.042231   \n",
       "SRR10022120         0.019656                      0.000186      0.033701   \n",
       "SRR10022238         0.006151                      0.000209      0.029289   \n",
       "SRR10022162         0.004280                      0.000132      0.032854   \n",
       "\n",
       "             _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "sample-id                                                                                  \n",
       "SRR10022259                                           0.028030                             \n",
       "SRR10022258                                           0.013373                             \n",
       "SRR10022120                                           0.008619                             \n",
       "SRR10022238                                           0.008452                             \n",
       "SRR10022162                                           0.000132                             \n",
       "\n",
       "             Streptococcus  Dorea_dash_longicatena  Roseburia  metagenome  \n",
       "sample-id                                                                  \n",
       "SRR10022259       0.003364                0.020616   0.023220    0.009223  \n",
       "SRR10022258       0.074362                0.013197   0.018371    0.005560  \n",
       "SRR10022120       0.017362                0.011037   0.004837    0.030477  \n",
       "SRR10022238       0.021297                0.010167   0.013933    0.017197  \n",
       "SRR10022162       0.025447                0.009152   0.013892    0.009547  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topX = 20\n",
    "df_16S_T = df_16S.loc[df_meta.index].T\n",
    "df_16S_T['mean'] = df_16S_T.mean(axis=1)\n",
    "df_16S_T = df_16S_T.sort_values(by=['mean'],axis=0,ascending=False)\n",
    "df_16S_T = df_16S_T.drop('mean', axis=1)\n",
    "df_16S_topX = df_16S_T.iloc[0:topX].T\n",
    "df_16S_topX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Bacteroides',\n",
       " 'Agathobacter',\n",
       " 'Faecalibacterium',\n",
       " 'Blautia',\n",
       " 'Bifidobacterium',\n",
       " 'gut_dash_metagenome',\n",
       " 'Prevotella',\n",
       " 'Faecalibacterium_dash_sp_dot_',\n",
       " 'Alistipes',\n",
       " 'Blautia_dash_sp_dot__dash_GD9',\n",
       " 'Fusicatenibacter',\n",
       " 'Ruminococcus',\n",
       " 'Lachnospiraceae',\n",
       " 'Bifidobacterium_dash_faecale',\n",
       " 'Anaerostipes',\n",
       " '_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group',\n",
       " 'Streptococcus',\n",
       " 'Dorea_dash_longicatena',\n",
       " 'Roseburia',\n",
       " 'metagenome']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_16S_topX.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize max value of 16S data to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject</th>\n",
       "      <th>Dose</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Agathobacter</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>...</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Bifidobacterium_dash_faecale</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>Streptococcus</th>\n",
       "      <th>Dorea_dash_longicatena</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR10022259</th>\n",
       "      <td>F01</td>\n",
       "      <td>baseline</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.758976</td>\n",
       "      <td>0.057121</td>\n",
       "      <td>0.144179</td>\n",
       "      <td>0.091900</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.118975</td>\n",
       "      <td>...</td>\n",
       "      <td>0.095092</td>\n",
       "      <td>0.088048</td>\n",
       "      <td>0.083756</td>\n",
       "      <td>0.005723</td>\n",
       "      <td>0.041493</td>\n",
       "      <td>0.085297</td>\n",
       "      <td>0.010236</td>\n",
       "      <td>0.062734</td>\n",
       "      <td>0.070659</td>\n",
       "      <td>0.028065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022258</th>\n",
       "      <td>F01</td>\n",
       "      <td>10g</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.244494</td>\n",
       "      <td>0.046800</td>\n",
       "      <td>0.313248</td>\n",
       "      <td>0.092957</td>\n",
       "      <td>0.036626</td>\n",
       "      <td>0.127013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.060186</td>\n",
       "      <td>0.055796</td>\n",
       "      <td>0.030093</td>\n",
       "      <td>0.000535</td>\n",
       "      <td>0.128512</td>\n",
       "      <td>0.040695</td>\n",
       "      <td>0.226288</td>\n",
       "      <td>0.040160</td>\n",
       "      <td>0.055903</td>\n",
       "      <td>0.016921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022120</th>\n",
       "      <td>F01</td>\n",
       "      <td>20g</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.298041</td>\n",
       "      <td>0.032361</td>\n",
       "      <td>0.265963</td>\n",
       "      <td>0.070571</td>\n",
       "      <td>0.020190</td>\n",
       "      <td>0.279266</td>\n",
       "      <td>...</td>\n",
       "      <td>0.086044</td>\n",
       "      <td>0.042928</td>\n",
       "      <td>0.059816</td>\n",
       "      <td>0.000566</td>\n",
       "      <td>0.102555</td>\n",
       "      <td>0.026228</td>\n",
       "      <td>0.052834</td>\n",
       "      <td>0.033587</td>\n",
       "      <td>0.014718</td>\n",
       "      <td>0.092743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022238</th>\n",
       "      <td>F01</td>\n",
       "      <td>35g</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.271711</td>\n",
       "      <td>0.084289</td>\n",
       "      <td>0.256050</td>\n",
       "      <td>0.133946</td>\n",
       "      <td>0.010950</td>\n",
       "      <td>0.288263</td>\n",
       "      <td>...</td>\n",
       "      <td>0.110900</td>\n",
       "      <td>0.007258</td>\n",
       "      <td>0.018717</td>\n",
       "      <td>0.000637</td>\n",
       "      <td>0.089127</td>\n",
       "      <td>0.025720</td>\n",
       "      <td>0.064808</td>\n",
       "      <td>0.030940</td>\n",
       "      <td>0.042399</td>\n",
       "      <td>0.052330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022162</th>\n",
       "      <td>F01</td>\n",
       "      <td>50g</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.338196</td>\n",
       "      <td>0.101078</td>\n",
       "      <td>0.184124</td>\n",
       "      <td>0.062811</td>\n",
       "      <td>0.033960</td>\n",
       "      <td>0.289009</td>\n",
       "      <td>...</td>\n",
       "      <td>0.043176</td>\n",
       "      <td>0.031155</td>\n",
       "      <td>0.013023</td>\n",
       "      <td>0.000401</td>\n",
       "      <td>0.099976</td>\n",
       "      <td>0.000401</td>\n",
       "      <td>0.077436</td>\n",
       "      <td>0.027849</td>\n",
       "      <td>0.042274</td>\n",
       "      <td>0.029051</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Subject      Dose  Day  Diet  Bacteroides  Agathobacter  \\\n",
       "sample-id                                                             \n",
       "SRR10022259     F01  baseline    0     0     0.758976      0.057121   \n",
       "SRR10022258     F01       10g    1     0     0.244494      0.046800   \n",
       "SRR10022120     F01       20g    2     0     0.298041      0.032361   \n",
       "SRR10022238     F01       35g    3     0     0.271711      0.084289   \n",
       "SRR10022162     F01       50g    4     0     0.338196      0.101078   \n",
       "\n",
       "             Faecalibacterium   Blautia  Bifidobacterium  gut_dash_metagenome  \\\n",
       "sample-id                                                                       \n",
       "SRR10022259          0.144179  0.091900         0.000220             0.118975   \n",
       "SRR10022258          0.313248  0.092957         0.036626             0.127013   \n",
       "SRR10022120          0.265963  0.070571         0.020190             0.279266   \n",
       "SRR10022238          0.256050  0.133946         0.010950             0.288263   \n",
       "SRR10022162          0.184124  0.062811         0.033960             0.289009   \n",
       "\n",
       "             ...  Fusicatenibacter  Ruminococcus  Lachnospiraceae  \\\n",
       "sample-id    ...                                                    \n",
       "SRR10022259  ...          0.095092      0.088048         0.083756   \n",
       "SRR10022258  ...          0.060186      0.055796         0.030093   \n",
       "SRR10022120  ...          0.086044      0.042928         0.059816   \n",
       "SRR10022238  ...          0.110900      0.007258         0.018717   \n",
       "SRR10022162  ...          0.043176      0.031155         0.013023   \n",
       "\n",
       "             Bifidobacterium_dash_faecale  Anaerostipes  \\\n",
       "sample-id                                                 \n",
       "SRR10022259                      0.005723      0.041493   \n",
       "SRR10022258                      0.000535      0.128512   \n",
       "SRR10022120                      0.000566      0.102555   \n",
       "SRR10022238                      0.000637      0.089127   \n",
       "SRR10022162                      0.000401      0.099976   \n",
       "\n",
       "             _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "sample-id                                                                                  \n",
       "SRR10022259                                           0.085297                             \n",
       "SRR10022258                                           0.040695                             \n",
       "SRR10022120                                           0.026228                             \n",
       "SRR10022238                                           0.025720                             \n",
       "SRR10022162                                           0.000401                             \n",
       "\n",
       "             Streptococcus  Dorea_dash_longicatena  Roseburia  metagenome  \n",
       "sample-id                                                                  \n",
       "SRR10022259       0.010236                0.062734   0.070659    0.028065  \n",
       "SRR10022258       0.226288                0.040160   0.055903    0.016921  \n",
       "SRR10022120       0.052834                0.033587   0.014718    0.092743  \n",
       "SRR10022238       0.064808                0.030940   0.042399    0.052330  \n",
       "SRR10022162       0.077436                0.027849   0.042274    0.029051  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_taxa = list(df_16S_topX.columns)\n",
    "df_16S_topX_w_meta = pd.merge(df_meta, df_16S_topX/df_16S_topX.max().max(), left_index=True, right_index=True, how='inner')\n",
    "df_16S_topX_w_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate log-derivatives (dlog(B)/dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject</th>\n",
       "      <th>Dose</th>\n",
       "      <th>Day</th>\n",
       "      <th>Diet</th>\n",
       "      <th>Bacteroides</th>\n",
       "      <th>Agathobacter</th>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <th>Blautia</th>\n",
       "      <th>Bifidobacterium</th>\n",
       "      <th>gut_dash_metagenome</th>\n",
       "      <th>...</th>\n",
       "      <th>Fusicatenibacter</th>\n",
       "      <th>Ruminococcus</th>\n",
       "      <th>Lachnospiraceae</th>\n",
       "      <th>Bifidobacterium_dash_faecale</th>\n",
       "      <th>Anaerostipes</th>\n",
       "      <th>_leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group</th>\n",
       "      <th>Streptococcus</th>\n",
       "      <th>Dorea_dash_longicatena</th>\n",
       "      <th>Roseburia</th>\n",
       "      <th>metagenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR10022259</th>\n",
       "      <td>F01</td>\n",
       "      <td>baseline</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.082849</td>\n",
       "      <td>0.029564</td>\n",
       "      <td>0.391247</td>\n",
       "      <td>0.076635</td>\n",
       "      <td>0.085754</td>\n",
       "      <td>-0.195101</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.083392</td>\n",
       "      <td>-0.066387</td>\n",
       "      <td>-0.168460</td>\n",
       "      <td>-0.009926</td>\n",
       "      <td>0.194823</td>\n",
       "      <td>-0.063178</td>\n",
       "      <td>0.665760</td>\n",
       "      <td>-0.035243</td>\n",
       "      <td>0.046220</td>\n",
       "      <td>-0.150409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022258</th>\n",
       "      <td>F01</td>\n",
       "      <td>10g</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.088291</td>\n",
       "      <td>-0.031294</td>\n",
       "      <td>0.003892</td>\n",
       "      <td>-0.042593</td>\n",
       "      <td>-0.001479</td>\n",
       "      <td>0.145661</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004528</td>\n",
       "      <td>-0.010339</td>\n",
       "      <td>0.024583</td>\n",
       "      <td>-0.001514</td>\n",
       "      <td>0.004873</td>\n",
       "      <td>-0.027780</td>\n",
       "      <td>-0.106178</td>\n",
       "      <td>-0.012240</td>\n",
       "      <td>-0.051851</td>\n",
       "      <td>0.080229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022120</th>\n",
       "      <td>F01</td>\n",
       "      <td>20g</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.053209</td>\n",
       "      <td>0.021330</td>\n",
       "      <td>-0.041461</td>\n",
       "      <td>0.029750</td>\n",
       "      <td>-0.019929</td>\n",
       "      <td>0.093331</td>\n",
       "      <td>...</td>\n",
       "      <td>0.038135</td>\n",
       "      <td>-0.027618</td>\n",
       "      <td>-0.001690</td>\n",
       "      <td>0.000511</td>\n",
       "      <td>-0.031130</td>\n",
       "      <td>-0.002909</td>\n",
       "      <td>-0.113253</td>\n",
       "      <td>-0.003240</td>\n",
       "      <td>-0.006637</td>\n",
       "      <td>0.023524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022238</th>\n",
       "      <td>F01</td>\n",
       "      <td>35g</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.042897</td>\n",
       "      <td>0.058442</td>\n",
       "      <td>-0.009643</td>\n",
       "      <td>0.046559</td>\n",
       "      <td>0.004167</td>\n",
       "      <td>-0.035233</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004929</td>\n",
       "      <td>-0.024805</td>\n",
       "      <td>-0.051952</td>\n",
       "      <td>-0.000226</td>\n",
       "      <td>0.001493</td>\n",
       "      <td>-0.005511</td>\n",
       "      <td>0.074751</td>\n",
       "      <td>-0.002462</td>\n",
       "      <td>0.037889</td>\n",
       "      <td>-0.068097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR10022162</th>\n",
       "      <td>F01</td>\n",
       "      <td>50g</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.238840</td>\n",
       "      <td>-0.048948</td>\n",
       "      <td>-0.165485</td>\n",
       "      <td>-0.239268</td>\n",
       "      <td>0.044571</td>\n",
       "      <td>0.076829</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.147023</td>\n",
       "      <td>0.091518</td>\n",
       "      <td>0.069121</td>\n",
       "      <td>-0.000102</td>\n",
       "      <td>0.017423</td>\n",
       "      <td>-0.052529</td>\n",
       "      <td>-0.111945</td>\n",
       "      <td>-0.004126</td>\n",
       "      <td>-0.062248</td>\n",
       "      <td>0.057791</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Subject      Dose  Day  Diet  Bacteroides  Agathobacter  \\\n",
       "sample-id                                                             \n",
       "SRR10022259     F01  baseline    0     0    -1.082849      0.029564   \n",
       "SRR10022258     F01       10g    1     0    -0.088291     -0.031294   \n",
       "SRR10022120     F01       20g    2     0     0.053209      0.021330   \n",
       "SRR10022238     F01       35g    3     0    -0.042897      0.058442   \n",
       "SRR10022162     F01       50g    4     0     0.238840     -0.048948   \n",
       "\n",
       "             Faecalibacterium   Blautia  Bifidobacterium  gut_dash_metagenome  \\\n",
       "sample-id                                                                       \n",
       "SRR10022259          0.391247  0.076635         0.085754            -0.195101   \n",
       "SRR10022258          0.003892 -0.042593        -0.001479             0.145661   \n",
       "SRR10022120         -0.041461  0.029750        -0.019929             0.093331   \n",
       "SRR10022238         -0.009643  0.046559         0.004167            -0.035233   \n",
       "SRR10022162         -0.165485 -0.239268         0.044571             0.076829   \n",
       "\n",
       "             ...  Fusicatenibacter  Ruminococcus  Lachnospiraceae  \\\n",
       "sample-id    ...                                                    \n",
       "SRR10022259  ...         -0.083392     -0.066387        -0.168460   \n",
       "SRR10022258  ...          0.004528     -0.010339         0.024583   \n",
       "SRR10022120  ...          0.038135     -0.027618        -0.001690   \n",
       "SRR10022238  ...         -0.004929     -0.024805        -0.051952   \n",
       "SRR10022162  ...         -0.147023      0.091518         0.069121   \n",
       "\n",
       "             Bifidobacterium_dash_faecale  Anaerostipes  \\\n",
       "sample-id                                                 \n",
       "SRR10022259                     -0.009926      0.194823   \n",
       "SRR10022258                     -0.001514      0.004873   \n",
       "SRR10022120                      0.000511     -0.031130   \n",
       "SRR10022238                     -0.000226      0.001493   \n",
       "SRR10022162                     -0.000102      0.017423   \n",
       "\n",
       "             _leftsquarebracket_Ruminococcus_rightsquarebracket__dash_torques_dash_group  \\\n",
       "sample-id                                                                                  \n",
       "SRR10022259                                          -0.063178                             \n",
       "SRR10022258                                          -0.027780                             \n",
       "SRR10022120                                          -0.002909                             \n",
       "SRR10022238                                          -0.005511                             \n",
       "SRR10022162                                          -0.052529                             \n",
       "\n",
       "             Streptococcus  Dorea_dash_longicatena  Roseburia  metagenome  \n",
       "sample-id                                                                  \n",
       "SRR10022259       0.665760               -0.035243   0.046220   -0.150409  \n",
       "SRR10022258      -0.106178               -0.012240  -0.051851    0.080229  \n",
       "SRR10022120      -0.113253               -0.003240  -0.006637    0.023524  \n",
       "SRR10022238       0.074751               -0.002462   0.037889   -0.068097  \n",
       "SRR10022162      -0.111945               -0.004126  -0.062248    0.057791  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_deriv = deepcopy(df_16S_topX_w_meta)\n",
    "for curr_sub in set(df_deriv.Subject):\n",
    "    curr_df = df_deriv[df_deriv.Subject==curr_sub].sort_values(by='Day')\n",
    "    for taxon in bacterial_taxa:\n",
    "        xdata = np.array(curr_df['Day'])\n",
    "        ydata = np.array(curr_df[taxon])\n",
    "        cs = CubicSpline(xdata, ydata)\n",
    "        csd1 = cs.derivative(nu=1)\n",
    "        ydata_d1 = csd1(xdata)\n",
    "        df_deriv.loc[df_deriv.Subject==curr_sub, taxon] = ydata_d1\n",
    "df_deriv.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct regression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-7.19185011, -0.5861302 ,  0.35369996, ..., -0.00742527,\n",
       "       -0.06780106,  0.1866441 ])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ymat should be samples by bacteria\n",
    "Ymat = df_deriv[bacterial_taxa].values\n",
    "Ymat = Ymat.flatten(order='F')\n",
    "Ymat = StandardScaler().fit_transform(Ymat.reshape(-1,1)).reshape(1,-1)[0] # standardize\n",
    "Ymat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2000"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Ymat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.        , 0.        , 0.75897617, ..., 0.        , 0.        ,\n",
       "        0.        ],\n",
       "       [1.        , 0.        , 0.2444942 , ..., 0.        , 0.        ,\n",
       "        0.        ],\n",
       "       [1.        , 0.        , 0.29804124, ..., 0.        , 0.        ,\n",
       "        0.        ],\n",
       "       ...,\n",
       "       [0.        , 0.        , 0.        , ..., 0.09230468, 0.03909039,\n",
       "        0.01626389],\n",
       "       [0.        , 0.        , 0.        , ..., 0.09130017, 0.08156949,\n",
       "        0.00660725],\n",
       "       [0.        , 0.        , 0.        , ..., 0.11636208, 0.0313341 ,\n",
       "        0.01156015]])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Xmat = np.zeros(shape=(topX*len(df_deriv.index), (topX+2)*topX))\n",
    "for k in np.arange(topX):\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)] = 1\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+1] = df_deriv.Diet.values\n",
    "    Xmat[k*len(df_deriv.index):(k+1)*len(df_deriv.index),k*(topX+2)+2:(k+1)*(topX+2)] = df_16S_topX_w_meta[bacterial_taxa].values\n",
    "Xmat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write stan program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "json_str = '{\\n\"N\" : %d,\\n'%(len(Ymat))\n",
    "\n",
    "# y variable\n",
    "json_str += '\\\"dlogX\\\" : [%s],\\n'%(','.join(list(Ymat.astype(str))))\n",
    "\n",
    "# x variable\n",
    "for k1,c1 in enumerate(bacterial_taxa):\n",
    "    # growth rate\n",
    "    json_str += '\\\"growth_rate_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)].astype(str))))\n",
    "    # diet response\n",
    "    json_str += '\\\"rs_response_%s\\\" : [%s],\\n'%(c1,','.join(list(Xmat[:,k1*(topX+2)+1].astype(str))))\n",
    "    # bacterial interactions\n",
    "    for k2,c2 in enumerate(bacterial_taxa):\n",
    "        v = list(Xmat[:,k1*(topX+2)+2+k2].astype(str))\n",
    "        json_str += '\\\"pairwise_interaction_%s_%s\\\" : [%s]'%(c1,c2,','.join(v))\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            json_str += '\\n}'\n",
    "        else:\n",
    "            json_str += ',\\n'\n",
    "#print(json_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa_3/mice_scfa_3.data.json\", \"w\")\n",
    "text_file.write(\"%s\" % json_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# data block\n",
    "model_str = 'data {\\n'\n",
    "model_str += '\\tint<lower=0> N;\\n'\n",
    "model_str += '\\tvector[N] dlogX;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\tvector[N] growth_rate_%s;\\n'%(c1)\n",
    "    model_str += '\\tvector[N] rs_response_%s;\\n'%(c1)\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tvector[N] pairwise_interaction_%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'\n",
    "\n",
    "# parameter block\n",
    "model_str += 'parameters {\\n\\treal<lower=0,upper=1> sigma;\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\treal alpha__%s;\\n'%(c1) # growth rate\n",
    "    model_str += '\\treal epsilon__%s;\\n'%(c1) # rs response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\treal beta__%s_%s;\\n'%(c1,c2)\n",
    "model_str += '}\\n'       \n",
    "        \n",
    "# model block\n",
    "model_str += 'model {\\n\\tsigma ~ uniform(0,1);\\n'\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += '\\talpha__%s ~ normal(0,1);\\n'%(c1) # growth rate\n",
    "    model_str += '\\tepsilon__%s ~ normal(0,1);\\n'%(c1) # rs response\n",
    "    for c2 in bacterial_taxa:\n",
    "        model_str += '\\tbeta__%s_%s ~ normal(0,1);\\n'%(c1,c2)\n",
    "model_str += '\\tdlogX ~ normal('\n",
    "for c1 in bacterial_taxa:\n",
    "    model_str += 'alpha__%s*growth_rate_%s+'%(c1,c1) # growth rate\n",
    "    model_str += 'epsilon__%s*rs_response_%s+'%(c1,c1) # rs response\n",
    "    for c2 in bacterial_taxa:\n",
    "        if c1 == bacterial_taxa[-1] and c2 == bacterial_taxa[-1]:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s'%(c1,c2,c1,c2)\n",
    "        else:\n",
    "            model_str += 'beta__%s_%s*pairwise_interaction_%s_%s+'%(c1,c2,c1,c2)\n",
    "model_str += ', sigma);\\n}'\n",
    "#print(model_str)\n",
    "\n",
    "text_file = open(\"/Users/liaoc/Documents/cmdstan-2.24.1/examples/mice_scfa_3/mice_scfa_3.stan\", \"w\")\n",
    "text_file.write(\"%s\" % model_str)\n",
    "text_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot stan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hpd_grid(sample, alpha=0.05, roundto=2):\n",
    "    \"\"\"Calculate highest posterior density (HPD) of array for given alpha. \n",
    "    The HPD is the minimum width Bayesian credible interval (BCI). \n",
    "    The function works for multimodal distributions, returning more than one mode\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    sample : Numpy array or python list\n",
    "        An array containing MCMC samples\n",
    "    alpha : float\n",
    "        Desired probability of type I error (defaults to 0.05)\n",
    "    roundto: integer\n",
    "        Number of digits after the decimal point for the results\n",
    "    Returns\n",
    "    ----------\n",
    "    hpd: array with the lower \n",
    "          \n",
    "    \"\"\"\n",
    "    sample = np.asarray(sample)\n",
    "    sample = sample[~np.isnan(sample)]\n",
    "    # get upper and lower bounds\n",
    "    l = np.min(sample)\n",
    "    u = np.max(sample)\n",
    "    density = kde.gaussian_kde(sample)\n",
    "    x = np.linspace(l, u, 2000)\n",
    "    y = density.evaluate(x)\n",
    "    #y = density.evaluate(x, l, u) waitting for PR to be accepted\n",
    "    xy_zipped = zip(x, y/np.sum(y))\n",
    "    xy = sorted(xy_zipped, key=lambda x: x[1], reverse=True)\n",
    "    xy_cum_sum = 0\n",
    "    hdv = []\n",
    "    for val in xy:\n",
    "        xy_cum_sum += val[1]\n",
    "        hdv.append(val[0])\n",
    "        if xy_cum_sum >= (1-alpha):\n",
    "            break\n",
    "    hdv.sort()\n",
    "    diff = (u-l)/20  # differences of 5%\n",
    "    hpd = []\n",
    "    hpd.append(round(min(hdv), roundto))\n",
    "    for i in range(1, len(hdv)):\n",
    "        if hdv[i]-hdv[i-1] >= diff:\n",
    "            hpd.append(round(hdv[i-1], roundto))\n",
    "            hpd.append(round(hdv[i], roundto))\n",
    "    hpd.append(round(max(hdv), roundto))\n",
    "    ite = iter(hpd)\n",
    "    hpd = list(zip(ite, ite))\n",
    "    modes = []\n",
    "    for value in hpd:\n",
    "        x_hpd = x[(x > value[0]) & (x < value[1])]\n",
    "        y_hpd = y[(x > value[0]) & (x < value[1])]\n",
    "        modes.append(round(x_hpd[np.argmax(y_hpd)], roundto))\n",
    "    return hpd, x, y, modes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = az.from_cmdstan([\"output_%d.csv\"%(i) for i in np.arange(1,4)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>-0.53</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.225</td>\n",
       "      <td>0.585639</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Agathobacter</td>\n",
       "      <td>-1.13</td>\n",
       "      <td>0.37</td>\n",
       "      <td>-0.380</td>\n",
       "      <td>-1.017260</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "      <td>-0.80</td>\n",
       "      <td>0.70</td>\n",
       "      <td>-0.050</td>\n",
       "      <td>-0.128562</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Blautia</td>\n",
       "      <td>-0.67</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.085</td>\n",
       "      <td>0.207836</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basal_growth_rate</td>\n",
       "      <td>Bifidobacterium</td>\n",
       "      <td>-0.81</td>\n",
       "      <td>0.68</td>\n",
       "      <td>-0.065</td>\n",
       "      <td>-0.194746</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Type              Taxa  Left  Right  Middle       SNR  \\\n",
       "0  basal_growth_rate       Bacteroides -0.53   0.98   0.225  0.585639   \n",
       "1  basal_growth_rate      Agathobacter -1.13   0.37  -0.380 -1.017260   \n",
       "2  basal_growth_rate  Faecalibacterium -0.80   0.70  -0.050 -0.128562   \n",
       "3  basal_growth_rate           Blautia -0.67   0.84   0.085  0.207836   \n",
       "4  basal_growth_rate   Bifidobacterium -0.81   0.68  -0.065 -0.194746   \n",
       "\n",
       "   Significant  \n",
       "0        False  \n",
       "1        False  \n",
       "2        False  \n",
       "3        False  \n",
       "4        False  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines = []\n",
    "    \n",
    "# basal growth rate\n",
    "for c in bacterial_taxa:\n",
    "    var = 'alpha__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['basal_growth_rate', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "# inulin response\n",
    "for c in bacterial_taxa:\n",
    "    var = 'epsilon__%s'%(c)\n",
    "    data = []\n",
    "    for i in np.arange(0,3):\n",
    "        data.extend(list(fit.posterior[var][i].values))\n",
    "    hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "    assert len(hpd_mu) == 1\n",
    "    (x0, x1) = hpd_mu[0]\n",
    "    lines.append(['rs_response', c, x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "\n",
    "# pairwise interactions\n",
    "for c1 in bacterial_taxa:\n",
    "    for c2 in bacterial_taxa:\n",
    "        var = 'beta__%s_%s'%(c1,c2)\n",
    "        data = []\n",
    "        for i in np.arange(0,3):\n",
    "            data.extend(list(fit.posterior[var][i].values))\n",
    "        hpd_mu, x_mu, y_mu, modes_mu = hpd_grid(data)\n",
    "        assert len(hpd_mu) == 1\n",
    "        (x0, x1) = hpd_mu[0]\n",
    "        lines.append(['pairwise_interaction', (c1,c2), x0, x1, (x0+x1)/2, np.mean(data)/np.std(data), x0*x1>0])\n",
    "    \n",
    "df_stan_output_summary = pd.DataFrame(lines, columns = ['Type','Taxa','Left','Right','Middle','SNR','Significant'])\n",
    "df_stan_output_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stan_output_summary.to_excel('bayesian_regression_summary_top20.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Taxa</th>\n",
       "      <th>Left</th>\n",
       "      <th>Right</th>\n",
       "      <th>Middle</th>\n",
       "      <th>SNR</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rs_response</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>-0.97</td>\n",
       "      <td>-0.11</td>\n",
       "      <td>-0.540</td>\n",
       "      <td>-2.482663</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Bacteroides')</td>\n",
       "      <td>-2.12</td>\n",
       "      <td>-0.12</td>\n",
       "      <td>-1.120</td>\n",
       "      <td>-2.226395</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bacteroides', 'Faecalibacterium_dash_sp_dot_')</td>\n",
       "      <td>0.13</td>\n",
       "      <td>3.70</td>\n",
       "      <td>1.915</td>\n",
       "      <td>2.123964</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Agathobacter', 'Bacteroides')</td>\n",
       "      <td>0.29</td>\n",
       "      <td>2.26</td>\n",
       "      <td>1.275</td>\n",
       "      <td>2.555938</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Agathobacter', 'Agathobacter')</td>\n",
       "      <td>0.38</td>\n",
       "      <td>2.41</td>\n",
       "      <td>1.395</td>\n",
       "      <td>2.719554</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>pairwise_interaction</td>\n",
       "      <td>('Bifidobacterium', 'Bifidobacterium')</td>\n",
       "      <td>-3.35</td>\n",
       "      <td>-0.62</td>\n",
       "      <td>-1.985</td>\n",
       "      <td>-2.849105</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Type                                              Taxa  \\\n",
       "20            rs_response                                       Bacteroides   \n",
       "40   pairwise_interaction                    ('Bacteroides', 'Bacteroides')   \n",
       "47   pairwise_interaction  ('Bacteroides', 'Faecalibacterium_dash_sp_dot_')   \n",
       "60   pairwise_interaction                   ('Agathobacter', 'Bacteroides')   \n",
       "61   pairwise_interaction                  ('Agathobacter', 'Agathobacter')   \n",
       "124  pairwise_interaction            ('Bifidobacterium', 'Bifidobacterium')   \n",
       "\n",
       "     Left  Right  Middle       SNR  Significant  \n",
       "20  -0.97  -0.11  -0.540 -2.482663         True  \n",
       "40  -2.12  -0.12  -1.120 -2.226395         True  \n",
       "47   0.13   3.70   1.915  2.123964         True  \n",
       "60   0.29   2.26   1.275  2.555938         True  \n",
       "61   0.38   2.41   1.395  2.719554         True  \n",
       "124 -3.35  -0.62  -1.985 -2.849105         True  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stan_output_summary = pd.read_excel('bayesian_regression_summary_top20.xlsx', index_col=0)\n",
    "df_stan_output_summary = df_stan_output_summary[df_stan_output_summary.Significant==True]\n",
    "df_stan_output_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
